---
title: "Accurate Inference of Genetic Ancestry from Cancer Sequencing"
author: Pascal Belleau, Astrid Deschênes and Alexander Krasnitz
output:
  BiocStyle::html_document:
    toc: true
bibliography: aicsBiblio.bibtex
vignette: >
  %\VignetteIndexEntry{Accurate Inference of Genetic Ancestry from Cancer Sequencing}
  %\VignettePackage{aicsPaper}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'hide', warning=FALSE, message=FALSE}
BiocStyle::markdown()

suppressPackageStartupMessages({
  library(knitr)
  library(aicsPaper)
})

set.seed(121444)
```

<br />
**Package**: `r Rpackage("aicsPaper")`<br />
**Authors**: `r packageDescription("aicsPaper")[["Author"]]`<br />
**Version**: `r packageDescription("aicsPaper")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("aicsPaper")[["License"]]`<br />


# Licensing 

The `r Githubpkg("belleau/aicsPaper")` package and the underlying 
`r Githubpkg("belleau/aicsPaper")` code are distributed under  
the Artistic license 2.0. You are free to use and 
redistribute this software.  

<br>
<br>

# Citing

If you use this package for a publication, we would ask you to cite the 
following:

> Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz. Accurate and robust inference of genetic ancestry from cancer-derived molecular data across genomic platforms. bioRxiv 2022.02.01.478737; doi: https://doi.org/10.1101/2022.02.01.478737 

<br>
<br>

# Introduction

Multiple methods have been implemented to infer ancestry from germline DNA 
sequence [@Price2006; @Pritchard2000; @Alexander2009]. However, genotyping of 
DNA from matched normal specimens is not part of standard clinical practice
and is not performed routinely outside academic clinical centers. 
In sum, matched germline DNA sequence is often missing for cancer-derived 
molecular data. In such cases, having the possibility to infer ancestry 
from tumor-derived data would be beneficial.

The **aicsPaper** package implements an inference procedure that has been
specifically developed to accurately infer genetic ancestry from 
cancer-derived sequencing. The covered cancer-derived sequencing are, more 
specifically, tumor exomes, targeted gene panels and RNA sequences.


<br>
<br>

# Installation

To install the latest version accessible on the 
[aicsPaper Github Website](https://github.com/belleau/aicsPaper "aicsPaper Github Site"), 
the `r CRANpkg("devtools")` package is required.

```{r installDemo01, eval=FALSE}
## Load required package
library(devtools)

## Install the latest version of aicsPaper
devtools::install_github('belleau/aicsPaper')
```

<br>
<br>



# Main Steps



<br>
<br>

## Formatting the information from 1000 Genomes (1KG)

### Download the required files from 1KG

First, the pedigree file with the description of the 1KG samples
needs to be downloaded from the 1KG ftp site:

> ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_g1k.ped

The genotyping files for the 1KG samples also needs to be downloaded. Beware 
that there is one file per chromosome:

> ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV

The genotyping files are split by chromosome. All files corresponding to this 
pattern must be downloaded:

> ALL.chr\*.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz\*

All files are related to GRCh38 genome.

### Split the 1KG genotyping file to one file per sample

TODO cleanup the section

For each chromosome we run the bash command

in directory PATHGENO
chrCur is the chromosome current in the form ex: chr1

FILECUR is the vcf file from 1000 genome for the chromosome current

```{r fileGenopart1, echo=T, eval=FALSE}

mkdir ${PATHGENO}/${chrCur}

FILECUR is the vcf file from 1000 genome

for i in `seq 1 2548`
do

    j=$(( $i + 9 ))
    SAMPLE=$(zcat $FILECUR||head -n 1000 |grep "#CHROM"|cut -d$'\t' -f$j)
    zcat $FILECUR|grep -v "##"|cut -d$'\t' -f$j |bzip2 -c > ${SAMPLE}.${CHR}.vcf.bz2
done
}```

In R
To combine the genotype for each chromosome in one file for all the genome

You can run
```{r fileGenopart2, echo=T, eval=FALSE}
## Load required package
library(aicsPaper)

PATHGENOCHR <- file.path("data", "pathgenochr")# the path
PATHOUT <- file.path("data", "pathgenoOUT")

groupChr1KGSNV(PATHGENOCHR, PATHOUT)

}```


### Create a predigree file in RDS format

The function __prepPed1KG()__ is used to create the pedigree file 
in RDS format:

```{r prepPed1KG, eval=FALSE, echo=TRUE}
## Load required package
library(aicsPaper)

## The path to the pedigree file from 1KG
## In this example, the file is in the current directory
filePED1KGOri <- "20130606_g1k.ped"

## Extract needed information from the pedigree file from 1KG into a data.frame
## Only the samples with genotyping information (sample file present 
## in PATHGENO parameter) are retained to create the final data.frame
ped <- prepPed1KG(pedFile=filePED1KGOri, 
                    PATHGENO=file.path("data", "sampleGeno"))

## Save the pedigree information data.frame as a RDS file 
## In this example, the file is saved here ./data/metadata/ped1KG.rds
filePED1KG <- file.path("data", "metadata","ped1KG.rds")
saveRDS(ped,  filePED1KG)
    
```


<br>
<br>

### Prepare a bulk SNP information file based on 1KG VCF

Some intermediate file containing the SNP information from 1KG need to be
generated so the information can ultimately be imported in a GDS file 
using Bioconductor [gdsfmt](https://bioconductor.org/packages/gdsfmt/).

The bulk intermediate SNP file contains the SNP position as well as the 
frequency in each super population.

```{r intermediateVCF, echo=T, eval=FALSE}

## This is not done in R
## The python script is in the 'scriptsPy' directory
    
    for i in `ls PATHVCF/*shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz`
    do
      chr=$(echo $FILECUR|perl -n -e '/ALL\.(chr[^\.]+)/;print $1')
      python PATH2SCRIPT/extract1000gFreq.py ${FILECUR} matFreq.${chr}
    done
    for i in `seq 1 22`
    do
      cat matFreq.chr${i}.txt >matFreqSNV.txt
      bzip2 matFreqSNV.txt
    done
    
```

The bulk SNP info file is called '__matFreqSNV.txt__'.

<br>
<br>

### Filter the bulk SNP information file

The bulk SNP file is filtered to only retain the SNPs with frequency 
higher then a specific cut-off (here >=0.01) for at least one super 
population. 

The filter SNP file is saved in RDS format ('__fileLSNP__' parameter). A 
second file containing the index of the retained SNPs is also 
created ('__fileLSNP__' parameter)

The function __generateMapSnvSel()__ is used to filter the SNP file and 
generated the needed RDS files:

```{r filterVCF, echo=T, eval=FALSE}
## Load required package
library(aicsPaper)

## The path to the intermediate SNP info file
fileSNV.v <- file.path(PATHGENO, "matFreqSNV.txt.bz2")

## The paths and names of the two output files
## One file contains the index of retained SNPs ("listSNP.rds")
## One file contains the filter SNP information ("mapSNVSel.rds")
fileLSNP.v <- file.path(PATHSEL, "listSNP.rds")
fileFREQ.v <- file.path(PATHSEL, "mapSNVSel.rds")

## Filter the bulk SNP file (fileSNV parameter)
## Create a RDS with filter SNPs (fileFREQ parameter)
## Also creates a RDS with the indexes of the retained SNPs (fileFREQ parameter)
generateMapSnvSel(cutOff=0.01, fileSNV=fileSNV.v,
                      fileLSNP = fileLSNP.v,
                      fileFREQ = fileFREQ.v)
```


<br>
<br>


### Generate the GDS file with the 1KG information

The CoreArray Genomic Data Structure (GDS) data files are files suited for 
large-scale datasets, especially for data which are much larger than the 
available random-access memory.

The function __generateGDS1KG()__ is used to generate the GDS file that will
contain the information related to 1KG:

```{r gdsCreation, echo=TRUE, eval=FALSE}
## TODO: what is PATHMETA and PATHSEL

## The path and file names of the required files
## First, the RDS file containing the pedigree information
## Second, the RDS file with the indexes of the retained SNPs
## Third, the RDS file with the filtered SNP information 
filePED1KG <- file.path(PATHMETA,"ped1KG.rds")
fileLSNP.v <- file.path(PATHSEL, "listSNP.rds")
fileFREQ.v <- file.path(PATHSEL, "mapSNVSel.rds")

## The name of the GDS file that will be created    
fileNameGDS <- "matGeno1000g.gds"
fileGDS <- file.path(PATHGDS, fileNameGDS)
    
## Generate GDS file containing the 1KG information
generateGDS1KG(PATHGENO=file.path("data", "sampleGeno"),
                    fileNamePED=filePED1KG,
                    fileSNPSel=fileFREQ.v,
                    fileListSNP=fileLSNP.v,
                    fileNameGDS=fileGDS)
```


<br>
<br>

### Identify genetically related patients in 1KG 

As only unrelated patients can be used in the following analyses, the 
genetically related patients in 1KG must be identified.

The function __identifyRelative()__ identifies patients that are genetically 
related in the 1KG files. It generates a list of unrelated as well as a list of
related patients.

```{r identifyRelative, echo=TRUE, eval=FALSE}
## TODO: what is PATHMETA 

## The name of the GDS file that contains the 1KG information    
fileNameGDS <- "matGeno1000g.gds"
fileGDS <- file.path(PATHGDS, fileNameGDS)

## TODO: whait are those?
fileIBD <- file.path(PATHMETA,"ibd.All.0.05.rds")
filePart <- file.path(PATHMETA,"part.All.0.05.rds")

## Identify the genetically related patients in 1KG
gds <- snpgdsOpen(fileGDS)
identifyRelative(gds=fileGDS, maf=0.05, thresh = 2^(-11/2),
                     fileIBD=fileIBD, filePart=filePart)
closefn.gds(gds)
```

<br>
<br>


### TODO 2

TODO

```{r addRef2GDS1KG, echo=TRUE, eval=FALSE}
 # create the ref sample list which correspond
    # to the list of sample unrelated
    
    addRef2GDS1KG(fileGDS=gds,
                  filePart=filePart)
    
```

<br>
<br>

# Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

<br>
<br>


# References

