---
title: "Accurate Inference of Genetic Ancestry from Cancer Sequencing"
author: Pascal Belleau, Astrid Deschênes and Alexander Krasnitz
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: true
  pkgdown:
    number_sections: yes
    as_is: true
urlcolor: darkred
linkcolor: darkred
bibliography: aicsBiblio.bibtex
vignette: >
  %\VignetteIndexEntry{Accurate Inference of Genetic Ancestry from Cancer Sequencing}
  %\VignettePackage{RAIDS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
BiocStyle::markdown()

suppressPackageStartupMessages({
  library(knitr)
  library(RAIDS)
})

set.seed(121444)
```

<br />
**Package**: `r Rpackage("RAIDS")`<br />
**Authors**: `r packageDescription("RAIDS")[["Author"]]`<br />
**Version**: `r packageDescription("RAIDS")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("RAIDS")[["License"]]`<br />


# Licensing 

The `r Githubpkg("KrasnitzLab/RAIDS")` package and the underlying 
`r Githubpkg("KrasnitzLab/RAIDS")` code are distributed under  
the https://opensource.org/licenses/Apache-2.0 license. You are free to use and 
redistribute this software.  

<br>
<br>

# Citing

If you use the **RAIDS** package for a publication, we would ask you to cite 
the following:

> Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz. Accurate and robust inference of genetic ancestry from cancer-derived molecular data across genomic platforms. bioRxiv 2022.02.01.478737; doi: https://doi.org/10.1101/2022.02.01.478737 

<br>
<br>

# Introduction

Multiple methods have been implemented to infer ancestry from germline DNA 
sequence [@Price2006; @Pritchard2000; @Alexander2009]. However, genotyping of 
DNA from matched normal specimens is not part of standard clinical practice
and is not performed routinely outside academic clinical centers. 
In sum, matched germline DNA sequence is often missing for cancer-derived 
molecular data. In such cases, having the possibility to infer ancestry 
from tumor-derived data would be beneficial.

The **RAIDS** package implements an inference procedure that has been
specifically developed to accurately infer genetic ancestry from 
cancer-derived sequencing. The covered cancer-derived sequencing are, more 
specifically, tumor exomes, targeted gene panels and RNA sequences.

The **RAIDS** package includes a data synthesis method that enables 
the selection of the
best parameters for a specific cancer profile and also permits a 
statistically assessment of inference accuracy for an individual 
cancer-derived molecular profile.

<br>
<br>

# Installation

To install the latest version accessible on the 
[RAIDS Github Website](https://github.com/KrasnitzLab/RAIDS "RAIDS Github Site"), 
the `r CRANpkg("devtools")` package is required.

```{r installDemo01, eval=FALSE}
## Load required package
library(devtools)

## Install the latest version of RAIDS
devtools::install_github('KrasnitzLab/RAIDS')
```

<br>
<br>


# Main Steps


This is an overview of genetic ancestry inference from cancer-derived 
molecular data:

```{r graphMainSteps, echo = FALSE, fig.align="center", fig.cap="An overview of the genetic ancestry inference process.", out.width = '120%', results='asis'}
knitr::include_graphics("MainSteps_v01.png")
```

The main steps are:

1. Format the information from 1000 Genomes (1KG) (optional)
2. Format the information from an external study
3. Find the optimized parameters for the ancestry inference
4. Run the ancestry inference on the external study

In the following sections, those main steps are described in details.

<br>
<br>

## Step 1 - Formatting the information from 1000 Genomes (optional)


```{r graphStep1, echo=FALSE, fig.align="center", fig.cap="Step 1 - Formatting the information from 1000 Genomes (optional)", out.width = '120%', results='asis'}
knitr::include_graphics("MainSteps_Step1_v01.png")
```


******

<span style="color:darkred;font-weight:bold">Beware that a pre-processed 1KG 
GDS file is available at this address:</span>

<span style="color:red">
[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds)
</span>

<span style="color:darkred;font-weight:bold">The size of the file is 15GB.</span>
<span style="color:darkred;font-weight:bold">This section can be skipped if you choose to use this file.</span>

******

There is a vignette  
[Formatting the information from 1000 Genomes (optional)](Create_1KG_GDS_file.html) 
that is dedicated to this step.


## Step 2 - Format information from an external study to get if ready for ancestry inference

To run the ancestry inference on samples from an external study, the sample
information must be formatted in a specific format.


```{r graphStep2, echo = FALSE, fig.align="center", fig.cap="Step 2 - Formatting the information from an external study", out.width = '120%', results='asis'}
knitr::include_graphics("MainSteps_Step2_v01.png")
```


Those steps shows how to prepare the sample information needed as input 
for the ancestry inference.

Those steps are:

1. Download or create the 3 reference files
2. Extract the read coverage at each SNP position in the 1KG GDS file
3. Pruned

Beware that a mapped BAM file is needed for each sample. The reference genome 
used for the mapping must correspond to the one used to generate the 
1KG GDS file. In this case, the 1KG GDS file is based on genome hg38.

### Step 1. Download or create the 3 reference files

wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds
wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matAnnot1000g.gds
wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/snvSel0.01.vcf.gz

In  a directory PATH_REF (I will use this name PATH_REF to refer at this directory)

### Step 2. Extract the read coverage at each SNP position in the 1KG GDS file

We used here snp-pileup from Facet

<br>
<br>


You need a __ped__ file (call later FILE_PED) with the column:

"Case.ID" Patient identifier     
"Name.ID"* This the unique id of the table
"Sample.Type"* Ex Primary Tumor or Blood Derived Normal (can be too WES or RNAseq) 
"Diagnosis"* C or N (cancer normal) this is for the sample not the patient
"Source"* The tissue ex Ovary 

must contain the information for all the samples (you can have more) process 
at the same time.


You need the extra code in the package Facet to extract the SNV
[snp-pileup](https://github.com/mskcc/facets/tree/master/inst/extcode) 
[@Shen2016].

The output from snp-pileup should be Name.ID.txt 
(snp-pileup add .gz to the filename) 

```{r snpPileup, echo=TRUE, eval=FALSE}

    snp-pileup -g -d5000 -q15 -Q20  -r0 PATH_REF/snvSel0.01.vcf.gz PATH_OUT/Name.ID.txt FILEBAM.bam
    
```


#### Step 3. Create a GDS file for each sample

We will call this file gdsSample

in R:
Here
```{r appendStudy2GDS1KG, echo=TRUE, eval=FALSE}

    library(RAIDS)
    
    ##########################################
    # Files related to the reference
    # Downloaded
    ##########################################
    PATH_1KG <- PATH_REF
    
    fileName.GDS <- "matGeno1000g.gds"
    fileName.GDS.Annot <- "matAnnot1000g.gds"
    file.GDS <- file.path(PATH_1KG, fileName.GDS)
    file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

    ##########################################
    
    # Need the path where the files generated by
    # snp-pileup are
    PATHGENO <- file.path(PATH_PILEUP)
    
    # Path to the directory where the gds file for each sample
    # will be create
    PATHSAMPLEGDS <- file.path("The path")
    
    # The file ped defined before
    filePED <- FILE_PED
    ped <- readRDS(filePED)
    
    
    # Need a data.frame with study description
    studyDF <- data.frame(study.id = "Id of the study",
                              study.desc = "Descripttion",
                              study.platform = "Platform string like Whole Exome",
                              stringsAsFactors = FALSE)
    
    listSamples <- ped[, "Name.ID"]
    
    # The next function will create a GDS for each 
    # sample in listSamples
    
    appendStudy2GDS1KG(PATHGENO = PATHGENO,
                       fileNamePED = filePED,
                       fileNameGDS = file.GDS,
                       listSamples = listSamples,
                       batch = 1,
                       studyDF = studyDF,
                       PATHSAMPLEGDS = PATHSAMPLEGDS)))
    
```


### Step 4. Select a pruned subset of the single nucleotide variant 

We select a subset of SNV pruned SNV
We add in each gdsSample for each sample the list of SNV in the pruning.
It use more disk space but the analysis for the samples can be run simulatanusly 
with multiple process (if we we all samples in one file only one process can 
write in the GDS). 


or in R for a s

```{r pruningSample, echo=TRUE, eval=FALSE}

    library(RAIDS)
    
    ##########################################
    # Files related to the reference
    # Downloaded
    ##########################################
    PATH_1KG <- PATH_REF
    
    fileName.GDS <- "matGeno1000g.gds"
    fileName.GDS.Annot <- "matAnnot1000g.gds"
    file.GDS <- file.path(PATH_1KG, fileName.GDS)
    file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

    ##########################################
    
    # Path to the directory where the gds file for each sample
    # will be create
    PATHSAMPLEGDS <- file.path("The path")
    
    gds <- snpgdsOpen(file.GDS)
    
    # The file ped defined before
    filePED <- FILE_PED
    ped <- readRDS(filePED)
    # You need a list of the sample Name.ID 
    # you can get it for other source than 
    # filePED 
    listSamples <- ped[, "Name.ID"]
    
    
    # Can be run in paralle or in different clusters...
    for(i in seq_len(length(listSamples))){
        # Compute the SNV pruned subset
        pruningSample(gds=gds,
                      method="corr",
                      sampleCurrent = listSamples[i],
                      listSNP = NULL,
                      slide.max.bp.v = 5e5,
                      ld.threshold.v=sqrt(0.1),
                      np = 1,
                      verbose.v=FALSE,
                      chr = NULL,
                      minAF.SuperPop = NULL,
                      keepGDSpruned = TRUE,
                      PATHSAMPLEGDS = PATHSAMPLEGDS,
                      keepFile = FALSE)
        
        # Add the genotype of 1KG and of the sample listSamples[i] 
        # for SNV pruned in the gdsSample
        addStudy1Kg(gds, file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds")))
        # Add the sample name from 1KG in the study.annot of gdsSample
        addStudy1Kg(gds, file.GDSSample)

    }
    
    
    closefn.gds(gds)
```

### Step 5. Estimate the allelic for the SNV pruned 



or in R for a s

```{r pruningSample, echo=TRUE, eval=FALSE}

    library(RAIDS)
    library(BSgenome)
    library(BSgenome.Hsapiens.UCSC.hg38)

    ##########################################
    # Files related to the reference
    # Downloaded
    ##########################################
    PATH_1KG <- PATH_REF
    
    fileName.GDS <- "matGeno1000g.gds"
    fileName.GDS.Annot <- "matAnnot1000g.gds"
    file.GDS <- file.path(PATH_1KG, fileName.GDS)
    file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

    ##########################################
    
    gds <- snpgdsOpen(file.GDS)
    
    # Path to the directory where the gds file for each sample
    # will be create
    PATHSAMPLEGDS <- file.path("The path")
    
    # Look at studyDF at step 3.
    study.id <- "Id of the study"
    
    # The file ped defined before
    filePED <- FILE_PED
    ped <- readRDS(filePED)
    # You need a list of the sample Name.ID 
    # you can get it from other source than 
    # filePED 
    listSamples <- ped[, "Name.ID"]
    
    # You need the autosomal chr length from the genome.
    chrInfo <- integer(22)
    
    for(i in seq_len(22)){
        chrInfo[i] <- length(Hsapiens[[paste0("chr", i)]])
    }

    # Can be run in paralle or in different clusters...
    for(i in seq_len(length(listSamples))){
        file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds"))
        gdsSample <- openfn.gds(file.GDSSample, readonly = FALSE)

        estimateAllelicFraction(gds, gdsSample,
                                listSamples[i], study.id,
                                chrInfo)
        closefn.gds(gdsSample)
    }
    
    
    closefn.gds(gds)
```

### Step 6. Generate the synthetic dataset



or in R 

```{r pruningSample, echo=TRUE, eval=FALSE}

    library(RAIDS)
    

    ##########################################
    # Files related to the reference
    # Downloaded
    ##########################################
    PATH_1KG <- PATH_REF
    
    fileName.GDS <- "matGeno1000g.gds"
    fileName.GDS.Annot <- "matAnnot1000g.gds"
    file.GDS <- file.path(PATH_1KG, fileName.GDS)
    file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

    ##########################################
    
    gds <- openfn.gds(file.GDS)
    gds.Annot <- openfn.gds(file.Annot)

    
    # Path to the directory where the gds file for each sample
    # will be create
    PATHSAMPLEGDS <- file.path("The path")
    
    
    # Need a data.frame with study description for the
    # synthetic data
    studyDF <- data.frame(study.id = "Id of the synthetic dataset",
                              study.desc = "Descripttion",
                              study.platform = "Platform string like Whole Exome",
                              stringsAsFactors = FALSE)
    
    
    
    # Look at studyDF at step 3.
    study.id <- "Id of the study"
    
    # The file ped defined before
    filePED <- FILE_PED
    ped <- readRDS(filePED)
    # You need a list of the sample Name.ID 
    # you can get it from other source than 
    # filePED 
    listSamples <- ped[, "Name.ID"]
    
    # You need a set of 1KG too use in the synthetic dataset
    # You need a list of id from 1KG get from file
    # or use the select1KGPop
    # The function select1KGPop select nbSamples 
    # of each 1KG pop.group (subcontinental)
    # with the same seed the function generate the same set
    set.seed(3043)
    dataRef <- select1KGPop(gds, nbSamples = 30L)
    listSampleRef <- dataRef$sample.id
    
    # Can be run in paralle or in different clusters...
    for(i in seq_len(length(listSamples))){
        
        file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds"))
        prepSynthetic(file.GDSSample,
              listSampleRef,
              listSamples[i],
              studyDF,
              nbSim = 1,
              prefId = "1")

        # synthetic data generator
        resG <- syntheticGeno(gds,
                              gds.Annot,
                              file.GDSSample,
                              listSamples[i],
                              listSampleRef,
                              nbSim = 1,
                              prefId = "1")

    }
    
    
    closefn.gds(gds)
    closefn.gds(gds.Annot)
    
```

### Step 7. Compute the PCA-KNN ancestry call for each synthetic data



or in R 

```{r pruningSample, echo=TRUE, eval=FALSE}

    library(RAIDS)
    

    ##########################################
    # Files related to the reference
    # Downloaded
    ##########################################
    PATH_1KG <- PATH_REF
    
    fileName.GDS <- "matGeno1000g.gds"
    fileName.GDS.Annot <- "matAnnot1000g.gds"
    file.GDS <- file.path(PATH_1KG, fileName.GDS)
    file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

    ##########################################
    
    gds <- openfn.gds(file.GDS)
    gds.Annot <- openfn.gds(file.Annot)

    
    # Path to the directory where the gds file for each sample
    # will be create
    PATHSAMPLEGDS <- file.path("The path")
    
    # directory which content directory name Name.ID in whihc the PCA is save
    # 
    PATH_OUT <- file.path("The path")
    if(! file.exists(PATH_OUT)){
        dir.create(PATH_OUT)
    }
    
    # Need a data.frame with study description for the
    # synthetic data
    studyDF <- data.frame(study.id = "Id of the synthetic dataset",
                              study.desc = "Descripttion",
                              study.platform = "Platform string like Whole Exome",
                              stringsAsFactors = FALSE)
    
    
    
    # Look at studyDF at step 3.
    study.id <- "Id of the study"
    
    # The file ped defined before
    filePED <- FILE_PED
    ped <- readRDS(filePED)
    # You need a list of the sample Name.ID 
    # you can get it from other source than 
    # filePED 
    listSamples <- ped[, "Name.ID"]
    
    # Get the known superPop for the reference
    spRef <- getRef1KGPop(gds, "superPop")
    
    
    # The set of 1KG too use in the synthetic dataset
    # if you use select1KGPop the seed must be the same 
    # as the step 6 
    set.seed(3043)
    dataRef <- select1KGPop(gds, nbSamples = 30L)
    
    
    
    # Can be run in paralle or in different clusters...
    for(i in seq_len(length(listSamples))){
        
        file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds"))
        
        PATH_OUT_SAMPLE <- file.path(PATH_OUT, listSamples[i])
        if(! file.exists(PATH_OUT_SAMPLE)){
            dir.create(PATH_OUT_SAMPLE)
        }
        
        gdsSample <- snpgdsOpen(file.GDSSample)
        
        # data.frame with the sample.id (Name.ID) with the study id
        study.annot <- read.gdsn(index.gdsn(gdsSample, "study.annot"))
        
        # list of sample genotyped
        sample.id <- read.gdsn(index.gdsn(gdsSample, "sample.id"))
        
        # The study id is Ref.1KG if you use addStudy1Kg in step 4
        sample.Unrel.All <- study.annot$data.id[study.annot$study.id == "Ref.1KG"]
        
        
        
        # get a data.frame with the synthetic sample description
        study.annot.synt <- study.annot[which(study.annot$study.id == studyDF$study.id),]
        
        # The list of SNV pruned
        pruned <-  read.gdsn(index.gdsn(gdsSample, "snp.id"))
        
        for(j in seq_len(nrow(study.annot.synt))){
            pcaCur <- computePCAsynthetic(gdsSample, pruned, sample.id, sample.Unrel.All, study.annot.synt[j,])

            knnCur <- computeKNNSuperPoprSynthetic(pcaCur, sample.Unrel.All, study.annot.synt[posSyn,], spRef)
            
            res <- list(resPCA = pcaCur,
                        resKNN = knnCur)
            
            saveRDS(res, file.path(PATH_OUT_SAMPLE, paste0(study.annot.synt[posSyn, "data.id"] , ".rds")))

        }
        closefn.gds(gdsSample)
    }
    
    
    closefn.gds(gds)
    closefn.gds(gds.Annot)
    
```

### Step 8. Select paramters for the PCA-KNN ancestry call



or in R 

```{r pruningSample, echo=TRUE, eval=FALSE}

    library(RAIDS)
    

    ##########################################
    # Files related to the reference
    # Downloaded
    ##########################################
    PATH_1KG <- PATH_REF
    
    fileName.GDS <- "matGeno1000g.gds"
    fileName.GDS.Annot <- "matAnnot1000g.gds"
    file.GDS <- file.path(PATH_1KG, fileName.GDS)
    file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

    ##########################################
    
    gds <- openfn.gds(file.GDS)
    gds.Annot <- openfn.gds(file.Annot)

    
    # Path to the directory where the gds file for each sample
    # will be create
    PATHSAMPLEGDS <- file.path("The path")
    
    # directory which content directory name Name.ID in whihc the PCA is save
    # 
    PATH_OUT <- file.path("The path")
    
    
    # Need a data.frame with study description for the
    # synthetic data
    studyDF <- data.frame(study.id = "Id of the synthetic dataset",
                              study.desc = "Descripttion",
                              study.platform = "Platform string like Whole Exome",
                              stringsAsFactors = FALSE)
    
    
    
    # Look at studyDF at step 3.
    study.id <- "Id of the study"
    
    # The file ped defined before
    filePED <- FILE_PED
    ped <- readRDS(filePED)
    # You need a list of the sample Name.ID 
    # you can get it from other source than 
    # filePED 
    listSamples <- ped[, "Name.ID"]
    
    
    spRef <- getRef1KGPop(gds, "superPop")
    

    # Can be run in paralle or in different clusters...
    listParaSample <- list()
    listPCASample <- list()
    listKNNSample <- list()
    listAncestrySample <- list()
    for(i in seq_len(length(listSamples))){
        
        file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds"))
        gdsSample <- snpgdsOpen(file.GDSSample)
        
        PATH_OUT_SAMPLE <- file.path(PATH_OUT, listSamples[i])
        # The directory PATH_OUT_SAMPLE must just content 
        # only the rds files from step 7
        listSynthetic <- dir(PATH_OUT_SAMPLE, ".rds")
        listKNN <- list()
        for(i in seq_len(length(listSynthetic))){
            tmp <- readRDS(file.path(PATHSYN, listSynthetic[i]))
            listKNN[[i]] <- tmp$resKNN$matKNN
        }
        matKNN.All <- do.call(rbind, listKNN)

        # The parameter superPop is define by the output of step 7
        # the function return data.frame with the known superPop 
        # associate with synthetic data.
        
        pedSyn <- prepPedSynthetic1KG(gds, gdsSample, studyDF$study.id, "superPop")
        
        # list of superPop in synthetic dataset
        listSP <- unique(pedSyn$superPop)
        
        # The SuperPop is the super population field in matKNN.All
        # superPop are the known super population from the ref
        # return the parameter selected.
        listParaSample[[listSamples[i]]] <- selParaPCAUpQuartile(matKNN.All, pedSyn, "superPop", "SuperPop", listSP)
        
        listPCASample[[listSamples[i]]] <- computePCARefSample(gdsSample, listSamples[i], study.id.ref = "Ref.1KG")
        
        
        listKNNSample[[listSamples[i]]] <- computeKNNSuperPopSample(gdsSample, listSamples[i], spRef)
        
        listAncestrySample[[listSamples[i]]] <- 
    }
    
    
    
    closefn.gds(gds)
    closefn.gds(gds.Annot)
    
```





<br>
<br>

# Pre-processed files are available

Pre-processed files, such as the 1KG GDS file, are available at this address:


[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper)

Beware that some of those files are voluminous.

<br>
<br>

# Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

<br>
<br>


# References

