---
title: "Accurate Inference of Genetic Ancestry from Cancer Sequencing"
author: Pascal Belleau, Astrid Deschênes and Alexander Krasnitz
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: true
  pkgdown:
    number_sections: yes
    as_is: true
urlcolor: darkred
linkcolor: darkred
bibliography: aicsBiblio.bibtex
vignette: >
  %\VignetteIndexEntry{Accurate Inference of Genetic Ancestry from Cancer Sequencing}
  %\VignettePackage{RAIDS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
BiocStyle::markdown()

suppressPackageStartupMessages({
  library(knitr)
  library(RAIDS)
})

set.seed(121444)
```

<br />
**Package**: `r Rpackage("RAIDS")`<br />
**Authors**: `r packageDescription("RAIDS")[["Author"]]`<br />
**Version**: `r packageDescription("RAIDS")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("RAIDS")[["License"]]`<br />


# Licensing 

The `r Githubpkg("KrasnitzLab/RAIDS")` package and the underlying 
`r Githubpkg("KrasnitzLab/RAIDS")` code are distributed under  
the https://opensource.org/licenses/Apache-2.0 license. You are free to use and 
redistribute this software.  

<br>
<br>

# Citing

If you use the **RAIDS** package for a publication, we would ask you to cite 
the following:

> Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz. Accurate and robust inference of genetic ancestry from cancer-derived molecular data across genomic platforms. bioRxiv 2022.02.01.478737; doi: https://doi.org/10.1101/2022.02.01.478737 

<br>
<br>

# Introduction

Multiple methods have been implemented to infer ancestry from germline DNA 
sequence [@Price2006; @Pritchard2000; @Alexander2009]. However, genotyping of 
DNA from matched normal specimens is not part of standard clinical practice
and is not performed routinely outside academic clinical centers. 
In sum, matched germline DNA sequence is often missing for cancer-derived 
molecular data. In such cases, having the possibility to infer ancestry 
from tumor-derived data would be beneficial.

The **RAIDS** package implements an inference procedure that has been
specifically developed to accurately infer genetic ancestry from 
cancer-derived sequencing. The covered cancer-derived sequencing are, more 
specifically, tumor exomes, targeted gene panels and RNA sequences.

The **RAIDS** package includes a data synthesis method that enables 
the selection of the
best parameters for a specific cancer profile and also permits a 
statistically assessment of inference accuracy for an individual 
cancer-derived molecular profile.

<br>
<br>

# Installation

To install the latest version accessible on the 
[RAIDS Github Website](https://github.com/KrasnitzLab/RAIDS "RAIDS Github Site"), 
the `r CRANpkg("devtools")` package is required.

```{r installDemo01, eval=FALSE}
## Load required package
library(devtools)

## Install the latest version of RAIDS
devtools::install_github('KrasnitzLab/RAIDS')
```

<br>
<br>


# Main Steps


This is an overview of genetic ancestry inference from cancer-derived 
molecular data:

```{r graphMainSteps, echo = FALSE, fig.align="center", fig.cap="An overview of the genetic ancestry inference process.", out.width = '120%', results='asis'}
knitr::include_graphics("MainSteps_v01.png")
```

The main steps are:

1. Format the information from 1000 Genomes (1KG) (optional)
2. Format the information from an external study
3. Find the optimized parameters for the ancestry inference
4. Run the ancestry inference on the external study

In the following sections, those main steps are described in details.

<br>
<br>

## Step 1 - Formatting the information from 1000 Genomes (optional)


```{r graphStep1, echo=FALSE, fig.align="center", fig.cap="Step 1 - Formatting the information from 1000 Genomes (optional)", out.width = '120%', results='asis'}
knitr::include_graphics("MainSteps_Step1_v01.png")
```


******

<span style="color:darkred;font-weight:bold">Beware that a pre-processed 1KG 
GDS file is available at this address:</span>

<span style="color:red">
[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds)
</span>

<span style="color:darkred;font-weight:bold">The size of the file is 15GB.</span>
<span style="color:darkred;font-weight:bold">This section can be skipped if you choose to use this file.</span>

******

There is a vignette  
[Formatting the information from 1000 Genomes (optional)](Create_1KG_GDS_file.html) 
that is dedicated to this step.


## Step 2 - Format information from an external study to get if ready for ancestry inference

To run the ancestry inference on samples from an external study, the sample
information must be formatted in a specific format.


```{r graphStep2, echo = FALSE, fig.align="center", fig.cap="Step 2 - Formatting the information from an external study", out.width = '120%', results='asis'}
knitr::include_graphics("MainSteps_Step2_v01.png")
```


Those steps shows how to prepare the sample information needed as input 
for the ancestry inference.

Those steps are:

1. Extract the read coverage at each SNP position in the 1KG GDS file
2. Create a VCF file containing the retained SNP positions from 1KG GDS file

Beware that a mapped BAM file is needed for each sample. The reference genome 
used for the mapping must correspond to the one used to generate the 
1KG GDS file. In this case, the 1KG GDS file is based on genome hg38.

### Extract the read coverage at each SNP position in the 1KG GDS file

We used here snp-pileup from Facet

First you a vcf file with the SNV you want to keep

<br>
<br>

### Create a VCF file containing the retained SNP positions from 1KG GDS file

The __snvListVCF()__ function is used to generate a VCF file that contains
the information of all retained SNPs from 1KG GDS file:

```{r snvListVCF, echo=TRUE, eval=FALSE}
## Load required package
library(RAIDS)

## Open the 1KG GDS file    
gds <- snpgdsOpen(fileGDS1kg)

## The VCF file that will be created
fileOUT <- "SNPretained.VCF"

## Generate the VCF with the retained SNP position
snvListVCF(gds=gds, fileOUT=fileOUT, offset=1, freqCutoff=NULL)

## Close the 1KG GDS file 
closefn.gds(gds)
``` 

You should compress and indexing the newly created VCF file. Do do so, 
you need to install [HTSlib](http://www.htslib.org/download/) [@Bonfield2021].

In a terminal:

```{r indexVCF, echo=TRUE, eval=FALSE}
## This in not a R script
## This script is in bash

## Compress the new VCF file (fileOUT parameter)
bgzip fileOUT

## Index the new VCF file
## HTSlib software is needed
tabix -p vcf fileOUT.gz
```


You need a __ped__ file with the column:

"sample.id"* Id of the sample but the sample can be genotyped more than one
"sex"*
"pop.group"* can be self declare or something else but must be there
"superPop"* can be self declare or something else but must be there
"Sample.Type"* Ex Primary Tumor or Blood Derived Normal 
"Diagnosis"* C or N (cancer normal) this is for the sample not the patient
"Source"* The tissue ex Ovary 
"Recurrent"  
"Case.ID" Patient identifier     
"Name.ID"* This the unique id of the table
"batch" Number

You need the extra code in the package Facet to extract the SNV
[snp-pileup](https://github.com/mskcc/facets/tree/master/inst/extcode) 
[@Shen2016].

The output from snp-pileup should be Name.ID.txt 
(snp-pileup add .gz to the filename) 

```{r snpPileup, echo=TRUE, eval=FALSE}

    snp-pileup -g -d5000 -q15 -Q20  -r0 VCFGenerateAbove snvSel0.01.vcf.gz OUTPUT.txt FILEBAM.bam
    
```

TODO change the path to something generic

in R:

```{r appendStudy2GDS1KG, echo=TRUE, eval=FALSE}

    library(RAIDS)
    
    PATHGENO <- file.path("data", "snpCancer")
    PATHMETA <- file.path("data", "metadata")
    PATHGDS <- file.path("data", "genoGDS1000gAF.All.TCGA_OV.WXS")
    
    fileNamePED <- "pedTCGA_OV_WXS_C.rds"
    fileNameGDS <- "matGeno1000g.gds"
    
    filePED <- file.path(PATHMETA, fileNamePED)
    fileNameGDS <- file.path(PATHGDS, fileNameGDS)
    
    ped <- readRDS(filePED)
    
    studyDF <- data.frame(study.id = "TCGA-OV.WXS.C",
                              study.desc = "Ovarian example",
                              study.platform = "WXS",
                              stringsAsFactors = FALSE)
    
    listSamples <- ped[, "Name.ID"]
    appendStudy2GDS1KG(PATHGENO = PATHGENO,
                       fileNamePED = filePED,
                       fileNameGDS = fileNameGDS,
                       listSamples = listSamples,
                       studyDesc = studyDF,
                       batch = 1)
```



We select a subset of SNV pruned SNV
We create a GDS for each sample with only the SNP in the pruning.
It use more disk space but the analysis for the samples can be run simulatanusly 
with multiple process (if we we all samples in one file only one process can 
write in the GDS). 


or in R for a s

```{r pruningSample, echo=TRUE, eval=FALSE}

    library(RAIDS)
    
    # fileNameGDS with the sample and 1KG genotype
    # listSamples list of sample.id from the study
    # PATHSAMPLEGDS is the path where the gds specific 
    # to the sample is created
    
    studyCur <- "Study.name"
    gds <- snpgdsOpen(fileNameGDS)
    
    
    
    sampleGDS <- index.gdsn(gds, "sample.id")
    sample.id <- read.gdsn(sampleGDS)
    
    study.offset <- read.gdsn(index.gdsn(gds, "study.offset"))
    
    study.id <- read.gdsn(index.gdsn(gds, "study.annot/study.id"))
    
    listSamples <- sample.id[which(study.id == studyCur) + study.offset]

    for(i in seq_len(length(listSamples))){
        print(system.time(pruned <- pruningSample(gds=gds,
                                            method="corr",
                                            sampleCurrent = listSamples[i],
                                            listSNP = NULL,
                                            slide.max.bp.v = 5e5,
                                            ld.threshold.v=sqrt(0.1),
                                            np = 1,
                                            verbose.v=FALSE,
                                            chr = NULL,
                                            minAF.SuperPop = NULL,
                                            keepGDSpruned = TRUE,
                                            PATHSAMPLEGDS = PATHSAMPLEGDS,
                                            keepFile = FALSE)))
    }
    
    Sys.time()
    closefn.gds(gds)
```


<br>
<br>

# Pre-processed files are available

Pre-processed files, such as the 1KG GDS file, are available at this address:


[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper)

Beware that some of those files are voluminous.

<br>
<br>

# Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

<br>
<br>


# References

