---
title: "Population reference dataset GDS files"
author: Pascal Belleau, Astrid Deschênes and Alexander Krasnitz
output:
    BiocStyle::html_document:
        number_sections: yes
        toc: true
    pkgdown:
        number_sections: yes
        as_is: true
urlcolor: darkred
linkcolor: darkred
bibliography: aicsBiblio.bibtex
vignette: >
    %\VignetteIndexEntry{Population reference dataset GDS files}
    %\VignettePackage{RAIDS}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
BiocStyle::markdown()

suppressPackageStartupMessages({
    library(knitr)
    library(RAIDS)
    library(SNPRelate)
    library(gdsfmt)
})

set.seed(121444)
```

<br />
**Package**: `r Rpackage("RAIDS")`<br />
**Authors**: `r packageDescription("RAIDS")[["Author"]]`<br />
**Version**: `r packageDescription("RAIDS")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("RAIDS")[["License"]]`<br />


<br>
<br>




This section explains in further details the format of the population 
reference files that are required to run the ancestry inference tool. 

Two different files are generated from a reference dataset:

- The Population Reference GDS File
- The Population Reference SNV Annotation GDS file


<br>
<br>

# Population Reference GDS File

The *Population Reference GDS file* should contain the genome-wide SNV 
information related to the population data set with known genetic ancestry. 
This reference data set will be used to generate the simulated samples. It is 
also used to generate the PCA on which the samples of interest are going to 
be projected. 

The *Population Reference GDS file* is a GDS object of class 
[SNPGDSFileClass](https://www.bioconductor.org/packages/release/bioc/vignettes/SNPRelate/inst/doc/SNPRelate.html#data-formats-used-in-snprelate) 
from [SNPRelate](https://www.bioconductor.org/packages/release/bioc/html/SNPRelate.html) 
package [@Zheng2012].

Beware that only unrelated samples should be present in the GDS file. 

```{r runRefGDS, echo=TRUE, eval=TRUE, collapse=TRUE, warning=FALSE, message=FALSE}
#############################################################################
## Load required packages
#############################################################################
library(RAIDS)    
library(SNPRelate)

pathReference <- system.file("extdata/example/", package="RAIDS")

fileReferenceGDS <- file.path(pathReference, "gdsRef", "ex1kg.gds")

gdsRef <- snpgdsOpen(fileReferenceGDS)

## Show the file format
print(gdsRef)

closefn.gds(gdsRef)
```

This output lists all variables stored in the *Population Reference GDS file". 
At the first level, it stores variables *sample.id*, *snp.id*, etc. 
The additional information displayed in the braces indicate the data type, 
size, compressed or not + compression ratio.

The mandatory fields are:

* *sample.id*:  a unique identifier for each sample.
* *snp.id*: a unique identifier for each SNV.
* *sample.annot*: a *data.frame* where each row correspond to a sample and containing those columns:
    * *sex*: a identifier of the sex of the sample
    * *pop.Group*: TODO
    * *superPop*: a identifier of the super-population ancestry of the sample
    * *batch*: a unique number representing the batch of provenance of the sample 
* *snp.chromosome*: an integer or character mapping for each chromosome. Integer: numeric values 1-26, mapped in order from 1-22, 23=X, 24=XY (the pseudoautosomal region), 25=Y, 26=M (the mitochondrial probes), and 0 for probes with unknown positions; it does not allow NA. Character: “X”, “XY”, “Y” and “M” can be used here, and a blank string indicating unknown position.
* *snp.position*: the base position of each SNV on the chromosome, and 0 for unknown position; it does not allow NA.
* *snp.allele*: TODO
* *snp.AF*: TODO
* *snp.[Population_ID]_AF*: TODO
* *snp.EUR_AF*: TODO
* *snp.AFR_AF*: TODO
* *snp.AMR_AF*: TODO
* *snp.SAS_AF*: TODO
* *genotype*: a SNV genotypic matrix (i.e., the number of A alleles) in individual-major mode nsnp×nsample
* *sample.ref*: TODO

<br>
<br>

# Population Reference Annotation GDS file


```{r runRefAnnotGDS, echo=TRUE, eval=TRUE, collapse=TRUE, warning=FALSE, message=FALSE}
#############################################################################
## Load required packages
#############################################################################
library(RAIDS)    
library(SNPRelate)

pathReference <- system.file("extdata/example/", package="RAIDS")

fileReferenceAnnotGDS <- file.path(pathReference, "gdsRef", "exAnnot1kg.gds")

gdsRefAnnot <- openfn.gds(fileReferenceAnnotGDS)

## Show the file format
print(gdsRefAnnot)

closefn.gds(gdsRefAnnot)
```


<br>
<br>


# Pre-processed files, from 1000 Genomes in hg38, are available

Pre-processed files used in the RAIDS associated publication, are 
available at this address:


[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper)

Beware that some of those files are voluminous.

<br>
<br>

# Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

<br>
<br>

# References
