---
title: "Accurate Inference of Genetic Ancestry from Cancer-derived Sequences"
author: Pascal Belleau, Astrid Deschênes and Alexander Krasnitz
output:
    BiocStyle::html_document:
        number_sections: yes
        toc: true
    pkgdown:
        number_sections: yes
        as_is: true
urlcolor: darkred
linkcolor: darkred
bibliography: aicsBiblio.bibtex
vignette: >
    %\VignetteIndexEntry{Accurate Inference of Genetic Ancestry from Cancer-derived Sequences}
    %\VignettePackage{RAIDS}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
BiocStyle::markdown()

suppressPackageStartupMessages({
    library(knitr)
    library(RAIDS)
})

set.seed(121444)
```

<br />
**Package**: `r Rpackage("RAIDS")`<br />
**Authors**: `r packageDescription("RAIDS")[["Author"]]`<br />
**Version**: `r packageDescription("RAIDS")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("RAIDS")[["License"]]`<br />


# Licensing 

The `r Githubpkg("KrasnitzLab/RAIDS")` package and the underlying 
`r Githubpkg("KrasnitzLab/RAIDS")` code are distributed under  
the https://opensource.org/licenses/Apache-2.0 license. You are free to use and 
redistribute this software.  

<br>
<br>

# Citing

If you use the **RAIDS** package for a publication, we would ask you to cite 
the following:

> Pascal Belleau, Astrid Deschênes, Nyasha Chambwe, David A. Tuveson, Alexander Krasnitz; Genetic Ancestry Inference from Cancer-Derived Molecular Data across Genomic and Transcriptomic Platforms. Cancer Res 1 January 2023; 83 (1): 49–58. https://doi.org/10.1158/0008-5472.CAN-22-0682
<br>
<br>

# Introduction

Multiple methods have been implemented to infer ancestry from germline DNA 
sequence [@Price2006; @Pritchard2000; @Alexander2009]. However, genotyping of 
DNA from matched normal specimens is not part of standard clinical practice
and is not performed routinely outside academic clinical centers. 
In sum, matched germline DNA sequence is often missing for cancer-derived 
molecular data. In such cases, having the possibility to infer ancestry 
from tumor-derived data would be beneficial.

The **RAIDS** package implements an inference procedure that has been
specifically developed to accurately infer genetic ancestry from 
cancer-derived sequences. The current version can handle cancer-derived 
sequences of:

* tumor exomes
* targeted gene panels 
* RNA

The **RAIDS** package implements a data synthesis method that, for any given
cancer-derived sequence profile, enables 
on the one hand, profile-specific inference
parameter optimization and on the other hand, a profile-specific inference
accuracy estimate.

<br>
<br>

# Installation

To install the latest version from the 
[RAIDS Github Website](https://github.com/KrasnitzLab/RAIDS "RAIDS Github Site"), 
the `r CRANpkg("devtools")` package is required.

```{r installDemo01, eval=FALSE, warning=FALSE, message=FALSE}
## Load required package
library(devtools)

## Install the latest version of RAIDS
devtools::install_github('KrasnitzLab/RAIDS')
```

<br>
<br>

<a name="mains"></a>

# Main Steps


This is an overview of genetic ancestry inference from cancer-derived 
molecular data:

```{r graphMainSteps, echo = FALSE, fig.align="center", fig.cap="An overview of the genetic ancestry inference process.", out.width = '120%', results='asis', warning=FALSE, message=FALSE}
knitr::include_graphics("MainSteps_v01.png")
```

The main steps are:

1. Format reference data from the 1000 Genomes (1KG) (optional)
2. Format cancer-derived data set starting from BAM files
3. Optimize ancestry inference parameters
4. Infer ancestry for the subjects of the external study

These main steps are described in detail in the following.

You can also run the steps from 5.2.4 to the end, with the default parameters,
in one command with the [wrapper](#wrapper) function.

<br>
<br>

## Step 1 - Format reference data from the 1000 Genomes (optional)


```{r graphStep1, echo=FALSE, fig.align="center", fig.cap="Step 1 - Formatting the information from 1000 Genomes (optional)", out.width = '120%', results='asis', warning=FALSE, message=FALSE}
knitr::include_graphics("MainSteps_Step1_v01.png")
```


******

At this step three important reference files are created:

- The 1KG GDS File
- The 1KG SNV Annotation GDS file
- The 1KG SNV Retained VCF file

<span style="color:darkred;font-weight:bold">Note that these pre-processed 
files are available at:</span>

<span style="color:red">
[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper)
</span>

<span style="color:darkred;font-weight:bold">The size of the 1KG GDS file 
is 15GB.</span>

<span style="color:darkred;font-weight:bold">The 1KG GDS file is mapped on 
hg38 [@Lowy-Gallego2019a].</span>

<span style="color:darkred;font-weight:bold">This section can be skipped if 
you choose to use the pre-processed files.</span>

The execution of this step is explained in the [Formatting the information from 1000 Genomes (optional)](Create_1KG_GDS_file.html) 
vignette.

<br>

## Step 2 - Prepare cancer-derived data for ancestry inference

Molecular profiles in a cancer-derived data set must be formatted 
following a series of sub-steps.

```{r graphStep2, echo = FALSE, fig.align="center", fig.cap="Step 2 - Formatting the information from an external study", out.width = '120%', results='asis', warning=FALSE, message=FALSE}
knitr::include_graphics("MainSteps_Step2_v01.png")
```

<br>

These are:

1. Create a directory containing the 3 reference files from 1KG
2. Make a SNP pileup file for the profile
3. Create an RDS file containing information about the samples
4. Create a Sample GDS file (1 GDS file per sample)
5. Generate a pruned subset of the single nucleotide variants (SNVs) 
6. Estimate the allelic frequency for the pruned SNVs

Note that a mapped BAM file is needed for each sample (step 2). 
The reference 
genome used for the mapping must be the same as the one used to generate the
1KG GDS file. The 1KG GDS file available for download (<span style="color:red">
[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper)
</span>) is based on the hg38 genome build.

<br>

### Sub-Step 1. Create a directory containing the 3 required reference files

The 3 required reference files may be downloaded:

```bash
####################################
## The 1KG GDS file
####################################
wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds

####################################
## The 1KG SNV Annotation GDS file
####################################
wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matAnnot1000g.gds

####################################
## The 1KG SNV Retained VCF file
####################################
wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/snvSel0.01.vcf.gz
```

The 3 files should be stored in the same directory. In the example below, 
this directory is referred to as **PATH_1KG**.

For more information on creating your own reference files, 
see the vignette 
[Formatting the information from 1000 Genomes (optional)](Create_1KG_GDS_file.html).

<br>

### Sub-Step 2. Generate a SNP pileup file (1 file per profile)

This step requires installation of the external code 
[snp-pileup](https://github.com/mskcc/facets/tree/master/inst/extcode) which 
is associated to the facets package [@Shen2016]. Given a VCF file 
containing SNP locations, 
[snp-pileup](https://github.com/mskcc/facets/tree/master/inst/extcode) 
application outputs, for each location, 
the read counts for the reference and alternative nucleotides from the BAM input.

At this point, 
[snp-pileup](https://github.com/mskcc/facets/tree/master/inst/extcode) is used 
to create,for each cancer-derived profile,
a **Profile SNP pileup file** containing the read counts for the reference and 
alternative nucleotides at each SNP position in the 
reference 1KG GDS file.
The **1KG SNV Retained VCF file**, which contains the list of 
retained 1KG SNVs, is required. 


```{r graphStep2_SubStep2, echo=FALSE, fig.align="center", fig.cap="Generate a Sample SNP pileup file (1 file per sample)", out.width='100%', results='asis', warning=FALSE, message=FALSE}
knitr::include_graphics("MainSteps_Step2_SubStep2_SNP-Pileup_v01.mod.png")
```

Note that the name assigned to the **Profile SNP pileup file** will correspond 
to the profile identifier (Name.ID) in the following analysis. For example, a 
SNP pileup file called "Sample.01.txt.gz" would be associated to the 
"Sample.01" profile. 

This is the command line to run 
[snp-pileup](https://github.com/mskcc/facets/tree/master/inst/extcode):

```bash
#####################################################################
## Description of the parameters
## -g : Compresses the output file with BGZF
## -d5000 : Sets the maximum depth to 5000
## -q15 : Sets the minimum threshold for mapping quality to 15
## -Q20 : Sets the minimum threshold for base quality to 20
## -r0 : Sets the minimum read counts for a position to be output to 0
## PATH_1KG/snvSel0.01.vcf.gz : The SNP Retained VCF file containing the 
##     positions of all retained 1KG SNPs
## PATH_OUT/Name.ID.txt : The name of the output Sample VCF file that will be
##     compressed by the application
## FILEBAM.bam : The aligned reads from the sample used as input
#####################################################################
snp-pileup -g -d5000 -q15 -Q20  -r0 PATH_1KG/snvSel0.01.vcf.gz PATH_OUT/Name.ID.txt FILEBAM.bam
```

<br>

### Sub-Step 3. Create a profile PED RDS file containing the information about the profiles

An RDS file describing all the profiles analyzed 
is required.

The RDS file must contain a *data.frame* with these 5 columns: 

- **Name.ID**: The unique sample identifier. The profile VCF file  
should be called "Name.ID.txt.gz". 
- **Case.ID**: The patient identifier associated to the sample.
- **Sample.Type**: The information about the profile tissue source (primary tumor,
metastatic tumor, normal, etc..). 
- **Diagnosis**: The donor's diagnosis.
- **Source**: The source of the profile sequence data (example: dbGAP_XYZ).

Important: The row names of the *data.frame* must be the profiles' **Name.ID**.

This file is referred to as the **Sample PED RDS file** (PED for pedigree). 
Alternatively, the PED information can be saved in another type of 
file (CVS, etc..) as long as the *data.frame* information can be regenerated 
in R (with _read.csv()_ or else). 


```{r pedCreation, echo=TRUE, warning=FALSE, message=FALSE}
##############################################################
## Location of the Sample PED RDS file to be created
##############################################################
data.dir <- system.file("extdata", package="RAIDS")
demo_PED_File <- file.path(data.dir, "Demo_PED.RDS")

##############################################################
## Create a data frame with the mandatory columns
## All columns are in character string format (no factor)
##############################################################
pedDF <- data.frame(Name.ID = c("Sample_01", "Sample_02", "Sample_03"),
                    Case.ID = c("Patient_h11", "Patient_h12", "Patient_h18"), 
                    Sample.Type = rep("Primary Tumor", 3),
                    Diagnosis = rep("Cancer", 3),
                    Source = rep("Databank B", 3),
                    stringsAsFactors = FALSE)

##############################################################
## The row names must correspond to the name of the samples
##############################################################
rownames(pedDF) <- pedDF$Name.ID

##############################################################
## Save the data frame into a RDS file
##############################################################
saveRDS(object=pedDF, file=demo_PED_File)

## Remove RDS file (created for demo purpose)
rm(demo_PED_File)
```

<br>


### Sub-Step 4 Create a profile GDS file (1 GDS file per profile) 

From here you can run directly the [wrapper](#wrapper) function 
or you can run each step separetly as explained here.

This step requires 3 files as input:

- The **1KG GDS file**
- The **Sample SNP pileup file** (one per sample present in the study)
- The **Sample PED RDS file** (one file with information for all samples in the study)

A *data.frame* containing the general information about the study is 
also required. The *data.frame* must contain those 3 columns:

- **study.id**: The study identifier (example: TCGA-BRCA).
- **study.desc**: The description of the study.
- **study.platform**: The type of sequencing (example: RNA-seq).

Using all those inputs, the *createStudy2GDS1KG()* function will 
generate a **Sample GDS file**. One **Sample GDS file** is created for each 
sample passed to the *listSamples* argument.

```{r appendStudy2GDS1KG, echo=TRUE, eval=TRUE, collapse=TRUE, warning=FALSE, message=FALSE}
#################################################################
## Load required package
#################################################################
library(RAIDS)
    
#################################################################
## The 1KG GDS file and the 1KG SNV Annotation GDS file 
## need to be located in the same directory
## Note that the 1KG GDS file used for this example is a  
## simplified version and CANNOT be used for any real analysis
#################################################################
data.dir <- system.file("extdata", package="RAIDS")
PATH_1KG <- file.path(data.dir, "example", "gdsRef")
    
fileName.GDS.1KG <- "ex1kg.gds"
file.GDS <- file.path(PATH_1KG, fileName.GDS.1KG)

#################################################################
## The Sample SNP pileup files (one per sample) need  
## to be located in the same directory.
#################################################################
PATHGENO <- file.path(data.dir, "example", "snpPileup")

#################################################################
## The path where the Sample GDS files (one per sample)
## will be created need to be specified.
#################################################################
PATHSAMPLEGDS <- file.path(data.dir, "example", "out")

#################################################################
## The path and file name for the PED RDS file
## will the information about the analyzed samples
#################################################################
filePED <- file.path(data.dir, "example", "pedEx.rds")
ped <- readRDS(filePED)
head(ped)

#################################################################
## A data frame containing general information about the study
## is also required. The data frame must have
## those 3 columns: "study.id", "study.desc", "study.platform"
#################################################################
studyDF <- data.frame(study.id = "MYDATA",
                        study.desc = "Description",
                        study.platform = "PLATFORM",
                        stringsAsFactors = FALSE)

#################################################################
## The list of profiles to analyzed is passed to the function.
## The profiles must be present in the Profile PED RDS file see
## sub-step 4 and must have an associated Profile SNP pileup file.
## Not all profiles present in the Profile PED file need to 
## be selected.
#################################################################
listSamples <- ped[, "Name.ID"]

#################################################################
## This function creates one Sample GDS file for each
## sample present in the 'listProfiles' parameter.
#################################################################
createStudy2GDS1KG(pathGeno = PATHGENO,
                        pedStudy = ped,
                        fileNameGDS = file.GDS,
                        listProfiles = listSamples,
                        studyDF = studyDF,
                        pathProfileGDS = PATHSAMPLEGDS)

#################################################################
## The Sample GDS file has been created in the 
## directory PATHSAMPLEGDS using the name of the sample (ex1) 
#################################################################
list.files(path = PATHSAMPLEGDS)
```

<br>

### Sub-Step 5. Generate a pruned subset of the single nucleotide variants (SNVs) 

The initial list of 1KG SNVs is pruned, using linkage disequilibrium analysis, 
and a sample-specific subset of SNVs is retained for each sample. This 
information is added to the **Sample GDS file**.

The __pruningSample()__ function requires the **1KG GDS file** as input. It 
also requires the path to the **Sample GDS file(s)**.

Note that this step can require large disk space.

```{r pruningSample, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
##############################################################
## Load required package
##############################################################
library(RAIDS)
    
##############################################################
## The 1KG GDS file is required (demo version)
## Note that the 1KG GDS file used for this example is a  
## simplified version and CANNOT be used for any real analyses
##############################################################
PATH_1KG <- file.path(data.dir, "example", "gdsRef")
    
fileName.GDS.1KG <- "ex1kg.gds"
file.GDS <- file.path(PATH_1KG, fileName.GDS.1KG)

## Open the 1KG GDS file (demo version)
gds_1KG <- snpgdsOpen(file.GDS)

##############################################################
## The pruning function is called with one sample as input at the time
##############################################################
for(i in seq_len(length(listSamples))) {
    ## Compute the SNV pruned subset
    ## study.id: Study identifier as defined in the preceding sub-step 4. 
    ##           The study identifier must be the same that the one present in
    ##           Profile GDS file.
    ## PATHSAMPLEGDS: All Profile GDS files must be in the same directory 
    ##           sub-step 4. 
    pruningSample(gds=gds_1KG,
                    currentProfile = listSamples[i],
                    study.id = studyDF$study.id, 
                    PATHSAMPLEGDS = PATHSAMPLEGDS)
    
    ## Profile GDS file for the current sample
    ## The file name corresponds to the path + profile identifier + ".gds"
    file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds"))
    
    ## Add the genotype information for the list of pruned SNVs 
    ## into the Profile GDS file
    ## The genotype information is extracted from the 1KG GDS file
    add1KG2SampleGDS(gds = gds_1KG, gdsSampleFile = file.GDSSample, 
                        sampleCurrent = listSamples[i], 
                        study.id = studyDF$study.id)

    ## Add annotation from the 1KG GDS file to the Sample GDS file
    ## This is required.
    addStudy1Kg(gds_1KG, file.GDSSample)
}
    
## Close the 1KG GDS file (it is important to always close the GDS files)
closefn.gds(gds_1KG)
```

<br>

### Sub-Step 6. Estimate the allelic fraction for the pruned SNVs

The __estimateAllelicFraction()__ estimates the allele fraction for all 
SNVs present in the pruned SNV dataset. Note that the function requires 
different inputs for DNA and RNA samples.

For the DNA samples, these 2 files are required:

1. The **Sample GDS file**
2. The **1KG GDS file**

For the RNA samples, these 3 files are required:

1. The **Sample GDS file**
2. The **1KG GDS file**
3. The **1KG SNV Annotation GDS file**

In both cases, the other required input is:

1. The information about the length of the chromosomes
2. The sample identifier (it corresponds to the Sample GDS file name)
3. The study identifier (it should correspond to the one used previously)

The information about the length of the chromosomes must be assigned into a 
*vector* object. This is an example on how to retrieve the information. 
There are alternative ways to retrieve this information, e.g., 

```{r extractChrLength, echo=TRUE, message=FALSE, warning=FALSE, collapse=TRUE}
###################################################################
## Load required library
###################################################################
library(BSgenome.Hsapiens.UCSC.hg38)

###################################################################
## The length of each chromosome is required
## Chromosomes X, Y and M need relabeling (see below)
## There are alternative ways to retrieve this information, e.g., 
## 
###################################################################
chrInfo <- integer(25L)
for(i in seq_len(22L)) {
    chrInfo[i] <- length(Hsapiens[[paste0("chr", i)]])
}
chrInfo[23] <- length(Hsapiens[["chrX"]])
chrInfo[24] <- length(Hsapiens[["chrY"]])
chrInfo[25] <- length(Hsapiens[["chrM"]])

chrInfo
```

<br>
<br>

The __estimateAllelicFraction()__ function processes one sample at the time, 
as shown in this example.  


```{r estimateAllelicFraction, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
###################################################################
## Load the required package
###################################################################
library(RAIDS)

###################################################################
## The 1KG GDS file is required
## The 1KG SNV Annotation GDS file is only required for RNA samples
###################################################################
PATH_1KG <- file.path(data.dir, "example", "gdsRef")
    
fileName.GDS.1KG <- "ex1kg.gds"
fileName.GDS.Annot <- "exAnnot1kg.gds"

file.GDS <- file.path(PATH_1KG, fileName.GDS.1KG)
file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

## Open the 1KG GDS file
gds <- snpgdsOpen(file.GDS)
    
###################################################################
## The information about the length of the chromosomes
###################################################################
head(chrInfo)

##############################################################
## The function must be called for each sample
## This example only uses one sample
##############################################################

## The first sample is used in the demo
sampleName <- listSamples[1]

###################################################################
## The Sample GDS file is required
###################################################################

## Name of the Sample GDS file
## The name must correspond to the sample identifier
file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(sampleName, ".gds"))

## Open the Sample GDS file in writing mode
gdsSample <- openfn.gds(file.GDSSample, readonly = FALSE)

###################################################################
## The estimation of the allelic fraction
###################################################################

## Estimate the allele fraction of the pruned SNVs
## The current example is for a DNA sample
## In the case of RNA sample, the function needs different inputs
## such as the 1KG SNV Annotation GDS file and
## the block.id should be as listed in the 1KG SNV Annotation GDS file 
## for the gene annotation of the SNVs
estimateAllelicFraction(gds = gds, gdsSample = gdsSample,
                                        sampleCurrent = sampleName,
                                        study.id = studyDF$study.id,
                                        chrInfo = chrInfo)

## Close both GDS files
closefn.gds(gdsSample)
closefn.gds(gds)
```

This step must be executed for each sample present in the study.

<br>


## Step 3 - Optimize the ancestry inference parameters

At this step, optimization of the parameters is required to maximize the 
the ancestry inference accuracy (next step).

```{r graphStep3, echo = FALSE, fig.align="center", fig.cap="Step 3 - Find the optimized parameters for the ancestry inference", out.width = '120%', results='asis', warning=FALSE, message=FALSE}
knitr::include_graphics("MainSteps_Step3_v01.png")
```

Two inference parameters to be optimized:

- _K_: the number of neighbors used to call the ancestry
- _D_: the number of PCA components retained

The accuracy is evaluated using a synthetic dataset created from merging one 
cancer profile with multiple 1KG samples of known ancestry. Using the synthetic 
profiles, a range of _K_ and _D_ values are tested. Through that process, the 
_K_ and _D_ values are tuned to maximize accuracy.

This step is consists of two sub-steps:

1. Generate the synthetic dataset
2. Compute the PCA-KNN ancestry call for each synthetic profile

<br>

### Sub-Step 1. Generate the synthetic dataset

A synthetic profile is generated through the merging of one cancer profile 
with one 1KG sample of known ancestry. Multiple 1KG samples of different 
ancestry are required to create a synthetic dataset that will be able to 
show the specific accuracy for each super-population. All the synthetic 
profiles are saved in the **Sample GDS file** corresponding to the 
cancer profile used to generate the dataset.

In summary, a fixed number of samples fo each super-population is extracted 
from the 1KG study. The information is saved in the **Sample GDS file** 
associated to the selected cancer profile. A synthetic profile is created for 
each combination of the 1KG samples and cancer profile. The synthetic profile 
is then saved in the **Sample GDS file**.

The three functions _select1KGPop()_, _prepSynthetic()_ and _syntheticGeno()_ 
are used for data synthesis.

These 3 files are required:

1. The **Sample GDS file**
2. The **1KG GDS file**
3. The **1KG SNv Annotation GDS file**


```{r generateSynthetic, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
####################################################################
## Load required package
####################################################################
library(RAIDS)    


####################################################################
## Randomly extract a fixed number of samples for each
## subcontinental population present in the 1KG GDS file.
## When not enough samples are available, all samples are selected.  
####################################################################   
gds <- snpgdsOpen(file.GDS)

####################################################################
## Fix RNG seed to ensure reproducible results
####################################################################
set.seed(3043)   

####################################################################
## Select the profiles from 1KG for the synthetic data.
## Here we select 2 profiles from 1KG for each subcontinental-level
## Normally, we use 30 profiles from 1KG for each 
## subcontinental-level but it is too big for the example.
## The 1KG files in this example only have 6 profiles for each 
## subcontinental-level (for demo purpose only)
####################################################################

dataRef <- select1KGPop(gds, nbSamples = 2L)

## Extract the list of selected 1KG sample identifiers
listSampleRef <- dataRef$sample.id

####################################################################
## A data.frame with the description of the study for the synthetic
## data is required.
## The column names must be as shown
#################################################################### 
studyDF.syn <- data.frame(study.id = "MYDATA.Synthetic",
                          study.desc = "MYDATA synthetic data",
                          study.platform = "PLATFORM",
                          stringsAsFactors = FALSE)

## The Sample GDS file is needed
file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(sampleName, ".gds"))

####################################################################
## The prepSynthetic() function prepares the annotation for
## the synthetic data
## The information is saved into the Sample GDS file
####################################################################
prepSynthetic(gdsSampleFile = file.GDSSample,
                listSampleRef = listSampleRef,
                data.id.profile = sampleName,
                studyDF = studyDF.syn,
                prefId = "1")

####################################################################
## Both the 1KG GDS file and the 1KG SNV Annotation GDS file
## are required
####################################################################
PATH_1KG <- file.path(data.dir, "example", "gdsRef")

fileName.GDS.Annot <- "exAnnot1kg.gds"

file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

## Open the 1KG GDS file and 1KG SNV Annotation file    
gds.Annot <- openfn.gds(file.Annot)

####################################################################
## The syntheticGeno() function generates the synthetic profiles.
## The synthetic profiles are saved in the Sample GDS file
####################################################################
resG <- syntheticGeno(gds = gds,
                        gdsRefAnnot = gds.Annot,
                        gdsSampleFile = file.GDSSample,
                        data.id.profile = sampleName,
                        listSampleRef = listSampleRef,
                        prefId = "1")

## Close both GDS files
closefn.gds(gds)
closefn.gds(gds.Annot)
```

<br>


### Sub-Step 2. Perform the PCA-KNN ancestry call for each synthetic profile



```{r PCA.KNN.Synthetic, collapse=TRUE, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
####################################################################
## Load required package
####################################################################
library(RAIDS)    
    
####################################################################
## The 1KG GDS file is required
####################################################################

## Open the 1KG GDS file     
gds <- openfn.gds(file.GDS)

####################################################################
## The path to the directory where the PCA results will be saved 
## in RDS files.
## The directory must exist.
####################################################################
PATH_OUT <- file.path(PATHSAMPLEGDS)

if(! file.exists(PATH_OUT)) {
    dir.create(PATH_OUT)
}

####################################################################
## Get the super-population information for the reference samples
## This is the ground truth for the 1KG samples
####################################################################
spRef <- getRef1KGPop(gds, "superPop")
    
####################################################################
## Fix the RNG seed as in step 6 to ensure the same results
####################################################################    
set.seed(3043)       
    
## Select the 1KG samples used to generate the synthetic dataset
## Already done in step 6, no need to repeat if the results have been saved
dataRef <- select1KGPop(gds, nbSamples = 2L)

####################################################################
## The function splitSelectByPop() generates a matrix with the
## reference samples split by sub-continental population
####################################################################  
sampleRM <- splitSelectByPop(dataRef)
    
## Loop for all cancer samples with associated synthetic data
for(i in seq_len(length(listSamples))) {

    ## The Sample GDS file associated to the cancer sample     
    file.GDSSample <- file.path(PATHSAMPLEGDS, 
                                    paste0(listSamples[i], ".gds"))

    ## A sub-directory is created for the cancer sample
    ## Beware that the number of files created will correspond to the 
    ## number of rows in the sampleRM matrix
    PATH_OUT_SAMPLE <- file.path(PATH_OUT, listSamples[i])
    if(! file.exists(PATH_OUT_SAMPLE)) {
            dir.create(PATH_OUT_SAMPLE)
    }
    
    ## Open the Sample GDS file      
    gdsSample <- snpgdsOpen(file.GDSSample)

    ## This variable will contain the results from the PCA analyses
    KNN.list <- list()

    ## For each row of the sampleRM matrix      
    for(j in seq_len(nrow(sampleRM))) {
        ## Run a PCA analysis using 1 synthetic profile from each 
        ##  sub-continental ancestry
        ## The synthetic profiles are projected on the 1KG PCA space 
        ##  (the reference samples used to generate the synthetic profiles are 
        ##  removed from this PCA)
        ## The K-nearest neighbor analysis is done using 
        ##  a range of K and D values
        KNN.synt <- computePoolSyntheticAncestryGr(gdsSample = gdsSample,
                                            sampleRM =  sampleRM[j,],
                                            study.id.syn = studyDF.syn$study.id,
                                            np = 4L,
                                            spRef = spRef,
                                            eigen.cnt = 15L)
        
        ## Results are saved
        saveRDS(KNN.synt$matKNN, file.path(PATH_OUT_SAMPLE, 
                        paste0("KNN.synt.", listSamples[i], ".", j, ".rds")))
    }
    
    ## Close Sample GDS file (important)
    closefn.gds(gdsSample)
}
    
## Close 1KG GDS file (important)
closefn.gds(gds)
```


<br>

## Step 4 - Run the ancestry inference in the input data set

The ancestry inference is done with the optimized K and D parameters. More 
specifically, a PCA is generated using the 1KG reference samples and the 
cancer sample. The _D_ parameter specifies the number of dimension for the 
PCA. Then, the ancestry of the cancer sample is inferred using
a k-nearest neighbors classification method. The _K_ parameter specifies the 
number of neighbors used for the classification.


```{r graphStep4, echo = FALSE, fig.align="center", fig.cap="Step 4 - Run the ancestry inference on the external study", out.width = '120%', results='asis', warning=FALSE, message=FALSE}
knitr::include_graphics("MainSteps_Step4_v01.png")
```

<br>

The PCA of the sample and KNN sample and Call the ancestry with the optimal 
_K_ and _D_ parameters.

Note: The formal selection of _K_ and _D_ parameters is done at this step but 
all the synthetic data are prepared in the step 3. 


```{r PCA.KNN.Sample, warning=FALSE, message=FALSE, collapse=TRUE, echo=TRUE, eval=TRUE}
####################################################################
## Load required package
####################################################################
library(RAIDS)    
    
####################################################################
## The 1KG GDS file is required
####################################################################

## Open the 1KG GDS file 
gds <- openfn.gds(file.GDS)
    
## A directory for the results, where a sub-directory 
## will be set up for each input profile 
PATH_OUT <- file.path(PATHSAMPLEGDS)

if(! file.exists(PATH_OUT)) {
    dir.create(PATH_OUT)
}

####################################################################
## The getRef1KGPop() function extract the known super-population 
## of the reference samples.
## We expect the ancestry call from a synthetic profile to 
## correspond to the known ancestry of the reference sample used to 
## synthesize it.
####################################################################
spRef <- getRef1KGPop(gds, "superPop")
    
## Loop on each sample 
## Can also be run in parallel or on different clusters...
for(i in seq_len(length(listSamples))){
    
    ## Extract the GDS file name and path for the current sample      
    file.GDSSample <- file.path(PATHSAMPLEGDS, paste0(listSamples[i], ".gds"))
    
    ## Directory where the KNN results have been saved      
    PATHKNN <- file.path(PATH_OUT, listSamples[i])
    listFilesName <- dir(file.path(PATHKNN), ".rds")

    ## List of the KNN result files from PCA on synthetic data
    listFiles <- file.path(file.path(PATHKNN) , listFilesName)

    ## Open the Sample GDS file
    gdsSample <- snpgdsOpen(file.GDSSample)

    resCall <- computeAncestryFromSyntheticFile(gds = gds, 
                                    gdsSample = gdsSample,
                                    listFiles = listFiles,
                                    sample.ana.id = listSamples[i],
                                    spRef = spRef,
                                    study.id.syn = studyDF.syn$study.id,
                                    np = 1L)

    saveRDS(resCall, file.path(PATH_OUT,
                                paste0(listSamples[i], ".infoCall", ".rds")))

    write.csv(resCall$Ancestry,
            file.path(PATH_OUT, paste0(listSamples[i], ".Ancestry",".csv")),
            quote = FALSE, row.names = FALSE)
    
    ## Close the Sample GDS file (important)
    closefn.gds(gdsSample)
}

## Close the 1KG GDS file (important)
closefn.gds(gds)

####################################################################
## Show the ancestry inference (SuperPop) and 
## optimal number of PCA components D
## optimal number of neighbours K
####################################################################
resAncestry <- read.csv(file.path(PATH_OUT, 
                    paste0(ped$Name.ID[1], ".Ancestry.csv")))
resAncestry

## Clean-up demo files
unlink(file.GDSSample, force=TRUE)
unlink(PATH_OUT, recursive = TRUE, force=TRUE)
```


The *computeAncestryFromSyntheticFile()* function generates 3 types of files 
in the *OUTPUT* directory.

* The ancestry inference CSV file (".Ancestry.csv" file) 
* The inference information RDS file (".infoCall.rds" file)
* The parameter information RDS files from the synthetic inference ("KNN.synt.*.rds" files in a sub-directory)

In addition, a sub-directory (named using the *profile ID*) is 
also created. 


<a name="wrapper"></a>

# Wrapper function to run ancestry inference in one command

The Steps 5.1 to 5.2.3 ( [Main Step](#mains) ) are required before 
this section.

The wrapper function requires 4 files as input:

- The **1KG GDS file**
- The **1KG SNV Annotation GDS file**
- The **Sample SNP pileup file** (one per sample present in the study)
- The **Sample PED RDS file** (one file with information for all samples in the study)

A *data.frame* containing the general information about the study is 
also required. The *data.frame* must contain those 3 columns:

- **study.id**: The study identifier (example: TCGA-BRCA).
- **study.desc**: The description of the study.
- **study.platform**: The type of sequencing (example: RNA-seq).


```{r runExomeAncestry, echo=TRUE, eval=TRUE, collapse=TRUE, warning=FALSE, message=FALSE}

## Path to the demo 1KG GDS file is located in this package
data.dir <- system.file("extdata", package="RAIDS")

#################################################################
## The path and file name for the PED RDS file
## will the information about the analyzed samples
#################################################################
filePED <- file.path(data.dir, "example", "pedEx.rds")
ped <- readRDS(filePED)
head(ped)

#################################################################
## The 1KG GDS file and the 1KG SNV Annotation GDS file
## need to be located in the same directory
## Note that the 1KG GDS file used for this example is a
## simplified version and CANNOT be used for any real analysis
#################################################################
PATH_1KG <- file.path(data.dir, "example", "gdsRef")

fileName.GDS.1KG <- "ex1kg.gds"
file.GDS <- file.path(PATH_1KG, fileName.GDS.1KG)

fileName.GDS.Annot <- "exAnnot1kg.gds"
file.Annot <- file.path(PATH_1KG, fileName.GDS.Annot)

#################################################################
## The Sample SNP pileup files (one per sample) need
## to be located in the same directory.
#################################################################
PATHGENO <- file.path(data.dir, "example", "snpPileup")

#################################################################
## The path where the Sample GDS files (one per sample)
## will be created need to be specified.
#################################################################
PATHPROFILEGDS <- file.path(data.dir, "example", "out.tmp")

PATHOUT <- file.path(data.dir, "example", "res.out")

#################################################################
## A data frame containing general information about the study
## is also required. The data frame must have
## those 3 columns: "study.id", "study.desc", "study.platform"
#################################################################
studyDF <- data.frame(study.id = "MYDATA",
                   study.desc = "Description",
                   study.platform = "PLATFORM",
                   stringsAsFactors = FALSE)

gds_1KG <- snpgdsOpen(file.GDS)

####################################################################
## Fix RNG seed to ensure reproducible results
####################################################################
set.seed(3043)

####################################################################
## Select the profiles from 1KG for the synthetic data.
## Here we select 2 profiles from 1KG for each subcontinental-level
## Normally, we use 30 profiles from 1KG for each 
## subcontinental-level but it is too big for the example.
## The 1KG files in this example only have 6 profiles for each 
## subcontinental-level (for demo purpose only)
####################################################################
dataRef <- select1KGPop(gds_1KG, nbSamples = 2L)

closefn.gds(gds_1KG)

## Chromosome length information
## chr23 is chrX, chr24 is chrY and chrM is 25
chrInfo <- c(248956422L, 242193529L, 198295559L, 190214555L,
 181538259L, 170805979L, 159345973L, 145138636L, 138394717L, 133797422L,
 135086622L, 133275309L, 114364328L, 107043718L, 101991189L, 90338345L,
 83257441L,  80373285L,  58617616L,  64444167L,  46709983L, 50818468L,
 156040895L, 57227415L,  16569L)

## A formal way to get the chormosome length information
## library(BSgenome.Hsapiens.UCSC.hg38)
## chrInfo <- integer(25L)
## for(i in seq_len(22L)){ chrInfo[i] <-
##                          length(Hsapiens[[paste0("chr", i)]])}
## chrInfo[23] <- length(Hsapiens[["chrX"]])
## chrInfo[24] <- length(Hsapiens[["chrY"]])
## chrInfo[25] <- length(Hsapiens[["chrM"]])

runExomeAncestry(pedStudy = ped, studyDF = studyDF,
                 PATHPROFILEGDS = PATHPROFILEGDS,
                 PATHGENO = PATHGENO,
                 PATH_OUT = PATHOUT,
                 File.gds_1KG = file.GDS,
                 File.gds_1KG.Annot = file.Annot,
                 chrInfo = chrInfo,
                 dataRefSyn = dataRef)
list.files(PATHOUT)
list.files(file.path(PATHOUT, ped$Name.ID[1]))

####################################################################
## Show the ancestry inference (SuperPop) and 
## optimal number of PCA component D
## optimal number neighbour K
####################################################################

resAncestry <- read.csv(file.path(PATHOUT, 
                                    paste0(ped$Name.ID[1], ".Ancestry.csv")))
resAncestry

## Remove temporary files created for this demo
unlink(PATHPROFILEGDS, recursive = TRUE, force=TRUE)
unlink(PATHOUT, recursive = TRUE, force=TRUE)
```

<br>
<br>

The *runExomeAncestry()* function generates 3 types of files 
in the *OUTPUT* directory.

* The ancestry inference CSV file (".Ancestry.csv" file) 
* The inference information RDS file (".infoCall.rds" file)
* The parameter information RDS files from the synthetic inference ("KNN.synt.*.rds" files in a sub-directory)

In addition, a sub-directory (named using the *profile ID*) is 
also created. 


<br>
<br>


# Pre-processed files are available

Pre-processed files, such as the 1KG GDS file, are available at this address:

[https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper](https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper)

Beware that some of these files are large.

<br>
<br>

# Session info

Here is the output of `sessionInfo()` in the enviroment in which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

<br>
<br>


# References

