<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate the allelic fraction of the pruned SNVs for a specific
DNA-seq profile — computeAllelicFractionDNA • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate the allelic fraction of the pruned SNVs for a specific
DNA-seq profile — computeAllelicFractionDNA"><meta property="og:description" content="The function creates a data.frame containing the
allelic fraction for the pruned SNV dataset specific to a DNA-seq profile"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Ancestry_Inference_Step_by_Step.html">Genetic Ancestry Inference Step by Step (optional)</a>
    </li>
    <li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from 1000 Genomes (optional)</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate the allelic fraction of the pruned SNVs for a specific
DNA-seq profile</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/HEAD/R/allelicFraction_internal.R" class="external-link"><code>R/allelicFraction_internal.R</code></a></small>
    <div class="hidden name"><code>computeAllelicFractionDNA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function creates a <code>data.frame</code> containing the
allelic fraction for the pruned SNV dataset specific to a DNA-seq profile</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">computeAllelicFractionDNA</span><span class="op">(</span></span>
<span>  <span class="va">gdsReference</span>,</span>
<span>  <span class="va">gdsSample</span>,</span>
<span>  <span class="va">currentProfile</span>,</span>
<span>  <span class="va">studyID</span>,</span>
<span>  <span class="va">chrInfo</span>,</span>
<span>  minCov <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  minProb <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  eProb <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  cutOffLOH <span class="op">=</span> <span class="op">-</span><span class="fl">5</span>,</span>
<span>  cutOffHomoScore <span class="op">=</span> <span class="op">-</span><span class="fl">3</span>,</span>
<span>  wAR <span class="op">=</span> <span class="fl">9L</span>,</span>
<span>  <span class="va">verbose</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gdsReference</dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened 1KG GDS file.</p></dd>


<dt>gdsSample</dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened Profile GDS file.</p></dd>


<dt>currentProfile</dt>
<dd><p>a <code>character</code> string corresponding to
the sample identifier as used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt>studyID</dt>
<dd><p>a <code>character</code> string corresponding to the name of
the study as
used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt>chrInfo</dt>
<dd><p>a <code>vector</code> of <code>integer</code> values representing
the length of the chromosomes.</p></dd>


<dt>minCov</dt>
<dd><p>a single positive <code>integer</code> representing the minimum
required coverage. Default: <code>10L</code>.</p></dd>


<dt>minProb</dt>
<dd><p>a single <code>numeric</code> between <code>0</code> and <code>1</code>
representing the probability that the calculated genotype call is correct.
Default: <code>0.999</code>.</p></dd>


<dt>eProb</dt>
<dd><p>a single <code>numeric</code> between 0 and 1 representing the
probability of sequencing error. Default: <code>0.001</code>.</p></dd>


<dt>cutOffLOH</dt>
<dd><p>a single <code>numeric</code> representing the cutoff, in log,
for the homozygote score to assign a region as LOH.
Default: <code>-5</code>.</p></dd>


<dt>cutOffHomoScore</dt>
<dd><p>a single <code>numeric</code> representing the cutoff, in
log, that the SNVs in a block are called homozygote by error.
Default: <code>-3</code>.</p></dd>


<dt>wAR</dt>
<dd><p>a single positive <code>integer</code> representing the size-1 of
the window used to compute an empty box. Default: <code>9L</code>.</p></dd>


<dt>verbose</dt>
<dd><p>a <code>logicial</code> indicating if the function should print
message when running.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>data.frame</code> containing the allelic information for the
pruned SNV dataset with coverage &gt; <code>minCov</code>. The <code>data.frame</code></p>


<p>contains those columns:</p><ul><li><p>cnt.tot a <code>integer</code> representing the total allele count</p></li>
<li><p>cnt.ref a <code>integer</code> representing the reference allele count</p></li>
<li><p>cnt.alt a <code>integer</code> representing the alternative allele count</p></li>
<li><p>snp.pos a <code>integer</code> representing the position on the chromosome</p></li>
<li><p>snp.chr a <code>integer</code> representing the chromosome</p></li>
<li><p>normal.geno a <code>integer</code> representing the genotype
(0=wild-type reference; 1=heterozygote; 2=homozygote alternative; 3=unkown)</p></li>
<li><p>pruned a <code>logical</code> indicating if the SNV is retained after
pruning</p></li>
<li><p>snp.index a <code>integer</code> representing the index position of the
SNV in the 1KG GDS file that contains all SNVs</p></li>
<li><p>keep a <code>logical</code> indicating if the genotype exists for the SNV</p></li>
<li><p>hetero a <code>logical</code> indicating if the SNV is heterozygote</p></li>
<li><p>homo a <code>logical</code> indicating if the SNV is homozygote</p></li>
<li><p>lap a <code>numeric</code> indicating lower allelic fraction</p></li>
<li><p>LOH a <code>integer</code> indicating if the SNV is in an LOH region
(0=not LOH, 1=in LOH)</p></li>
<li><p>imbAR a <code>integer</code> indicating if the SNV is in an imbalanced
region (-1=not classified as imbalanced or LOH, 0=in LOH; 1=tested
positive for imbalance in at least 1 window)</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Path to the demo 1KG GDS file is located in this package</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tests"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1_good_small_1KG_GDS.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Copy the Profile GDS file demo that has been pruned and annotated</span></span></span>
<span class="r-in"><span><span class="co">## into a test directory (deleted after the example has been run)</span></span></span>
<span class="r-in"><span><span class="va">dataDirAllelicFraction</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 <span class="st">"demoAllelicFraction"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">dataDirAllelicFraction</span>, showWarnings<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                 recursive<span class="op">=</span><span class="cn">FALSE</span>, mode<span class="op">=</span><span class="st">"0777"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1_demo_with_pruning_and_1KG_annot.gds"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDirAllelicFraction</span>, <span class="st">"ex1.gds"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Open the reference GDS file (demo version)</span></span></span>
<span class="r-in"><span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Profile GDS file for one profile</span></span></span>
<span class="r-in"><span><span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDirAllelicFraction</span>, <span class="st">"ex1.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">profileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileProfile</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Chromosome length information</span></span></span>
<span class="r-in"><span><span class="co">## chr23 is chrX, chr24 is chrY and chrM is 25</span></span></span>
<span class="r-in"><span><span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">248956422L</span>, <span class="fl">242193529L</span>, <span class="fl">198295559L</span>, <span class="fl">190214555L</span>,</span></span>
<span class="r-in"><span>    <span class="fl">181538259L</span>, <span class="fl">170805979L</span>, <span class="fl">159345973L</span>, <span class="fl">145138636L</span>, <span class="fl">138394717L</span>, <span class="fl">133797422L</span>,</span></span>
<span class="r-in"><span>    <span class="fl">135086622L</span>, <span class="fl">133275309L</span>, <span class="fl">114364328L</span>, <span class="fl">107043718L</span>, <span class="fl">101991189L</span>, <span class="fl">90338345L</span>,</span></span>
<span class="r-in"><span>    <span class="fl">83257441L</span>,  <span class="fl">80373285L</span>,  <span class="fl">58617616L</span>,  <span class="fl">64444167L</span>,  <span class="fl">46709983L</span>, <span class="fl">50818468L</span>,</span></span>
<span class="r-in"><span>    <span class="fl">156040895L</span>, <span class="fl">57227415L</span>,  <span class="fl">16569L</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The function returns a data frame containing the allelic fraction info</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">RAIDS</span><span class="fu">:::</span><span class="fu">computeAllelicFractionDNA</span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>,</span></span>
<span class="r-in"><span>    gdsSample<span class="op">=</span><span class="va">profileGDS</span>,</span></span>
<span class="r-in"><span>    currentProfile<span class="op">=</span><span class="st">"ex1"</span>, studyID<span class="op">=</span><span class="st">"MYDATA"</span>, chrInfo<span class="op">=</span><span class="va">chrInfo</span>, minCov<span class="op">=</span><span class="fl">10L</span>,</span></span>
<span class="r-in"><span>    minProb<span class="op">=</span><span class="fl">0.999</span>, eProb<span class="op">=</span><span class="fl">0.001</span>, cutOffLOH<span class="op">=</span><span class="op">-</span><span class="fl">5</span>,</span></span>
<span class="r-in"><span>    cutOffHomoScore<span class="op">=</span><span class="op">-</span><span class="fl">3</span>, wAR<span class="op">=</span><span class="fl">9L</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   cnt.tot cnt.ref cnt.alt snp.pos snp.chr normal.geno pruned snp.index keep</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      41      40       0 3722256       1           3   TRUE       160 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      17      17       0 3722328       1           3   TRUE       162 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      27      27       0 3767522       1           3  FALSE       204 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      15      15       0 3868160       1           3   TRUE       256 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      11       4       7 3869467       1           3  FALSE       259 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      37      14      23 4712655       1           3   TRUE       288 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   hetero  homo lap LOH imbAR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  FALSE  TRUE 0.5   0    -1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  FALSE  TRUE 0.5   0    -1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  FALSE  TRUE 0.5   0    -1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  FALSE  TRUE 0.5   0    -1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   TRUE FALSE 0.5   0    -1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   TRUE FALSE 0.5   0    -1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Close both GDS files (important)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">profileGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Unlink Profile GDS file (created for demo purpose)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDirAllelicFraction</span>, <span class="st">"ex1.gds"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">dataDirAllelicFraction</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

