<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate the allelic fraction of the pruned SNVs for a specific profile — estimateAllelicFraction • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate the allelic fraction of the pruned SNVs for a specific profile — estimateAllelicFraction"><meta property="og:description" content="The function estimates the allelic fraction of the
SNVs for a specific profile and add the information to the associated
Profile GDS file. The allelic fraction estimation method is adapted to
the type of study (DNA or RNA)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
    <li>
      <a href="../articles/Wrappers.html">Using wrappper functions</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate the allelic fraction of the pruned SNVs for a specific profile</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/R/allelicFraction.R" class="external-link"><code>R/allelicFraction.R</code></a></small>
    <div class="hidden name"><code>estimateAllelicFraction.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function estimates the allelic fraction of the
SNVs for a specific profile and add the information to the associated
Profile GDS file. The allelic fraction estimation method is adapted to
the type of study (DNA or RNA).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">estimateAllelicFraction</span><span class="op">(</span></span>
<span>  <span class="va">gdsReference</span>,</span>
<span>  <span class="va">gdsProfile</span>,</span>
<span>  <span class="va">currentProfile</span>,</span>
<span>  <span class="va">studyID</span>,</span>
<span>  <span class="va">chrInfo</span>,</span>
<span>  studyType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DNA"</span>, <span class="st">"RNA"</span><span class="op">)</span>,</span>
<span>  minCov <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  minProb <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  eProb <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  cutOffLOH <span class="op">=</span> <span class="op">-</span><span class="fl">5</span>,</span>
<span>  cutOffHomoScore <span class="op">=</span> <span class="op">-</span><span class="fl">3</span>,</span>
<span>  wAR <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  cutOffAR <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  gdsRefAnnot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blockID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-gdsreference">gdsReference<a class="anchor" aria-label="anchor" href="#arg-gdsreference"></a></dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened Reference GDS file.</p></dd>


<dt id="arg-gdsprofile">gdsProfile<a class="anchor" aria-label="anchor" href="#arg-gdsprofile"></a></dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened Profile GDS file.</p></dd>


<dt id="arg-currentprofile">currentProfile<a class="anchor" aria-label="anchor" href="#arg-currentprofile"></a></dt>
<dd><p>a <code>character</code> string corresponding to
the sample identifier as used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt id="arg-studyid">studyID<a class="anchor" aria-label="anchor" href="#arg-studyid"></a></dt>
<dd><p>a <code>character</code> string corresponding to the name of
the study as
used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt id="arg-chrinfo">chrInfo<a class="anchor" aria-label="anchor" href="#arg-chrinfo"></a></dt>
<dd><p>a <code>vector</code> of <code>integer</code> values representing
the length of the chromosomes. See 'details' section.</p></dd>


<dt id="arg-studytype">studyType<a class="anchor" aria-label="anchor" href="#arg-studytype"></a></dt>
<dd><p>a <code>character</code> string representing the type of study.
The possible choices are: "DNA" and "RNA". The type of study affects the
way the estimation of the allelic fraction is done. Default: <code>"DNA"</code>.</p></dd>


<dt id="arg-mincov">minCov<a class="anchor" aria-label="anchor" href="#arg-mincov"></a></dt>
<dd><p>a single positive <code>integer</code> representing the minimum
required coverage. Default: <code>10L</code>.</p></dd>


<dt id="arg-minprob">minProb<a class="anchor" aria-label="anchor" href="#arg-minprob"></a></dt>
<dd><p>a single <code>numeric</code> between 0 and 1 representing the
probability that the calculated genotype call is correct.
Default: <code>0.999</code>.</p></dd>


<dt id="arg-eprob">eProb<a class="anchor" aria-label="anchor" href="#arg-eprob"></a></dt>
<dd><p>a single <code>numeric</code> between 0 and 1 representing the
probability of sequencing error. Default: <code>0.001</code>.</p></dd>


<dt id="arg-cutoffloh">cutOffLOH<a class="anchor" aria-label="anchor" href="#arg-cutoffloh"></a></dt>
<dd><p>a single <code>numeric</code> representing the cutoff, in log,
for the homozygote score to assign a region as LOH.
Default: <code>-5</code>.</p></dd>


<dt id="arg-cutoffhomoscore">cutOffHomoScore<a class="anchor" aria-label="anchor" href="#arg-cutoffhomoscore"></a></dt>
<dd><p>a single <code>numeric</code> representing the cutoff, in
log, that the SNVs in a block are called homozygote by error.
Default: <code>-3</code>.</p></dd>


<dt id="arg-war">wAR<a class="anchor" aria-label="anchor" href="#arg-war"></a></dt>
<dd><p>a single positive <code>integer</code> representing the size-1 of
the window used to compute an empty box. Default: <code>9</code>.</p></dd>


<dt id="arg-cutoffar">cutOffAR<a class="anchor" aria-label="anchor" href="#arg-cutoffar"></a></dt>
<dd><p>a single <code>numeric</code> representing the cutoff, in
log score, that the SNVs in a gene are allelic fraction different 0.5
Default: <code>3</code>.</p></dd>


<dt id="arg-gdsrefannot">gdsRefAnnot<a class="anchor" aria-label="anchor" href="#arg-gdsrefannot"></a></dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened Reference SNV Annotation GDS file.
This parameter is RNA specific.
Default: <code>NULL</code>.</p></dd>


<dt id="arg-blockid">blockID<a class="anchor" aria-label="anchor" href="#arg-blockid"></a></dt>
<dd><p>a <code>character</code> string corresponding to the block
identifier in <code>gdsRefAnnot</code>. <strong>This parameter is RNA specific.</strong>
Default: <code>NULL</code></p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>a <code>logicial</code> indicating if the function should print
message when running. Default: <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The integer <code>0L</code> when successful.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The <code>chrInfo</code> parameter contains the length of the chromosomes. The
length of the chromosomes can be obtain through the
<code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></code>
library.</p>
<p>As example, for hg38 genome:</p>
<p></p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">requireNamespace</span>(<span class="st">"GenomeInfoDb"</span>, <span class="at">quietly=</span><span class="cn">TRUE</span>) <span class="sc">&amp;&amp;</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>     <span class="fu">requireNamespace</span>(<span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, <span class="at">quietly=</span><span class="cn">TRUE</span>)) {</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>     chrInfo <span class="ot">&lt;-</span> GenomeInfoDb<span class="sc">::</span><span class="fu">seqlengths</span>(BSgenome.Hsapiens.UCSC.hg38<span class="sc">::</span>Hsapiens)[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>]</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>}</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span></code></pre><p></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Required library for GDS</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Path to the demo 1KG GDS file located in this package</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tests"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Profile GDS file for one profile</span></span></span>
<span class="r-in"><span><span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"ex1.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Copy the Profile GDS file demo that has been pruned and annotated</span></span></span>
<span class="r-in"><span><span class="co">## into current directory</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1_demo_with_pruning_and_1KG_annot.gds"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                            <span class="va">fileProfile</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Open the reference GDS file (demo version)</span></span></span>
<span class="r-in"><span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Profile GDS file for one profile</span></span></span>
<span class="r-in"><span><span class="va">profileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileProfile</span>, readonly<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Required library for this example to run correctly</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GenomeInfoDb"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Chromosome length information</span></span></span>
<span class="r-in"><span>    <span class="co">## chr23 is chrX, chr24 is chrY and chrM is 25</span></span></span>
<span class="r-in"><span>    <span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Estimate the allelic fraction of the pruned SNVs</span></span></span>
<span class="r-in"><span>    <span class="fu">estimateAllelicFraction</span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>, gdsProfile<span class="op">=</span><span class="va">profileGDS</span>,</span></span>
<span class="r-in"><span>        currentProfile<span class="op">=</span><span class="st">"ex1"</span>, studyID<span class="op">=</span><span class="st">"MYDATA"</span>, chrInfo<span class="op">=</span><span class="va">chrInfo</span>,</span></span>
<span class="r-in"><span>        studyType<span class="op">=</span><span class="st">"DNA"</span>, minCov<span class="op">=</span><span class="fl">10L</span>, minProb<span class="op">=</span><span class="fl">0.999</span>, eProb<span class="op">=</span><span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>        cutOffLOH<span class="op">=</span><span class="op">-</span><span class="fl">5</span>, cutOffHomoScore<span class="op">=</span><span class="op">-</span><span class="fl">3</span>, wAR<span class="op">=</span><span class="fl">9</span>, cutOffAR<span class="op">=</span><span class="fl">3</span>,</span></span>
<span class="r-in"><span>        gdsRefAnnot<span class="op">=</span><span class="cn">NULL</span>, blockID<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## The allelic fraction is saved in the 'lap' node of Profile GDS file</span></span></span>
<span class="r-in"><span>    <span class="co">## The 'lap' entry should be present</span></span></span>
<span class="r-in"><span>    <span class="va">profileGDS</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Close both GDS files (important)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">profileGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Remove Profile GDS file (created for demo purpose)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">fileProfile</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

      </footer></div>






  </body></html>

