<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Append the genotype information from a profile into the associated
Profile GDS File — generateGDS1KGgenotypeFromSNPPileup • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Append the genotype information from a profile into the associated
Profile GDS File — generateGDS1KGgenotypeFromSNPPileup"><meta property="og:description" content="This function append the genotype information from a specific
profile into the Profile GDS file. The genotype information is extracted
from a SNV file as generated by SNP-pileup or other tools."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Ancestry_Inference_Step_by_Step.html">Genetic Ancestry Inference Step by Step (optional)</a>
    </li>
    <li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from 1000 Genomes (optional)</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Append the genotype information from a profile into the associated
Profile GDS File</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/HEAD/R/gdsWrapper_internal.R" class="external-link"><code>R/gdsWrapper_internal.R</code></a></small>
    <div class="hidden name"><code>generateGDS1KGgenotypeFromSNPPileup.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function append the genotype information from a specific
profile into the Profile GDS file. The genotype information is extracted
from a SNV file as generated by SNP-pileup or other tools.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateGDS1KGgenotypeFromSNPPileup</span><span class="op">(</span></span>
<span>  <span class="va">pathGeno</span>,</span>
<span>  <span class="va">listSamples</span>,</span>
<span>  <span class="va">listPos</span>,</span>
<span>  <span class="va">offset</span>,</span>
<span>  minCov <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minProb <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  seqError <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  <span class="va">dfPedProfile</span>,</span>
<span>  <span class="va">batch</span>,</span>
<span>  <span class="va">studyDF</span>,</span>
<span>  <span class="va">pathProfileGDS</span>,</span>
<span>  <span class="va">genoSource</span>,</span>
<span>  <span class="va">verbose</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>pathGeno</dt>
<dd><p>a <code>character</code> string representing the path to a
directory with the genotype files for the profiles, as generated by
snp-pileup. The path must exist.</p></dd>


<dt>listSamples</dt>
<dd><p>a <code>vector</code> of <code>character</code> string corresponding
to the sample identifiers that will have a Profile GDS file created. The
sample identifiers must be present in the "Name.ID" column of the
<code>data.frame</code> passed to the <code>dfPedProfile</code> parameter.</p></dd>


<dt>listPos</dt>
<dd><p>a <code>data.frame</code> containing 2 columns. The first column,
called "snp.chromosome" contains the name of the chromosome where the
SNV is located. The second column, called "snp.position" contains the
position of the SNV on the chromosome.</p></dd>


<dt>offset</dt>
<dd><p>a <code>integer</code> to adjust if the genome start at 0 or 1.</p></dd>


<dt>minCov</dt>
<dd><p>a single positive <code>integer</code> representing the minimum
coverage needed to keep the SNVs in the analysis. Default: <code>10</code>.</p></dd>


<dt>minProb</dt>
<dd><p>a single positive <code>numeric</code> between 0 and 1
representing the probability that the base change at the SNV position
is not an error.
Default: <code>0.999</code>.</p></dd>


<dt>seqError</dt>
<dd><p>a single positive <code>numeric</code> between 0 and 1
representing the sequencing error rate. Default: <code>0.001</code>.</p></dd>


<dt>dfPedProfile</dt>
<dd><p>a <code>data.frame</code> with the information about
the sample(s).
Those are mandatory columns: "Name.ID",
"Case.ID", "Sample.Type", "Diagnosis", "Source". All columns must be in
<code>character</code> strings. The <code>data.frame</code>
must contain the information for all the samples passed in the
<code>listSamples</code> parameter.</p></dd>


<dt>batch</dt>
<dd><p>a single positive <code>integer</code> representing the current
identifier for the batch. Beware, this field is not stored anymore.</p></dd>


<dt>studyDF</dt>
<dd><p>a <code>data.frame</code> containing the information about the
study associated to the analysed sample(s). The <code>data.frame</code> must have
those 3 columns: "study.id", "study.desc", "study.platform". All columns
must be in <code>character</code> strings.</p></dd>


<dt>pathProfileGDS</dt>
<dd><p>a <code>character</code> string representing the path to
the directory where the GDS Sample files will be created.</p></dd>


<dt>genoSource</dt>
<dd><p>a <code>character</code> string with two possible values:
'snp-pileup' or 'generic'. It specifies if the genotype files
are generated by snp-pileup (Facets) or are a generic format CSV file
with at least those columns:
'Chromosome', 'Position', 'Ref', 'Alt', 'Count', 'File1R' and 'File1A'.
The 'Count' is the depth at the specified position;
'FileR' is the depth of the reference allele and
'File1A' is the depth of the specific alternative allele.</p></dd>


<dt>verbose</dt>
<dd><p>a <code>logical</code> indicating if the function must print
messages when running.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The  function returns <code>0L</code> when successful.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Path to the files in this package</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tests"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The data.frame containing the information about the study</span></span></span>
<span class="r-in"><span><span class="co">## The 3 mandatory columns: "study.id", "study.desc", "study.platform"</span></span></span>
<span class="r-in"><span><span class="co">## The entries should be strings, not factors (stringsAsFactors=FALSE)</span></span></span>
<span class="r-in"><span><span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"MYDATA"</span>,</span></span>
<span class="r-in"><span>                        study.desc <span class="op">=</span> <span class="st">"Description"</span>,</span></span>
<span class="r-in"><span>                        study.platform <span class="op">=</span> <span class="st">"PLATFORM"</span>,</span></span>
<span class="r-in"><span>                        stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The data.frame containing the information about the samples</span></span></span>
<span class="r-in"><span><span class="co">## The entries should be strings, not factors (stringsAsFactors=FALSE)</span></span></span>
<span class="r-in"><span><span class="va">samplePED</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Name.ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1"</span>, <span class="st">"ex2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    Case.ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient_h11"</span>, <span class="st">"Patient_h12"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    Diagnosis<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Cancer"</span>, <span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    Sample.Type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Primary Tumor"</span>, <span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    Source<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Databank B"</span>, <span class="fl">2</span><span class="op">)</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">samplePED</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samplePED</span><span class="op">$</span><span class="va">Name.ID</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## List of SNV positions</span></span></span>
<span class="r-in"><span><span class="va">listPositions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>snp.chromosome<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    snp.position<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3467333</span>, <span class="fl">3467428</span>, <span class="fl">3469375</span>, <span class="fl">3469387</span>, <span class="fl">3469502</span>, <span class="fl">3469527</span>,</span></span>
<span class="r-in"><span>    <span class="fl">3469737</span>, <span class="fl">3471497</span>, <span class="fl">3471565</span>, <span class="fl">3471618</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Append genotype information to the Profile GDS file</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">RAIDS</span><span class="fu">:::</span><span class="fu">generateGDS1KGgenotypeFromSNPPileup</span><span class="op">(</span>pathGeno<span class="op">=</span><span class="va">dataDir</span>,</span></span>
<span class="r-in"><span>            listSamples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1"</span><span class="op">)</span>, listPos<span class="op">=</span><span class="va">listPositions</span>,</span></span>
<span class="r-in"><span>            offset<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, minCov<span class="op">=</span><span class="fl">10</span>, minProb<span class="op">=</span><span class="fl">0.999</span>, seqError<span class="op">=</span><span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>            dfPedProfile<span class="op">=</span><span class="va">samplePED</span>, batch<span class="op">=</span><span class="fl">1</span>, studyDF<span class="op">=</span><span class="va">studyDF</span>,</span></span>
<span class="r-in"><span>            pathProfileGDS<span class="op">=</span><span class="va">dataDir</span>, genoSource<span class="op">=</span><span class="st">"snp-pileup"</span>,</span></span>
<span class="r-in"><span>            verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The function returns OL when successful</span></span></span>
<span class="r-in"><span><span class="va">result</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The Profile GDS file 'ex1.gds' has been created in the</span></span></span>
<span class="r-in"><span><span class="co">## specified directory</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dataDir</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "1KG_Test.gds"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2] "1KG_Test_02.gds"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "Sample_Info_Test.RDS"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "ex1.gds"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "ex1.txt.gz"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "ex1_demo.gds"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "ex1_demo_with_pruning.gds"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "ex1_demo_with_pruning_and_1KG_annot.gds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "ex1_good_small_1KG_Annot_GDS.gds"       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "ex1_good_small_1KG_GDS.gds"             </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Unlink Profile GDS file (created for demo purpose)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1.gds"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

