<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Extract the genotype information for a SNV dataset using the Profile GDS file and the Reference GDS file — getTableSNV • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Extract the genotype information for a SNV dataset using the Profile GDS file and the Reference GDS file — getTableSNV"><meta property="og:description" content="The function generates a data.frame containing the
genotype information from a initial list of SNVs associated to a specific
profile. The function uses the information present in the Reference GDS file
and the Profile GDS file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
    <li>
      <a href="../articles/Wrappers.html">Using wrappper functions</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract the genotype information for a SNV dataset using the Profile GDS file and the Reference GDS file</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/R/allelicFraction_internal.R" class="external-link"><code>R/allelicFraction_internal.R</code></a></small>
    <div class="hidden name"><code>getTableSNV.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function generates a <code>data.frame</code> containing the
genotype information from a initial list of SNVs associated to a specific
profile. The function uses the information present in the Reference GDS file
and the Profile GDS file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getTableSNV</span><span class="op">(</span></span>
<span>  <span class="va">gdsReference</span>,</span>
<span>  <span class="va">gdsSample</span>,</span>
<span>  <span class="va">currentProfile</span>,</span>
<span>  <span class="va">studyID</span>,</span>
<span>  minCov <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minProb <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  eProb <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  <span class="va">verbose</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-gdsreference">gdsReference<a class="anchor" aria-label="anchor" href="#arg-gdsreference"></a></dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code> (a
GDS file), the opened Reference GDS file.</p></dd>


<dt id="arg-gdssample">gdsSample<a class="anchor" aria-label="anchor" href="#arg-gdssample"></a></dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened Profile GDS file.</p></dd>


<dt id="arg-currentprofile">currentProfile<a class="anchor" aria-label="anchor" href="#arg-currentprofile"></a></dt>
<dd><p>a <code>character</code> string corresponding to
the sample identifier used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt id="arg-studyid">studyID<a class="anchor" aria-label="anchor" href="#arg-studyid"></a></dt>
<dd><p>a <code>character</code> string corresponding to the study
identifier used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt id="arg-mincov">minCov<a class="anchor" aria-label="anchor" href="#arg-mincov"></a></dt>
<dd><p>a single positive <code>integer</code> representing the minimum
required coverage. Default: <code>10L</code>.</p></dd>


<dt id="arg-minprob">minProb<a class="anchor" aria-label="anchor" href="#arg-minprob"></a></dt>
<dd><p>a single <code>numeric</code> between <code>0</code> and <code>1</code>
representing the probability that the calculated genotype call is correct.
Default: <code>0.999</code>.</p></dd>


<dt id="arg-eprob">eProb<a class="anchor" aria-label="anchor" href="#arg-eprob"></a></dt>
<dd><p>a single <code>numeric</code> between <code>0</code> and <code>1</code>
representing the probability of sequencing error. Default: <code>0.001</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>a <code>logicial</code> indicating if messages should be printed
when the function is running.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a <code>data.frame</code> containing:</p><dl><dt>cnt.tot</dt>
<dd><p>a single <code>integer</code> representing the total coverage for
the SNV.</p></dd>

<dt>cnt.ref</dt>
<dd><p>a single <code>integer</code> representing the coverage for
the reference allele.</p></dd>

<dt>cnt.alt</dt>
<dd><p>a single <code>integer</code> representing the coverage for
the alternative allele.</p></dd>

<dt>snpPos</dt>
<dd><p>a single <code>integer</code> representing the SNV position.</p></dd>

<dt>snp.chr</dt>
<dd><p>a single <code>integer</code> representing the SNV chromosome.</p></dd>

<dt>normal.geno</dt>
<dd><p>a single <code>numeric</code> indicating the genotype of the
SNV. The possibles are: <code>0</code> (wild-type homozygote), <code>1</code>
(heterozygote), <code>2</code> (altenative homozygote), <code>3</code> indicating that
the normal genotype is unknown.</p></dd>

<dt>pruned</dt>
<dd><p>a <code>logical</code></p></dd>

<dt>snp.index</dt>
<dd><p>a <code>vector</code> of <code>integer</code> representing the
position of the SNVs in the Reference GDS file.</p></dd>

<dt>keep</dt>
<dd><p>a <code>logical</code></p></dd>

<dt>hetero</dt>
<dd><p>a <code>logical</code></p></dd>

<dt>homo</dt>
<dd><p>a <code>logical</code></p></dd>


</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Required library</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Path to the demo 1KG GDS file is located in this package</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tests"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Temporary Profile GDS file for one profile in temporary directory</span></span></span>
<span class="r-in"><span><span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"ex1.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Copy the Profile GDS file demo that has been pruned and annotated</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"ex1_demo_with_pruning_and_1KG_annot.gds"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 <span class="va">fileProfile</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Open the reference GDS file (demo version)</span></span></span>
<span class="r-in"><span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Open Profile GDS file for one profile</span></span></span>
<span class="r-in"><span><span class="va">profileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileProfile</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The function returns a data frame containing the SNVs information</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">RAIDS</span><span class="fu">:::</span><span class="fu">getTableSNV</span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>, gdsSample<span class="op">=</span><span class="va">profileGDS</span>,</span></span>
<span class="r-in"><span>        currentProfile<span class="op">=</span><span class="st">"ex1"</span>, studyID<span class="op">=</span><span class="st">"MYDATA"</span>, minCov<span class="op">=</span><span class="fl">10L</span>, minProb<span class="op">=</span><span class="fl">0.999</span>,</span></span>
<span class="r-in"><span>        eProb<span class="op">=</span><span class="fl">0.001</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   cnt.tot cnt.ref cnt.alt snp.pos snp.chr normal.geno pruned snp.index keep</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      41      40       0 3722256       1           3   TRUE       160 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      17      17       0 3722328       1           3   TRUE       162 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      27      27       0 3767522       1           3  FALSE       204 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      15      15       0 3868160       1           3   TRUE       256 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      11       4       7 3869467       1           3  FALSE       259 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      37      14      23 4712655       1           3   TRUE       288 TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   hetero  homo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  FALSE  TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  FALSE  TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  FALSE  TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  FALSE  TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   TRUE FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   TRUE FALSE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Close both GDS files (important)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">profileGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Remove Profile GDS file (created for demo purpose)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">fileProfile</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

