<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Extract the genotype information for a SNV dataset using
the GDS Sample file and the 1KG GDS file — getTableSNV • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Extract the genotype information for a SNV dataset using
the GDS Sample file and the 1KG GDS file — getTableSNV"><meta property="og:description" content="The function generates a data.frame containing the
genotype information from a initial list of SNVs associated to a specific
sample. The function uses the
information present in the 1KG GDS file and the GDS Sample file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from 1000 Genomes (optional)</a>
    </li>
    <li>
      <a href="../articles/aicsPaper.html">Accurate Inference of Genetic Ancestry from Cancer Sequencing</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract the genotype information for a SNV dataset using
the GDS Sample file and the 1KG GDS file</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/HEAD/R/allelicFraction.R" class="external-link"><code>R/allelicFraction.R</code></a></small>
    <div class="hidden name"><code>getTableSNV.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function generates a <code>data.frame</code> containing the
genotype information from a initial list of SNVs associated to a specific
sample. The function uses the
information present in the 1KG GDS file and the GDS Sample file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getTableSNV</span><span class="op">(</span></span>
<span>  <span class="va">gds</span>,</span>
<span>  <span class="va">gdsSample</span>,</span>
<span>  <span class="va">sampleCurrent</span>,</span>
<span>  <span class="va">study.id</span>,</span>
<span>  minCov <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minProb <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  eProb <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gds</dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code> (a GDS file),
the opened 1KG GDS file.</p></dd>


<dt>gdsSample</dt>
<dd><p>an object of class <code><a href="https://rdrr.io/pkg/gdsfmt/man/gds.class.html" class="external-link">gds.class</a></code>
(a GDS file), the opened GDS Sample file.</p></dd>


<dt>sampleCurrent</dt>
<dd><p>a <code>character</code> string corresponding to
the sample identifier used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt>study.id</dt>
<dd><p>a <code>character</code> string corresponding to the study
identifier used in <code><a href="pruningSample.html">pruningSample</a></code> function.</p></dd>


<dt>minCov</dt>
<dd><p>a single positive <code>integer</code> representing the minimum
required coverage. Default: <code>10L</code>.</p></dd>


<dt>minProb</dt>
<dd><p>a single <code>numeric</code> between <code>0</code> and <code>1</code>
representing the probability that the calculated genotype call is correct.
Default: <code>0.999</code>.</p></dd>


<dt>eProb</dt>
<dd><p>a single <code>numeric</code> between <code>0</code> and <code>1</code>
representing the probability of sequencing error. Default: <code>0.001</code>.</p></dd>


<dt>verbose</dt>
<dd><p>a <code>logicial</code> indicating if messages should be printed
when the function is running. Default: <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>data.frame</code> containing:</p><ul><li><p>cnt.tot a single <code>integer</code> representing the total coverage for
the SNV.</p></li>
<li><p>cnt.ref a single <code>integer</code> representing the coverage for
the reference allele.</p></li>
<li><p>cnt.alt a single <code>integer</code> representing the coverage for
the alternative allele.</p></li>
<li><p>snp.pos a single <code>integer</code> representing the SNV position.</p></li>
<li><p>snp.chr a single <code>integer</code> representing the SNV chromosome.</p></li>
<li><p>normal.geno a single <code>numeric</code> indicating the genotype of the
SNV. The possibles are: <code>0</code> (wild-type homozygote), <code>1</code>
(heterozygote), <code>2</code> (altenative homozygote), <code>3</code> indicating that
the normal genotype is unknown.</p></li>
<li><p>snp.index a <code>vector</code> of <code>integer</code> representing the
position of the SNVs in the 1KG GDS file.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Path to the demo pedigree file is located in this package</span></span></span>
<span class="r-in"><span><span class="va">data.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## TODO</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

