<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add1KG2SampleGDS.html">add1KG2SampleGDS()</a></code> </p>
        </td>
        <td><p>Add the genotype information for the list of pruned SNVs into the Profile GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneBlockGDSRefAnnot.html">addGeneBlockGDSRefAnnot()</a></code> </p>
        </td>
        <td><p>Append information associated to blocks, as indexes, into the Population Reference SNV Annotation GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneBlockRefAnnot.html">addGeneBlockRefAnnot()</a></code> </p>
        </td>
        <td><p>Append information associated to blocks, as indexes, into the Population Reference SNV Annotation GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="addRef2GDS1KG.html">addRef2GDS1KG()</a></code> </p>
        </td>
        <td><p>Add the information about the unrelated patients to the Reference GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="addStudy1Kg.html">addStudy1Kg()</a></code> </p>
        </td>
        <td><p>Append information about the 1KG samples into the Profile GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="computeAncestryFromSynthetic.html">computeAncestryFromSynthetic()</a></code> </p>
        </td>
        <td><p>Select the optimal K and D parameters using the synthetic data and infer the ancestry of a specific profile</p></td>
      </tr><tr><td>
          <p><code><a href="computeAncestryFromSyntheticFile.html">computeAncestryFromSyntheticFile()</a></code> </p>
        </td>
        <td><p>Select the optimal K and D parameters using the synthetic data and infer the ancestry of a specific profile</p></td>
      </tr><tr><td>
          <p><code><a href="computeKNNRefSample.html">computeKNNRefSample()</a></code> </p>
        </td>
        <td><p>Run a k-nearest neighbors analysis on one specific profile</p></td>
      </tr><tr><td>
          <p><code><a href="computeKNNRefSynthetic.html">computeKNNRefSynthetic()</a></code> </p>
        </td>
        <td><p>Run a k-nearest neighbors analysis on a subset of the synthetic dataset</p></td>
      </tr><tr><td>
          <p><code><a href="computePCAMultiSynthetic.html">computePCAMultiSynthetic()</a></code> </p>
        </td>
        <td><p>Project synthetic profiles onto existing principal component axes generated using the reference 1KG profiles</p></td>
      </tr><tr><td>
          <p><code><a href="computePCARefSample.html">computePCARefSample()</a></code> </p>
        </td>
        <td><p>Project specified profile onto PCA axes generated using known reference profiles</p></td>
      </tr><tr><td>
          <p><code><a href="computePoolSyntheticAncestryGr.html">computePoolSyntheticAncestryGr()</a></code> </p>
        </td>
        <td><p>Run a PCA analysis and a K-nearest neighbors analysis on a small set of synthetic data using all 1KG profiles except the ones used to generate the synthetic profiles</p></td>
      </tr><tr><td>
          <p><code><a href="computeSyntheticROC.html">computeSyntheticROC()</a></code> </p>
        </td>
        <td><p>Calculate the AUROC of the inferences for specific values of D and K using the inferred ancestry results from the synthetic profiles.</p></td>
      </tr><tr><td>
          <p><code><a href="createAUROCGraph.html">createAUROCGraph()</a></code> </p>
        </td>
        <td><p>Generate accuracy graph</p></td>
      </tr><tr><td>
          <p><code><a href="createAccuracyGraph.html">createAccuracyGraph()</a></code> </p>
        </td>
        <td><p>Generate accuracy graph</p></td>
      </tr><tr><td>
          <p><code><a href="createStudy2GDS1KG.html">createStudy2GDS1KG()</a></code> </p>
        </td>
        <td><p>Create the Profile GDS file(s) for one or multiple specific profiles using the information from a RDS Sample description file and the 1KG GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="demoKnownSuperPop1KG.html">demoKnownSuperPop1KG</a></code> </p>
        </td>
        <td><p>The known super population ancestry of the demo 1KG reference profiles.</p></td>
      </tr><tr><td>
          <p><code><a href="demoPCA1KG.html">demoPCA1KG</a></code> </p>
        </td>
        <td><p>The PCA results of the demo 1KG reference dataset for demonstration purpose. Beware that the PCA has been run on a very small subset of the 1KG reference dataset and should not be used to call ancestry inference on a real profile.</p></td>
      </tr><tr><td>
          <p><code><a href="demoPCASyntheticProfiles.html">demoPCASyntheticProfiles</a></code> </p>
        </td>
        <td><p>The PCA result of demo synthetic profiles projected on the demo subset 1KG reference PCA.</p></td>
      </tr><tr><td>
          <p><code><a href="demoPedigreeEx1.html">demoPedigreeEx1</a></code> </p>
        </td>
        <td><p>The pedigree information about a demo profile called 'ex1'.</p></td>
      </tr><tr><td>
          <p><code><a href="estimateAllelicFraction.html">estimateAllelicFraction()</a></code> </p>
        </td>
        <td><p>Estimate the allelic fraction of the pruned SNVs for a specific profile</p></td>
      </tr><tr><td>
          <p><code><a href="generateGDS1KG.html">generateGDS1KG()</a></code> </p>
        </td>
        <td><p>Generate the GDS file that will contain the information from Reference data set (reference data set)</p></td>
      </tr><tr><td>
          <p><code><a href="generateMapSnvSel.html">generateMapSnvSel()</a></code> </p>
        </td>
        <td><p>Generate the filter SNP information file in RDS format</p></td>
      </tr><tr><td>
          <p><code><a href="generatePhase1KG2GDS.html">generatePhase1KG2GDS()</a></code> </p>
        </td>
        <td><p>Adding the phase information into the Reference GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="generatePhaseRef.html">generatePhaseRef()</a></code> </p>
        </td>
        <td><p>Adding the phase information into the Reference GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="getRef1KGPop.html">getRef1KGPop()</a></code> </p>
        </td>
        <td><p>Extract the specified column from the 1KG GDS 'sample.ref' node for the reference profiles (real ancestry assignation)</p></td>
      </tr><tr><td>
          <p><code><a href="getRefSuperPop.html">getRefSuperPop()</a></code> </p>
        </td>
        <td><p>Extract the from the 1KG GDS 'sample.ref' node for the reference profiles (real ancestry assignation)</p></td>
      </tr><tr><td>
          <p><code><a href="groupChr1KGSNV.html">groupChr1KGSNV()</a></code> </p>
        </td>
        <td><p>Merge the genotyping files per chromosome into one file</p></td>
      </tr><tr><td>
          <p><code><a href="identifyRelative.html">identifyRelative()</a></code> </p>
        </td>
        <td><p>Identify genetically unrelated patients in GDS Reference file</p></td>
      </tr><tr><td>
          <p><code><a href="identifyRelativeRef.html">identifyRelativeRef()</a></code> </p>
        </td>
        <td><p>Identify genetically unrelated patients in GDS Reference file</p></td>
      </tr><tr><td>
          <p><code><a href="inferAncestry.html">inferAncestry()</a></code> </p>
        </td>
        <td><p>Run most steps leading to the ancestry inference call on a specific DNA profile</p></td>
      </tr><tr><td>
          <p><code><a href="inferAncestryGeneAware.html">inferAncestryGeneAware()</a></code> </p>
        </td>
        <td><p>Run most steps leading to the ancestry inference call on a specific RNA profile</p></td>
      </tr><tr><td>
          <p><code><a href="matKNNSynthetic.html">matKNNSynthetic</a></code> </p>
        </td>
        <td><p>A small <code>data.frame</code> containing the inferred ancestry on the synthetic profiles.</p></td>
      </tr><tr><td>
          <p><code><a href="pedSynthetic.html">pedSynthetic</a></code> </p>
        </td>
        <td><p>A small <code>data.frame</code> containing the information related to synthetic profiles. The ancestry of the profiles used to generate the synthetic profiles must be present.</p></td>
      </tr><tr><td>
          <p><code><a href="prepPed1KG.html">prepPed1KG()</a></code> </p>
        </td>
        <td><p>Prepare the pedigree file using pedigree information from Reference</p></td>
      </tr><tr><td>
          <p><code><a href="prepSynthetic.html">prepSynthetic()</a></code> </p>
        </td>
        <td><p>Add information related to the synthetic profiles (study and synthetic reference profiles information) into a Profile GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="pruningSample.html">pruningSample()</a></code> </p>
        </td>
        <td><p>Compute the list of pruned SNVs for a specific profile using the information from the Reference GDS file and a linkage disequilibrium analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runExomeAncestry.html">runExomeAncestry()</a></code> </p>
        </td>
        <td><p>Run most steps leading to the ancestry inference call on a specific exome profile</p></td>
      </tr><tr><td>
          <p><code><a href="runRNAAncestry.html">runRNAAncestry()</a></code> </p>
        </td>
        <td><p>Run most steps leading to the ancestry inference call on a specific RNA profile</p></td>
      </tr><tr><td>
          <p><code><a href="select1KGPop.html">select1KGPop()</a></code> </p>
        </td>
        <td><p>Random selection of a specific number of reference profiles in each subcontinental population present in the 1KG GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="select1KGPopForSynthetic.html">select1KGPopForSynthetic()</a></code> </p>
        </td>
        <td><p>Random selection of a specific number of reference profiles in each subcontinental population present in the 1KG GDS file ( same as select1KGPop but the function doesn't need gds object as parameters but the file name of the referenceGDS )</p></td>
      </tr><tr><td>
          <p><code><a href="snpPositionDemo.html">snpPositionDemo</a></code> </p>
        </td>
        <td><p>A small <code>data.frame</code> containing the SNV information.</p></td>
      </tr><tr><td>
          <p><code><a href="snvListVCF.html">snvListVCF()</a></code> </p>
        </td>
        <td><p>Generate a VCF with the information from the SNPs that pass a cut-off threshold</p></td>
      </tr><tr><td>
          <p><code><a href="splitSelectByPop.html">splitSelectByPop()</a></code> </p>
        </td>
        <td><p>Group samples per subcontinental population</p></td>
      </tr><tr><td>
          <p><code><a href="syntheticGeno.html">syntheticGeno()</a></code> </p>
        </td>
        <td><p>Generate synthetic profiles for each cancer profile and 1KG reference profile combination and add them to the Profile GDS file</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

