<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from 1000 Genomes (optional)</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="RAIDS-package.html">RAIDS-package</a></code> <code><a href="RAIDS-package.html">RAIDS</a></code> </p>
        </td>
        <td><p>RAIDS: Accurate Inference of Genetic Ancestry from Cancer Sequences</p></td>
      </tr><tr><td>
          <p><code><a href="add1KG2SampleGDS.html">add1KG2SampleGDS()</a></code> </p>
        </td>
        <td><p>Add the genotype information for the list of pruned SNVs
into the Profile GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="addBlockFromPlink2GDS.html">addBlockFromPlink2GDS()</a></code> </p>
        </td>
        <td><p>TODO contain the information from 1KG</p></td>
      </tr><tr><td>
          <p><code><a href="addGeneBlockGDSRefAnnot.html">addGeneBlockGDSRefAnnot()</a></code> </p>
        </td>
        <td><p>Generate two indexes based on gene annotation for gdsAnnot1KG
block and add the indexes into the
gdsAnnot1KG</p></td>
      </tr><tr><td>
          <p><code><a href="addPhase1KG2SampleGDSFromFile.html">addPhase1KG2SampleGDSFromFile()</a></code> </p>
        </td>
        <td><p>TODO</p></td>
      </tr><tr><td>
          <p><code><a href="addPhase1KG2SampleGDSFromGDS.html">addPhase1KG2SampleGDSFromGDS()</a></code> </p>
        </td>
        <td><p>TODO</p></td>
      </tr><tr><td>
          <p><code><a href="addRef2GDS1KG.html">addRef2GDS1KG()</a></code> </p>
        </td>
        <td><p>Add the information about the unrelated patients to the 1KG GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="addStudy1Kg.html">addStudy1Kg()</a></code> </p>
        </td>
        <td><p>Append information about the 1KG samples into
the Profile GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="appendStudy2GDS1KG.html">appendStudy2GDS1KG()</a></code> </p>
        </td>
        <td><p>Create the GDS Sample file(s) for one or multiple specific samples
using the information from a RDS Sample description file and the 1KG
GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="basePCASample.html">basePCASample()</a></code> </p>
        </td>
        <td><p>Compute principal component axes (PCA) on SNV data using the
reference samples</p></td>
      </tr><tr><td>
          <p><code><a href="computeAncestryFromSyntheticFile.html">computeAncestryFromSyntheticFile()</a></code> </p>
        </td>
        <td><p>Select the optimal K and D parameters using the synthetic data and
infer the ancestry of a specific sample</p></td>
      </tr><tr><td>
          <p><code><a href="computeKNNRefSample.html">computeKNNRefSample()</a></code> </p>
        </td>
        <td><p>Run a k-nearest neighbors analysis on a 1KG reference profile</p></td>
      </tr><tr><td>
          <p><code><a href="computeKNNRefSynthetic.html">computeKNNRefSynthetic()</a></code> </p>
        </td>
        <td><p>Run a k-nearest neighbors analysis on a subset of the
synthetic dataset</p></td>
      </tr><tr><td>
          <p><code><a href="computePCAMultiSynthetic.html">computePCAMultiSynthetic()</a></code> </p>
        </td>
        <td><p>Project synthetic profiles onto existing principal component axes
generated using the reference 1KG profiles</p></td>
      </tr><tr><td>
          <p><code><a href="computePCARefSample.html">computePCARefSample()</a></code> </p>
        </td>
        <td><p>Compute Principal Component Analysis (PCA) on SNV genotype for
data from a sample</p></td>
      </tr><tr><td>
          <p><code><a href="computePoolSyntheticAncestry.html">computePoolSyntheticAncestry()</a></code> </p>
        </td>
        <td><p>TODO</p></td>
      </tr><tr><td>
          <p><code><a href="computePoolSyntheticAncestryGr.html">computePoolSyntheticAncestryGr()</a></code> </p>
        </td>
        <td><p>Run a PCA analysis and a K-nearest neighbors analysis on a small set
of synthetic data using all 1KG profiles except the ones used to generate
the synthetic profiles</p></td>
      </tr><tr><td>
          <p><code><a href="computePrunedPCARef.html">computePrunedPCARef()</a></code> </p>
        </td>
        <td><p>Compute principal component axes (PCA) on pruned SNV with the
reference samples</p></td>
      </tr><tr><td>
          <p><code><a href="computeSyntheticROC.html">computeSyntheticROC()</a></code> </p>
        </td>
        <td><p>TODO</p></td>
      </tr><tr><td>
          <p><code><a href="createStudy2GDS1KG.html">createStudy2GDS1KG()</a></code> </p>
        </td>
        <td><p>Create the Profile GDS file(s) for one or multiple specific profiles
using the information from a RDS Sample description file and the 1KG
GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="estimateAllelicFraction.html">estimateAllelicFraction()</a></code> </p>
        </td>
        <td><p>Estimate the allelic fraction of the pruned SNVs for a specific
profile</p></td>
      </tr><tr><td>
          <p><code><a href="gds2tfam.html">gds2tfam()</a></code> </p>
        </td>
        <td><p>create a file tfam file for plink from the gdsReference file</p></td>
      </tr><tr><td>
          <p><code><a href="gds2tfamSample.html">gds2tfamSample()</a></code> </p>
        </td>
        <td><p>create a file tfam file for plink from the gdsProfile file</p></td>
      </tr><tr><td>
          <p><code><a href="gds2tped.html">gds2tped()</a></code> </p>
        </td>
        <td><p>create a file tped file for plink from the gds file</p></td>
      </tr><tr><td>
          <p><code><a href="generateGDS1KG.html">generateGDS1KG()</a></code> </p>
        </td>
        <td><p>Generate the GDS file that will contain the information from 1KG
data set (reference data set)</p></td>
      </tr><tr><td>
          <p><code><a href="generateMapSnvSel.html">generateMapSnvSel()</a></code> </p>
        </td>
        <td><p>Generate the filter SNP information file in RDS format</p></td>
      </tr><tr><td>
          <p><code><a href="generatePhase1KG2GDS.html">generatePhase1KG2GDS()</a></code> </p>
        </td>
        <td><p>Adding the phase information into the Reference GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="getRef1KGPop.html">getRef1KGPop()</a></code> </p>
        </td>
        <td><p>Extract the specified column from the 1KG GDS 'sample.ref' node
for the reference profiles (real ancestry assignation)</p></td>
      </tr><tr><td>
          <p><code><a href="groupChr1KGSNV.html">groupChr1KGSNV()</a></code> </p>
        </td>
        <td><p>Merge the genotyping files per chromosome into one file</p></td>
      </tr><tr><td>
          <p><code><a href="identifyRelative.html">identifyRelative()</a></code> </p>
        </td>
        <td><p>Identify genetically unrelated patients in GDS 1KG file</p></td>
      </tr><tr><td>
          <p><code><a href="prepPed1KG.html">prepPed1KG()</a></code> </p>
        </td>
        <td><p>Prepare the pedigree file using pedigree information from 1KG</p></td>
      </tr><tr><td>
          <p><code><a href="prepSynthetic.html">prepSynthetic()</a></code> </p>
        </td>
        <td><p>Add information related to the synthetic profiles
(study and synthetic reference profiles information) into a Profile GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="projectSample2PCA.html">projectSample2PCA()</a></code> </p>
        </td>
        <td><p>Project profile onto existing principal component axes (PCA)</p></td>
      </tr><tr><td>
          <p><code><a href="pruningSample.html">pruningSample()</a></code> </p>
        </td>
        <td><p>Compute the list of pruned SNVs for a specific profile using the
information from the 1KG GDS file and a linkage disequilibrium analysis</p></td>
      </tr><tr><td>
          <p><code><a href="runExomeAncestry.html">runExomeAncestry()</a></code> </p>
        </td>
        <td><p>TODO</p></td>
      </tr><tr><td>
          <p><code><a href="select1KGPop.html">select1KGPop()</a></code> </p>
        </td>
        <td><p>Random selection of a specific number of reference profiles in each
subcontinental population present in the 1KG GDS file</p></td>
      </tr><tr><td>
          <p><code><a href="snvListVCF.html">snvListVCF()</a></code> </p>
        </td>
        <td><p>Generate a VCF with the information from the SNPs that pass
a cut-off threshold</p></td>
      </tr><tr><td>
          <p><code><a href="splitSelectByPop.html">splitSelectByPop()</a></code> </p>
        </td>
        <td><p>Group samples per subcontinental population</p></td>
      </tr><tr><td>
          <p><code><a href="syntheticGeno.html">syntheticGeno()</a></code> </p>
        </td>
        <td><p>Generate synthetic profiles for each cancer profile and 1KG
reference profile combination and add them to the Profile GDS file</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

