<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run most steps leading to the ancestry inference call on a specific DNA profile — inferAncestry • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run most steps leading to the ancestry inference call on a specific DNA profile — inferAncestry"><meta property="og:description" content="This function runs most steps leading to the ancestry inference
call on a specific RNA profile. First, the function creates the
Profile GDS file for the specific profile using the information from a
RDS Sample description file and the Population Reference GDS file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run most steps leading to the ancestry inference call on a specific DNA profile</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/R/processStudy.R" class="external-link"><code>R/processStudy.R</code></a></small>
    <div class="hidden name"><code>inferAncestry.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function runs most steps leading to the ancestry inference
call on a specific RNA profile. First, the function creates the
Profile GDS file for the specific profile using the information from a
RDS Sample description file and the Population Reference GDS file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">inferAncestry</span><span class="op">(</span></span>
<span>  <span class="va">profileFile</span>,</span>
<span>  <span class="va">pathProfileGDS</span>,</span>
<span>  <span class="va">fileReferenceGDS</span>,</span>
<span>  <span class="va">fileReferenceAnnotGDS</span>,</span>
<span>  <span class="va">chrInfo</span>,</span>
<span>  <span class="va">syntheticRefDF</span>,</span>
<span>  genoSource <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"snp-pileup"</span>, <span class="st">"generic"</span>, <span class="st">"VCF"</span>, <span class="st">"bam"</span><span class="op">)</span>,</span>
<span>  np <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-profilefile">profileFile<a class="anchor" aria-label="anchor" href="#arg-profilefile"></a></dt>
<dd><p>a <code>character</code> string representing the path and the
file name of the genotype file or the bam if genoSource is snp-pileup the
fine extension must be .txt.gz, if VCF the extension must be .vcf.gz</p></dd>


<dt id="arg-pathprofilegds">pathProfileGDS<a class="anchor" aria-label="anchor" href="#arg-pathprofilegds"></a></dt>
<dd><p>a <code>character</code> string representing the path to
the directory where the GDS Profile files will be created.
Default: <code>NULL</code>.</p></dd>


<dt id="arg-filereferencegds">fileReferenceGDS<a class="anchor" aria-label="anchor" href="#arg-filereferencegds"></a></dt>
<dd><p>a <code>character</code> string representing the file
name of the Population Reference GDS file. The file must exist.</p></dd>


<dt id="arg-filereferenceannotgds">fileReferenceAnnotGDS<a class="anchor" aria-label="anchor" href="#arg-filereferenceannotgds"></a></dt>
<dd><p>a <code>character</code> string representing the
file name of the Population Reference GDS Annotation file. The file
must exist.</p></dd>


<dt id="arg-chrinfo">chrInfo<a class="anchor" aria-label="anchor" href="#arg-chrinfo"></a></dt>
<dd><p>a <code>vector</code> of positive <code>integer</code> values
representing the length of the chromosomes. See 'details' section.</p></dd>


<dt id="arg-syntheticrefdf">syntheticRefDF<a class="anchor" aria-label="anchor" href="#arg-syntheticrefdf"></a></dt>
<dd><p>a <code>data.frame</code> containing a subset of
reference profiles for each sub-population present in the Reference GDS
file. The <code>data.frame</code> must have those columns:</p><dl><dt>sample.id</dt>
<dd><p>a <code>character</code> string representing the sample
identifier.</p></dd>

<dt>pop.group</dt>
<dd><p>a <code>character</code> string representing the
subcontinental population assigned to the sample.</p></dd>

<dt>superPop</dt>
<dd><p>a <code>character</code> string representing the
super-population assigned to the sample.</p></dd>


</dl></dd>


<dt id="arg-genosource">genoSource<a class="anchor" aria-label="anchor" href="#arg-genosource"></a></dt>
<dd><p>a <code>character</code> string with two possible values:
'snp-pileup', 'generic' or 'VCF'. It specifies if the genotype files
are generated by snp-pileup (Facets) or are a generic format CSV file
with at least those columns:
'Chromosome', 'Position', 'Ref', 'Alt', 'Count', 'File1R' and 'File1A'.
The 'Count' is the depth at the specified position;
'FileR' is the depth of the reference allele and
'File1A' is the depth of the specific alternative allele.
Finally the file can be a VCF file with at least those genotype
fields: GT, AD, DP.</p></dd>


<dt id="arg-np">np<a class="anchor" aria-label="anchor" href="#arg-np"></a></dt>
<dd><p>a single positive <code>integer</code> specifying the number of
threads to be used. Default: <code>1L</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>a <code>logical</code> indicating if messages should be printed
to show how the different steps in the function. Default: <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>TODO a <code>list</code> containing 4 entries:</p><dl><dt><code>pcaSample</code></dt>
<dd><p>a <code>list</code> containing the information related
to the eigenvectors. The <code>list</code> contains those 3 entries:</p><dl><dt><code>sample.id</code></dt>
<dd><p>a <code>character</code> string representing the unique
identifier of the current profile.</p></dd>

<dt><code>eigenvector.ref</code></dt>
<dd><p>a <code>matrix</code> of <code>numeric</code> containing
the eigenvectors for the reference profiles.</p></dd>

<dt><code>eigenvector</code></dt>
<dd><p>a <code>matrix</code> of <code>numeric</code> containing the
eigenvectors for the current profile projected on the PCA from the
reference profiles.</p></dd>


</dl></dd>

<dt><code>paraSample</code></dt>
<dd><p>a <code>list</code> containing the results with
different <code>D</code> and <code>K</code> values that lead to optimal parameter
selection. The <code>list</code> contains those entries:</p><dl><dt><code>dfPCA</code></dt>
<dd><p>a <code>data.frame</code> containing statistical results
on all combined synthetic results done with a fixed value of <code>D</code> (the
number of dimensions). The <code>data.frame</code> contains those columns:</p><dl><dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>D</code> (the
number of dimensions).</p></dd>

<dt><code>median</code></dt>
<dd><p>a <code>numeric</code> representing the median of the
minimum AUROC obtained (within super populations) for all combination of
the fixed <code>D</code> value and all tested <code>K</code> values.</p></dd>

<dt><code>mad</code></dt>
<dd><p>a <code>numeric</code> representing the MAD of the minimum
AUROC obtained (within super populations) for all combination of the fixed
<code>D</code> value and all tested <code>K</code> values.</p></dd>

<dt><code>upQuartile</code></dt>
<dd><p>a <code>numeric</code> representing the upper quartile
of the minimum AUROC obtained (within super populations) for all
combination of the fixed <code>D</code> value and all tested <code>K</code> values.</p></dd>

<dt><code>k</code></dt>
<dd><p>a <code>numeric</code> representing the optimal <code>K</code> value
(the number of neighbors) for a fixed <code>D</code> value.</p></dd>


</dl></dd>

<dt><code>dfPop</code></dt>
<dd><p>a <code>data.frame</code> containing statistical results on
all combined synthetic results done with different values of <code>D</code> (the
number of dimensions) and <code>K</code> (the number of neighbors).
The <code>data.frame</code> contains those columns:</p><dl><dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>D</code> (the
number of dimensions).</p></dd>

<dt><code>K</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>K</code> (the
number of neighbors).</p></dd>

<dt><code>AUROC.min</code></dt>
<dd><p>a <code>numeric</code> representing the minimum accuracy
obtained by grouping all the synthetic results by super-populations, for
the specified values of <code>D</code> and <code>K</code>.</p></dd>

<dt><code>AUROC</code></dt>
<dd><p>a <code>numeric</code> representing the accuracy obtained
by grouping all the synthetic results for the specified values of <code>D</code>
and <code>K</code>.</p></dd>

<dt><code>Accu.CM</code></dt>
<dd><p>a <code>numeric</code> representing the value of accuracy
of the confusion matrix obtained by grouping all the synthetic results for
the specified values of <code>D</code> and <code>K</code>.</p></dd>


</dl></dd>

<dt><code>dfAUROC</code></dt>
<dd><p>a <code>data.frame</code> the summary of the results by
super-population. The <code>data.frame</code> contains
those columns:</p><dl><dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>D</code> (the
number of dimensions).</p></dd>

<dt><code>K</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>K</code> (the
number of neighbors).</p></dd>

<dt><code>Call</code></dt>
<dd><p>a <code>character</code> string representing the
super-population.</p></dd>

<dt><code>L</code></dt>
<dd><p>a <code>numeric</code> representing the lower value of the 95%
confidence interval for the AUROC obtained for the fixed values of
super-population, <code>D</code> and <code>K</code>.</p></dd>

<dt><code>AUROC</code></dt>
<dd><p>a <code>numeric</code> representing  the AUROC obtained for the
fixed values of super-population, <code>D</code> and <code>K</code>.</p></dd>

<dt><code>H</code></dt>
<dd><p>a <code>numeric</code> representing the higher value of the 95%
confidence interval for the AUROC obtained for the fixed values of
super-population, <code>D</code> and <code>K</code>.</p></dd>


</dl></dd>

<dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the optimal <code>D</code> value
(the number of dimensions) for the specific profile.</p></dd>

<dt><code>K</code></dt>
<dd><p>a <code>numeric</code> representing the optimal <code>K</code> value
(the number of neighbors) for the specific profile.</p></dd>

<dt><code>listD</code></dt>
<dd><p>a <code>numeric</code> representing the optimal <code>D</code>
values (the number of dimensions) for the specific profile. More than one
<code>D</code> is possible.</p></dd>


</dl></dd>

<dt><code>KNNSample</code></dt>
<dd><p>a <code>data.frame</code> containing the inferred ancestry
for different values of <code>K</code> and <code>D</code>. The <code>data.frame</code>
contains those columns:</p><dl><dt><code>sample.id</code></dt>
<dd><p>a <code>character</code> string representing the unique
identifier of the current profile.</p></dd>

<dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>D</code> (the
number of dimensions) used to infer the ancestry.</p></dd>

<dt><code>K</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>K</code> (the
number of neighbors) used to infer the ancestry.</p></dd>

<dt><code>SuperPop</code></dt>
<dd><p>a <code>character</code> string representing the inferred
ancestry for the specified <code>D</code> and <code>K</code> values.</p></dd>


</dl></dd>

<dt><code>KNNSynthetic</code></dt>
<dd><p>a <code>data.frame</code> containing the inferred ancestry
for each synthetic data for different values of <code>K</code> and <code>D</code>.
The <code>data.frame</code>
contains those columns: "sample.id", "D", "K", "infer.superPop", "ref.superPop"</p><dl><dt><code>sample.id</code></dt>
<dd><p>a <code>character</code> string representing the unique
identifier of the current synthetic data.</p></dd>

<dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>D</code> (the
number of dimensions) used to infer the ancestry.</p></dd>

<dt><code>K</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>K</code> (the
number of neighbors) used to infer the ancestry.</p></dd>

<dt><code>infer.superPop</code></dt>
<dd><p>a <code>character</code> string representing the inferred
ancestry for the specified <code>D</code> and <code>K</code> values.</p></dd>

<dt><code>ref.superPop</code></dt>
<dd><p>a <code>character</code> string representing the known
ancestry from the reference</p></dd>


</dl></dd>

<dt><code>Ancestry</code></dt>
<dd><p>a <code>data.frame</code> containing the inferred
ancestry for the current profile. The <code>data.frame</code> contains those
columns:</p><dl><dt><code>sample.id</code></dt>
<dd><p>a <code>character</code> string representing the unique
identifier of the current profile.</p></dd>

<dt><code>D</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>D</code> (the
number of dimensions) used to infer the ancestry.</p></dd>

<dt><code>K</code></dt>
<dd><p>a <code>numeric</code> representing the value of <code>K</code> (the
number of neighbors) used to infer the ancestry.</p></dd>

<dt><code>SuperPop</code></dt>
<dd><p>a <code>character</code> string representing the inferred
ancestry.</p></dd>


</dl></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>The profileAncestry() generates list <code>list</code>
TODO update the description</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Galinsky KJ, Bhatia G, Loh PR, Georgiev S, Mukherjee S, Patterson NJ,
Price AL. Fast Principal-Component Analysis Reveals Convergent Evolution
of ADH1B in Europe and East Asia. Am J Hum Genet. 2016 Mar 3;98(3):456-72.
doi: 10.1016/j.ajhg.2015.12.022. Epub 2016 Feb 25.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Required library for GDS</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Path to the demo 1KG GDS file is located in this package</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#################################################################</span></span></span>
<span class="r-in"><span><span class="co">## The 1KG GDS file and the 1KG SNV Annotation GDS file</span></span></span>
<span class="r-in"><span><span class="co">## need to be located in the same directory</span></span></span>
<span class="r-in"><span><span class="co">## Note that the 1KG GDS file used for this example is a</span></span></span>
<span class="r-in"><span><span class="co">## simplified version and CANNOT be used for any real analysis</span></span></span>
<span class="r-in"><span><span class="co">#################################################################</span></span></span>
<span class="r-in"><span><span class="va">path1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"tests"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fileReferenceGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fileAnnotGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"ex1_good_small_1KG_Annot.gds"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#################################################################</span></span></span>
<span class="r-in"><span><span class="co">## The Sample SNP pileup files (one per sample) need</span></span></span>
<span class="r-in"><span><span class="co">## to be located in the same directory.</span></span></span>
<span class="r-in"><span><span class="co">#################################################################</span></span></span>
<span class="r-in"><span><span class="va">demoProfileEx1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"snpPileup"</span>, <span class="st">"ex1.txt.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#################################################################</span></span></span>
<span class="r-in"><span><span class="co">## The path where the Profile GDS Files (one per sample)</span></span></span>
<span class="r-in"><span><span class="co">## will be created need to be specified.</span></span></span>
<span class="r-in"><span><span class="co">#################################################################</span></span></span>
<span class="r-in"><span><span class="va">pathProfileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"out.tmp"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">####################################################################</span></span></span>
<span class="r-in"><span><span class="co">## Fix seed to ensure reproducible results</span></span></span>
<span class="r-in"><span><span class="co">####################################################################</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileReferenceGDS</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds1KG</span>, nbProfiles<span class="op">=</span><span class="fl">2L</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Required library for this example to run correctly</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GenomeInfoDb"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Chromosome length information</span></span></span>
<span class="r-in"><span>    <span class="co">## chr23 is chrX, chr24 is chrY and chrM is 25</span></span></span>
<span class="r-in"><span>    <span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co"># \donttest{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>        <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">inferAncestry</span><span class="op">(</span>profileFile<span class="op">=</span><span class="va">demoProfileEx1</span>,</span></span>
<span class="r-in"><span>            pathProfileGDS<span class="op">=</span><span class="va">pathProfileGDS</span>,</span></span>
<span class="r-in"><span>            fileReferenceGDS<span class="op">=</span><span class="va">fileReferenceGDS</span>,</span></span>
<span class="r-in"><span>            fileReferenceAnnotGDS<span class="op">=</span><span class="va">fileAnnotGDS</span>,</span></span>
<span class="r-in"><span>            chrInfo<span class="op">=</span><span class="va">chrInfo</span>,</span></span>
<span class="r-in"><span>            syntheticRefDF<span class="op">=</span><span class="va">dataRef</span>,</span></span>
<span class="r-in"><span>            genoSource<span class="op">=</span><span class="st">"snp-pileup"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>        <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co"># }</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

