<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>A small data.frame containing the SNV information. — snpPositionDemo • RAIDS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="A small data.frame containing the SNV information. — snpPositionDemo"><meta property="og:description" content="The object is a data.frame with 17 columns."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
    <li>
      <a href="../articles/Wrappers.html">Using wrappper functions</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A small <code>data.frame</code> containing the SNV information.</h1>
    <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/R/RAIDS.R" class="external-link"><code>R/RAIDS.R</code></a></small>
    <div class="hidden name"><code>snpPositionDemo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The object is a <code>data.frame</code> with 17 columns.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">snpPositionDemo</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>The <code>data.frame</code> containing the information about the
synthetic profiles. The <code>data.frame</code> contains 4 columns:</p><dl><dt><code>cnt.tot</code></dt>
<dd><p>a <code>integer</code> representing the number of reads at
the SNV position.</p></dd>

<dt><code>cnt.ref</code></dt>
<dd><p>a <code>integer</code> representing the number of reads
corresponding to the reference at the SNV position.</p></dd>

<dt><code>cnt.alt</code></dt>
<dd><p>a <code>integer</code> representing the number of reads
different than the reference at the SNV position.</p></dd>

<dt><code>snp.pos</code></dt>
<dd><p>a <code>integer</code> representing the position of the
SNV on the chromosome.</p></dd>

<dt><code>snp.chr</code></dt>
<dd><p>a <code>integer</code> representing the chromosome on which
the SNV is located.</p></dd>

<dt><code>normal.geno</code></dt>
<dd><p>a <code>integer</code> representing the genotype
(0=wild-type reference; 1=heterozygote; 2=homozygote alternative;
3=unkown).</p></dd>

<dt><code>pruned</code></dt>
<dd><p>a <code>logical</code> indicated if the SNV is pruned.</p></dd>

<dt><code>snp.index</code></dt>
<dd><p>a <code>integer</code> representing the index of the
SNV in the reference SNV GDS file.</p></dd>

<dt><code>keep</code></dt>
<dd><p>a <code>logical</code> indicated if the genotype
exists for the SNV.</p></dd>

<dt><code>hetero</code></dt>
<dd><p>a <code>logical</code> indicated if the SNV is
heterozygote.</p></dd>

<dt><code>homo</code></dt>
<dd><p>a <code>logical</code> indicated if the SNV is homozygote.</p></dd>

<dt><code>block.id</code></dt>
<dd><p>a <code>integer</code> representing the block identifier
associated to the current SNV.</p></dd>

<dt><code>phase</code></dt>
<dd><p>a <code>integer</code> representing the block identifier
associated to the current SNV.</p></dd>

<dt><code>lap</code></dt>
<dd><p>a <code>numeric</code> representing the lower allelic
fraction.</p></dd>

<dt><code>LOH</code></dt>
<dd><p>a <code>integer</code> indicating if the SNV is in an LOH
region (0=not LOH, 1=in LOH).</p></dd>

<dt><code>imbAR</code></dt>
<dd><p>a <code>integer</code> indicating if the SNV is in an
imbalanced region (-1=not classified as imbalanced or LOH, 0=in LOH;
1=tested positive for imbalance in at least 1 window).</p></dd>

<dt><code>freq</code></dt>
<dd><p>a <code>numeric</code> representing the frequency of the
variant in the the reference.</p></dd>


</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The <code>data.frame</code> containing the information about the
synthetic profiles. The <code>data.frame</code> contains 4 columns:</p><dl><dt><code>cnt.tot</code></dt>
<dd><p>a <code>integer</code> representing the number of reads at
the SNV position.</p></dd>

<dt><code>cnt.ref</code></dt>
<dd><p>a <code>integer</code> representing the number of reads
corresponding to the reference at the SNV position.</p></dd>

<dt><code>cnt.alt</code></dt>
<dd><p>a <code>integer</code> representing the number of reads
different than the reference at the SNV position.</p></dd>

<dt><code>snp.pos</code></dt>
<dd><p>a <code>integer</code> representing the position of the
SNV on the chromosome.</p></dd>

<dt><code>snp.chr</code></dt>
<dd><p>a <code>integer</code> representing the chromosome on which
the SNV is located.</p></dd>

<dt><code>normal.geno</code></dt>
<dd><p>a <code>integer</code> representing the genotype
(0=wild-type reference; 1=heterozygote; 2=homozygote alternative; 3=unkown).</p></dd>

<dt><code>pruned</code></dt>
<dd><p>a <code>logical</code> indicated if the SNV is pruned.</p></dd>

<dt><code>snp.index</code></dt>
<dd><p>a <code>integer</code> representing the index of the
SNV in the reference SNV GDS file.</p></dd>

<dt><code>keep</code></dt>
<dd><p>a <code>logical</code> indicated if the genotype
exists for the SNV.</p></dd>

<dt><code>hetero</code></dt>
<dd><p>a <code>logical</code> indicated if the SNV is heterozygote.</p></dd>

<dt><code>homo</code></dt>
<dd><p>a <code>logical</code> indicated if the SNV is homozygote.</p></dd>

<dt><code>block.id</code></dt>
<dd><p>a <code>integer</code> representing the block identifier
associated to the current SNV.</p></dd>

<dt><code>phase</code></dt>
<dd><p>a <code>integer</code> representing the block identifier
associated to the current SNV.</p></dd>

<dt><code>lap</code></dt>
<dd><p>a <code>numeric</code> representing the lower allelic fraction.</p></dd>

<dt><code>LOH</code></dt>
<dd><p>a <code>integer</code> indicating if the SNV is in an LOH region
(0=not LOH, 1=in LOH).</p></dd>

<dt><code>imbAR</code></dt>
<dd><p>a <code>integer</code> indicating if the SNV is in an
imbalanced region (-1=not classified as imbalanced or LOH, 0=in LOH;
1=tested positive for imbalance in at least 1 window).</p></dd>

<dt><code>freq</code></dt>
<dd><p>a <code>numeric</code> representing the frequency of the
variant in the the reference.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This dataset can be
used to test the <code><a href="calcAFMLRNA.html">calcAFMLRNA</a></code> and <code><a href="tableBlockAF.html">tableBlockAF</a></code>
internal functions.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Loading demo dataset containing SNV information</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">snpPositionDemo</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Only use a subset of heterozygote SNVs related to one block</span></span></span>
<span class="r-in"><span><span class="va">subset</span> <span class="op">&lt;-</span> <span class="va">snpPositionDemo</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">snpPositionDemo</span><span class="op">$</span><span class="va">block.id</span> <span class="op">==</span> <span class="fl">2750</span> <span class="op">&amp;</span></span></span>
<span class="r-in"><span>                  <span class="va">snpPositionDemo</span><span class="op">$</span><span class="va">hetero</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cnt.ref"</span>, <span class="st">"cnt.alt"</span>, <span class="st">"phase"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Compute the log likelihood ratio based on the coverage of</span></span></span>
<span class="r-in"><span><span class="co">## each allele in a specific block</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">RAIDS</span><span class="fu">:::</span><span class="fu"><a href="calcAFMLRNA.html">calcAFMLRNA</a></span><span class="op">(</span><span class="va">subset</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $lR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 4.057862</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $aFraction</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.3389831</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nPhase</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sumAlleleLow</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 60</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sumAlleleHigh</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

