<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate Inference of Genetic Ancestry from Cancer Sequencing • aicsPaper</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate Inference of Genetic Ancestry from Cancer Sequencing">
<meta property="og:description" content="aicsPaper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">aicsPaper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/aicsPaper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate Inference of Genetic Ancestry from Cancer Sequencing</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      
      <div class="hidden name"><code>aicsPaper.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>aicsPaper</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes [aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), Alexander Krasnitz [aut]<br><strong>Version</strong>: 0.50.0<br><strong>Compiled date</strong>: 2022-04-01<br><strong>License</strong>: Artistic-2.0<br></p>
<div class="section level2">
<h2 id="licensing">Licensing<a class="anchor" aria-label="anchor" href="#licensing"></a>
</h2>
<p>The <em><a href="https://github.com/belleau/aicsPaper" class="external-link">aicsPaper</a></em> package and the underlying <em><a href="https://github.com/belleau/aicsPaper" class="external-link">aicsPaper</a></em> code are distributed under<br>
the Artistic license 2.0. You are free to use and redistribute this software.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>If you use this package for a publication, we would ask you to cite the following:</p>
<blockquote>
<p>Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz. Accurate and robust inference of genetic ancestry from cancer-derived molecular data across genomic platforms. bioRxiv 2022.02.01.478737; doi: <a href="https://doi.org/10.1101/2022.02.01.478737" class="external-link uri">https://doi.org/10.1101/2022.02.01.478737</a></p>
</blockquote>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple methods have been implemented to infer ancestry from germline DNA sequence <span class="citation">(Price et al. 2006; Pritchard, Stephens, and Donnelly 2000; Alexander, Novembre, and Lange 2009)</span>. However, genotyping of DNA from matched normal specimens is not part of standard clinical practice and is not performed routinely outside academic clinical centers. In sum, matched germline DNA sequence is often missing for cancer-derived molecular data. In such cases, having the possibility to infer ancestry from tumor-derived data would be beneficial.</p>
<p>The <strong>aicsPaper</strong> package implements an inference procedure that has been specifically developed to accurately infer genetic ancestry from cancer-derived sequencing. The covered cancer-derived sequencing are, more specifically, tumor exomes, targeted gene panels and RNA sequences.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version accessible on the <a href="https://github.com/belleau/aicsPaper" title="aicsPaper Github Site" class="external-link">aicsPaper Github Website</a>, the <em><a href="https://CRAN.R-project.org/package=devtools" class="external-link">devtools</a></em> package is required.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load required package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span>

<span class="co">## Install the latest version of aicsPaper</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'belleau/aicsPaper'</span><span class="op">)</span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="main-steps">Main Steps<a class="anchor" aria-label="anchor" href="#main-steps"></a>
</h2>
<p><br><br></p>
<div class="section level3">
<h3 id="formatting-the-information-from-1000-genomes-1kg">Formatting the information from 1000 Genomes (1KG)<a class="anchor" aria-label="anchor" href="#formatting-the-information-from-1000-genomes-1kg"></a>
</h3>
<div class="section level4">
<h4 id="download-the-required-files-from-1kg">Download the required files from 1KG<a class="anchor" aria-label="anchor" href="#download-the-required-files-from-1kg"></a>
</h4>
<p>First, the pedigree file with the description of the 1KG samples needs to be downloaded from the 1KG ftp site:</p>
<blockquote>
<p><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_g1k.ped" class="uri">ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_g1k.ped</a></p>
</blockquote>
<p>The genotyping files for the 1KG samples also needs to be downloaded. Beware that there is one file per chromosome:</p>
<blockquote>
<p>ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV</p>
</blockquote>
<p>The genotyping files are split by chromosome. All files corresponding to this pattern must be downloaded:</p>
<blockquote>
<p>ALL.chr*.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz*</p>
</blockquote>
<p>All files are related to GRCh38 genome.</p>
</div>
<div class="section level4">
<h4 id="split-the-1kg-genotyping-file-to-one-file-per-sample">Split the 1KG genotyping file to one file per sample<a class="anchor" aria-label="anchor" href="#split-the-1kg-genotyping-file-to-one-file-per-sample"></a>
</h4>
<p>TODO</p>
</div>
<div class="section level4">
<h4 id="create-a-predigree-file-in-rds-format">Create a predigree file in RDS format<a class="anchor" aria-label="anchor" href="#create-a-predigree-file-in-rds-format"></a>
</h4>
<p>The function <strong>prepPed1KG()</strong> is used to create the pedigree file in RDS format:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load required package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">aicsPaper</span><span class="op">)</span>

<span class="co">## The path to the pedigree file from 1KG</span>
<span class="co">## In this example, the file is in the current directory</span>
<span class="va">filePED1KGOri</span> <span class="op">&lt;-</span> <span class="st">"20130606_g1k.ped"</span>

<span class="co">## Extract needed information from the pedigree file from 1KG into a data.frame</span>
<span class="co">## Only the samples with genotyping information (sample file present </span>
<span class="co">## in PATHGENO parameter) are retained to create the final data.frame</span>
<span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepPed1KG.html">prepPed1KG</a></span><span class="op">(</span>pedFile<span class="op">=</span><span class="va">filePED1KGOri</span>, 
                    PATHGENO<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"sampleGeno"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Save the pedigree information data.frame as a RDS file </span>
<span class="co">## In this example, the file is saved here ./data/metadata/ped1KG.rds</span>
<span class="va">filePED1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"metadata"</span>,<span class="st">"ped1KG.rds"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ped</span>,  <span class="va">filePED1KG</span><span class="op">)</span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="prepare-a-bulk-snp-information-file-based-on-1kg-vcf">Prepare a bulk SNP information file based on 1KG VCF<a class="anchor" aria-label="anchor" href="#prepare-a-bulk-snp-information-file-based-on-1kg-vcf"></a>
</h4>
<p>Some intermediate file containing the SNP information from 1KG need to be generated so the information can ultimately be imported in a GDS file using Bioconductor <a href="https://bioconductor.org/packages/gdsfmt/" class="external-link">gdsfmt</a>.</p>
<p>The bulk intermediate SNP file contains the SNP position as well as the frequency in each super population.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This is not done in R</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="do">## The python script is in the 'scriptsPy' directory</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="cf">in</span> <span class="st">`</span><span class="at">ls PATHVCF/*shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz</span><span class="st">`</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    do</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>      chr<span class="ot">=</span><span class="er">$</span>(echo <span class="sc">$</span>FILECUR<span class="sc">|</span>perl <span class="sc">-</span>n <span class="sc">-</span>e <span class="st">'/ALL\.(chr[^\.]+)/;print $1'</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>      python PATH2SCRIPT<span class="sc">/</span>extract1000gFreq.py <span class="sc">$</span>{FILECUR} matFreq.<span class="sc">$</span>{chr}</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    done</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="cf">in</span> <span class="st">`</span><span class="at">seq 1 22</span><span class="st">`</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    do</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>      cat matFreq.chr<span class="sc">$</span>{i}.txt <span class="sc">&gt;</span>matFreqSNV.txt</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>      bzip2 matFreqSNV.txt</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    done</span></code></pre></div>
<p>The bulk SNP info file is called ‘<strong>matFreqSNV.txt</strong>.’</p>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="filter-the-bulk-snp-information-file">Filter the bulk SNP information file<a class="anchor" aria-label="anchor" href="#filter-the-bulk-snp-information-file"></a>
</h4>
<p>The bulk SNP file is filtered to only retain the SNPs with frequency higher then a specific cut-off (here &gt;=0.01) for at least one super population.</p>
<p>The filter SNP file is saved in RDS format (‘<strong>fileLSNP</strong>’ parameter). A second file containing the index of the retained SNPs is also created (‘<strong>fileLSNP</strong>’ parameter)</p>
<p>The function <strong>generateMapSnvSel()</strong> is used to filter the SNP file and generated the needed RDS files:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load required package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">aicsPaper</span><span class="op">)</span>

<span class="co">## The path to the intermediate SNP info file</span>
<span class="va">fileSNV.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGENO</span>, <span class="st">"matFreqSNV.txt.bz2"</span><span class="op">)</span>

<span class="co">## The paths and names of the two output files</span>
<span class="co">## One file contains the index of retained SNPs ("listSNP.rds")</span>
<span class="co">## One file contains the filter SNP information ("mapSNVSel.rds")</span>
<span class="va">fileLSNP.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSEL</span>, <span class="st">"listSNP.rds"</span><span class="op">)</span>
<span class="va">fileFREQ.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSEL</span>, <span class="st">"mapSNVSel.rds"</span><span class="op">)</span>

<span class="co">## Filter the bulk SNP file (fileSNV parameter)</span>
<span class="co">## Create a RDS with filter SNPs (fileFREQ parameter)</span>
<span class="co">## Also creates a RDS with the indexes of the retained SNPs (fileFREQ parameter)</span>
<span class="fu"><a href="../reference/generateMapSnvSel.html">generateMapSnvSel</a></span><span class="op">(</span>cutOff<span class="op">=</span><span class="fl">0.01</span>, fileSNV<span class="op">=</span><span class="va">fileSNV.v</span>,
                      fileLSNP <span class="op">=</span> <span class="va">fileLSNP.v</span>,
                      fileFREQ <span class="op">=</span> <span class="va">fileFREQ.v</span><span class="op">)</span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="generate-the-gds-file-with-the-1kg-information">Generate the GDS file with the 1KG information<a class="anchor" aria-label="anchor" href="#generate-the-gds-file-with-the-1kg-information"></a>
</h4>
<p>The CoreArray Genomic Data Structure (GDS) data files are files suited for large-scale datasets, especially for data which are much larger than the available random-access memory.</p>
<p>The function <strong>generateGDS1KG()</strong> is used to generate the GDS file that will contain the information related to 1KG:</p>
<p><br><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on which this document was compiled:</p>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19043)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=English_Canada.1252  LC_CTYPE=English_Canada.1252   </span>
<span class="co">## [3] LC_MONETARY=English_Canada.1252 LC_NUMERIC=C                   </span>
<span class="co">## [5] LC_TIME=English_Canada.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    parallel  stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] aicsPaper_0.50.0          GENESIS_2.22.2           </span>
<span class="co">##  [3] SNPRelate_1.26.0          EnsDb.Hsapiens.v86_2.99.0</span>
<span class="co">##  [5] ensembldb_2.16.4          AnnotationFilter_1.16.0  </span>
<span class="co">##  [7] GenomicFeatures_1.44.2    AnnotationDbi_1.54.1     </span>
<span class="co">##  [9] Biobase_2.52.0            GenomicRanges_1.44.0     </span>
<span class="co">## [11] GenomeInfoDb_1.28.4       IRanges_2.26.0           </span>
<span class="co">## [13] S4Vectors_0.30.2          BiocGenerics_0.38.0      </span>
<span class="co">## [15] gdsfmt_1.28.1             knitr_1.38               </span>
<span class="co">## [17] BiocStyle_2.20.2         </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] rjson_0.2.21                ellipsis_0.3.2             </span>
<span class="co">##   [3] rprojroot_2.0.2             DNAcopy_1.66.0             </span>
<span class="co">##   [5] XVector_0.32.0              GWASTools_1.38.0           </span>
<span class="co">##   [7] fs_1.5.2                    mice_3.14.0                </span>
<span class="co">##   [9] rstudioapi_0.13             MatrixModels_0.5-0         </span>
<span class="co">##  [11] bit64_4.0.5                 fansi_1.0.3                </span>
<span class="co">##  [13] xml2_1.3.2                  codetools_0.2-18           </span>
<span class="co">##  [15] splines_4.1.2               cachem_1.0.6               </span>
<span class="co">##  [17] SeqVarTools_1.30.0          jsonlite_1.8.0             </span>
<span class="co">##  [19] logistf_1.24.1              Rsamtools_2.8.0            </span>
<span class="co">##  [21] broom_0.7.12                dbplyr_2.1.1               </span>
<span class="co">##  [23] png_0.1-7                   BiocManager_1.30.16        </span>
<span class="co">##  [25] compiler_4.1.2              httr_1.4.2                 </span>
<span class="co">##  [27] backports_1.4.1             assertthat_0.2.1           </span>
<span class="co">##  [29] Matrix_1.3-4                fastmap_1.1.0              </span>
<span class="co">##  [31] lazyeval_0.2.2              cli_3.1.0                  </span>
<span class="co">##  [33] quantreg_5.88               htmltools_0.5.2            </span>
<span class="co">##  [35] prettyunits_1.1.1           tools_4.1.2                </span>
<span class="co">##  [37] glue_1.4.2                  GenomeInfoDbData_1.2.6     </span>
<span class="co">##  [39] dplyr_1.0.8                 rappdirs_0.3.3             </span>
<span class="co">##  [41] Rcpp_1.0.7                  jquerylib_0.1.4            </span>
<span class="co">##  [43] pkgdown_2.0.2               vctrs_0.3.8                </span>
<span class="co">##  [45] Biostrings_2.60.2           nlme_3.1-153               </span>
<span class="co">##  [47] rtracklayer_1.52.1          iterators_1.0.14           </span>
<span class="co">##  [49] lmtest_0.9-40               xfun_0.29                  </span>
<span class="co">##  [51] stringr_1.4.0               lifecycle_1.0.1            </span>
<span class="co">##  [53] restfulr_0.0.13             formula.tools_1.7.1        </span>
<span class="co">##  [55] XML_3.99-0.9                zlibbioc_1.38.0            </span>
<span class="co">##  [57] zoo_1.8-9                   ragg_1.2.2                 </span>
<span class="co">##  [59] hms_1.1.1                   MatrixGenerics_1.4.3       </span>
<span class="co">##  [61] ProtGenerics_1.24.0         SummarizedExperiment_1.22.0</span>
<span class="co">##  [63] sandwich_3.0-1              SparseM_1.81               </span>
<span class="co">##  [65] yaml_2.3.5                  curl_4.3.2                 </span>
<span class="co">##  [67] memoise_2.0.1               sass_0.4.1                 </span>
<span class="co">##  [69] biomaRt_2.48.3              stringi_1.7.3              </span>
<span class="co">##  [71] RSQLite_2.2.8               BiocIO_1.2.0               </span>
<span class="co">##  [73] desc_1.4.1                  foreach_1.5.2              </span>
<span class="co">##  [75] filelock_1.0.2              BiocParallel_1.26.2        </span>
<span class="co">##  [77] operator.tools_1.6.3        rlang_1.0.2                </span>
<span class="co">##  [79] pkgconfig_2.0.3             systemfonts_1.0.4          </span>
<span class="co">##  [81] matrixStats_0.61.0          bitops_1.0-7               </span>
<span class="co">##  [83] SeqArray_1.32.0             evaluate_0.15              </span>
<span class="co">##  [85] lattice_0.20-45             purrr_0.3.4                </span>
<span class="co">##  [87] GenomicAlignments_1.28.0    bit_4.0.4                  </span>
<span class="co">##  [89] tidyselect_1.1.2            magrittr_2.0.1             </span>
<span class="co">##  [91] bookdown_0.25               R6_2.5.1                   </span>
<span class="co">##  [93] generics_0.1.2              DelayedArray_0.18.0        </span>
<span class="co">##  [95] DBI_1.1.2                   withr_2.5.0                </span>
<span class="co">##  [97] mgcv_1.8-38                 pillar_1.7.0               </span>
<span class="co">##  [99] GWASExactHW_1.01            survival_3.2-13            </span>
<span class="co">## [101] KEGGREST_1.32.0             RCurl_1.98-1.5             </span>
<span class="co">## [103] tibble_3.1.6                crayon_1.5.0               </span>
<span class="co">## [105] utf8_1.2.2                  BiocFileCache_2.0.0        </span>
<span class="co">## [107] rmarkdown_2.13              progress_1.2.2             </span>
<span class="co">## [109] quantsmooth_1.58.0          grid_4.1.2                 </span>
<span class="co">## [111] data.table_1.14.2           blob_1.2.2                 </span>
<span class="co">## [113] digest_0.6.29               tidyr_1.2.0                </span>
<span class="co">## [115] textshaping_0.3.6           bslib_0.3.1</span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Alexander2009" class="csl-entry">
Alexander, D. H., J. Novembre, and K. Lange. 2009. <span>“Fast Model-Based Estimation of Ancestry in Unrelated Individuals.”</span> <em>Genome Res</em> 19 (9): 1655–64. <a href="https://doi.org/10.1101/gr.094052.109" class="external-link">https://doi.org/10.1101/gr.094052.109</a>.
</div>
<div id="ref-Price2006" class="csl-entry">
Price, A. L., N. J. Patterson, R. M. Plenge, M. E. Weinblatt, N. A. Shadick, and D. Reich. 2006. <span>“Principal Components Analysis Corrects for Stratification in Genome-Wide Association Studies.”</span> <em>Nat Genet</em> 38 (8): 904–9. <a href="https://doi.org/10.1038/ng1847" class="external-link">https://doi.org/10.1038/ng1847</a>.
</div>
<div id="ref-Pritchard2000" class="csl-entry">
Pritchard, J. K., M. Stephens, and P. Donnelly. 2000. <span>“Inference of Population Structure Using Multilocus Genotype Data.”</span> <em>Genetics</em> 155 (2): 945–59. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10835412" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/10835412</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
