<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate Inference of Genetic Ancestry from Cancer Sequencing • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate Inference of Genetic Ancestry from Cancer Sequencing">
<meta property="og:description" content="RAIDS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from 1000 Genomes (optional)</a>
    </li>
    <li>
      <a href="../articles/aicsPaper.html">Accurate Inference of Genetic Ancestry from Cancer Sequencing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate Inference of Genetic Ancestry from
Cancer Sequencing</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      
      <div class="hidden name"><code>aicsPaper.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 0.50.0<br><strong>Compiled date</strong>: 2022-08-12<br><strong>License</strong>: Apache License (== 2.0) | file
LICENSE<br></p>
<div class="section level2">
<h2 id="licensing">Licensing<a class="anchor" aria-label="anchor" href="#licensing"></a>
</h2>
<p>The <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em>
package and the underlying <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em> code are
distributed under<br>
the <a href="https://opensource.org/licenses/Apache-2.0" class="external-link uri">https://opensource.org/licenses/Apache-2.0</a> license. You
are free to use and redistribute this software.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>If you use the <strong>RAIDS</strong> package for a publication, we
would ask you to cite the following:</p>
<blockquote>
<p>Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander
Krasnitz. Accurate and robust inference of genetic ancestry from
cancer-derived molecular data across genomic platforms. bioRxiv
2022.02.01.478737; doi: <a href="https://doi.org/10.1101/2022.02.01.478737" class="external-link uri">https://doi.org/10.1101/2022.02.01.478737</a></p>
</blockquote>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple methods have been implemented to infer ancestry from
germline DNA sequence <span class="citation">(Price et al. 2006;
Pritchard, Stephens, and Donnelly 2000; Alexander, Novembre, and Lange
2009)</span>. However, genotyping of DNA from matched normal specimens
is not part of standard clinical practice and is not performed routinely
outside academic clinical centers. In sum, matched germline DNA sequence
is often missing for cancer-derived molecular data. In such cases,
having the possibility to infer ancestry from tumor-derived data would
be beneficial.</p>
<p>The <strong>RAIDS</strong> package implements an inference procedure
that has been specifically developed to accurately infer genetic
ancestry from cancer-derived sequencing. The covered cancer-derived
sequencing are, more specifically, tumor exomes, targeted gene panels
and RNA sequences.</p>
<p>The <strong>RAIDS</strong> package includes a data synthesis method
that enables the selection of the best parameters for a specific cancer
profile and also permits a statistically assessment of inference
accuracy for an individual cancer-derived molecular profile.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version accessible on the <a href="https://github.com/KrasnitzLab/RAIDS" title="RAIDS Github Site" class="external-link">RAIDS Github Website</a>, the <em><a href="https://CRAN.R-project.org/package=devtools" class="external-link">devtools</a></em>
package is required.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Install the latest version of RAIDS</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'KrasnitzLab/RAIDS'</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="main-steps">Main Steps<a class="anchor" aria-label="anchor" href="#main-steps"></a>
</h2>
<p>This is an overview of genetic ancestry inference from cancer-derived
molecular data:</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_v01.png" alt="An overview of the genetic ancestry inference process." width="120%"><p class="caption">
An overview of the genetic ancestry inference process.
</p>
</div>
<p>The main steps are:</p>
<ol style="list-style-type: decimal">
<li>Format the information from 1000 Genomes (1KG) (optional)</li>
<li>Format the information from an external study</li>
<li>Find the optimized parameters for the ancestry inference</li>
<li>Run the ancestry inference on the external study</li>
</ol>
<p>In the following sections, those main steps are described in
details.</p>
<p><br><br></p>
<div class="section level3">
<h3 id="step-1---formatting-the-information-from-1000-genomes-optional">Step 1 - Formatting the information from 1000 Genomes
(optional)<a class="anchor" aria-label="anchor" href="#step-1---formatting-the-information-from-1000-genomes-optional"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step1_v01.png" alt="Step 1 - Formatting the information from 1000 Genomes (optional)" width="120%"><p class="caption">
Step 1 - Formatting the information from 1000 Genomes (optional)
</p>
</div>
<hr>
<p><span style="color:darkred;font-weight:bold">Beware that a
pre-processed 1KG GDS file is available at this address:</span></p>
<p><span style="color:red"> <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds</a>
</span></p>
<p><span style="color:darkred;font-weight:bold">The size of the file is
15GB.</span> <span style="color:darkred;font-weight:bold">This section
can be skipped if you choose to use this file.</span></p>
<hr>
<p>The vignette <a href="Create_1KG_GDS_file.html">Formatting the
information from 1000 Genomes (optional)</a> is dedicated to this
step.</p>
</div>
<div class="section level3">
<h3 id="step-2---format-information-from-an-external-study-to-get-it-ready-for-ancestry-inference">Step 2 - Format information from an external study to get it ready
for ancestry inference<a class="anchor" aria-label="anchor" href="#step-2---format-information-from-an-external-study-to-get-it-ready-for-ancestry-inference"></a>
</h3>
<p>To run the ancestry inference on samples from an external study. The
information about the samples needs to follow a specific format.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step2_v01.png" alt="Step 2 - Formatting the information from an external study" width="120%"><p class="caption">
Step 2 - Formatting the information from an external study
</p>
</div>
<p>The required steps are:</p>
<ol style="list-style-type: decimal">
<li>Download or create the 3 reference files</li>
<li>Extract the read coverage at each SNP position in the 1KG GDS
file</li>
<li>Pruned</li>
</ol>
<p>Beware that a mapped BAM file is needed for each sample. The
reference genome used for the mapping must correspond to the one used to
generate the 1KG GDS file. In this case, the 1KG GDS file is based on
genome hg38.</p>
<div class="section level4">
<h4 id="step-1--download-or-create-the-3-reference-files">Step 1. Download or create the 3 reference files<a class="anchor" aria-label="anchor" href="#step-1--download-or-create-the-3-reference-files"></a>
</h4>
<p>wget <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds" class="external-link uri">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds</a>
wget <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matAnnot1000g.gds" class="external-link uri">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matAnnot1000g.gds</a>
wget <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/snvSel0.01.vcf.gz" class="external-link uri">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/snvSel0.01.vcf.gz</a></p>
<p>In a directory PATH_REF (I will use this name PATH_REF to refer at
this directory)</p>
</div>
<div class="section level4">
<h4 id="step-2--extract-the-read-coverage-at-each-snp-position-in-the-1kg-gds-file">Step 2. Extract the read coverage at each SNP position in the 1KG
GDS file<a class="anchor" aria-label="anchor" href="#step-2--extract-the-read-coverage-at-each-snp-position-in-the-1kg-gds-file"></a>
</h4>
<p>We used here snp-pileup from Facet</p>
<p><br><br></p>
<p>You need a <strong>ped</strong> file (call later FILE_PED) with the
column:</p>
<p>“Case.ID” Patient identifier<br>
“Name.ID”* This the unique id of the table “Sample.Type”* Ex Primary
Tumor or Blood Derived Normal (can be too WES or RNAseq) “Diagnosis”* C
or N (cancer normal) this is for the sample not the patient “Source”*
The tissue ex Ovary</p>
<p>must contain the information for all the samples (you can have more)
process at the same time.</p>
<p>You need the extra code in the package Facet to extract the SNV <a href="https://github.com/mskcc/facets/tree/master/inst/extcode" class="external-link">snp-pileup</a>
<span class="citation">(Shen and Seshan 2016)</span>.</p>
<p>The output from snp-pileup should be Name.ID.txt (snp-pileup add .gz
to the filename)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>    snp<span class="sc">-</span>pileup <span class="sc">-</span>g <span class="sc">-</span>d5000 <span class="sc">-</span>q15 <span class="sc">-</span>Q20  <span class="sc">-</span>r0 PATH_REF<span class="sc">/</span>snvSel0.<span class="fl">01.</span>vcf.gz PATH_OUT<span class="sc">/</span>Name.ID.txt FILEBAM.bam</span></code></pre></div>
<div class="section level5">
<h5 id="step-3--create-a-gds-file-for-each-sample">Step 3. Create a GDS file for each sample<a class="anchor" aria-label="anchor" href="#step-3--create-a-gds-file-for-each-sample"></a>
</h5>
<p>We will call this file gdsSample</p>
<p>in R: Here</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(RAIDS)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="do">##########################################</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Files related to the reference</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Downloaded</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="do">##########################################</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    PATH_1KG <span class="ot">&lt;-</span> PATH_REF</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    fileName.GDS <span class="ot">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    fileName.GDS.Annot <span class="ot">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    file.GDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_1KG, fileName.GDS)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    file.Annot <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_1KG, fileName.GDS.Annot)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="do">##########################################</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Need the path where the files generated by</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># snp-pileup are</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    PATHGENO <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_PILEUP)</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Path to the directory where the gds file for each sample</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># will be create</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    PATHSAMPLEGDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"The path"</span>)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># The file ped defined before</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>    filePED <span class="ot">&lt;-</span> FILE_PED</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>    ped <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(filePED)</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Need a data.frame with study description</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>    studyDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">study.id =</span> <span class="st">"Id of the study"</span>,</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>                              <span class="at">study.desc =</span> <span class="st">"Descripttion"</span>,</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>                              <span class="at">study.platform =</span> <span class="st">"Platform string like Whole Exome"</span>,</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>                              <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>    listSamples <span class="ot">&lt;-</span> ped[, <span class="st">"Name.ID"</span>]</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># The next function will create a GDS for each </span></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># sample in listSamples</span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">appendStudy2GDS1KG</span>(<span class="at">PATHGENO =</span> PATHGENO,</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>                       <span class="at">fileNamePED =</span> filePED,</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>                       <span class="at">fileNameGDS =</span> file.GDS,</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>                       <span class="at">listSamples =</span> listSamples,</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>                       <span class="at">batch =</span> <span class="dv">1</span>,</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>                       <span class="at">studyDF =</span> studyDF,</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>                       <span class="at">PATHSAMPLEGDS =</span> PATHSAMPLEGDS)<span class="er">))</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="step-4--select-a-pruned-subset-of-the-single-nucleotide-variants">Step 4. Select a pruned subset of the single nucleotide
variants<a class="anchor" aria-label="anchor" href="#step-4--select-a-pruned-subset-of-the-single-nucleotide-variants"></a>
</h4>
<p>A subset of pruned SNVs is retained for the following analyses. The
list of selected SNVs is specific to each sample and the information
about the retained SNVs is added to the GDS Sample file.</p>
<p>This step is requiring a lot of disk space and the process can be run
on multiple threads.</p>
<p>???:It use more disk space but the analysis for the samples can be
run simultaneously with multiple process (if we we all samples in one
file only one process can write in the GDS).</p>
<p>or in R for a s</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="co"># Files related to the reference</span></span>
<span>    <span class="co"># Downloaded</span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_WHERE_THE_REFERENCE_1KG_IS_LOCATED</span></span>
<span>    </span>
<span>    <span class="va">fileName.GDS.1KG</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span>    <span class="va">fileName.GDS.Annot</span> <span class="op">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span>    <span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.1KG</span><span class="op">)</span></span>
<span>    <span class="va">file.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.Annot</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    </span>
<span>    <span class="co"># Path to the directory where the GDS Sample files are located </span></span>
<span>    <span class="co"># will be create</span></span>
<span>    <span class="va">PATHSAMPLEGDS</span> <span class="op">&lt;-</span> <span class="va">PATH_WHERE_THE_GDS_SAMPLES_FILES_ARE_LOCATED</span></span>
<span>    </span>
<span>    <span class="va">gds_1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># The file ped defined before</span></span>
<span>    <span class="va">filePED</span> <span class="op">&lt;-</span> <span class="va">FILE_PED</span></span>
<span>    <span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span>    <span class="co"># You need a list of the sample Name.ID </span></span>
<span>    <span class="co"># you can get it for other source than </span></span>
<span>    <span class="co"># filePED </span></span>
<span>    <span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Can be run in parallel or in different clusters...</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="co"># Compute the SNV pruned subset</span></span>
<span>        <span class="fu"><a href="../reference/pruningSample.html">pruningSample</a></span><span class="op">(</span>gds<span class="op">=</span><span class="va">gds_1KG</span>,</span>
<span>                      method<span class="op">=</span><span class="st">"corr"</span>,</span>
<span>                      sampleCurrent <span class="op">=</span> <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                      listSNP <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      slide.max.bp.v <span class="op">=</span> <span class="fl">5e5</span>,</span>
<span>                      ld.threshold.v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>,</span>
<span>                      np <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      verbose.v<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                      chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      minAF.SuperPop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      keepGDSpruned <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      PATHSAMPLEGDS <span class="op">=</span> <span class="va">PATHSAMPLEGDS</span>,</span>
<span>                      keepFile <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="co"># Add the genotype of 1KG and of the sample listSamples[i] </span></span>
<span>        <span class="co"># for SNV pruned in the gdsSample</span></span>
<span>        <span class="fu"><a href="../reference/addStudy1Kg.html">addStudy1Kg</a></span><span class="op">(</span><span class="va">gds_1KG</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSAMPLEGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="co"># Add the sample name from 1KG in the study.annot of gdsSample</span></span>
<span>        <span class="fu"><a href="../reference/addStudy1Kg.html">addStudy1Kg</a></span><span class="op">(</span><span class="va">gds_1KG</span>, <span class="va">file.GDSSample</span><span class="op">)</span></span>
<span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-5--estimate-the-allelic-for-the-snv-pruned">Step 5. Estimate the allelic for the SNV pruned<a class="anchor" aria-label="anchor" href="#step-5--estimate-the-allelic-for-the-snv-pruned"></a>
</h4>
<p>or in R for a s</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="co"># Files related to the reference</span></span>
<span>    <span class="co"># Downloaded</span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span>    </span>
<span>    <span class="va">fileName.GDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span>    <span class="va">fileName.GDS.Annot</span> <span class="op">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span>    <span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span>    <span class="va">file.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.Annot</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    </span>
<span>    <span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Path to the directory where the gds file for each sample</span></span>
<span>    <span class="co"># will be create</span></span>
<span>    <span class="va">PATHSAMPLEGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Look at studyDF at step 3.</span></span>
<span>    <span class="va">study.id</span> <span class="op">&lt;-</span> <span class="st">"Id of the study"</span></span>
<span>    </span>
<span>    <span class="co"># The file ped defined before</span></span>
<span>    <span class="va">filePED</span> <span class="op">&lt;-</span> <span class="va">FILE_PED</span></span>
<span>    <span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span>    <span class="co"># You need a list of the sample Name.ID </span></span>
<span>    <span class="co"># you can get it from other source than </span></span>
<span>    <span class="co"># filePED </span></span>
<span>    <span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># You need the autosomal chr length from the genome.</span></span>
<span>    <span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">integer</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">chrInfo</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Hsapiens</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">i</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co"># Can be run in paralle or in different clusters...</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSAMPLEGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">gdsSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.GDSSample</span>, readonly <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>        <span class="fu"><a href="../reference/estimateAllelicFraction.html">estimateAllelicFraction</a></span><span class="op">(</span><span class="va">gds</span>, <span class="va">gdsSample</span>,</span>
<span>                                <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">study.id</span>,</span>
<span>                                <span class="va">chrInfo</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsSample</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-6--generate-the-synthetic-dataset">Step 6. Generate the synthetic dataset<a class="anchor" aria-label="anchor" href="#step-6--generate-the-synthetic-dataset"></a>
</h4>
<p>or in R</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span>    </span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="co"># Files related to the reference</span></span>
<span>    <span class="co"># Downloaded</span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span>    </span>
<span>    <span class="va">fileName.GDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span>    <span class="va">fileName.GDS.Annot</span> <span class="op">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span>    <span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span>    <span class="va">file.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.Annot</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    </span>
<span>    <span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span>    <span class="va">gds.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.Annot</span><span class="op">)</span></span>
<span></span>
<span>    </span>
<span>    <span class="co"># Path to the directory where the gds file for each sample</span></span>
<span>    <span class="co"># will be create</span></span>
<span>    <span class="va">PATHSAMPLEGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Need a data.frame with study description for the</span></span>
<span>    <span class="co"># synthetic data</span></span>
<span>    <span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"Id of the synthetic dataset"</span>,</span>
<span>                              study.desc <span class="op">=</span> <span class="st">"Descripttion"</span>,</span>
<span>                              study.platform <span class="op">=</span> <span class="st">"Platform string like Whole Exome"</span>,</span>
<span>                              stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Look at studyDF at step 3.</span></span>
<span>    <span class="va">study.id</span> <span class="op">&lt;-</span> <span class="st">"Id of the study"</span></span>
<span>    </span>
<span>    <span class="co"># The file ped defined before</span></span>
<span>    <span class="va">filePED</span> <span class="op">&lt;-</span> <span class="va">FILE_PED</span></span>
<span>    <span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span>    <span class="co"># You need a list of the sample Name.ID </span></span>
<span>    <span class="co"># you can get it from other source than </span></span>
<span>    <span class="co"># filePED </span></span>
<span>    <span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># You need a set of 1KG too use in the synthetic dataset</span></span>
<span>    <span class="co"># You need a list of id from 1KG get from file</span></span>
<span>    <span class="co"># or use the select1KGPop</span></span>
<span>    <span class="co"># The function select1KGPop select nbSamples </span></span>
<span>    <span class="co"># of each 1KG pop.group (subcontinental)</span></span>
<span>    <span class="co"># with the same seed the function generate the same set</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span></span>
<span>    <span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, nbSamples <span class="op">=</span> <span class="fl">30L</span><span class="op">)</span></span>
<span>    <span class="va">listSampleRef</span> <span class="op">&lt;-</span> <span class="va">dataRef</span><span class="op">$</span><span class="va">sample.id</span></span>
<span>    </span>
<span>    <span class="co"># Can be run in paralle or in different clusters...</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        </span>
<span>        <span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSAMPLEGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="../reference/prepSynthetic.html">prepSynthetic</a></span><span class="op">(</span><span class="va">file.GDSSample</span>,</span>
<span>              <span class="va">listSampleRef</span>,</span>
<span>              <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>              <span class="va">studyDF</span>,</span>
<span>              nbSim <span class="op">=</span> <span class="fl">1</span>,</span>
<span>              prefId <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co"># synthetic data generator</span></span>
<span>        <span class="va">resG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/syntheticGeno.html">syntheticGeno</a></span><span class="op">(</span><span class="va">gds</span>,</span>
<span>                              <span class="va">gds.Annot</span>,</span>
<span>                              <span class="va">file.GDSSample</span>,</span>
<span>                              <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                              <span class="va">listSampleRef</span>,</span>
<span>                              nbSim <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                              prefId <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds.Annot</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-7--compute-the-pca-knn-ancestry-call-for-each-synthetic-data">Step 7. Compute the PCA-KNN ancestry call for each synthetic
data<a class="anchor" aria-label="anchor" href="#step-7--compute-the-pca-knn-ancestry-call-for-each-synthetic-data"></a>
</h4>
<p>or in R</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span>    </span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="co"># Files related to the reference</span></span>
<span>    <span class="co"># Downloaded</span></span>
<span>    <span class="co">##########################################</span></span>
<span>    <span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span>    </span>
<span>    <span class="va">fileName.GDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span>    <span class="va">fileName.GDS.Annot</span> <span class="op">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span>    <span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span>    <span class="va">file.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.Annot</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">##########################################</span></span>
<span>    </span>
<span>    <span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span>    <span class="va">gds.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.Annot</span><span class="op">)</span></span>
<span></span>
<span>    </span>
<span>    <span class="co"># Path to the directory where the gds file for each sample</span></span>
<span>    <span class="co"># will be create</span></span>
<span>    <span class="va">PATHSAMPLEGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># directory which content directory name Name.ID in whihc the PCA is save</span></span>
<span>    <span class="co"># </span></span>
<span>    <span class="va">PATH_OUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">PATH_OUT</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">PATH_OUT</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co"># Need a data.frame with study description for the</span></span>
<span>    <span class="co"># synthetic data</span></span>
<span>    <span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"Id of the synthetic dataset"</span>,</span>
<span>                              study.desc <span class="op">=</span> <span class="st">"Descripttion"</span>,</span>
<span>                              study.platform <span class="op">=</span> <span class="st">"Platform string like Whole Exome"</span>,</span>
<span>                              stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Look at studyDF at step 3.</span></span>
<span>    <span class="va">study.id</span> <span class="op">&lt;-</span> <span class="st">"Id of the study"</span></span>
<span>    </span>
<span>    <span class="co"># The file ped defined before</span></span>
<span>    <span class="va">filePED</span> <span class="op">&lt;-</span> <span class="va">FILE_PED</span></span>
<span>    <span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span>    <span class="co"># You need a list of the sample Name.ID </span></span>
<span>    <span class="co"># you can get it from other source than </span></span>
<span>    <span class="co"># filePED </span></span>
<span>    <span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># Get the known superPop for the reference</span></span>
<span>    <span class="va">spRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRef1KGPop.html">getRef1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"superPop"</span><span class="op">)</span></span>
<span>    <span class="va">listSuperPop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EAS"</span>, <span class="st">"EUR"</span>, <span class="st">"AFR"</span>, <span class="st">"AMR"</span>, <span class="st">"SAS"</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># The set of 1KG use in the synthetic dataset</span></span>
<span>    <span class="co"># if you use select1KGPop the seed must be the same </span></span>
<span>    <span class="co"># as the step 6 </span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span></span>
<span>    <span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, nbSamples <span class="op">=</span> <span class="fl">30L</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Split dataRef by pop </span></span>
<span>    <span class="co">#</span></span>
<span>    <span class="va">sampleRM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/splitSelectByPop.html">splitSelectByPop</a></span><span class="op">(</span><span class="va">dataRef</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="co"># Can be run in paralle or in different clusters...</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        </span>
<span>        <span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSAMPLEGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="va">PATH_OUT_SAMPLE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_OUT</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">PATH_OUT_SAMPLE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">PATH_OUT_SAMPLE</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        </span>
<span>        <span class="va">gdsSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDSSample</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="co"># data.frame with the sample.id (Name.ID) with the study id</span></span>
<span>        </span>
<span>        <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sampleRM</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="va">pca1KG</span> <span class="op">&lt;-</span> <span class="fu">ccomputePCARefRMMulti</span><span class="op">(</span><span class="va">gdsSample</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">spRef</span><span class="op">)</span>,<span class="va">sampleRM</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">)</span></span>
<span>            <span class="va">resPCA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computePCAMultiSynthetic.html">computePCAMultiSynthetic</a></span><span class="op">(</span><span class="va">gdsSample</span>, <span class="va">pca1KG</span>,</span>
<span>                                               <span class="va">sampleRM</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>, <span class="st">"CRC.Synthetic"</span><span class="op">)</span></span>
<span>            </span>
<span>            </span>
<span></span>
<span>            <span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeKNNRefSynthetic.html">computeKNNRefSynthetic</a></span><span class="op">(</span><span class="va">gdsSample</span>, <span class="va">resPCA</span>, <span class="va">listSuperPop</span>,</span>
<span>                                               <span class="va">studyDF</span><span class="op">$</span><span class="va">study.id</span>, <span class="va">spRef</span><span class="op">)</span></span>
<span>            <span class="va">KNN.list</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">KNN.synt</span><span class="op">$</span><span class="va">matKNN</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="va">KNN.sample.syn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">KNN.list</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="va">pedSyn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepPedSynthetic1KG.html">prepPedSynthetic1KG</a></span><span class="op">(</span><span class="va">gds</span>, <span class="va">gdsSample</span>, <span class="va">studyDF</span><span class="op">$</span><span class="va">study.id</span>, <span class="st">"superPop"</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="va">fieldPopIn1KG</span> <span class="op">&lt;-</span> <span class="st">"superPop"</span></span>
<span>        <span class="va">fieldPopInfAnc</span> <span class="op">&lt;-</span> <span class="st">"SuperPop"</span></span>
<span>        <span class="va">listParaSample</span><span class="op">[[</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selParaPCAUpQuartile.html">selParaPCAUpQuartile</a></span><span class="op">(</span><span class="va">KNN.sample.syn</span>, <span class="va">pedSyn</span>, <span class="va">fieldPopIn1KG</span>, <span class="va">fieldPopInfAnc</span>, <span class="va">listSuperPop</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="va">listPCASample</span><span class="op">[[</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computePCARefSample.html">computePCARefSample</a></span><span class="op">(</span><span class="va">gdsSample</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, study.id.ref <span class="op">=</span> <span class="st">"Ref.1KG"</span><span class="op">)</span></span>
<span>        </span>
<span>        </span>
<span>        <span class="va">listKNNSample</span><span class="op">[[</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeKNNSuperPopSample.html">computeKNNSuperPopSample</a></span><span class="op">(</span><span class="va">gdsSample</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">spRef</span><span class="op">)</span></span>
<span></span>
<span>        </span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsSample</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds.Annot</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-8--select-paramters-for-the-pca-knn-ancestry-call">Step 8. Select paramters for the PCA-KNN ancestry call<a class="anchor" aria-label="anchor" href="#step-8--select-paramters-for-the-pca-knn-ancestry-call"></a>
</h4>
<p>or in R</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(RAIDS)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="do">##########################################</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Files related to the reference</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Downloaded</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="do">##########################################</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    PATH_1KG <span class="ot">&lt;-</span> PATH_REF</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    fileName.GDS <span class="ot">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    fileName.GDS.Annot <span class="ot">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    file.GDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_1KG, fileName.GDS)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    file.Annot <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_1KG, fileName.GDS.Annot)</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    <span class="do">##########################################</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    gds <span class="ot">&lt;-</span> <span class="fu">openfn.gds</span>(file.GDS)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>    gds.Annot <span class="ot">&lt;-</span> <span class="fu">openfn.gds</span>(file.Annot)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Path to the directory where the gds file for each sample</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>    <span class="co"># will be create</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    PATHSAMPLEGDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"The path"</span>)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># directory which content directory name Name.ID in whihc the PCA is save</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>    <span class="co"># </span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>    PATH_OUT <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"The path"</span>)</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Need a data.frame with study description for the</span></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># synthetic data</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>    studyDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">study.id =</span> <span class="st">"Id of the synthetic dataset"</span>,</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>                              <span class="at">study.desc =</span> <span class="st">"Descripttion"</span>,</span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>                              <span class="at">study.platform =</span> <span class="st">"Platform string like Whole Exome"</span>,</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>                              <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Look at studyDF at step 3.</span></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>    study.id <span class="ot">&lt;-</span> <span class="st">"Id of the study"</span></span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>    <span class="co"># The file ped defined before</span></span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>    filePED <span class="ot">&lt;-</span> FILE_PED</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>    ped <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(filePED)</span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># You need a list of the sample Name.ID </span></span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># you can get it from other source than </span></span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># filePED </span></span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>    listSamples <span class="ot">&lt;-</span> ped[, <span class="st">"Name.ID"</span>]</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>    spRef <span class="ot">&lt;-</span> <span class="fu">getRef1KGPop</span>(gds, <span class="st">"superPop"</span>)</span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Can be run in paralle or in different clusters...</span></span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a>    listParaSample <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>    listPCASample <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a>    listKNNSample <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a>    listAncestrySample <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">length</span>(listSamples))){</span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>        file.GDSSample <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATHSAMPLEGDS, <span class="fu">paste0</span>(listSamples[i], <span class="st">".gds"</span>))</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>        gdsSample <span class="ot">&lt;-</span> <span class="fu">snpgdsOpen</span>(file.GDSSample)</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>        PATH_OUT_SAMPLE <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_OUT, listSamples[i])</span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a>        <span class="co"># The directory PATH_OUT_SAMPLE must just content </span></span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a>        <span class="co"># only the rds files from step 7</span></span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a>        listSynthetic <span class="ot">&lt;-</span> <span class="fu">dir</span>(PATH_OUT_SAMPLE, <span class="st">".rds"</span>)</span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a>        listKNN <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">length</span>(listSynthetic))){</span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a>            tmp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(PATHSYN, listSynthetic[i]))</span>
<span id="cb8-71"><a href="#cb8-71" aria-hidden="true" tabindex="-1"></a>            listKNN[[i]] <span class="ot">&lt;-</span> tmp<span class="sc">$</span>resKNN<span class="sc">$</span>matKNN</span>
<span id="cb8-72"><a href="#cb8-72" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb8-73"><a href="#cb8-73" aria-hidden="true" tabindex="-1"></a>        matKNN.All <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, listKNN)</span>
<span id="cb8-74"><a href="#cb8-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-75"><a href="#cb8-75" aria-hidden="true" tabindex="-1"></a>        <span class="co"># The parameter superPop is define by the output of step 7</span></span>
<span id="cb8-76"><a href="#cb8-76" aria-hidden="true" tabindex="-1"></a>        <span class="co"># the function return data.frame with the known superPop </span></span>
<span id="cb8-77"><a href="#cb8-77" aria-hidden="true" tabindex="-1"></a>        <span class="co"># associate with synthetic data.</span></span>
<span id="cb8-78"><a href="#cb8-78" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-79"><a href="#cb8-79" aria-hidden="true" tabindex="-1"></a>        pedSyn <span class="ot">&lt;-</span> <span class="fu">prepPedSynthetic1KG</span>(gds, gdsSample, studyDF<span class="sc">$</span>study.id, <span class="st">"superPop"</span>)</span>
<span id="cb8-80"><a href="#cb8-80" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-81"><a href="#cb8-81" aria-hidden="true" tabindex="-1"></a>        <span class="co"># list of superPop in synthetic dataset</span></span>
<span id="cb8-82"><a href="#cb8-82" aria-hidden="true" tabindex="-1"></a>        listSP <span class="ot">&lt;-</span> <span class="fu">unique</span>(pedSyn<span class="sc">$</span>superPop)</span>
<span id="cb8-83"><a href="#cb8-83" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-84"><a href="#cb8-84" aria-hidden="true" tabindex="-1"></a>        <span class="co"># The SuperPop is the super population field in matKNN.All</span></span>
<span id="cb8-85"><a href="#cb8-85" aria-hidden="true" tabindex="-1"></a>        <span class="co"># superPop are the known super population from the ref</span></span>
<span id="cb8-86"><a href="#cb8-86" aria-hidden="true" tabindex="-1"></a>        <span class="co"># return the parameter selected.</span></span>
<span id="cb8-87"><a href="#cb8-87" aria-hidden="true" tabindex="-1"></a>        listParaSample[[listSamples[i]]] <span class="ot">&lt;-</span> <span class="fu">selParaPCAUpQuartile</span>(matKNN.All, pedSyn, <span class="st">"superPop"</span>, <span class="st">"SuperPop"</span>, listSP)</span>
<span id="cb8-88"><a href="#cb8-88" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-89"><a href="#cb8-89" aria-hidden="true" tabindex="-1"></a>        listPCASample[[listSamples[i]]] <span class="ot">&lt;-</span> <span class="fu">computePCARefSample</span>(gdsSample, listSamples[i], <span class="at">study.id.ref =</span> <span class="st">"Ref.1KG"</span>)</span>
<span id="cb8-90"><a href="#cb8-90" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-91"><a href="#cb8-91" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-92"><a href="#cb8-92" aria-hidden="true" tabindex="-1"></a>        listKNNSample[[listSamples[i]]] <span class="ot">&lt;-</span> <span class="fu">computeKNNSuperPopSample</span>(gdsSample, listSamples[i], spRef)</span>
<span id="cb8-93"><a href="#cb8-93" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-94"><a href="#cb8-94" aria-hidden="true" tabindex="-1"></a>        listAncestrySample[[listSamples[i]]] <span class="ot">&lt;-</span> </span>
<span id="cb8-95"><a href="#cb8-95" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb8-96"><a href="#cb8-96" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-97"><a href="#cb8-97" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-98"><a href="#cb8-98" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-99"><a href="#cb8-99" aria-hidden="true" tabindex="-1"></a>    <span class="fu">closefn.gds</span>(gds)</span>
<span id="cb8-100"><a href="#cb8-100" aria-hidden="true" tabindex="-1"></a>    <span class="fu">closefn.gds</span>(gds.Annot)</span></code></pre></div>
<p><br><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="pre-processed-files-are-available">Pre-processed files are available<a class="anchor" aria-label="anchor" href="#pre-processed-files-are-available"></a>
</h2>
<p>Pre-processed files, such as the 1KG GDS file, are available at this
address:</p>
<p><a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a></p>
<p>Beware that some of those files are voluminous.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.1.3 (2022-03-10)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] RAIDS_0.50.0     GENESIS_2.24.2   SNPRelate_1.28.0 gdsfmt_1.30.0   </span></span>
<span><span class="co">## [5] knitr_1.39       BiocStyle_2.22.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] nlme_3.1-159           bitops_1.0-7           fs_1.5.2              </span></span>
<span><span class="co">##  [4] bit64_4.0.5            rprojroot_2.0.3        GenomeInfoDb_1.30.1   </span></span>
<span><span class="co">##  [7] tools_4.1.3            backports_1.4.1        bslib_0.4.0           </span></span>
<span><span class="co">## [10] utf8_1.2.2             R6_2.5.1               DBI_1.1.3             </span></span>
<span><span class="co">## [13] BiocGenerics_0.40.0    mgcv_1.8-40            DNAcopy_1.68.0        </span></span>
<span><span class="co">## [16] withr_2.5.0            tidyselect_1.1.2       bit_4.0.4             </span></span>
<span><span class="co">## [19] compiler_4.1.3         textshaping_0.3.6      cli_3.3.0             </span></span>
<span><span class="co">## [22] quantreg_5.94          Biobase_2.54.0         mice_3.14.0           </span></span>
<span><span class="co">## [25] SparseM_1.81           desc_1.4.1             sandwich_3.0-2        </span></span>
<span><span class="co">## [28] bookdown_0.28          sass_0.4.2             lmtest_0.9-40         </span></span>
<span><span class="co">## [31] quantsmooth_1.60.0     pkgdown_2.0.6          systemfonts_1.0.4     </span></span>
<span><span class="co">## [34] stringr_1.4.0          digest_0.6.29          GWASExactHW_1.01      </span></span>
<span><span class="co">## [37] rmarkdown_2.14         XVector_0.34.0         pkgconfig_2.0.3       </span></span>
<span><span class="co">## [40] htmltools_0.5.3        highr_0.9              fastmap_1.1.0         </span></span>
<span><span class="co">## [43] rlang_1.0.4            RSQLite_2.2.15         jquerylib_0.1.4       </span></span>
<span><span class="co">## [46] generics_0.1.3         zoo_1.8-10             jsonlite_1.8.0        </span></span>
<span><span class="co">## [49] BiocParallel_1.28.3    dplyr_1.0.9            RCurl_1.98-1.8        </span></span>
<span><span class="co">## [52] magrittr_2.0.3         GenomeInfoDbData_1.2.7 Matrix_1.4-1          </span></span>
<span><span class="co">## [55] Rcpp_1.0.9             S4Vectors_0.32.4       fansi_1.0.3           </span></span>
<span><span class="co">## [58] lifecycle_1.0.1        pROC_1.18.0            stringi_1.7.8         </span></span>
<span><span class="co">## [61] yaml_2.3.5             GWASTools_1.40.0       zlibbioc_1.40.0       </span></span>
<span><span class="co">## [64] MASS_7.3-58.1          plyr_1.8.7             grid_4.1.3            </span></span>
<span><span class="co">## [67] formula.tools_1.7.1    blob_1.2.3             parallel_4.1.3        </span></span>
<span><span class="co">## [70] crayon_1.5.1           lattice_0.20-45        Biostrings_2.62.0     </span></span>
<span><span class="co">## [73] splines_4.1.3          pillar_1.8.0           GenomicRanges_1.46.1  </span></span>
<span><span class="co">## [76] logistf_1.24.1         stats4_4.1.3           glue_1.6.2            </span></span>
<span><span class="co">## [79] evaluate_0.16          data.table_1.14.2      BiocManager_1.30.18   </span></span>
<span><span class="co">## [82] operator.tools_1.6.3   vctrs_0.4.1            MatrixModels_0.5-0    </span></span>
<span><span class="co">## [85] purrr_0.3.4            tidyr_1.2.0            SeqArray_1.34.0       </span></span>
<span><span class="co">## [88] assertthat_0.2.1       cachem_1.0.6           xfun_0.32             </span></span>
<span><span class="co">## [91] broom_1.0.0            SeqVarTools_1.32.0     class_7.3-20          </span></span>
<span><span class="co">## [94] ragg_1.2.2             survival_3.4-0         tibble_3.1.8          </span></span>
<span><span class="co">## [97] memoise_2.0.1          IRanges_2.28.0</span></span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Alexander2009" class="csl-entry">
Alexander, D. H., J. Novembre, and K. Lange. 2009. <span>“Fast
Model-Based Estimation of Ancestry in Unrelated Individuals.”</span>
<em>Genome Res</em> 19 (9): 1655–64. <a href="https://doi.org/10.1101/gr.094052.109" class="external-link">https://doi.org/10.1101/gr.094052.109</a>.
</div>
<div id="ref-Price2006" class="csl-entry">
Price, A. L., N. J. Patterson, R. M. Plenge, M. E. Weinblatt, N. A.
Shadick, and D. Reich. 2006. <span>“Principal Components Analysis
Corrects for Stratification in Genome-Wide Association Studies.”</span>
<em>Nat Genet</em> 38 (8): 904–9. <a href="https://doi.org/10.1038/ng1847" class="external-link">https://doi.org/10.1038/ng1847</a>.
</div>
<div id="ref-Pritchard2000" class="csl-entry">
Pritchard, J. K., M. Stephens, and P. Donnelly. 2000. <span>“Inference
of Population Structure Using Multilocus Genotype Data.”</span>
<em>Genetics</em> 155 (2): 945–59. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10835412" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/10835412</a>.
</div>
<div id="ref-Shen2016" class="csl-entry">
Shen, R, and V E Seshan. 2016. <span>“<span class="nocase">FACETS:
allele-specific copy number and clonal heterogeneity analysis tool for
high-throughput DNA sequencing</span>.”</span> <em>Nucleic Acids
Res</em> 44 (16): e131. <a href="https://doi.org/10.1093/nar/gkw520" class="external-link">https://doi.org/10.1093/nar/gkw520</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
