<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate Inference of Genetic Ancestry from Cancer Sequencing • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate Inference of Genetic Ancestry from Cancer Sequencing">
<meta property="og:description" content="RAIDS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from 1000 Genomes (optional)</a>
    </li>
    <li>
      <a href="../articles/aicsPaper.html">Accurate Inference of Genetic Ancestry from Cancer Sequencing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate Inference of Genetic Ancestry from
Cancer Sequencing</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/HEAD/vignettes/aicsPaper.Rmd" class="external-link"><code>vignettes/aicsPaper.Rmd</code></a></small>
      <div class="hidden name"><code>aicsPaper.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 0.50.0<br><strong>Compiled date</strong>: 2022-09-14<br><strong>License</strong>: Apache License (== 2.0) | file
LICENSE<br></p>
<div class="section level2">
<h2 id="licensing">Licensing<a class="anchor" aria-label="anchor" href="#licensing"></a>
</h2>
<p>The <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em>
package and the underlying <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em> code are
distributed under<br>
the <a href="https://opensource.org/licenses/Apache-2.0" class="external-link uri">https://opensource.org/licenses/Apache-2.0</a> license. You
are free to use and redistribute this software.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>If you use the <strong>RAIDS</strong> package for a publication, we
would ask you to cite the following:</p>
<blockquote>
<p>Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander
Krasnitz. Accurate and robust inference of genetic ancestry from
cancer-derived molecular data across genomic platforms. bioRxiv
2022.02.01.478737; doi: <a href="https://doi.org/10.1101/2022.02.01.478737" class="external-link uri">https://doi.org/10.1101/2022.02.01.478737</a></p>
</blockquote>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple methods have been implemented to infer ancestry from
germline DNA sequence <span class="citation">(Price et al. 2006;
Pritchard, Stephens, and Donnelly 2000; Alexander, Novembre, and Lange
2009)</span>. However, genotyping of DNA from matched normal specimens
is not part of standard clinical practice and is not performed routinely
outside academic clinical centers. In sum, matched germline DNA sequence
is often missing for cancer-derived molecular data. In such cases,
having the possibility to infer ancestry from tumor-derived data would
be beneficial.</p>
<p>The <strong>RAIDS</strong> package implements an inference procedure
that has been specifically developed to accurately infer genetic
ancestry from cancer-derived sequencing. The covered cancer-derived
sequencing are, more specifically, tumor exomes, targeted gene panels
and RNA sequences.</p>
<p>The <strong>RAIDS</strong> package includes a data synthesis method
that enables the selection of the best parameters for a specific cancer
profile and also permits a statistically assessment of inference
accuracy for an individual cancer-derived molecular profile.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version accessible on the <a href="https://github.com/KrasnitzLab/RAIDS" title="RAIDS Github Site" class="external-link">RAIDS Github Website</a>, the <em><a href="https://CRAN.R-project.org/package=devtools" class="external-link">devtools</a></em>
package is required.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Install the latest version of RAIDS</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'KrasnitzLab/RAIDS'</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="main-steps">Main Steps<a class="anchor" aria-label="anchor" href="#main-steps"></a>
</h2>
<p>This is an overview of genetic ancestry inference from cancer-derived
molecular data:</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_v01.png" alt="An overview of the genetic ancestry inference process." width="120%"><p class="caption">
An overview of the genetic ancestry inference process.
</p>
</div>
<p>The main steps are:</p>
<ol style="list-style-type: decimal">
<li>Format the information from 1000 Genomes (1KG) (optional)</li>
<li>Format the information from an external study</li>
<li>Find the optimized parameters for the ancestry inference</li>
<li>Run the ancestry inference on the external study</li>
</ol>
<p>In the following sections, those main steps are described in
details.</p>
<p><br><br></p>
<div class="section level3">
<h3 id="step-1---format-the-information-from-1000-genomes-optional">Step 1 - Format the information from 1000 Genomes (optional)<a class="anchor" aria-label="anchor" href="#step-1---format-the-information-from-1000-genomes-optional"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step1_v01.png" alt="Step 1 - Formatting the information from 1000 Genomes (optional)" width="120%"><p class="caption">
Step 1 - Formatting the information from 1000 Genomes (optional)
</p>
</div>
<hr>
<p>This step generates 3 important files:</p>
<ul>
<li>The 1KG GDS File</li>
<li>The SNV Annotation GDS file</li>
<li>The SNV Retained VCF file</li>
</ul>
<p><span style="color:darkred;font-weight:bold">Beware that the 3
pre-processed files are available at this address:</span></p>
<p><span style="color:red"> <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a>
</span></p>
<p><span style="color:darkred;font-weight:bold">The size of the 1KG GDS
file is 15GB.</span></p>
<p><span style="color:darkred;font-weight:bold">The 1KG GDS file is
mapped on hg38 <span class="citation">(Lowy-Gallego et al.
2019)</span>.</span></p>
<p><span style="color:darkred;font-weight:bold">This section can be
skipped if you choose to use the pre-processed files.</span></p>
<p>The vignette <a href="Create_1KG_GDS_file.html">Formatting the
information from 1000 Genomes (optional)</a> is dedicated to this
section.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step-2---format-information-from-an-external-study-to-get-it-ready-for-ancestry-inference">Step 2 - Format information from an external study to get it ready
for ancestry inference<a class="anchor" aria-label="anchor" href="#step-2---format-information-from-an-external-study-to-get-it-ready-for-ancestry-inference"></a>
</h3>
<p>To be able to run the ancestry inference on samples from an external
study, the information about the samples needs to follow a specific
format.</p>
<p>The preparation of the external study data is divided is
sub-steps.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step2_v01.png" alt="Step 2 - Formatting the information from an external study" width="120%"><p class="caption">
Step 2 - Formatting the information from an external study
</p>
</div>
<p>The required sub-steps are:</p>
<ol style="list-style-type: decimal">
<li>Create a directory containing the 3 required reference files</li>
<li>Generate a Sample SNP pileup file (1 file per sample)</li>
<li>Create a RDS file containing information about the samples</li>
<li>Create a GDS Sample file (1 GDS file per sample)</li>
<li>Generate a pruned subset of the single nucleotide variants
(SNVs)</li>
<li>Estimate the allelic frequency for the pruned SNVs</li>
</ol>
<p>Beware that a mapped BAM file is needed for each sample. The
reference genome used for the mapping must correspond to the one used to
generate the 1KG GDS file. In this case, the 1KG GDS file is based on
hg38 genome.</p>
<p><br></p>
<div class="section level4">
<h4 id="sub-step-1--create-a-directory-containing-the-3-required-reference-files">Sub-Step 1. Create a directory containing the 3 required reference
files<a class="anchor" aria-label="anchor" href="#sub-step-1--create-a-directory-containing-the-3-required-reference-files"></a>
</h4>
<p>The 3 required reference files can be downloaded:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">####################################</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="do">## The 1KG GDS file</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="do">####################################</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>wget https<span class="sc">:</span><span class="er">//</span>labshare.cshl.edu<span class="sc">/</span>shares<span class="sc">/</span>krasnitzlab<span class="sc">/</span>aicsPaper<span class="sc">/</span>matGeno1000g.gds</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="do">####################################</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="do">## The 1KG SNV Annotation GDS file</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="do">####################################</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>wget https<span class="sc">:</span><span class="er">//</span>labshare.cshl.edu<span class="sc">/</span>shares<span class="sc">/</span>krasnitzlab<span class="sc">/</span>aicsPaper<span class="sc">/</span>matAnnot1000g.gds</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="do">####################################</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="do">## The 1KG SNV Retained VCF file</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="do">####################################</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>wget https<span class="sc">:</span><span class="er">//</span>labshare.cshl.edu<span class="sc">/</span>shares<span class="sc">/</span>krasnitzlab<span class="sc">/</span>aicsPaper<span class="sc">/</span>snvSel0.<span class="fl">01.</span>vcf.gz</span></code></pre></div>
<p>The 3 files should be stored in the same directory. For the current
demonstration, the directory will be refered as
<strong>PATH_REF</strong>.</p>
<p>For more information on how to generate your own reference files, you
can refer to the vignette <a href="Create_1KG_GDS_file.html">Formatting
the information from 1000 Genomes (optional)</a>.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-2--generate-a-sample-snp-pileup-file-1-file-per-sample">Sub-Step 2. Generate a Sample SNP pileup file (1 file per
sample)<a class="anchor" aria-label="anchor" href="#sub-step-2--generate-a-sample-snp-pileup-file-1-file-per-sample"></a>
</h4>
<p>This step requires the installation of the external code <a href="https://github.com/mskcc/facets/tree/master/inst/extcode" class="external-link">snp-pileup</a>
which is associated to the facets package <span class="citation">(Shen
and Seshan 2016)</span>. Given a VCF file containing SNP locations, <a href="https://github.com/mskcc/facets/tree/master/inst/extcode" class="external-link">snp-pileup</a>
application outputs, for each location, the counts of the reference and
alternative nucleotide from an BAM input.</p>
<p>At this step, <a href="https://github.com/mskcc/facets/tree/master/inst/extcode" class="external-link">snp-pileup</a>
is used to generate a Sample SNP pileup file that will contain the
reference and alternative nucleotides counts at each SNP location
present in the reference 1KG GDS file. The 1KG VCF file, which contains
the list of retained 1KG SNVs, is required.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step2_SubStep2_SNP-Pileup_v01.mod.png" alt="Generate a Sample SNP pileup file (1 file per sample)" width="100%"><p class="caption">
Generate a Sample SNP pileup file (1 file per sample)
</p>
</div>
<p>Beware that the name assigned to the Sample SNP pileup file will
correspond to the Sample identifier (Name.ID) in the following analysis.
For example, a SNP pileupe file called “Sample.01.txt.gz” would be
associated to the sample “Sample.01”.</p>
<p>This is the command line to run <a href="https://github.com/mskcc/facets/tree/master/inst/extcode" class="external-link">snp-pileup</a>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">#####################################################################</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Description of the parameters</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="do">## -g : Compresses the output file with BGZF</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">## -d5000 : Sets the maximum depth to 5000</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="do">## -q15 : Sets the minimum threshold for mapping quality to 15</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="do">## -Q20 : Sets the minimum threshold for base quality to 20</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="do">## -r0 : Sets the minimum read counts for a position to be output to 0</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="do">## PATH_REF/snvSel0.01.vcf.gz : The SNP Retained VCF file containing the </span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="do">##     positions of all retained 1KG SNPs</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="do">## PATH_OUT/Name.ID.txt : The name of the output Sample VCF file that will be</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="do">##     compressed by the application</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="do">## FILEBAM.bam : The aligned reads from the sample used as input</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="do">#####################################################################</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>snp<span class="sc">-</span>pileup <span class="sc">-</span>g <span class="sc">-</span>d5000 <span class="sc">-</span>q15 <span class="sc">-</span>Q20  <span class="sc">-</span>r0 PATH_REF<span class="sc">/</span>snvSel0.<span class="fl">01.</span>vcf.gz PATH_OUT<span class="sc">/</span>Name.ID.txt FILEBAM.bam</span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-3--create-a-rds-file-containing-the-information-about-the-samples">Sub-Step 3. Create a RDS file containing the information about the
samples<a class="anchor" aria-label="anchor" href="#sub-step-3--create-a-rds-file-containing-the-information-about-the-samples"></a>
</h4>
<p>A RDS file describing the samples that are part of the analysis is
required.</p>
<p>The RDS file must contain a <em>data.frame</em> with those 5
columns:</p>
<ul>
<li>
<strong>Name.ID</strong>: The unique sample identifier. The GDS
Sample file associated to this sample should be called
“Name.ID.txt.gz”.</li>
<li>
<strong>Case.ID</strong>: The patient identifier associated to the
sample.</li>
<li>
<strong>Sample.Type</strong>: The information about the sample type
(primary tumor, metastasis, normal, etc..).</li>
<li>
<strong>Diagnosis</strong>: The diagnosis associated to the
patient.</li>
<li>
<strong>Source</strong>: The source of the sample (example:
TCGA-BRCA).</li>
</ul>
<p>This file is referred as the PED RDS file (PED for pedigree).
Alternatively, the PED information can be saved in another type of file
(CVS, etc..) as long as the data.frame information can be regenerated in
R (with <em>read.csv()</em> or else).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##############################################################</span></span>
<span><span class="co">## Location of the new PED RDS file</span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="va">data.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span><span class="va">demo_PED_File</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data.dir</span>, <span class="st">"Demo_PED.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="co">## Create a data frame with the mandatory columns</span></span>
<span><span class="co">## All columns are in character string format (no factor)</span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="va">pedDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Name.ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample_01"</span>, <span class="st">"Sample_02"</span>, <span class="st">"Sample_03"</span><span class="op">)</span>,</span>
<span>                    Case.ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient_h11"</span>, <span class="st">"Patient_h12"</span>, <span class="st">"Patient_h18"</span><span class="op">)</span>, </span>
<span>                    Sample.Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Primary Tumor"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                    Diagnosis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Cancer"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                    Source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Databank B"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="co">## Save the data frame into a RDS file</span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span>object<span class="op">=</span><span class="va">pedDF</span>, file<span class="op">=</span><span class="va">demo_PED_File</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Remove RDS file (created for demo purpose)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">demo_PED_File</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-4--create-a-gds-sample-file-1-gds-file-per-sample">Sub-Step 4. Create a GDS Sample file (1 GDS file per sample)<a class="anchor" aria-label="anchor" href="#sub-step-4--create-a-gds-sample-file-1-gds-file-per-sample"></a>
</h4>
<p>This step requires 4 files as input:</p>
<ul>
<li>The 1KG GDS file</li>
<li>The Sample SNP pileup file (one per sample present in the
study)</li>
<li>The PED RDS file</li>
</ul>
<p>A <em>data.frame</em> containing the general information about the
study is also required. The <em>data.frame</em> must contain those 3
columns:</p>
<ul>
<li>
<strong>study.id</strong>: The study identifier (example:
TCGA-BRCA).</li>
<li>
<strong>study.desc</strong>: The description of the study.</li>
<li>
<strong>study.platform</strong>: The type of sequencing (example:
RNA-seq).</li>
</ul>
<p>Using all those inputs, the <em>createStudy2GDS1KG()</em> function
will generate a GDS Sample file. One GDS Sample file is created for each
sample passed to the <em>listSamples</em> argument.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Load required package</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RAIDS)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="do">## The 1KG GDS file and the 1KG SNV Annotation GDS file </span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="do">## need to be located in the same directory</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>PATH_1KG <span class="ot">&lt;-</span> PATH_REF</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>fileName.GDS <span class="ot">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>file.GDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_1KG, fileName.GDS)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="do">## The Sample SNP pileup files (one per sample) need  </span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="do">## to be located in the same directory.</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>PATHGENO <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_SAMPLE_SNP_PILEUP)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="do">## The path where the Sample GDS files (one per sample)</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="do">## will be created need to be specified.</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>PATHSAMPLEGDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_SAMPLE_GDS_FILES)</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="do">## The path and file name for the PED RDS file </span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="do">## will the information about the analyzed samples</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>filePED <span class="ot">&lt;-</span> FILE_PED</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>ped <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(filePED)</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a><span class="do">## A data frame containing general information about the study</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a><span class="do">## is also required. The data frame must have</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="do">## those 3 columns: "study.id", "study.desc", "study.platform"</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>studyDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">study.id =</span> <span class="st">"Id of the study"</span>,</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>                  <span class="at">study.desc =</span> <span class="st">"Description"</span>,</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>                  <span class="at">study.platform =</span> <span class="st">"Whole-Exome"</span>,</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>                  <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="do">## The list of samples to analyzed is passed to the</span></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a><span class="do">## function. The samples must be present in the PED RDS file</span></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a><span class="do">## and must have an associated Sample SNP pileup file.</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a><span class="do">## Not all samples present in the PED file need to be selected.</span></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>listSamples <span class="ot">&lt;-</span> ped[, <span class="st">"Name.ID"</span>]</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="do">## This function creates one GDS Sample file for each</span></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a><span class="do">## sample present in the 'listSamples' parameter.</span></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a><span class="fu">createStudy2GDS1KG</span>(<span class="at">PATHGENO =</span> PATHGENO,</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>                      <span class="at">pedStudy =</span> ped,</span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>                      <span class="at">fileNameGDS =</span> file.GDS,</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a>                      <span class="at">listSamples =</span> listSamples,</span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>                      <span class="at">studyDF =</span> studyDF,</span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>                      <span class="at">PATHSAMPLEGDS =</span> PATHSAMPLEGDS)<span class="er">))</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-5--generate-a-pruned-subset-of-the-single-nucleotide-variants-snvs">Sub-Step 5. Generate a pruned subset of the single nucleotide
variants (SNVs)<a class="anchor" aria-label="anchor" href="#sub-step-5--generate-a-pruned-subset-of-the-single-nucleotide-variants-snvs"></a>
</h4>
<p>The initial list of 1KG SNVs is pruned, using linkage disequilibrium
analysis, to retained as subset of SNVs of interest for each sample. The
list of selected SNVs is specific to each sample and the information
about the retained SNVs is added to the GDS Sample file.</p>
<p>The <strong>pruningSample()</strong> function requires the 1KG GDS
file as input. It also requires the path to the GDS Sample files.</p>
<p>Beware that this step can requisite large disk space.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Load required package</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RAIDS)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="do">## The 1KG GDS file is required </span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>PATH_1KG <span class="ot">&lt;-</span> PATH_REF</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>fileName.GDS<span class="fl">.1</span>KG <span class="ot">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>file.GDS <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATH_1KG, fileName.GDS<span class="fl">.1</span>KG)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Open the 1KG GDS file</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>gds_1KG <span class="ot">&lt;-</span> <span class="fu">snpgdsOpen</span>(file.GDS)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="do">## All GDS Sample files need to be located in the same </span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="do">## directory.</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>PATHSAMPLEGDS <span class="ot">&lt;-</span> PATH_WHERE_THE_GDS_SAMPLES_FILES_ARE_LOCATED</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="do">## The list of samples to analyse as obtained by the </span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="do">## PED RDS file. The samples must have an </span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="do">## associated GDS Sample file.</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>filePED <span class="ot">&lt;-</span> FILE_PED</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>ped <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(filePED)</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>listSamples <span class="ot">&lt;-</span> ped[, <span class="st">"Name.ID"</span>]</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a><span class="do">## The study information passed to the function must correspond</span></span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a><span class="do">## the information present in the previously defined study</span></span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a><span class="do">## data.frame (see previous step).</span></span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a><span class="do">#################################################################</span></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>studyDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">study.id =</span> <span class="st">"Id of the study"</span>,  <span class="do">## IMPORTANT INFORMATION</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>                  <span class="at">study.desc =</span> <span class="st">"Description"</span>,</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>                  <span class="at">study.platform =</span> <span class="st">"Whole-Exome"</span>,</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>                  <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a><span class="do">## The pruning function is called on one sample at the time</span></span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a><span class="do">##############################################################</span></span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">length</span>(listSamples))) {</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Compute the SNV pruned subset</span></span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true" tabindex="-1"></a>  <span class="do">## study.id: Study identifier as defined in the previous step.</span></span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true" tabindex="-1"></a>  <span class="do">##           The study identifier must be the same that the one present in</span></span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true" tabindex="-1"></a>  <span class="do">##           GDS Sample file.</span></span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PATHSAMPLEGDS: All GDS Sample files must be in the same directory.</span></span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pruningSample</span>(<span class="at">gds=</span>gds_1KG,</span>
<span id="cb6-52"><a href="#cb6-52" aria-hidden="true" tabindex="-1"></a>          <span class="at">sampleCurrent =</span> listSamples[i],</span>
<span id="cb6-53"><a href="#cb6-53" aria-hidden="true" tabindex="-1"></a>          <span class="at">study.id =</span> studyDF<span class="sc">$</span>study.id, </span>
<span id="cb6-54"><a href="#cb6-54" aria-hidden="true" tabindex="-1"></a>          <span class="at">PATHSAMPLEGDS =</span> PATHSAMPLEGDS)</span>
<span id="cb6-55"><a href="#cb6-55" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-56"><a href="#cb6-56" aria-hidden="true" tabindex="-1"></a>  <span class="do">## GDS Sample file for the current sample</span></span>
<span id="cb6-57"><a href="#cb6-57" aria-hidden="true" tabindex="-1"></a>  <span class="do">## The file name corresponds to the path + sample identifier + ".gds"</span></span>
<span id="cb6-58"><a href="#cb6-58" aria-hidden="true" tabindex="-1"></a>  file.GDSSample <span class="ot">&lt;-</span> <span class="fu">file.path</span>(PATHSAMPLEGDS, <span class="fu">paste0</span>(listSamples[i], <span class="st">".gds"</span>)</span>
<span id="cb6-59"><a href="#cb6-59" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb6-60"><a href="#cb6-60" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Add the genotype information for the list of pruned SNVs </span></span>
<span id="cb6-61"><a href="#cb6-61" aria-hidden="true" tabindex="-1"></a>  <span class="do">## into the GDS Sample file</span></span>
<span id="cb6-62"><a href="#cb6-62" aria-hidden="true" tabindex="-1"></a>  <span class="do">## The genotype information is extracted from the 1KG GDS file</span></span>
<span id="cb6-63"><a href="#cb6-63" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add1KG2SampleGDS</span>(<span class="at">gds =</span> gds_1KG, <span class="at">gdsSampleFile =</span> file.GDSSample, </span>
<span id="cb6-64"><a href="#cb6-64" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sampleCurrent =</span> listSamples[<span class="dv">1</span>], </span>
<span id="cb6-65"><a href="#cb6-65" aria-hidden="true" tabindex="-1"></a>                      <span class="at">study.id =</span> studyDF<span class="sc">$</span>study.id))</span>
<span id="cb6-66"><a href="#cb6-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-67"><a href="#cb6-67" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Add annotation from the 1KG GDS file to the GDS Sample file</span></span>
<span id="cb6-68"><a href="#cb6-68" aria-hidden="true" tabindex="-1"></a>  <span class="do">## This is required.</span></span>
<span id="cb6-69"><a href="#cb6-69" aria-hidden="true" tabindex="-1"></a>  <span class="fu">addStudy1Kg</span>(gds_1KG, file.GDSSample)</span>
<span id="cb6-70"><a href="#cb6-70" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-71"><a href="#cb6-71" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-72"><a href="#cb6-72" aria-hidden="true" tabindex="-1"></a><span class="do">## Close the 1KG GDS file (it is important to always close the GDS files)</span></span>
<span id="cb6-73"><a href="#cb6-73" aria-hidden="true" tabindex="-1"></a><span class="fu">closefn.gds</span>(gds_1KG)</span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-6--estimate-the-allelic-fraction-for-the-pruned-snvs">Sub-Step 6. Estimate the allelic fraction for the pruned SNVs<a class="anchor" aria-label="anchor" href="#sub-step-6--estimate-the-allelic-fraction-for-the-pruned-snvs"></a>
</h4>
<p>The <strong>estimateAllelicFraction()</strong> estimates the allelic
fraction for all SNVs present in the pruned SNV dataset. Beware that the
function requires different inputs for DNA and RNA samples.</p>
<p>For the DNA samples, these 2 files are required:</p>
<ol style="list-style-type: decimal">
<li>The GDS Sample file</li>
<li>The 1KG GDS file</li>
</ol>
<p>For the RNA samples, these 3 files are required:</p>
<ol style="list-style-type: decimal">
<li>The GDS Sample file</li>
<li>The 1KG GDS file</li>
<li>The 1KG SNv Annotation GDS file</li>
</ol>
<p>In both cases, the other required inputs are:</p>
<ol style="list-style-type: decimal">
<li>The information about the length of the chromosomes</li>
<li>The sample identifier (it corresponds to the GDS Sample file
name)</li>
<li>The study identifier (it should correspond to the one used
previously)</li>
</ol>
<p>The <strong>estimateAllelicFraction()</strong> function processes one
sample at the time.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###################################################################</span></span>
<span><span class="co">## Load required package</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/BSgenome" class="external-link">BSgenome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## The 1KG GDS file is required</span></span>
<span><span class="co">## The 1KG SNV Annotation GDS file is only required for RNA samples</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span>    </span>
<span><span class="va">fileName.GDS.1KG</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span><span class="va">fileName.GDS.Annot</span> <span class="op">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span></span>
<span><span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span><span class="va">file.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.Annot</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file</span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## The length of each chromosome is required</span></span>
<span><span class="co">## Chromosomes X, Y and M need relabeling</span></span>
<span><span class="co">## There is other ways to extract the information</span></span>
<span><span class="co">## This is one example</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">integer</a></span><span class="op">(</span><span class="fl">25L</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">22L</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">chrInfo</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Hsapiens</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">i</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">chrInfo</span><span class="op">[</span><span class="fl">23</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Hsapiens</span><span class="op">[[</span><span class="st">"chrX"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">chrInfo</span><span class="op">[</span><span class="fl">24</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Hsapiens</span><span class="op">[[</span><span class="st">"chrY"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">chrInfo</span><span class="op">[</span><span class="fl">25</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Hsapiens</span><span class="op">[[</span><span class="st">"chrM"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## The GDS Sample file is required</span></span>
<span><span class="co">###################################################################</span></span>
<span></span>
<span><span class="co">## Name of the GDS Sample file</span></span>
<span><span class="co">## The name must corresponds to the sample identifier</span></span>
<span><span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSAMPLEGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sampleName</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the GDS Sample file in writing mode</span></span>
<span><span class="va">gdsSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.GDSSample</span>, readonly <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## The estimation of the allelic fraction</span></span>
<span><span class="co">###################################################################</span></span>
<span></span>
<span><span class="co">## Estimate the allelic fraction of the pruned SNVs</span></span>
<span><span class="co">## The current example is for a DNA sample</span></span>
<span><span class="co">## In the case of RNA sample, the function needs different inputs</span></span>
<span><span class="co">## such as the 1KG SNV Annotation GDS file and</span></span>
<span><span class="co">## the block.id which refers to block.id in  the 1KG SNV Annotation GDS file </span></span>
<span><span class="co">## for the gene annotation of the SNVs</span></span>
<span><span class="fu"><a href="../reference/estimateAllelicFraction.html">estimateAllelicFraction</a></span><span class="op">(</span>gds <span class="op">=</span> <span class="va">gds</span>, gdsSample <span class="op">=</span> <span class="va">gdsSample</span>,</span>
<span>                                      sampleCurrent <span class="op">=</span> <span class="va">sampleName</span>,</span>
<span>                                      study.id <span class="op">=</span> <span class="va">studyDF</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                                      chrInfo <span class="op">=</span> <span class="va">chrInfo</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Close both GDS files</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsSample</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-3---find-the-optimized-parameters-for-the-ancestry-inference">Step 3 - Find the optimized parameters for the ancestry
inference<a class="anchor" aria-label="anchor" href="#step-3---find-the-optimized-parameters-for-the-ancestry-inference"></a>
</h3>
<p>At this step, the optimization of the parameters is required to
obtained the most accurate results at the ancestry inference step.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step3_v01.png" alt="Step 3 - Find the optimized parameters for the ancestry inference" width="120%"><p class="caption">
Step 3 - Find the optimized parameters for the ancestry inference
</p>
</div>
<p>Two parameters can be optimized for the ancestry inference
process:</p>
<ul>
<li>
<em>K</em>: the number of neighbors used to call the ancestry</li>
<li>
<em>D</em>: the number of dimensions used to create the PCA</li>
</ul>
<p>The accuracy is evaluated through a synthetic dataset created from
merging one cancer profile with multiple 1KG samples of known ancestry.
Using the synthetic profiles, a range of <em>K</em> and <em>D</em>
values are tested. Through that process, the <em>K</em> and <em>D</em>
values with satisfying accuracy can be selected.</p>
<p>This step is divided in two sub-steps:</p>
<ol style="list-style-type: decimal">
<li>Generate the synthetic dataset</li>
<li>Compute the PCA-KNN ancestry call for each synthetic profile</li>
</ol>
<p><br></p>
<div class="section level4">
<h4 id="sub-step-1--generate-the-synthetic-dataset">Sub-Step 1. Generate the synthetic dataset<a class="anchor" aria-label="anchor" href="#sub-step-1--generate-the-synthetic-dataset"></a>
</h4>
<p>A synthetic profile is generated through the merging of one cancer
profile with one 1KG sample of known ancestry. Multiple 1KG samples of
different ancestry are required to create a synthetic dataset that will
be able to show the specific accuracy for each super-population. All the
synthetic profiles are saved in the GDS Sample file corresponding to the
cancer profile used to generate the dataset.</p>
<p>In summary, a fixed number of samples fo each super-population is
extracted from the 1KG study. The information is saved in the GDS Sample
file associated to the selected cancer profile. A synthetic profile is
created for each combination of the 1KG samples and cancer profile. The
synthetic profile is then saved in the GDS Sample file.</p>
<p>The three functions <em>select1KGPop()</em>, <em>prepSynthetic()</em>
and <em>syntheticGeno()</em> are association to the creation of the
synthetic dataset.</p>
<p>Those 3 files are required:</p>
<ol style="list-style-type: decimal">
<li>The GDS Sample file</li>
<li>The 1KG GDS file</li>
<li>The 1KG SNv Annotation GDS file</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################################################################</span></span>
<span><span class="co">## Load required package</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Fix seed to ensure reproducible results</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span>   </span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Randomly extract a fixed number of samples for each</span></span>
<span><span class="co">## subcontinental population present in the 1KG GDS file.</span></span>
<span><span class="co">## When not enough samples are available, all samples are selected.  </span></span>
<span><span class="co">####################################################################   </span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, nbSamples <span class="op">=</span> <span class="fl">30L</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the list of selected 1KG sample identifiers</span></span>
<span><span class="va">listSampleRef</span> <span class="op">&lt;-</span> <span class="va">dataRef</span><span class="op">$</span><span class="va">sample.id</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## A data.frame with the description of the study for the synthetic</span></span>
<span><span class="co">## data is required.</span></span>
<span><span class="co">## The column names must be as shown</span></span>
<span><span class="co">#################################################################### </span></span>
<span><span class="va">studyDF.syn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"TCGA.BRCA.Synthetic"</span>,</span>
<span>                            study.desc <span class="op">=</span> <span class="st">"TCGA.BRCA synthetic data"</span>,</span>
<span>                            study.platform <span class="op">=</span> <span class="st">"WXS"</span>,</span>
<span>                            stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The GDS Sample file is needed</span></span>
<span><span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="va">PATH_TO_GDS_SAMPLE_FILE</span></span>
<span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="co">## The prepSynthetic() function prepares the annotations for</span></span>
<span><span class="co">## the synthetic data</span></span>
<span><span class="co">## The information is saved into the GDS sample file</span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="fu"><a href="../reference/prepSynthetic.html">prepSynthetic</a></span><span class="op">(</span>gdsSampleFile <span class="op">=</span> <span class="va">file.GDSSample</span>,</span>
<span>                listSampleRef <span class="op">=</span> <span class="va">listSampleRef</span>,</span>
<span>                data.id.profile <span class="op">=</span> <span class="va">sampleName</span>,</span>
<span>                studyDF <span class="op">=</span> <span class="va">studyDF.syn</span>,</span>
<span>                prefId <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="co">## Both the 1KG GDS file and the 1KG SNV Annotation GDS file</span></span>
<span><span class="co">## are required</span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span>    </span>
<span><span class="va">fileName.GDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span><span class="va">fileName.GDS.Annot</span> <span class="op">&lt;-</span> <span class="st">"matAnnot1000g.gds"</span></span>
<span><span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span><span class="va">file.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS.Annot</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file and 1KG SNV Annotation file    </span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span><span class="va">gds.Annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.Annot</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="co">## The syntheticGeno() function generates the synthetic </span></span>
<span><span class="co">## profiles.</span></span>
<span><span class="co">## The synthetic profiles are saved in the GDS Sample file</span></span>
<span><span class="co">##############################################################</span></span>
<span><span class="va">resG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/syntheticGeno.html">syntheticGeno</a></span><span class="op">(</span>gds <span class="op">=</span> <span class="va">gds</span>,</span>
<span>                        gdsRefAnnot <span class="op">=</span> <span class="va">gds.Annot</span>,</span>
<span>                        gdsSampleFile <span class="op">=</span> <span class="va">file.GDSSample</span>,</span>
<span>                        data.id.profile <span class="op">=</span> <span class="va">sampleName</span>,</span>
<span>                        listSampleRef <span class="op">=</span> <span class="va">listSampleRef</span>,</span>
<span>                        prefId <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Close both GDS files</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds.Annot</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-2--compute-the-pca-knn-ancestry-call-for-each-synthetic-profile">Sub-Step 2. Compute the PCA-KNN ancestry call for each synthetic
profile<a class="anchor" aria-label="anchor" href="#sub-step-2--compute-the-pca-knn-ancestry-call-for-each-synthetic-profile"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################################################################</span></span>
<span><span class="co">## Load required package</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The 1KG GDS file is required</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span>    </span>
<span><span class="va">fileName.GDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span><span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file     </span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The path to the directory containing the GDS Sample files</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">PATH_SAMPLE_GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The path to the directory where the PCA results will be saved </span></span>
<span><span class="co">## in RDS files.</span></span>
<span><span class="co">## The directory must exist.</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">PATH_OUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">PATH_OUT</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">PATH_OUT</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## A data.frame with the description of the study for the synthetic</span></span>
<span><span class="co">## data is required.</span></span>
<span><span class="co">## The column names must be as shown </span></span>
<span><span class="co">#################################################################### </span></span>
<span><span class="va">studyDF.syn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"Id of the synthetic dataset"</span>,</span>
<span>                          study.desc <span class="op">=</span> <span class="st">"Description"</span>,</span>
<span>                          study.platform <span class="op">=</span> <span class="st">"Platform string like Whole Exome"</span>,</span>
<span>                          stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The list of cancer samples used to generate the synthetic data </span></span>
<span><span class="co">## as obtained by the PED RDS file. The samples must have an </span></span>
<span><span class="co">## associated GDS Sample file.</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">filePED</span> <span class="op">&lt;-</span> <span class="va">FILE_PED</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span><span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Get the super-population information for the reference samples</span></span>
<span><span class="co">## This is the ground truth for the 1KG samples</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">spRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRef1KGPop.html">getRef1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"superPop"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Fix seed as in step 6 to ensure the same results</span></span>
<span><span class="co">####################################################################    </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span>       </span>
<span>    </span>
<span><span class="co">## Extract the 1KG samples used to generate the synthetic dataset</span></span>
<span><span class="co">## Already done in step 6, no need to repeat if the results have been saved</span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, nbSamples <span class="op">=</span> <span class="fl">30L</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The function splitSelectByPop() generates a matrix with the</span></span>
<span><span class="co">## reference samples split by sub-continental population</span></span>
<span><span class="co">####################################################################  </span></span>
<span><span class="va">sampleRM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/splitSelectByPop.html">splitSelectByPop</a></span><span class="op">(</span><span class="va">dataRef</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## Loop for all cancer samples with associated synthetic data</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="co">## The GDS Sample associated to the cancer sample     </span></span>
<span>  <span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_SAMPLE_GDS</span>, </span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## A sub-directory is created for the cancer sample</span></span>
<span>  <span class="co">## Beware that the number of files created will correspond to the </span></span>
<span>  <span class="co">## number of rows in the sampleRM matrix</span></span>
<span>  <span class="va">PATH_OUT_SAMPLE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_OUT</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">PATH_OUT_SAMPLE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">PATH_OUT_SAMPLE</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">## Open GDS Sample file      </span></span>
<span>  <span class="va">gdsSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDSSample</span><span class="op">)</span></span>
<span>        </span>
<span>  <span class="co">## This variable will contain the results from the PCA analyses</span></span>
<span>  <span class="va">KNN.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## For each row of the sampleRM matrix      </span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sampleRM</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="co">## Run a PCA analysis using 1 synthetic profile for each sub-continental</span></span>
<span>      <span class="co">## The synthetic profiles are projected on the 1KG PCA space </span></span>
<span>      <span class="co">## (the reference samples used to generate the synthetic profiles are </span></span>
<span>      <span class="co">##  removed from this PCA)</span></span>
<span>      <span class="co">## The K-nearest neighbor analysis is done using </span></span>
<span>      <span class="co">##  a range of K and D values</span></span>
<span>      <span class="va">KNN.synt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computePoolSyntheticAncestryGr.html">computePoolSyntheticAncestryGr</a></span><span class="op">(</span>gds <span class="op">=</span> <span class="va">gds</span>, </span>
<span>                                          gdsSample <span class="op">=</span> <span class="va">gdsSample</span>,</span>
<span>                                          sampleRM <span class="op">=</span>  <span class="va">sampleRM</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span>,</span>
<span>                                          study.id.syn <span class="op">=</span> <span class="va">studyDF.syn</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                                          np <span class="op">=</span> <span class="fl">4L</span>,</span>
<span>                                          spRef <span class="op">=</span> <span class="va">spRef</span>,</span>
<span>                                          eigen.cnt <span class="op">=</span> <span class="fl">15L</span><span class="op">)</span></span>
<span>      </span>
<span>      <span class="co">## Results are saved</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">KNN.synt</span><span class="op">$</span><span class="va">matKNN</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_OUT_SAMPLE</span>, </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"KNN.synt."</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"."</span>, <span class="va">j</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">## Close GDS Sample file (important)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsSample</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>    </span>
<span><span class="co">## Close 1KG GDS file (important)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-4---run-the-ancestry-inference-on-the-external-study">Step 4 - Run the ancestry inference on the external study<a class="anchor" aria-label="anchor" href="#step-4---run-the-ancestry-inference-on-the-external-study"></a>
</h3>
<p>The ancestry inference is done with the optimized K and D parameters.
More specifically, a PCA is generated using the 1KG reference samples
and the cancer sample. The <em>D</em> parameter specifies the number of
dimension for the PCA. Then, the ancestry of the cancer sample is
inferred using a k-nearest neighbors classification method. The
<em>K</em> parameter specifies the number of neighbors used for the
classification.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step4_v01.png" alt="Step 4 - Run the ancestry inference on the external study" width="120%"><p class="caption">
Step 4 - Run the ancestry inference on the external study
</p>
</div>
<p>The PCA of the sample and KNN sample and Call the ancestry with the
optimal <em>K</em> and <em>D</em> parameters.</p>
<p>Note: The formal selection of <em>K</em> and <em>D</em> parameters is
done at this step but all the synthetic data is prepared in the step
3.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################################################################</span></span>
<span><span class="co">## Load required package</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The 1KG GDS file is required</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">PATH_1KG</span> <span class="op">&lt;-</span> <span class="va">PATH_REF</span></span>
<span></span>
<span><span class="va">fileName.GDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span><span class="va">file.GDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_1KG</span>, <span class="va">fileName.GDS</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## Open the 1KG GDS file </span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file.GDS</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span><span class="co"># Path to the directory where the gds file for each sample</span></span>
<span><span class="co"># will be create</span></span>
<span><span class="va">PATHSAMPLEGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># directory which content directory name Name.ID in which the PCA is save</span></span>
<span><span class="co"># </span></span>
<span><span class="va">PATH_OUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"The path"</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">PATH_OUT</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">PATH_OUT</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## A data.frame with the description of the study for the synthetic</span></span>
<span><span class="co">## data is required.</span></span>
<span><span class="co">## The column names must be as shown </span></span>
<span><span class="co">#################################################################### </span></span>
<span><span class="va">studyDF.syn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"Id of the synthetic dataset"</span>,</span>
<span>                          study.desc <span class="op">=</span> <span class="st">"Description"</span>,</span>
<span>                          study.platform <span class="op">=</span> <span class="st">"Platform string like Whole Exome"</span>,</span>
<span>                          stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The list of samples to analyse as obtained by the PED RDS file.</span></span>
<span><span class="co">## The samples must have an associated GDS Sample files.</span></span>
<span><span class="co">## Not needed if the list of samples is already available.</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">filePED</span> <span class="op">&lt;-</span> <span class="va">FILE_PED</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span><span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The getRef1KGPop() function extract the known super-population </span></span>
<span><span class="co">## of the reference samples.</span></span>
<span><span class="co">## We expect the ancestry call from a synthetic profile to </span></span>
<span><span class="co">## correspond to the known ancestry of the reference sample used to </span></span>
<span><span class="co">## generate it.</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">spRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRef1KGPop.html">getRef1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"superPop"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co"># Can be run in parallel or in different clusters...</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        </span>
<span>  <span class="va">file.GDSSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSAMPLEGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>        </span>
<span>  <span class="va">PATHKNN</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATH_OUT</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">listFilesName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHKNN</span><span class="op">)</span>, <span class="st">".rds"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># List of the KNN result from PCA on synthetic data with path</span></span>
<span>  <span class="va">listFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHKNN</span><span class="op">)</span> , <span class="va">listFilesName</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gdsSample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file.GDSSample</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">resCall</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeAncestryFromSyntheticFile.html">computeAncestryFromSyntheticFile</a></span><span class="op">(</span>gds <span class="op">=</span> <span class="va">gds</span>, </span>
<span>                                    gdsSample <span class="op">=</span> <span class="va">gdsSample</span>,</span>
<span>                                    listFiles <span class="op">=</span> <span class="va">listFiles</span>,</span>
<span>                                    sample.ana.id <span class="op">=</span> <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                                    spRef <span class="op">=</span> <span class="va">spRef</span>,</span>
<span>                                    study.id.syn <span class="op">=</span> <span class="va">studyDF.syn</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                                    np <span class="op">=</span> <span class="fl">4L</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">resCall</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHOUT</span>,</span>
<span>                              <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".infoCall"</span>,<span class="st">".rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">resCall</span><span class="op">$</span><span class="va">Ancestry</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHOUT</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".Ancestry"</span>,<span class="st">".csv"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## Close GDS Sample file</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsSample</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Close both 1KG GDS file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="pre-processed-files-are-available">Pre-processed files are available<a class="anchor" aria-label="anchor" href="#pre-processed-files-are-available"></a>
</h2>
<p>Pre-processed files, such as the 1KG GDS file, are available at this
address:</p>
<p><a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a></p>
<p>Beware that some of those files are voluminous.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.1.3 (2022-03-10)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] RAIDS_0.50.0     GENESIS_2.24.2   SNPRelate_1.28.0 gdsfmt_1.30.0   </span></span>
<span><span class="co">## [5] knitr_1.40       BiocStyle_2.22.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] backports_1.4.1             BiocFileCache_2.2.1        </span></span>
<span><span class="co">##   [3] systemfonts_1.0.4           plyr_1.8.7                 </span></span>
<span><span class="co">##   [5] lazyeval_0.2.2              splines_4.1.3              </span></span>
<span><span class="co">##   [7] operator.tools_1.6.3        BiocParallel_1.28.3        </span></span>
<span><span class="co">##   [9] GenomeInfoDb_1.30.1         digest_0.6.29              </span></span>
<span><span class="co">##  [11] ensembldb_2.18.4            htmltools_0.5.3            </span></span>
<span><span class="co">##  [13] fansi_1.0.3                 magrittr_2.0.3             </span></span>
<span><span class="co">##  [15] memoise_2.0.1               BSgenome_1.62.0            </span></span>
<span><span class="co">##  [17] Biostrings_2.62.0           matrixStats_0.62.0         </span></span>
<span><span class="co">##  [19] formula.tools_1.7.1         sandwich_3.0-2             </span></span>
<span><span class="co">##  [21] pkgdown_2.0.6               prettyunits_1.1.1          </span></span>
<span><span class="co">##  [23] blob_1.2.3                  rappdirs_0.3.3             </span></span>
<span><span class="co">##  [25] textshaping_0.3.6           xfun_0.33                  </span></span>
<span><span class="co">##  [27] dplyr_1.0.10                crayon_1.5.1               </span></span>
<span><span class="co">##  [29] RCurl_1.98-1.8              jsonlite_1.8.0             </span></span>
<span><span class="co">##  [31] survival_3.4-0              zoo_1.8-10                 </span></span>
<span><span class="co">##  [33] glue_1.6.2                  zlibbioc_1.40.0            </span></span>
<span><span class="co">##  [35] XVector_0.34.0              MatrixModels_0.5-1         </span></span>
<span><span class="co">##  [37] DelayedArray_0.20.0         SeqVarTools_1.32.0         </span></span>
<span><span class="co">##  [39] BiocGenerics_0.40.0         SparseM_1.81               </span></span>
<span><span class="co">##  [41] DBI_1.1.3                   Rcpp_1.0.9                 </span></span>
<span><span class="co">##  [43] progress_1.2.2              quantsmooth_1.60.0         </span></span>
<span><span class="co">##  [45] bit_4.0.4                   stats4_4.1.3               </span></span>
<span><span class="co">##  [47] httr_1.4.4                  logistf_1.24.1             </span></span>
<span><span class="co">##  [49] ellipsis_0.3.2              mice_3.14.0                </span></span>
<span><span class="co">##  [51] pkgconfig_2.0.3             XML_3.99-0.10              </span></span>
<span><span class="co">##  [53] sass_0.4.2                  dbplyr_2.2.1               </span></span>
<span><span class="co">##  [55] utf8_1.2.2                  SeqArray_1.34.0            </span></span>
<span><span class="co">##  [57] DNAcopy_1.68.0              tidyselect_1.1.2           </span></span>
<span><span class="co">##  [59] rlang_1.0.5                 AnnotationDbi_1.56.2       </span></span>
<span><span class="co">##  [61] tools_4.1.3                 cachem_1.0.6               </span></span>
<span><span class="co">##  [63] cli_3.4.0                   generics_0.1.3             </span></span>
<span><span class="co">##  [65] RSQLite_2.2.17              broom_1.0.1                </span></span>
<span><span class="co">##  [67] evaluate_0.16               stringr_1.4.1              </span></span>
<span><span class="co">##  [69] fastmap_1.1.0               yaml_2.3.5                 </span></span>
<span><span class="co">##  [71] ragg_1.2.2                  bit64_4.0.5                </span></span>
<span><span class="co">##  [73] fs_1.5.2                    purrr_0.3.4                </span></span>
<span><span class="co">##  [75] KEGGREST_1.34.0             AnnotationFilter_1.18.0    </span></span>
<span><span class="co">##  [77] nlme_3.1-159                GWASExactHW_1.01           </span></span>
<span><span class="co">##  [79] quantreg_5.94               xml2_1.3.3                 </span></span>
<span><span class="co">##  [81] biomaRt_2.50.3              compiler_4.1.3             </span></span>
<span><span class="co">##  [83] GWASTools_1.40.0            filelock_1.0.2             </span></span>
<span><span class="co">##  [85] curl_4.3.2                  png_0.1-7                  </span></span>
<span><span class="co">##  [87] tibble_3.1.8                bslib_0.4.0                </span></span>
<span><span class="co">##  [89] stringi_1.7.8               highr_0.9                  </span></span>
<span><span class="co">##  [91] GenomicFeatures_1.46.5      desc_1.4.2                 </span></span>
<span><span class="co">##  [93] lattice_0.20-45             ProtGenerics_1.26.0        </span></span>
<span><span class="co">##  [95] Matrix_1.5-0                vctrs_0.4.1                </span></span>
<span><span class="co">##  [97] pillar_1.8.1                lifecycle_1.0.2            </span></span>
<span><span class="co">##  [99] BiocManager_1.30.18         lmtest_0.9-40              </span></span>
<span><span class="co">## [101] jquerylib_0.1.4             data.table_1.14.2          </span></span>
<span><span class="co">## [103] bitops_1.0-7                rtracklayer_1.54.0         </span></span>
<span><span class="co">## [105] GenomicRanges_1.46.1        R6_2.5.1                   </span></span>
<span><span class="co">## [107] BiocIO_1.4.0                bookdown_0.29              </span></span>
<span><span class="co">## [109] IRanges_2.28.0              MASS_7.3-58.1              </span></span>
<span><span class="co">## [111] assertthat_0.2.1            SummarizedExperiment_1.24.0</span></span>
<span><span class="co">## [113] rprojroot_2.0.3             rjson_0.2.21               </span></span>
<span><span class="co">## [115] withr_2.5.0                 GenomicAlignments_1.30.0   </span></span>
<span><span class="co">## [117] Rsamtools_2.10.0            S4Vectors_0.32.4           </span></span>
<span><span class="co">## [119] GenomeInfoDbData_1.2.7      mgcv_1.8-40                </span></span>
<span><span class="co">## [121] parallel_4.1.3              hms_1.1.2                  </span></span>
<span><span class="co">## [123] grid_4.1.3                  tidyr_1.2.1                </span></span>
<span><span class="co">## [125] class_7.3-20                rmarkdown_2.16             </span></span>
<span><span class="co">## [127] MatrixGenerics_1.6.0        pROC_1.18.0                </span></span>
<span><span class="co">## [129] Biobase_2.54.0              restfulr_0.0.15</span></span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Alexander2009" class="csl-entry">
Alexander, D. H., J. Novembre, and K. Lange. 2009. <span>“Fast
Model-Based Estimation of Ancestry in Unrelated Individuals.”</span>
<em>Genome Res</em> 19 (9): 1655–64. <a href="https://doi.org/10.1101/gr.094052.109" class="external-link">https://doi.org/10.1101/gr.094052.109</a>.
</div>
<div id="ref-Lowy-Gallego2019a" class="csl-entry">
Lowy-Gallego, Ernesto, Susan Fairley, Xiangqun Zheng-Bradley, Magali
Ruffier, Laura Clarke, and Paul Flicek. 2019. <span>“<span class="nocase">Variant calling on the grch38 assembly with the data from
phase three of the 1000 genomes project [version 2; peer review: 1
approved, 1 not approved]</span>.”</span> <em>Wellcome Open
Research</em> 4: 1–42. <a href="https://doi.org/10.12688/wellcomeopenres.15126.1" class="external-link">https://doi.org/10.12688/wellcomeopenres.15126.1</a>.
</div>
<div id="ref-Price2006" class="csl-entry">
Price, A. L., N. J. Patterson, R. M. Plenge, M. E. Weinblatt, N. A.
Shadick, and D. Reich. 2006. <span>“Principal Components Analysis
Corrects for Stratification in Genome-Wide Association Studies.”</span>
<em>Nat Genet</em> 38 (8): 904–9. <a href="https://doi.org/10.1038/ng1847" class="external-link">https://doi.org/10.1038/ng1847</a>.
</div>
<div id="ref-Pritchard2000" class="csl-entry">
Pritchard, J. K., M. Stephens, and P. Donnelly. 2000. <span>“Inference
of Population Structure Using Multilocus Genotype Data.”</span>
<em>Genetics</em> 155 (2): 945–59. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10835412" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/10835412</a>.
</div>
<div id="ref-Shen2016" class="csl-entry">
Shen, R, and V E Seshan. 2016. <span>“<span class="nocase">FACETS:
allele-specific copy number and clonal heterogeneity analysis tool for
high-throughput DNA sequencing</span>.”</span> <em>Nucleic Acids
Res</em> 44 (16): e131. <a href="https://doi.org/10.1093/nar/gkw520" class="external-link">https://doi.org/10.1093/nar/gkw520</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
