<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate Inference of Genetic Ancestry from Cancer Sequencing • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate Inference of Genetic Ancestry from Cancer Sequencing">
<meta property="og:description" content="RAIDS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/aicsPaper.html">Accurate Inference of Genetic Ancestry from Cancer Sequencing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate Inference of Genetic Ancestry from Cancer Sequencing</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau, Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      
      <div class="hidden name"><code>aicsPaper.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes [aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), Alexander Krasnitz [aut]<br><strong>Version</strong>: 0.50.0<br><strong>Compiled date</strong>: 2022-07-30<br><strong>License</strong>: Apache License (== 2.0) | file LICENSE<br></p>
<div class="section level2">
<h2 id="licensing">Licensing<a class="anchor" aria-label="anchor" href="#licensing"></a>
</h2>
<p>The <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em> package and the underlying <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em> code are distributed under<br>
the <a href="https://opensource.org/licenses/Apache-2.0" class="external-link uri">https://opensource.org/licenses/Apache-2.0</a> license. You are free to use and redistribute this software.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>If you use the <strong>RAIDS</strong> package for a publication, we would ask you to cite the following:</p>
<blockquote>
<p>Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz. Accurate and robust inference of genetic ancestry from cancer-derived molecular data across genomic platforms. bioRxiv 2022.02.01.478737; doi: <a href="https://doi.org/10.1101/2022.02.01.478737" class="external-link uri">https://doi.org/10.1101/2022.02.01.478737</a></p>
</blockquote>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple methods have been implemented to infer ancestry from germline DNA sequence <span class="citation">(Price et al. 2006; Pritchard, Stephens, and Donnelly 2000; Alexander, Novembre, and Lange 2009)</span>. However, genotyping of DNA from matched normal specimens is not part of standard clinical practice and is not performed routinely outside academic clinical centers. In sum, matched germline DNA sequence is often missing for cancer-derived molecular data. In such cases, having the possibility to infer ancestry from tumor-derived data would be beneficial.</p>
<p>The <strong>RAIDS</strong> package implements an inference procedure that has been specifically developed to accurately infer genetic ancestry from cancer-derived sequencing. The covered cancer-derived sequencing are, more specifically, tumor exomes, targeted gene panels and RNA sequences.</p>
<p>The <strong>RAIDS</strong> package includes a data synthesis method that enables the selection of the best parameters for a specific cancer profile and also permits a statistically assessment of inference accuracy for an individual cancer-derived molecular profile.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the latest version accessible on the <a href="https://github.com/KrasnitzLab/RAIDS" title="RAIDS Github Site" class="external-link">RAIDS Github Website</a>, the <em><a href="https://CRAN.R-project.org/package=devtools" class="external-link">devtools</a></em> package is required.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Install the latest version of RAIDS</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'KrasnitzLab/RAIDS'</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="main-steps">Main Steps<a class="anchor" aria-label="anchor" href="#main-steps"></a>
</h2>
<p>This is an overview of genetic ancestry inference from cancer-derived molecular data:</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_v01.png" alt="An overview of the genetic ancestry inference process." width="120%"><p class="caption">
An overview of the genetic ancestry inference process.
</p>
</div>
<p>The main steps are:</p>
<ol style="list-style-type: decimal">
<li>Format the information from 1000 Genomes (1KG) (optional)</li>
<li>Format the information from an external study</li>
<li>Find the optimized parameters for the ancestry inference</li>
<li>Run the ancestry inference on the external study</li>
</ol>
<p>In the following sections, those main steps are described in details.</p>
<p><br><br></p>
<div class="section level3">
<h3 id="step-1---formatting-the-information-from-1000-genomes-optional">Step 1 - Formatting the information from 1000 Genomes (optional)<a class="anchor" aria-label="anchor" href="#step-1---formatting-the-information-from-1000-genomes-optional"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step1_v01.png" alt="Step 1 - Formatting the information from 1000 Genomes (optional)" width="120%"><p class="caption">
Step 1 - Formatting the information from 1000 Genomes (optional)
</p>
</div>
<hr>
<p><span style="color:darkred;font-weight:bold">Beware that a pre-processed 1KG GDS file is available at this address:</span></p>
<p><span style="color:red"> <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds</a> </span></p>
<p><span style="color:darkred;font-weight:bold">The size of the file is 15GB.</span> <span style="color:darkred;font-weight:bold">This section can be skipped if you choose to use this file.</span></p>
<hr>
<p>This section explains in further details how to generate the 1000 Genomes (1KG) GDS file that is needed to run the ancestry inference tool. Beware that it is unnecessary to re-run those steps as the 1KG GDS file is publicly available.</p>
<p>More specifically, the formatting of the 1KG information includes all those sub-steps:</p>
<ol style="list-style-type: decimal">
<li>Download the required files from 1KG</li>
<li>Split the 1KG genotyping file to one file per sample per chromosome</li>
<li>Combine the chromosomes to obtain one genotyping file per sample</li>
<li>Create a pedigree file in RDS format</li>
<li>Prepare a bulk SNP information file based on 1KG VCF</li>
<li>Filter the bulk SNP information file</li>
<li>Generate the GDS file with the 1KG information</li>
<li>Identify genetically related patients present in 1KG GDS file</li>
<li>Add information about unrelated patients to the 1KG GDS file</li>
<li>Create a second GDS file containing the 1KG phase information</li>
</ol>
<p>Beware that those sub-steps are time, as well as, space consuming.</p>
<p>In the following sections, those sub-steps are described in details.</p>
<p><br></p>
<div class="section level4">
<h4 id="download-the-required-files-from-1kg">Download the required files from 1KG<a class="anchor" aria-label="anchor" href="#download-the-required-files-from-1kg"></a>
</h4>
<p>First, the pedigree file with the description of the 1KG samples needs to be downloaded from the 1KG ftp site:</p>
<blockquote>
<p><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_g1k.ped" class="uri">ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_g1k.ped</a></p>
</blockquote>
<p>The genotyping files for the 1KG samples also needs to be downloaded. Beware that there is one file per chromosome:</p>
<blockquote>
<p>ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV</p>
</blockquote>
<p>As the genotyping files are split by chromosome, all files corresponding to this pattern must be downloaded:</p>
<blockquote>
<p>ALL.chr*.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz</p>
</blockquote>
<p>All files are related to GRCh38 genome.</p>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="split-the-1kg-genotyping-file-to-one-file-per-sample-per-chromosome">Split the 1KG genotyping file to one file per sample per chromosome<a class="anchor" aria-label="anchor" href="#split-the-1kg-genotyping-file-to-one-file-per-sample-per-chromosome"></a>
</h4>
<p>To facilitate the manipulation of the genotyping data, the large 1KG genotyping files are split into smaller files (one per sample).</p>
<p>As there is one genotyping file per chromosome, the process must be executed for each of chromosome</p>
<p>The splitting is done through a bash script:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This in not a R script</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="do">## This script is in bash</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="do">## The script has to be run separately for each chromosome</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Create one directory for the specific chromosome</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Two variables need to be assigned</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="do">## The variable PATHGENO is the path where the split 1KG genotyping </span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="do">##   files will be located</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="do">## The variable chrCur is the current chromosome (ex: "chr1"), a sub-directory </span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="do">##   with the name of the chromosome will be created. The variable must </span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="do">##   include the prefix "chr".</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="do">## Ex: PATHGENO=./genotypingPerSample</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="do">## Ex: chrCur=chr1</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>mkdir <span class="sc">$</span>{PATHGENO}<span class="sc">/</span><span class="er">$</span>{chrCur}</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>cd <span class="sc">$</span>{PATHGENO}<span class="sc">/</span><span class="er">$</span>{chrCur}</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="do">## The variable FILECUR is the associated 1KG genotyping VCF file for the </span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="do">##     specific chromosome</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="do">## The variable FILECUR must also contain the relative or complete path to </span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="do">##     the VCF file</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="do">## Ex: FILECUR=../../1KG_files/ALL.chr1.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a><span class="do">## There is 2548 samples in one phase VCF file</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="do">## The information associated to each sample is extracted and a VCF file </span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="do">## specific to each sample for the current chromosome is generated</span></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="cf">in</span> <span class="st">`</span><span class="at">seq 1 2548</span><span class="st">`</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>do</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>    j<span class="ot">=</span><span class="er">$</span>(( <span class="sc">$</span>i <span class="sc">+</span> <span class="dv">9</span> ))</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>    SAMPLE<span class="ot">=</span><span class="er">$</span>(zcat <span class="sc">$</span>FILECUR<span class="sc">||</span>head <span class="sc">-</span>n <span class="dv">1000</span> <span class="sc">|</span>grep <span class="st">"#CHROM"</span><span class="sc">|</span>cut <span class="sc">-</span>d<span class="sc">$</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span> <span class="sc">-</span>f<span class="sc">$</span>j)</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>    zcat <span class="sc">$</span>FILECUR<span class="sc">|</span>grep <span class="sc">-</span>v <span class="st">"##"</span><span class="sc">|</span>cut <span class="sc">-</span>d<span class="sc">$</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span> <span class="sc">-</span>f<span class="sc">$</span>j <span class="sc">|</span>bzip2 <span class="sc">-</span>c <span class="sc">&gt;</span> <span class="er">$</span>{SAMPLE}.<span class="sc">$</span>{CHR}.vcf.bz2</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>done</span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="combine-the-chromosomes-to-obtain-one-genotyping-file-per-sample">Combine the chromosomes to obtain one genotyping file per sample<a class="anchor" aria-label="anchor" href="#combine-the-chromosomes-to-obtain-one-genotyping-file-per-sample"></a>
</h4>
<p>The genotyping information for all chromosomes is merged so that there is only one genotyping file per sample.</p>
<div class="figure" style="text-align: center">
<img src="groupChr1KGSNV_v01.png" alt="The function groupChr1KGSNV() merge samples that are split by chromosomes into one file." width="120%"><p class="caption">
The function groupChr1KGSNV() merge samples that are split by chromosomes into one file.
</p>
</div>
<p>The <strong>groupChr1KGSNV()</strong> function is used to combine the genotyping information from multiple files.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The path where the genotyping files are located</span></span>
<span><span class="va">PATHGENOCHR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"pathgenochr"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The path where the merged genotyping files will be created</span></span>
<span><span class="va">PATHOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"pathgenoOUT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Combining the genotyping information for all chromosome in one file per</span></span>
<span><span class="co">## sample</span></span>
<span><span class="fu"><a href="../reference/groupChr1KGSNV.html">groupChr1KGSNV</a></span><span class="op">(</span><span class="va">PATHGENOCHR</span>, <span class="va">PATHOUT</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="create-a-predigree-file-in-rds-format">Create a predigree file in RDS format<a class="anchor" aria-label="anchor" href="#create-a-predigree-file-in-rds-format"></a>
</h4>
<p>The function <strong>prepPed1KG()</strong> is used to create the pedigree file in RDS format:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The path to the pedigree file from 1KG</span></span>
<span><span class="co">## In this example, the file is in the current directory</span></span>
<span><span class="va">filePED1KGOri</span> <span class="op">&lt;-</span> <span class="st">"20130606_g1k.ped"</span></span>
<span></span>
<span><span class="co">## Extract needed information from the pedigree file from 1KG into a data.frame</span></span>
<span><span class="co">## Only the samples with genotyping information (sample file present </span></span>
<span><span class="co">## in PATHGENO parameter) are retained to create the final data.frame</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepPed1KG.html">prepPed1KG</a></span><span class="op">(</span>pedFile<span class="op">=</span><span class="va">filePED1KGOri</span>, </span>
<span>                    PATHGENO<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"sampleGeno"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Save the pedigree information data.frame as a RDS file </span></span>
<span><span class="co">## In this example, the file is saved here ./data/metadata/ped1KG.rds</span></span>
<span><span class="va">filePED1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"metadata"</span>,<span class="st">"ped1KG.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ped</span>,  <span class="va">filePED1KG</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="prepare-a-bulk-snp-information-file-based-on-1kg-vcf">Prepare a bulk SNP information file based on 1KG VCF<a class="anchor" aria-label="anchor" href="#prepare-a-bulk-snp-information-file-based-on-1kg-vcf"></a>
</h4>
<p>Some intermediate file containing the SNP information from 1KG need to be generated so the information can ultimately be imported in a GDS file using Bioconductor <a href="https://bioconductor.org/packages/gdsfmt/" class="external-link">gdsfmt</a>.</p>
<p>The bulk intermediate SNP file contains the SNP position as well as the frequency in each super population.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This is not done in R</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="do">## The python script is in the 'scriptsPy' directory</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="cf">in</span> <span class="st">`</span><span class="at">ls PATHVCF/*shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz</span><span class="st">`</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>do</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  chr<span class="ot">=</span><span class="er">$</span>(echo <span class="sc">$</span>FILECUR<span class="sc">|</span>perl <span class="sc">-</span>n <span class="sc">-</span>e <span class="st">'/ALL\.(chr[^\.]+)/;print $1'</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  python PATH2SCRIPT<span class="sc">/</span>extract1000gFreq.py <span class="sc">$</span>{FILECUR} matFreq.<span class="sc">$</span>{chr}</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>done</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="cf">in</span> <span class="st">`</span><span class="at">seq 1 22</span><span class="st">`</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>do</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  cat matFreq.chr<span class="sc">$</span>{i}.txt <span class="sc">&gt;</span>matFreqSNV.txt</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  bzip2 matFreqSNV.txt</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>done</span></code></pre></div>
<p>The bulk SNP info file is called <strong>matFreqSNV.txt.bz2</strong>.</p>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="filter-the-bulk-snp-information-file">Filter the bulk SNP information file<a class="anchor" aria-label="anchor" href="#filter-the-bulk-snp-information-file"></a>
</h4>
<p>The bulk SNP file is filtered to only retain the SNPs with frequency higher then a specific cut-off (here &gt;=0.01) for at least one super population.</p>
<p>The filter SNP file is saved in RDS format (<em>fileLSNP</em> parameter). A second file containing the index of the retained SNPs is also created (<em>fileLSNP</em> parameter)</p>
<p>The function <strong>generateMapSnvSel()</strong> is used to filter the SNP file and generated the needed RDS files:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The path to the bulk SNP info file</span></span>
<span><span class="va">fileSNV.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGENO</span>, <span class="st">"matFreqSNV.txt.bz2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The paths and names of the two output files</span></span>
<span><span class="co">## One file contains the index of retained SNPs ("listSNP.rds")</span></span>
<span><span class="co">## One file contains the filter SNP information ("mapSNVSel.rds")</span></span>
<span><span class="va">fileLSNP.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSEL</span>, <span class="st">"listSNP.rds"</span><span class="op">)</span></span>
<span><span class="va">fileFREQ.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSEL</span>, <span class="st">"mapSNVSel.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Filter the bulk SNP file (fileSNV parameter)</span></span>
<span><span class="co">## Create a RDS with filter SNPs (fileFREQ parameter)</span></span>
<span><span class="co">## Also creates a RDS with the indexes of the retained SNPs (fileFREQ parameter)</span></span>
<span><span class="fu"><a href="../reference/generateMapSnvSel.html">generateMapSnvSel</a></span><span class="op">(</span>cutOff<span class="op">=</span><span class="fl">0.01</span>, fileSNV<span class="op">=</span><span class="va">fileSNV.v</span>,</span>
<span>                      fileLSNP<span class="op">=</span><span class="va">fileLSNP.v</span>,</span>
<span>                      fileFREQ<span class="op">=</span><span class="va">fileFREQ.v</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="generate-the-gds-file-with-the-1kg-information">Generate the GDS file with the 1KG information<a class="anchor" aria-label="anchor" href="#generate-the-gds-file-with-the-1kg-information"></a>
</h4>
<p>The CoreArray Genomic Data Structure (GDS) data files are files suited for large-scale datasets, especially for data which are much larger than the available random-access memory.</p>
<p>The function <strong>generateGDS1KG()</strong> is used to generate the GDS file that will contain the information related to 1KG:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## TODO: what is PATHMETA and PATHSEL</span></span>
<span></span>
<span><span class="co">## The path and file names of the required files</span></span>
<span><span class="co">## First, the RDS file containing the pedigree information</span></span>
<span><span class="co">## Second, the RDS file with the indexes of the retained SNPs</span></span>
<span><span class="co">## Third, the RDS file with the filtered SNP information </span></span>
<span><span class="va">filePED1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHMETA</span>,<span class="st">"ped1KG.rds"</span><span class="op">)</span></span>
<span><span class="va">fileLSNP.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSEL</span>, <span class="st">"listSNP.rds"</span><span class="op">)</span></span>
<span><span class="va">fileFREQ.v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHSEL</span>, <span class="st">"mapSNVSel.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The name of the GDS file that will be created    </span></span>
<span><span class="va">fileNameGDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGDS</span>, <span class="va">fileNameGDS</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## Generate GDS file containing the 1KG information</span></span>
<span><span class="fu"><a href="../reference/generateGDS1KG.html">generateGDS1KG</a></span><span class="op">(</span>PATHGENO<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"sampleGeno"</span><span class="op">)</span>,</span>
<span>                    fileNamePED<span class="op">=</span><span class="va">filePED1KG</span>,</span>
<span>                    fileSNPSel<span class="op">=</span><span class="va">fileFREQ.v</span>,</span>
<span>                    fileListSNP<span class="op">=</span><span class="va">fileLSNP.v</span>,</span>
<span>                    fileNameGDS<span class="op">=</span><span class="va">fileGDS</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="identify-genetically-related-patients-present-in-1kg-gds-file">Identify genetically related patients present in 1KG GDS file<a class="anchor" aria-label="anchor" href="#identify-genetically-related-patients-present-in-1kg-gds-file"></a>
</h4>
<p>As only unrelated patients can be used in the following analyses, the genetically related patients in 1KG must be identified.</p>
<p>The function <strong>identifyRelative()</strong> identifies patients that are genetically related in the 1KG files. It generates a RDS file with the unrelated patient information (<em>filePart</em> parameter) as well as a RDS file with the kinship coefficient between the patients (<em>fileIBD</em> parameter).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## TODO: what is PATHMETA </span></span>
<span></span>
<span><span class="co">## The name of the GDS file that contains the 1KG information    </span></span>
<span><span class="va">fileNameGDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGDS</span>, <span class="va">fileNameGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Files that will be created by the identifyRelative() function</span></span>
<span><span class="co">## The first RDS file will contain the kinship information between patients</span></span>
<span><span class="co">## The second RDS file will contain the list of unrelated patients</span></span>
<span><span class="va">fileIBD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHMETA</span>,<span class="st">"ibd.All.0.05.rds"</span><span class="op">)</span></span>
<span><span class="va">filePart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHMETA</span>,<span class="st">"part.All.0.05.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file</span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Identify the genetically related patients in 1KG</span></span>
<span><span class="fu"><a href="../reference/identifyRelative.html">identifyRelative</a></span><span class="op">(</span>gds<span class="op">=</span><span class="va">fileGDS</span>, maf<span class="op">=</span><span class="fl">0.05</span>, thresh <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">11</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                     fileIBD<span class="op">=</span><span class="va">fileIBD</span>, filePart<span class="op">=</span><span class="va">filePart</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Close the 1KG GDS file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="add-information-about-unrelated-patients-to-the-1kg-gds-file">Add information about unrelated patients to the 1KG GDS file<a class="anchor" aria-label="anchor" href="#add-information-about-unrelated-patients-to-the-1kg-gds-file"></a>
</h4>
<p>Only the unrelated patients from 1KG are used in the following ancestry inference and kept in the GDS 1KG file. The function <strong>identifyRelative()</strong> identifies the unrelated patients and saves the information about those patients in an intermediate external file.</p>
<p>The function <strong>addRef2GDS1KG()</strong> is adding the information about the unrelated patients to the GDS 1KG file using the intermediate external file.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add the information about the unrelated 1KG patients to the 1KG GDS file </span></span>
<span><span class="fu"><a href="../reference/addRef2GDS1KG.html">addRef2GDS1KG</a></span><span class="op">(</span>fileGDS<span class="op">=</span><span class="va">gds</span>,  filePart<span class="op">=</span><span class="va">filePart</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="create-a-second-gds-containing-the-1kg-phase-information">Create a second GDS containing the 1KG phase information<a class="anchor" aria-label="anchor" href="#create-a-second-gds-containing-the-1kg-phase-information"></a>
</h4>
<p>We generate a GDS with the phase information.</p>
<p>The function <strong>generatePhase1KG2GDS()</strong> is adding the phase information into the newly created 1KG GDS phase file.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The name of the 1KG GDS file that already exists</span></span>
<span><span class="va">fileGDS1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGDS</span>, <span class="st">"matGeno1000g.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The name of the 1KG GDS phase file that will be created</span></span>
<span><span class="va">fileGDSPhase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGDS</span>, <span class="st">"matPhase1000g.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file</span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileGDS1KG</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create the 1KG GDS phase file</span></span>
<span><span class="va">gdsPhase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/createfn.gds.html" class="external-link">createfn.gds</a></span><span class="op">(</span><span class="va">fileGDSPhase</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add the phase information to the 1KG GDS phase file</span></span>
<span><span class="fu"><a href="../reference/generatePhase1KG2GDS.html">generatePhase1KG2GDS</a></span><span class="op">(</span><span class="va">gds</span>, <span class="va">gdsPhase</span>, <span class="va">PATHGENO</span>, <span class="va">fileLSNP.v</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Close both files</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsPhase</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-2---format-information-from-an-external-study-to-get-if-ready-for-ancestry-inference">Step 2 - Format information from an external study to get if ready for ancestry inference<a class="anchor" aria-label="anchor" href="#step-2---format-information-from-an-external-study-to-get-if-ready-for-ancestry-inference"></a>
</h3>
<p>To run the ancestry inference on samples from an external study, the sample information must be formatted in a specific format.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step2_v01.png" alt="Step 2 - Formatting the information from an external study" width="120%"><p class="caption">
Step 2 - Formatting the information from an external study
</p>
</div>
<p>Those steps shows how to prepare the sample information needed as input for the ancestry inference.</p>
<p>Those steps are:</p>
<ol style="list-style-type: decimal">
<li>Extract the read coverage at each SNP position in the 1KG GDS file</li>
<li>Create a VCF file containing the retained SNP positions from 1KG GDS file</li>
</ol>
<p>Beware that a mapped BAM file is needed for each sample. The reference genome used for the mapping must correspond to the one used to generate the 1KG GDS file. In this case, the 1KG GDS file is based on genome hg38.</p>
<div class="section level4">
<h4 id="extract-the-read-coverage-at-each-snp-position-in-the-1kg-gds-file">Extract the read coverage at each SNP position in the 1KG GDS file<a class="anchor" aria-label="anchor" href="#extract-the-read-coverage-at-each-snp-position-in-the-1kg-gds-file"></a>
</h4>
<p>We used here snp-pileup from Facet</p>
<p>First you a vcf file with the SNV you want to keep</p>
<p><br><br></p>
</div>
<div class="section level4">
<h4 id="create-a-vcf-file-containing-the-retained-snp-positions-from-1kg-gds-file">Create a VCF file containing the retained SNP positions from 1KG GDS file<a class="anchor" aria-label="anchor" href="#create-a-vcf-file-containing-the-retained-snp-positions-from-1kg-gds-file"></a>
</h4>
<p>The <strong>snvListVCF()</strong> function is used to generate a VCF file that contains the information of all retained SNPs from 1KG GDS fil:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load required package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file    </span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS1kg</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The VCF file that will be created</span></span>
<span><span class="va">fileOUT</span> <span class="op">&lt;-</span> <span class="st">"SNPretained.VCF"</span></span>
<span></span>
<span><span class="co">## Generate the VCF with the retained SNP position</span></span>
<span><span class="fu"><a href="../reference/snvListVCF.html">snvListVCF</a></span><span class="op">(</span>gds<span class="op">=</span><span class="va">gds</span>, fileOUT<span class="op">=</span><span class="va">fileOUT</span>, offset<span class="op">=</span><span class="fl">1</span>, freqCutoff<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Close the 1KG GDS file </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p>You should compress and indexing the newly created VCF file. Do do so, you need to install <a href="http://www.htslib.org/download/" class="external-link">HTSlib</a> <span class="citation">(Bonfield et al. 2021)</span>.</p>
<p>In a terminal:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This in not a R script</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="do">## This script is in bash</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Compress the new VCF file (fileOUT parameter)</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>bgzip fileOUT</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Index the new VCF file</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="do">## HTSlib software is needed</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>tabix <span class="sc">-</span>p vcf fileOUT.gz</span></code></pre></div>
<p>You need a <strong>ped</strong> file with the column:</p>
<p>“sample.id”* Id of the sample but the sample can be genotyped more than one “sex”<em> “pop.group”</em> can be self declare or something else but must be there “superPop”* can be self declare or something else but must be there “Sample.Type”* Ex Primary Tumor or Blood Derived Normal “Diagnosis”* C or N (cancer normal) this is for the sample not the patient “Source”* The tissue ex Ovary “Recurrent”<br>
“Case.ID” Patient identifiant<br>
“Name.ID”* This the unique id of the table “batch” Number</p>
<p>You need the extra code in the package Facet to extract the SNV <a href="https://github.com/mskcc/facets/tree/master/inst/extcode" class="external-link">snp-pileup</a> <span class="citation">(Shen and Seshan 2016)</span>.</p>
<p>The output from snp-pileup should be Name.ID.txt (snp-pileup add .gz to the filename)</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>    snp<span class="sc">-</span>pileup <span class="sc">-</span>g <span class="sc">-</span>d5000 <span class="sc">-</span>q15 <span class="sc">-</span>Q20  <span class="sc">-</span>r0 VCFGenerateAbove snvSel0.<span class="fl">01.</span>vcf.gz OUTPUT.txt FILEBAM.bam</span></code></pre></div>
<p>TODO change the path to something generic</p>
<p>in R:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">PATHGENO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"snpCancer"</span><span class="op">)</span></span>
<span>    <span class="va">PATHMETA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"metadata"</span><span class="op">)</span></span>
<span>    <span class="va">PATHGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"genoGDS1000gAF.All.TCGA_OV.WXS"</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">fileNamePED</span> <span class="op">&lt;-</span> <span class="st">"pedTCGA_OV_WXS_C.rds"</span></span>
<span>    <span class="va">fileNameGDS</span> <span class="op">&lt;-</span> <span class="st">"matGeno1000g.gds"</span></span>
<span>    </span>
<span>    <span class="va">filePED</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHMETA</span>, <span class="va">fileNamePED</span><span class="op">)</span></span>
<span>    <span class="va">fileNameGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PATHGDS</span>, <span class="va">fileNameGDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id <span class="op">=</span> <span class="st">"TCGA-OV.WXS.C"</span>,</span>
<span>                              study.desc <span class="op">=</span> <span class="st">"Ovarian example"</span>,</span>
<span>                              study.platform <span class="op">=</span> <span class="st">"WXS"</span>,</span>
<span>                              stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="../reference/appendStudy2GDS1KG.html">appendStudy2GDS1KG</a></span><span class="op">(</span>PATHGENO <span class="op">=</span> <span class="va">PATHGENO</span>,</span>
<span>                       fileNamePED <span class="op">=</span> <span class="va">filePED</span>,</span>
<span>                       fileNameGDS <span class="op">=</span> <span class="va">fileNameGDS</span>,</span>
<span>                       listSamples <span class="op">=</span> <span class="va">listSamples</span>,</span>
<span>                       studyDesc <span class="op">=</span> <span class="va">studyDF</span>,</span>
<span>                       batch <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>We select a subset of SNV pruned SNV We create a GDS for each sample with only the SNP in the pruning. It use more disk space but the analysis for the samples can be run simulatanusly with multiple process (if we we all samples in one file only one process can write in the GDS).</p>
<p>or in R for a s</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RAIDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># fileNameGDS with the sample and 1KG genotype</span></span>
<span>    <span class="co"># listSamples list of sample.id from the study</span></span>
<span>    <span class="co"># PATHSAMPLEGDS is the path where the gds specific </span></span>
<span>    <span class="co"># to the sample is created</span></span>
<span>    </span>
<span>    <span class="va">studyCur</span> <span class="op">&lt;-</span> <span class="st">"Study.name"</span></span>
<span>    <span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileNameGDS</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    </span>
<span>    <span class="va">sampleGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/index.gdsn.html" class="external-link">index.gdsn</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"sample.id"</span><span class="op">)</span></span>
<span>    <span class="va">sample.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/read.gdsn.html" class="external-link">read.gdsn</a></span><span class="op">(</span><span class="va">sampleGDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">study.offset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/read.gdsn.html" class="external-link">read.gdsn</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/index.gdsn.html" class="external-link">index.gdsn</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"study.offset"</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">study.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/read.gdsn.html" class="external-link">read.gdsn</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/index.gdsn.html" class="external-link">index.gdsn</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"study.annot/study.id"</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">sample.id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">study.id</span> <span class="op">==</span> <span class="va">studyCur</span><span class="op">)</span> <span class="op">+</span> <span class="va">study.offset</span><span class="op">]</span></span>
<span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">pruned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pruningSample.html">pruningSample</a></span><span class="op">(</span>gds<span class="op">=</span><span class="va">gds</span>,</span>
<span>                                            method<span class="op">=</span><span class="st">"corr"</span>,</span>
<span>                                            sampleCurrent <span class="op">=</span> <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                                            listSNP <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            slide.max.bp.v <span class="op">=</span> <span class="fl">5e5</span>,</span>
<span>                                            ld.threshold.v<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>,</span>
<span>                                            np <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                            verbose.v<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                                            chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            minAF.SuperPop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            keepGDSpruned <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                            PATHSAMPLEGDS <span class="op">=</span> <span class="va">PATHSAMPLEGDS</span>,</span>
<span>                                            keepFile <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="pre-processed-files-are-available">Pre-processed files are available<a class="anchor" aria-label="anchor" href="#pre-processed-files-are-available"></a>
</h2>
<p>Pre-processed files, such as the 1KG GDS file, are available at this address:</p>
<p><a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a></p>
<p>Beware that some of those files are voluminous.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.1.1 (2021-08-10)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] RAIDS_0.50.0     GENESIS_2.22.2   SNPRelate_1.26.0 gdsfmt_1.28.1   </span></span>
<span><span class="co">## [5] knitr_1.39       BiocStyle_2.20.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] nlme_3.1-158           bitops_1.0-7           fs_1.5.2              </span></span>
<span><span class="co">##   [4] bit64_4.0.5            GenomeInfoDb_1.28.4    rprojroot_2.0.3       </span></span>
<span><span class="co">##   [7] tools_4.1.1            backports_1.4.1        bslib_0.4.0           </span></span>
<span><span class="co">##  [10] utf8_1.2.2             R6_2.5.1               DBI_1.1.3             </span></span>
<span><span class="co">##  [13] BiocGenerics_0.38.0    mgcv_1.8-40            DNAcopy_1.66.0        </span></span>
<span><span class="co">##  [16] withr_2.5.0            tidyselect_1.1.2       bit_4.0.4             </span></span>
<span><span class="co">##  [19] compiler_4.1.1         textshaping_0.3.6      cli_3.3.0             </span></span>
<span><span class="co">##  [22] quantreg_5.94          Biobase_2.52.0         mice_3.14.0           </span></span>
<span><span class="co">##  [25] SparseM_1.81           desc_1.4.1             sandwich_3.0-2        </span></span>
<span><span class="co">##  [28] bookdown_0.27          sass_0.4.2             lmtest_0.9-40         </span></span>
<span><span class="co">##  [31] quantsmooth_1.58.0     pkgdown_2.0.6          systemfonts_1.0.4     </span></span>
<span><span class="co">##  [34] stringr_1.4.0          digest_0.6.29          GWASExactHW_1.01      </span></span>
<span><span class="co">##  [37] rmarkdown_2.14         XVector_0.32.0         pkgconfig_2.0.3       </span></span>
<span><span class="co">##  [40] htmltools_0.5.3        highr_0.9              fastmap_1.1.0         </span></span>
<span><span class="co">##  [43] rlang_1.0.4            RSQLite_2.2.15         jquerylib_0.1.4       </span></span>
<span><span class="co">##  [46] generics_0.1.3         zoo_1.8-10             jsonlite_1.8.0        </span></span>
<span><span class="co">##  [49] dplyr_1.0.9            RCurl_1.98-1.7         magrittr_2.0.3        </span></span>
<span><span class="co">##  [52] GenomeInfoDbData_1.2.6 Matrix_1.4-1           Rcpp_1.0.9            </span></span>
<span><span class="co">##  [55] S4Vectors_0.30.2       fansi_1.0.3            lifecycle_1.0.1       </span></span>
<span><span class="co">##  [58] pROC_1.18.0            stringi_1.7.8          yaml_2.3.5            </span></span>
<span><span class="co">##  [61] GWASTools_1.38.0       zlibbioc_1.38.0        MASS_7.3-58           </span></span>
<span><span class="co">##  [64] plyr_1.8.7             grid_4.1.1             formula.tools_1.7.1   </span></span>
<span><span class="co">##  [67] blob_1.2.3             parallel_4.1.1         crayon_1.5.1          </span></span>
<span><span class="co">##  [70] lattice_0.20-45        Biostrings_2.60.2      splines_4.1.1         </span></span>
<span><span class="co">##  [73] pillar_1.8.0           GenomicRanges_1.44.0   logistf_1.24.1        </span></span>
<span><span class="co">##  [76] codetools_0.2-18       stats4_4.1.1           glue_1.6.2            </span></span>
<span><span class="co">##  [79] evaluate_0.15          data.table_1.14.2      BiocManager_1.30.18   </span></span>
<span><span class="co">##  [82] operator.tools_1.6.3   vctrs_0.4.1            foreach_1.5.2         </span></span>
<span><span class="co">##  [85] MatrixModels_0.5-0     purrr_0.3.4            tidyr_1.2.0           </span></span>
<span><span class="co">##  [88] SeqArray_1.32.0        assertthat_0.2.1       cachem_1.0.6          </span></span>
<span><span class="co">##  [91] xfun_0.31              broom_1.0.0            SeqVarTools_1.30.0    </span></span>
<span><span class="co">##  [94] class_7.3-20           ragg_1.2.2             survival_3.3-1        </span></span>
<span><span class="co">##  [97] tibble_3.1.8           iterators_1.0.14       memoise_2.0.1         </span></span>
<span><span class="co">## [100] IRanges_2.26.0</span></span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Alexander2009" class="csl-entry">
Alexander, D. H., J. Novembre, and K. Lange. 2009. <span>“Fast Model-Based Estimation of Ancestry in Unrelated Individuals.”</span> <em>Genome Res</em> 19 (9): 1655–64. <a href="https://doi.org/10.1101/gr.094052.109" class="external-link">https://doi.org/10.1101/gr.094052.109</a>.
</div>
<div id="ref-Bonfield2021" class="csl-entry">
Bonfield, James K, John Marshall, Petr Danecek, Heng Li, Valeriu Ohan, Andrew Whitwham, Thomas Keane, and Robert M Davies. 2021. <span>“<span class="nocase">HTSlib: C library for reading/writing high-throughput sequencing data</span>.”</span> <em>GigaScience</em> 10 (2). <a href="https://doi.org/10.1093/gigascience/giab007" class="external-link">https://doi.org/10.1093/gigascience/giab007</a>.
</div>
<div id="ref-Price2006" class="csl-entry">
Price, A. L., N. J. Patterson, R. M. Plenge, M. E. Weinblatt, N. A. Shadick, and D. Reich. 2006. <span>“Principal Components Analysis Corrects for Stratification in Genome-Wide Association Studies.”</span> <em>Nat Genet</em> 38 (8): 904–9. <a href="https://doi.org/10.1038/ng1847" class="external-link">https://doi.org/10.1038/ng1847</a>.
</div>
<div id="ref-Pritchard2000" class="csl-entry">
Pritchard, J. K., M. Stephens, and P. Donnelly. 2000. <span>“Inference of Population Structure Using Multilocus Genotype Data.”</span> <em>Genetics</em> 155 (2): 945–59. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10835412" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/10835412</a>.
</div>
<div id="ref-Shen2016" class="csl-entry">
Shen, R, and V E Seshan. 2016. <span>“<span class="nocase">FACETS: allele-specific copy number and clonal heterogeneity analysis tool for high-throughput DNA sequencing</span>.”</span> <em>Nucleic Acids Res</em> 44 (16): e131. <a href="https://doi.org/10.1093/nar/gkw520" class="external-link">https://doi.org/10.1093/nar/gkw520</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
