<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genetic Ancestry Inference Step by Step (optional) • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Genetic Ancestry Inference Step by Step (optional)">
<meta property="og:description" content="RAIDS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Ancestry_Inference_Step_by_Step.html">Genetic Ancestry Inference Step by Step (optional)</a>
    </li>
    <li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from the population reference dataset (optional)</a>
    </li>
    <li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genetic Ancestry Inference Step by Step
(optional)</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/HEAD/vignettes/Ancestry_Inference_Step_by_Step.Rmd" class="external-link"><code>vignettes/Ancestry_Inference_Step_by_Step.Rmd</code></a></small>
      <div class="hidden name"><code>Ancestry_Inference_Step_by_Step.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 0.99.6<br><strong>Compiled date</strong>: 2023-08-15<br><strong>License</strong>: Apache License (&gt;= 2)<br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This is an overview of genetic ancestry inference from cancer-derived
molecular data:</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_v03.png" alt="An overview of the genetic ancestry inference process." width="120%"><p class="caption">
An overview of the genetic ancestry inference process.
</p>
</div>
<p>The main steps are:</p>
<ol style="list-style-type: decimal">
<li>Format reference data from the 1000 Genomes (1KG) (optional)</li>
<li>Format cancer-derived data set starting from BAM files</li>
<li>Optimize ancestry inference parameters</li>
<li>Infer ancestry for the subjects of the external study</li>
</ol>
<p>All steps from Step 2, sub-step 4, to Step 4 can be run through a
wrapper (as described in <a href="RAIDS.html">Main vignette</a>. Those
steps can also be run separately; this has the advantage of allowing
finer parameter selection.</p>
<p>This vignette describes, in details, the process done by the wrapper
from Step 2, sub-step 4, to Step 4.</p>
<p><br><br></p>
<div class="section level3">
<h3 id="step-1---format-reference-data-from-the-1000-genomes-optional">Step 1 - Format reference data from the 1000 Genomes (optional)<a class="anchor" aria-label="anchor" href="#step-1---format-reference-data-from-the-1000-genomes-optional"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step1_v03.png" alt="Step 1 - Formatting the information from the population reference dataset (optional)" width="120%"><p class="caption">
Step 1 - Formatting the information from the population reference
dataset (optional)
</p>
</div>
<p>The detail execution of this step is presented in the <a href="Create_1KG_GDS_file.html">Formatting the information from 1000
Genomes (optional)</a> vignette.</p>
<p>In addition, the available pre-processed files are described in the
<a href="RAIDS.html">Main vignette</a>.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step-2---prepare-cancer-derived-data-for-ancestry-inference">Step 2 - Prepare cancer-derived data for ancestry inference<a class="anchor" aria-label="anchor" href="#step-2---prepare-cancer-derived-data-for-ancestry-inference"></a>
</h3>
<p>Molecular profiles in a cancer-derived data set must be formatted
following a series of sub-steps.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step2_v03.png" alt="Step 2 - Formatting the information from an external study" width="120%"><p class="caption">
Step 2 - Formatting the information from an external study
</p>
</div>
<p><br></p>
<p>These are:</p>
<ol style="list-style-type: decimal">
<li>Create a directory containing the 3 reference files from 1KG</li>
<li>Make a SNP pileup file for the profile</li>
<li>Create an RDS file containing information about the samples</li>
<li>Create a Profile GDS file (one GDS file per profile)</li>
<li>Generate a pruned subset of the single nucleotide variants
(SNVs)</li>
<li>Estimate the allelic frequency for the pruned SNVs</li>
</ol>
<p>The first steps, from Sub-step 1 to Sub-step 3, are described in the
<a href="RAIDS.html">Main vigette</a>.</p>
<div class="section level4">
<h4 id="sub-step-2-4-create-a-profile-gds-file-one-gds-file-per-profile">Sub-Step 2.4 Create a Profile GDS file (one GDS file per
profile)<a class="anchor" aria-label="anchor" href="#sub-step-2-4-create-a-profile-gds-file-one-gds-file-per-profile"></a>
</h4>
<p>This step requires 3 files as input:</p>
<ul>
<li>The <strong>1KG GDS file</strong>
</li>
<li>The <strong>Profile SNP pileup file</strong> (one per profile
present in the study)</li>
<li>The <strong>Profile PED RDS file</strong> (one file with information
for all profiles in the study)</li>
</ul>
<p>A <em>data.frame</em> containing the general information about the
study is also required. The <em>data.frame</em> must contain those 3
columns:</p>
<ul>
<li>
<strong>study.id</strong>: The study identifier (example:
TCGA-BRCA).</li>
<li>
<strong>study.desc</strong>: The description of the study.</li>
<li>
<strong>study.platform</strong>: The type of sequencing (example:
RNA-seq).</li>
</ul>
<p>Using all those inputs, the <em>createStudy2GDS1KG()</em> function
will generate a <strong>Profile GDS file</strong>. One <strong>Profile
GDS file</strong> is created for each profile passed to the
<em>listProfiles</em> argument.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#################################################################</span></span>
<span><span class="co">## Load required package</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## The 1KG GDS file location</span></span>
<span><span class="co">## Note that the 1KG GDS file used for this example is a  </span></span>
<span><span class="co">## simplified version and CANNOT be used for any real analysis</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span><span class="va">path1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"gdsRef"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"ex1kg.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## The Profile SNP pileup files (one per profile) need  </span></span>
<span><span class="co">## to be located in the same directory.</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="va">pathGeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"snpPileup"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## The path where the Profile GDS files (one per profile)</span></span>
<span><span class="co">## will be created need to be specified.</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="va">pathProfileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"out"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## The path and file name for the PED RDS file</span></span>
<span><span class="co">## will the information about the analyzed samples</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="va">filePED</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"pedEx.rds"</span><span class="op">)</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">filePED</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span></span>
<span><span class="co">##     Name.ID Case.ID   Sample.Type Diagnosis     Source</span></span>
<span><span class="co">## ex1     ex1     ex1 Primary Tumor    Cancer Databank B</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## A data frame containing general information about the study</span></span>
<span><span class="co">## is also required. The data frame must have</span></span>
<span><span class="co">## those 3 columns: "study.id", "study.desc", "study.platform"</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id<span class="op">=</span><span class="st">"MYDATA"</span>,</span>
<span>                        study.desc<span class="op">=</span><span class="st">"Description"</span>,</span>
<span>                        study.platform<span class="op">=</span><span class="st">"PLATFORM"</span>,</span>
<span>                        stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## The list of profiles to analyzed is passed to the function.</span></span>
<span><span class="co">## The profiles must be present in the Profile PED RDS file see</span></span>
<span><span class="co">## sub-step 4 and must have an associated Profile SNP pileup file.</span></span>
<span><span class="co">## Not all profiles present in the Profile PED file need to </span></span>
<span><span class="co">## be selected.</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="va">listSamples</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span>, <span class="st">"Name.ID"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## This function creates one Profile GDS file for each</span></span>
<span><span class="co">## profile present in the 'listProfiles' parameter.</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="fu"><a href="../reference/createStudy2GDS1KG.html">createStudy2GDS1KG</a></span><span class="op">(</span>pathGeno<span class="op">=</span><span class="va">pathGeno</span>,</span>
<span>                        pedStudy<span class="op">=</span><span class="va">ped</span>,</span>
<span>                        fileNameGDS<span class="op">=</span><span class="va">gds1KG</span>,</span>
<span>                        listProfiles<span class="op">=</span><span class="va">listSamples</span>,</span>
<span>                        studyDF<span class="op">=</span><span class="va">studyDF</span>,</span>
<span>                        pathProfileGDS<span class="op">=</span><span class="va">pathProfileGDS</span>,</span>
<span>                        genoSource<span class="op">=</span><span class="st">"snp-pileup"</span><span class="op">)</span></span>
<span><span class="co">## [1] 0</span></span>
<span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="co">## The Profile GDS file has been created in the </span></span>
<span><span class="co">## directory pathProfileGDS using the name of the sample (ex1)</span></span>
<span><span class="co">#################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span><span class="co">## [1] "ex1.gds"</span></span></code></pre></div>
<p><br></p>
<p>A <strong>Profile GDS file</strong> is created for each profile. The
GDS file contains those nodes:</p>
<div class="figure" style="text-align: left">
<img src="ProfileGDSdemo_when_created.png" alt="Step 2.4 - The nodes present in the newly created Profile GDS file (one GDS file per profile)." width="100%"><p class="caption">
Step 2.4 - The nodes present in the newly created Profile GDS file (one
GDS file per profile).
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-2-5-generate-a-pruned-subset-of-the-single-nucleotide-variants-snvs">Sub-Step 2.5 Generate a pruned subset of the single nucleotide
variants (SNVs)<a class="anchor" aria-label="anchor" href="#sub-step-2-5-generate-a-pruned-subset-of-the-single-nucleotide-variants-snvs"></a>
</h4>
<p>The initial list of 1KG SNVs is pruned, using linkage disequilibrium
analysis, and a profile-specific subset of SNVs is retained for each
profile. This information is added to the <strong>Profile GDS
file</strong>.</p>
<p>The <strong>pruningSample()</strong> function requires the
<strong>1KG GDS file</strong> as input. It also requires the path to the
<strong>Profile GDS file(s)</strong>.</p>
<p>Note that this step can require large disk space.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required package</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The 1KG GDS file is required (demo version)</span></span>
<span><span class="co">## Note that the 1KG GDS file used for this example is a  </span></span>
<span><span class="co">## simplified version and CANNOT be used for any real analyses</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">path1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"gdsRef"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"ex1kg.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file (demo version)</span></span>
<span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The pruning function is called with one profile as input at the time</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Compute the SNV pruned subset</span></span>
<span>    <span class="co">## studyID: Study identifier as defined in the preceding sub-step. </span></span>
<span>    <span class="co">##           The study identifier must be the same that the one present in</span></span>
<span>    <span class="co">##           Profile GDS file.</span></span>
<span>    <span class="co">## pathProfileGDS: All Profile GDS files must be in the same directory. </span></span>
<span>    <span class="fu"><a href="../reference/pruningSample.html">pruningSample</a></span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>,</span>
<span>                    currentProfile<span class="op">=</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                    studyID<span class="op">=</span><span class="va">studyDF</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                    pathProfileGDS<span class="op">=</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co">## Profile GDS file for the current profile</span></span>
<span>    <span class="co">## The file name corresponds to the path + profile identifier + ".gds"</span></span>
<span>    <span class="va">fileGDSProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co">## Add the genotype information for the list of pruned SNVs </span></span>
<span>    <span class="co">## into the Profile GDS file</span></span>
<span>    <span class="co">## The genotype information is extracted from the 1KG GDS file</span></span>
<span>    <span class="fu"><a href="../reference/add1KG2SampleGDS.html">add1KG2SampleGDS</a></span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>, fileProfileGDS<span class="op">=</span><span class="va">fileGDSProfile</span>,</span>
<span>                        currentProfile<span class="op">=</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                        studyID<span class="op">=</span><span class="va">studyDF</span><span class="op">$</span><span class="va">study.id</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Add annotation from the 1KG GDS file to the Profile GDS file</span></span>
<span>    <span class="co">## This is required.</span></span>
<span>    <span class="fu"><a href="../reference/addStudy1Kg.html">addStudy1Kg</a></span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>, fileProfileGDS<span class="op">=</span><span class="va">fileGDSProfile</span>, </span>
<span>                        verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>    </span>
<span><span class="co">## Close the 1KG GDS file (it is important to always close the GDS files)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>The nodes, with the genotype information for the pruned SNVs, are
added into<br>
the <strong>Profile GDS file</strong>:</p>
<div class="figure" style="text-align: left">
<img src="ProfileGDSdemo_after_add1KG2SampleGDS.png" alt="Step 2.5 - The nodes added to the Profile GDS file are in light blue." width="100%"><p class="caption">
Step 2.5 - The nodes added to the Profile GDS file are in light blue.
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-2-6-estimate-the-allelic-fraction-for-the-pruned-snvs">Sub-Step 2.6 Estimate the allelic fraction for the pruned SNVs<a class="anchor" aria-label="anchor" href="#sub-step-2-6-estimate-the-allelic-fraction-for-the-pruned-snvs"></a>
</h4>
<p>The <strong>estimateAllelicFraction()</strong> estimates the allele
fraction for all SNVs present in the pruned SNV list. Note that the
function requires different inputs for DNA and RNA profiles.</p>
<p>For the DNA profiles, these 2 files are required:</p>
<ol style="list-style-type: decimal">
<li>The <strong>Profile GDS file</strong>
</li>
<li>The <strong>1KG GDS file</strong>
</li>
</ol>
<p>For the RNA profiles, these 3 files are required:</p>
<ol style="list-style-type: decimal">
<li>The <strong>Profile GDS file</strong>
</li>
<li>The <strong>1KG GDS file</strong>
</li>
<li>The <strong>1KG SNV Annotation GDS file</strong>
</li>
</ol>
<p>In both cases, the other required inputs are:</p>
<ol style="list-style-type: decimal">
<li>The information about the length of the chromosomes</li>
<li>The profile identifier (it corresponds to the Profile GDS file
name)</li>
<li>The study identifier (it should correspond to the one used
previously)</li>
</ol>
<p>The information about the length of the chromosomes must be assigned
into a <em>vector</em> object. This is an example on how to retrieve the
information. There are alternative ways to retrieve this information,
e.g.,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###################################################################</span></span>
<span><span class="co">## Load required library</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## The length of each chromosome is required</span></span>
<span><span class="co">## Chromosomes X, Y and M need relabeling (see below)</span></span>
<span><span class="co">## There are alternative ways to retrieve this information</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">Hsapiens</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span></code></pre></div>
<p><br><br></p>
<p>The <strong>estimateAllelicFraction()</strong> function processes one
profile at the time, as shown in this example.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#####################################################################</span></span>
<span><span class="co">## Load the required packages</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The 1KG GDS file is required</span></span>
<span><span class="co">## The 1KG SNV Annotation GDS file is only required for RNA profiles</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="va">path1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"gdsRef"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">file1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"ex1kg.gds"</span><span class="op">)</span></span>
<span><span class="va">fileAnnot1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"exAnnot1kg.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file</span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file1KG</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The information about the length of the chromosomes</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">chrInfo</span><span class="op">)</span></span>
<span><span class="co">##      chr1      chr2      chr3      chr4      chr5      chr6 </span></span>
<span><span class="co">## 248956422 242193529 198295559 190214555 181538259 170805979</span></span>
<span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The function must be called for each profile</span></span>
<span><span class="co">## This example only uses one profile</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The first profile is used in the demo</span></span>
<span><span class="va">profileName</span> <span class="op">&lt;-</span> <span class="va">listSamples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The Profile GDS file is required</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The name must correspond to the profile identifier</span></span>
<span><span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">profileName</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open the Profile GDS file in writing mode</span></span>
<span><span class="va">gdsProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileProfile</span>, readonly<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## The estimation of the allelic fraction</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="co">## Estimate the allele fraction of the pruned SNVs</span></span>
<span><span class="co">## The current example is for a DNA sample</span></span>
<span><span class="co">## In the case of RNA sample, the function needs different inputs</span></span>
<span><span class="co">## such as the 1KG Annotation GDS file and</span></span>
<span><span class="co">## the 'blockID' should be as listed in the 1KG Annotation GDS file </span></span>
<span><span class="co">## for the gene annotation of the SNVs</span></span>
<span><span class="fu"><a href="../reference/estimateAllelicFraction.html">estimateAllelicFraction</a></span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds</span>, gdsProfile<span class="op">=</span><span class="va">gdsProfile</span>,</span>
<span>                                        currentProfile<span class="op">=</span><span class="va">profileName</span>,</span>
<span>                                        studyID<span class="op">=</span><span class="va">studyDF</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                                        chrInfo<span class="op">=</span><span class="va">chrInfo</span><span class="op">)</span></span>
<span><span class="co">## [1] 0</span></span>
<span></span>
<span><span class="co">## Close both GDS files (important)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsProfile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>This step must be executed for each profile present in the study.</p>
<p>The information is added into the ‘lap’ and ‘segment’ nodes:</p>
<div class="figure" style="text-align: left">
<img src="ProfileGDSdemo_after_estimateAllelicFraction.png" alt="Step 2.6 - The information is added into the nodes in light blue." width="100%"><p class="caption">
Step 2.6 - The information is added into the nodes in light blue.
</p>
</div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-3---optimize-the-ancestry-inference-parameters">Step 3 - Optimize the ancestry inference parameters<a class="anchor" aria-label="anchor" href="#step-3---optimize-the-ancestry-inference-parameters"></a>
</h3>
<p>At this step, optimization of the parameters is required to maximize
the the ancestry inference accuracy (next step).</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step3_v03.png" alt="Step 3 - Find the optimized parameters for the ancestry inference" width="120%"><p class="caption">
Step 3 - Find the optimized parameters for the ancestry inference
</p>
</div>
<p>Two inference parameters to be optimized:</p>
<ul>
<li>
<em>K</em>: the number of neighbors used to call the ancestry</li>
<li>
<em>D</em>: the number of PCA components retained</li>
</ul>
<p>The accuracy is evaluated using a synthetic data set created from
merging one cancer profile with multiple population referecne profiles
of known ancestry. Using the synthetic profiles, a range of <em>K</em>
and <em>D</em> values are tested. Through that process, the <em>K</em>
and <em>D</em> values are tuned to maximize accuracy.</p>
<p>This step consists of two sub-steps:</p>
<ol style="list-style-type: decimal">
<li>Generate the synthetic dataset</li>
<li>Compute the PCA-KNN ancestry call for each synthetic profile</li>
</ol>
<p><br></p>
<div class="section level4">
<h4 id="sub-step-3-1-generate-the-synthetic-dataset">Sub-Step 3.1 Generate the synthetic dataset<a class="anchor" aria-label="anchor" href="#sub-step-3-1-generate-the-synthetic-dataset"></a>
</h4>
<p>A synthetic profile is generated through the merging of one cancer
profile with one 1KG profile of known ancestry. Multiple 1KG profiles of
different ancestry are required to create a synthetic data set that will
be able to show the specific accuracy for each super-population. All the
synthetic profiles are saved in the <strong>Profile GDS file</strong>
corresponding to the cancer profile used to generate the synthetic
dataset.</p>
<p>In summary, a fixed number of profiles for each super-population are
extracted from the 1KG study. The information is saved into the
<strong>Profile GDS file</strong> associated to the selected cancer
profile. A synthetic profile is created for each combination of one 1KG
profile and cancer profile. All synthetic profiles are then saved into
the <strong>Profile GDS file</strong>.</p>
<p>The three functions <em>select1KGPop()</em>, <em>prepSynthetic()</em>
and <em>syntheticGeno()</em> are required to process the synthetic data
synthesis step.</p>
<p>These 3 files are required:</p>
<ol style="list-style-type: decimal">
<li>The <strong>Profile GDS file</strong>
</li>
<li>The <strong>1KG GDS file</strong>
</li>
<li>The <strong>1KG Annotation GDS file</strong>
</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Randomly extract a fixed number of profiles for each</span></span>
<span><span class="co">## subcontinental population present in the 1KG GDS file.</span></span>
<span><span class="co">## When not enough profiles are available, all profiles are selected.  </span></span>
<span><span class="co">####################################################################   </span></span>
<span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">file1KG</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Fix seed to ensure reproducible results</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span>   </span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Select the profiles from 1KG for the synthetic data.</span></span>
<span><span class="co">## Here we select 2 profiles from 1KG for each subcontinental-level</span></span>
<span><span class="co">## Normally, we use 30 profiles from 1KG for each </span></span>
<span><span class="co">## subcontinental-level but it is too big for the example.</span></span>
<span><span class="co">## The 1KG GDS file in this example only has 6 profiles for each </span></span>
<span><span class="co">## subcontinental-level (for demo purpose only)</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds1KG</span>, nbProfiles<span class="op">=</span><span class="fl">2L</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the list of selected 1KG sample identifiers</span></span>
<span><span class="va">listProfileRef</span> <span class="op">&lt;-</span> <span class="va">dataRef</span><span class="op">$</span><span class="va">sample.id</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## A data.frame with the description of the study for the synthetic</span></span>
<span><span class="co">## data is required.</span></span>
<span><span class="co">## The column names must be as shown</span></span>
<span><span class="co">#################################################################### </span></span>
<span><span class="va">syntheticStudyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id<span class="op">=</span><span class="st">"MYDATA.Synthetic"</span>,</span>
<span>                          study.desc<span class="op">=</span><span class="st">"MYDATA synthetic data"</span>,</span>
<span>                          study.platform<span class="op">=</span><span class="st">"PLATFORM"</span>,</span>
<span>                          stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The Profile GDS file is needed</span></span>
<span><span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">profileName</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The prepSynthetic() function prepares the annotation for</span></span>
<span><span class="co">## the synthetic data</span></span>
<span><span class="co">## The information is saved into the Profile GDS file</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="fu"><a href="../reference/prepSynthetic.html">prepSynthetic</a></span><span class="op">(</span>fileProfileGDS<span class="op">=</span><span class="va">fileProfile</span>,</span>
<span>                listSampleRef<span class="op">=</span><span class="va">listProfileRef</span>,</span>
<span>                profileID<span class="op">=</span><span class="va">profileName</span>,</span>
<span>                studyDF<span class="op">=</span><span class="va">syntheticStudyDF</span>,</span>
<span>                prefix<span class="op">=</span><span class="st">"1"</span><span class="op">)</span></span>
<span><span class="co">## [1] 0</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Both the 1KG GDS file and the 1KG Annotation GDS file</span></span>
<span><span class="co">## are required</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">path1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"gdsRef"</span><span class="op">)</span></span>
<span><span class="va">fileRefAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path1KG</span>, <span class="st">"exAnnot1kg.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Open  1KG Annotation GDS file    </span></span>
<span><span class="va">gdsRefAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileRefAnnot</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The syntheticGeno() function generates the synthetic profiles.</span></span>
<span><span class="co">## The synthetic profiles are saved in the Profile GDS file</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">resG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/syntheticGeno.html">syntheticGeno</a></span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gds1KG</span>,</span>
<span>                        gdsRefAnnot<span class="op">=</span><span class="va">gdsRefAnnot</span>,</span>
<span>                        fileProfileGDS<span class="op">=</span><span class="va">fileProfile</span>,</span>
<span>                        profileID<span class="op">=</span><span class="va">profileName</span>,</span>
<span>                        listSampleRef<span class="op">=</span><span class="va">listProfileRef</span>,</span>
<span>                        prefix<span class="op">=</span><span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Close both GDS files</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRefAnnot</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sub-step-3-2-perform-the-pca-knn-ancestry-call-for-each-synthetic-profile">Sub-Step 3.2 Perform the PCA-KNN ancestry call for each synthetic
profile<a class="anchor" aria-label="anchor" href="#sub-step-3-2-perform-the-pca-knn-ancestry-call-for-each-synthetic-profile"></a>
</h4>
<p>The ancestry is inferred for each synthetic profile. As the ancestry
of origin of the 1KG profile used to generate the synthetic profile is
known, the accuracy of the calls will be assessed for different
parameters.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#####################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span>   </span>
<span>    </span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The 1KG GDS file is required</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## Open the 1KG GDS file     </span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file1KG</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The path to the directory where the PCA results will be saved </span></span>
<span><span class="co">## in RDS files.</span></span>
<span><span class="co">## The directory must exist.</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="va">pathOut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## Get the super-population information (known ancestry) for the </span></span>
<span><span class="co">## reference profiles. This is the ground truth for the 1KG profiles.</span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="va">refKnownSuperPop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRef1KGPop.html">getRef1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, <span class="st">"superPop"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## Fix the RNG seed as in step 6 to ensure the same results</span></span>
<span><span class="co">#####################################################################    </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span>       </span>
<span>    </span>
<span><span class="co">## Select the 1KG samples used to generate the synthetic dataset</span></span>
<span><span class="co">## Already done in step 6, no need to repeat if the results have been saved</span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds</span>, nbProfiles<span class="op">=</span><span class="fl">2L</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#####################################################################</span></span>
<span><span class="co">## The function splitSelectByPop() generates a matrix with the</span></span>
<span><span class="co">## reference samples split by sub-continental population</span></span>
<span><span class="co">#####################################################################  </span></span>
<span><span class="va">sampleRM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/splitSelectByPop.html">splitSelectByPop</a></span><span class="op">(</span><span class="va">dataRef</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## Loop for all cancer samples with associated synthetic data</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="co">## The Profile GDS file associated to the cancer profile     </span></span>
<span>    <span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>,</span>
<span>                                    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## A sub-directory is created for the cancer sample</span></span>
<span>    <span class="co">## Beware that the number of files created will correspond to the </span></span>
<span>    <span class="co">## number of rows in the sampleRM matrix</span></span>
<span>    <span class="va">pathOutProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">pathOutProfile</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathOutProfile</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co">## Open the Profile GDS file      </span></span>
<span>    <span class="va">gdsProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileProfile</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## For each row of the sampleRM matrix      </span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sampleRM</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co">## Run a PCA analysis using 1 synthetic profile from each </span></span>
<span>        <span class="co">##  sub-continental ancestry</span></span>
<span>        <span class="co">## The synthetic profiles are projected on the 1KG PCA space </span></span>
<span>        <span class="co">##  (the 1KG reference profiles used to generate the synthetic profiles </span></span>
<span>        <span class="co">##  are removed from this PCA)</span></span>
<span>        <span class="co">## The K-nearest neighbors analysis is done using </span></span>
<span>        <span class="co">##  a range of K and D values</span></span>
<span>        <span class="va">syntKNN</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computePoolSyntheticAncestryGr.html">computePoolSyntheticAncestryGr</a></span><span class="op">(</span>gdsProfile<span class="op">=</span><span class="va">gdsProfile</span>,</span>
<span>                                        sampleRM<span class="op">=</span><span class="va">sampleRM</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span>,</span>
<span>                                        studyIDSyn<span class="op">=</span><span class="va">syntheticStudyDF</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                                        np<span class="op">=</span><span class="fl">4L</span>,</span>
<span>                                        spRef<span class="op">=</span><span class="va">refKnownSuperPop</span>,</span>
<span>                                        eigenCount<span class="op">=</span><span class="fl">15L</span><span class="op">)</span></span>
<span>        </span>
<span>        <span class="co">## Results are saved</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">syntKNN</span><span class="op">$</span><span class="va">matKNN</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOutProfile</span>, </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"KNN.synt."</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"."</span>, <span class="va">j</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co">## Close Sample GDS file (important)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsProfile</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span>    </span>
<span><span class="co">## Close 1KG GDS file (important)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-4---run-the-ancestry-inference-in-the-input-data-set">Step 4 - Run the ancestry inference in the input data set<a class="anchor" aria-label="anchor" href="#step-4---run-the-ancestry-inference-in-the-input-data-set"></a>
</h3>
<p>The ancestry inference is done with the optimized <em>K</em> and
<em>D</em> parameters. More specifically, a PCA is generated using the
1KG reference samples and the cancer sample. The <em>D</em> parameter
specifies the number of dimension for the PCA. Then, the ancestry of the
cancer sample is inferred using a k-nearest neighbors classification
method. The <em>K</em> parameter specifies the number of neighbors used
for the classification.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step4_v02.png" alt="Step 4 - Run the ancestry inference on the external study" width="120%"><p class="caption">
Step 4 - Run the ancestry inference on the external study
</p>
</div>
<p><br></p>
<p>The PCA of the sample and KNN sample and Call the ancestry with the
optimal <em>K</em> and <em>D</em> parameters.</p>
<p>Note: The formal selection of <em>K</em> and <em>D</em> parameters is
done at this step but all the synthetic data are prepared in the
previous step.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span>   </span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The reference 1KG GDS file is required</span></span>
<span><span class="co">####################################################################</span></span>
<span></span>
<span><span class="co">## Open the 1KG GDS file </span></span>
<span><span class="va">gdsReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">file1KG</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## A directory where result files are going to be saved, </span></span>
<span><span class="co">## where a sub-directory will be set up for each input profile </span></span>
<span><span class="va">pathOut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## The getRef1KGPop() function extract the known super-population </span></span>
<span><span class="co">## of the reference samples.</span></span>
<span><span class="co">## We expect the ancestry call from a synthetic profile to </span></span>
<span><span class="co">## correspond to the known ancestry of the reference sample used to </span></span>
<span><span class="co">## synthesize it.</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">refKnownSuperPop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRef1KGPop.html">getRef1KGPop</a></span><span class="op">(</span><span class="va">gdsReference</span>, <span class="st">"superPop"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## Loop on each profile </span></span>
<span><span class="co">## Can also be run in parallel or on different clusters...</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    </span>
<span>    <span class="co">## Extract the GDS file name and path for the current profile      </span></span>
<span>    <span class="va">fileProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".gds"</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co">## Directory where the KNN results of the synthetic profiles have been saved      </span></span>
<span>    <span class="va">pathKNN</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, <span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">listFilesName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathKNN</span><span class="op">)</span>, <span class="st">".rds"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## List of the KNN result files from PCA run on synthetic data</span></span>
<span>    <span class="va">listFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathKNN</span><span class="op">)</span> , <span class="va">listFilesName</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Open the Profile GDS file</span></span>
<span>    <span class="va">gdsProfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileProfile</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Select the optimal K and D parameters from the synthetic data results</span></span>
<span>    <span class="co">## Use those parameter to infer the ancestry of the specific profile</span></span>
<span>    <span class="va">resCall</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeAncestryFromSyntheticFile.html">computeAncestryFromSyntheticFile</a></span><span class="op">(</span>gdsReference<span class="op">=</span><span class="va">gdsReference</span>, </span>
<span>                                    gdsProfile<span class="op">=</span><span class="va">gdsProfile</span>,</span>
<span>                                    listFiles<span class="op">=</span><span class="va">listFiles</span>,</span>
<span>                                    currentProfile<span class="op">=</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>                                    spRef<span class="op">=</span><span class="va">refKnownSuperPop</span>,</span>
<span>                                    studyIDSyn<span class="op">=</span><span class="va">syntheticStudyDF</span><span class="op">$</span><span class="va">study.id</span>,</span>
<span>                                    np<span class="op">=</span><span class="fl">1L</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">resCall</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>,</span>
<span>                                <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".infoCall"</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">resCall</span><span class="op">$</span><span class="va">Ancestry</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">listSamples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".Ancestry"</span>,<span class="st">".csv"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            quote<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co">## Close the Profile GDS file (important)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsProfile</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Close the 1KG GDS file (important)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsReference</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="co">## Show the ancestry inference (SuperPop) and </span></span>
<span><span class="co">## optimal number of PCA components D</span></span>
<span><span class="co">## optimal number of neighbours K</span></span>
<span><span class="co">####################################################################</span></span>
<span><span class="va">resAncestry</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, </span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ped</span><span class="op">$</span><span class="va">Name.ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">".Ancestry.csv"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">resAncestry</span></span>
<span><span class="co">##   sample.id D K SuperPop</span></span>
<span><span class="co">## 1       ex1 7 8      EAS</span></span>
<span></span>
<span><span class="co">## Clean-up demo files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">fileProfile</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathOut</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The <em>computeAncestryFromSyntheticFile()</em> function generates 3
types of files in the <em>OUTPUT</em> directory.</p>
<ul>
<li>The ancestry inference CSV file (“.Ancestry.csv” file)</li>
<li>The inference information RDS file (“.infoCall.rds” file)</li>
<li>The parameter information RDS files from the synthetic inference
(“KNN.synt.*.rds” files in a sub-directory)</li>
</ul>
<p>In addition, a sub-directory (named using the <em>profile ID</em>) is
also created.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> in the environment
in which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] Matrix_1.6-0                      BSgenome.Hsapiens.UCSC.hg38_1.4.5</span></span>
<span><span class="co">##  [3] BSgenome_1.68.0                   rtracklayer_1.60.0               </span></span>
<span><span class="co">##  [5] Biostrings_2.68.1                 XVector_0.40.0                   </span></span>
<span><span class="co">##  [7] GenomicRanges_1.52.0              GenomeInfoDb_1.36.1              </span></span>
<span><span class="co">##  [9] IRanges_2.34.1                    S4Vectors_0.38.1                 </span></span>
<span><span class="co">## [11] BiocGenerics_0.46.0               RAIDS_0.99.6                     </span></span>
<span><span class="co">## [13] GENESIS_2.30.0                    SNPRelate_1.34.1                 </span></span>
<span><span class="co">## [15] gdsfmt_1.36.1                     knitr_1.43                       </span></span>
<span><span class="co">## [17] BiocStyle_2.28.0                 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.7              shape_1.4.6                </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              jomo_2.7-6                 </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.52.1      logistf_1.25.0             </span></span>
<span><span class="co">##   [7] nloptr_2.0.3                rmarkdown_2.24             </span></span>
<span><span class="co">##   [9] fs_1.6.3                    BiocIO_1.10.0              </span></span>
<span><span class="co">##  [11] zlibbioc_1.46.0             ragg_1.2.5                 </span></span>
<span><span class="co">##  [13] vctrs_0.6.3                 memoise_2.0.1              </span></span>
<span><span class="co">##  [15] minqa_1.2.5                 Rsamtools_2.16.0           </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.12             quantsmooth_1.66.0         </span></span>
<span><span class="co">##  [19] htmltools_0.5.6             S4Arrays_1.0.5             </span></span>
<span><span class="co">##  [21] progress_1.2.2              curl_5.0.2                 </span></span>
<span><span class="co">##  [23] broom_1.0.5                 pROC_1.18.4                </span></span>
<span><span class="co">##  [25] mitml_0.4-5                 sass_0.4.7                 </span></span>
<span><span class="co">##  [27] bslib_0.5.1                 desc_1.4.2                 </span></span>
<span><span class="co">##  [29] plyr_1.8.8                  sandwich_3.0-2             </span></span>
<span><span class="co">##  [31] zoo_1.8-12                  cachem_1.0.8               </span></span>
<span><span class="co">##  [33] GenomicAlignments_1.36.0    lifecycle_1.0.3            </span></span>
<span><span class="co">##  [35] iterators_1.0.14            pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [37] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">##  [39] GenomeInfoDbData_1.2.10     MatrixGenerics_1.12.3      </span></span>
<span><span class="co">##  [41] digest_0.6.33               GWASTools_1.46.0           </span></span>
<span><span class="co">##  [43] AnnotationDbi_1.62.2        rprojroot_2.0.3            </span></span>
<span><span class="co">##  [45] textshaping_0.3.6           RSQLite_2.3.1              </span></span>
<span><span class="co">##  [47] filelock_1.0.2              fansi_1.0.4                </span></span>
<span><span class="co">##  [49] httr_1.4.6                  abind_1.4-5                </span></span>
<span><span class="co">##  [51] mgcv_1.8-42                 compiler_4.3.0             </span></span>
<span><span class="co">##  [53] bit64_4.0.5                 backports_1.4.1            </span></span>
<span><span class="co">##  [55] BiocParallel_1.34.2         DBI_1.1.3                  </span></span>
<span><span class="co">##  [57] highr_0.10                  pan_1.8                    </span></span>
<span><span class="co">##  [59] biomaRt_2.56.1              MASS_7.3-59                </span></span>
<span><span class="co">##  [61] quantreg_5.96               rappdirs_0.3.3             </span></span>
<span><span class="co">##  [63] DelayedArray_0.26.7         rjson_0.2.21               </span></span>
<span><span class="co">##  [65] DNAcopy_1.74.1              tools_4.3.0                </span></span>
<span><span class="co">##  [67] lmtest_0.9-40               nnet_7.3-18                </span></span>
<span><span class="co">##  [69] glue_1.6.2                  restfulr_0.0.15            </span></span>
<span><span class="co">##  [71] nlme_3.1-162                grid_4.3.0                 </span></span>
<span><span class="co">##  [73] generics_0.1.3              operator.tools_1.6.3       </span></span>
<span><span class="co">##  [75] formula.tools_1.7.1         class_7.3-21               </span></span>
<span><span class="co">##  [77] tidyr_1.3.0                 ensembldb_2.24.0           </span></span>
<span><span class="co">##  [79] hms_1.1.3                   data.table_1.14.8          </span></span>
<span><span class="co">##  [81] xml2_1.3.5                  utf8_1.2.3                 </span></span>
<span><span class="co">##  [83] GWASExactHW_1.01            foreach_1.5.2              </span></span>
<span><span class="co">##  [85] pillar_1.9.0                stringr_1.5.0              </span></span>
<span><span class="co">##  [87] splines_4.3.0               dplyr_1.1.2                </span></span>
<span><span class="co">##  [89] BiocFileCache_2.8.0         lattice_0.21-8             </span></span>
<span><span class="co">##  [91] survival_3.5-5              bit_4.0.5                  </span></span>
<span><span class="co">##  [93] SparseM_1.81                tidyselect_1.2.0           </span></span>
<span><span class="co">##  [95] SeqVarTools_1.38.0          bookdown_0.35              </span></span>
<span><span class="co">##  [97] ProtGenerics_1.32.0         SummarizedExperiment_1.30.2</span></span>
<span><span class="co">##  [99] xfun_0.40                   Biobase_2.60.0             </span></span>
<span><span class="co">## [101] matrixStats_1.0.0           stringi_1.7.12             </span></span>
<span><span class="co">## [103] lazyeval_0.2.2              yaml_2.3.7                 </span></span>
<span><span class="co">## [105] boot_1.3-28.1               evaluate_0.21              </span></span>
<span><span class="co">## [107] codetools_0.2-19            tibble_3.2.1               </span></span>
<span><span class="co">## [109] BiocManager_1.30.22         cli_3.6.1                  </span></span>
<span><span class="co">## [111] rpart_4.1.19                systemfonts_1.0.4          </span></span>
<span><span class="co">## [113] jquerylib_0.1.4             Rcpp_1.0.11                </span></span>
<span><span class="co">## [115] dbplyr_2.3.3                png_0.1-8                  </span></span>
<span><span class="co">## [117] XML_3.99-0.14               parallel_4.3.0             </span></span>
<span><span class="co">## [119] MatrixModels_0.5-2          pkgdown_2.0.7              </span></span>
<span><span class="co">## [121] blob_1.2.4                  prettyunits_1.1.1          </span></span>
<span><span class="co">## [123] AnnotationFilter_1.24.0     bitops_1.0-7               </span></span>
<span><span class="co">## [125] lme4_1.1-34                 glmnet_4.1-7               </span></span>
<span><span class="co">## [127] SeqArray_1.40.1             purrr_1.0.2                </span></span>
<span><span class="co">## [129] crayon_1.5.2                rlang_1.1.1                </span></span>
<span><span class="co">## [131] KEGGREST_1.40.0             mice_3.16.0</span></span></code></pre>
<p><br><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
