<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Population reference dataset GDS files • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Population reference dataset GDS files">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
    <li>
      <a href="../articles/Wrappers.html">Using wrappper functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Population reference dataset GDS files</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/vignettes/Create_Reference_GDS_File.Rmd" class="external-link"><code>vignettes/Create_Reference_GDS_File.Rmd</code></a></small>
      <div class="hidden name"><code>Create_Reference_GDS_File.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), David A. Tuveson
[aut] (<a href="https://orcid.org/0000-0002-8017-2712" class="external-link uri">https://orcid.org/0000-0002-8017-2712</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 1.3.2<br><strong>Compiled date</strong>: 2024-10-22<br><strong>License</strong>: Apache License (&gt;= 2)<br></p>
<p><br><br></p>
<p>This vignette explains, in further details, the format of the
population reference files that are required to run the ancestry
inference tool.</p>
<p>Two different files are generated from a reference dataset:</p>
<ul>
<li>The Population Reference GDS File</li>
<li>The Population Reference SNV Annotation GDS file</li>
</ul>
<p><br><br></p>
<div class="section level2">
<h2 id="population-reference-gds-file">Population Reference GDS File<a class="anchor" aria-label="anchor" href="#population-reference-gds-file"></a>
</h2>
<p>The <em>Population Reference GDS file</em> should contain the
genome-wide SNV information related to the population data set with
known genetic ancestry. This reference data set will be used to generate
the simulated samples. It is also used to generate the PCA on which the
samples of interest are going to be projected.</p>
<p>The <em>Population Reference GDS file</em> is a GDS object of class
<a href="https://www.bioconductor.org/packages/release/bioc/vignettes/" class="external-link">SNPGDSFileClass</a>
from <a href="https://www.bioconductor.org/packages/release/bioc/html/SNPRelate.html" class="external-link">SNPRelate</a>
package <span class="citation">(Zheng et al. 2012)</span>.</p>
<p>Beware that related profiles should be flagged in the <em>Population
Reference GDS file</em> files.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pathRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileReferenceGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathRef</span>, <span class="st">"PopulationReferenceDemo.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdsRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileReferenceGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show the file format</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gdsRef</span><span class="op">)</span></span>
<span><span class="co">## File: /__w/_temp/Library/RAIDS/extdata/PopulationReferenceDemo.gds <span style="color: #555555;">(3.2K)</span></span></span>
<span><span class="co">## +    [  ]</span></span>
<span><span class="co">## |--+ sample.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 80B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ sample.annot   [ <span style="color: #555555;">data.frame</span> ] *</span></span>
<span><span class="co">## |  |--+ sex   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 20B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ pop.group   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ superPop   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  \--+ batch   <span style="color: #555555;">{</span> <span style="color: #555555;">Float64</span> <span style="color: #555555;">10, 80B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.chromosome   <span style="color: #555555;">{</span> <span style="color: #555555;">UInt16</span> <span style="color: #555555;">7, 14B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.position   <span style="color: #555555;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">7, 28B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.allele   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">7, 28B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EUR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AFR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AMR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.SAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">7, 21B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ genotype   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">7x10, 18B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## \--+ sample.ref   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit1</span> <span style="color: #555555;">10, 2B</span> <span style="color: #555555;">}</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRef</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>This output lists all variables stored in the <em>Population
Reference GDS file</em>. At the first level, it stores variables
<em>sample.id</em>, <em>snp.id</em>, etc. The additional information
displayed in the braces indicate the data type, size, compressed or not
with compression ratio.</p>
<p>The mandatory fields are:</p>
<ul>
<li>
<strong>sample.id</strong>: a <em>character</em> string (saved in
<em>Str8</em> format) used as unique identifier for each sample</li>
<li>
<strong>sample.annot</strong>: a <em>data.frame</em> where each row
correspond to a sample and containing those columns:
<ul>
<li>
<strong>sex</strong>: a <em>character</em> string (saved in
<em>Str8</em> format) used as identifier of the sex of the sample</li>
<li>
<strong>pop.Group</strong>: a <em>character</em> string (saved in
<em>Str8</em> format) representing the sub-population ancestry of the
sample (ex:GBR, etc)</li>
<li>
<strong>superPop</strong>: a <em>character</em> string (saved in
<em>Str8</em> format) representing the super-population ancestry of the
sample (ex:EUR, AFR, EAS, SAS, AMR)</li>
<li>
<strong>batch</strong>: an <em>integer</em> (saved in
<em>Float64</em> format) representing the batch of provenance of the
sample</li>
</ul>
</li>
<li>
<strong>snp.id</strong>: a a <em>character</em> string (saved in
<em>Str8</em> format) used as unique identifier for each SNV</li>
<li>
<strong>snp.chromosome</strong>: an <em>integer</em> or
<em>character</em> (saved in <em>UInt16 </em> format) mapping for each
chromosome. Integer: numeric values 1-26, mapped in order from 1-22,
23=X, 24=XY (the pseudoautosomal region), 25=Y, 26=M (the mitochondrial
probes), and 0 for probes with unknown positions; it does not allow NA.
Character: “X”, “XY”, “Y” and “M” can be used here, and a blank string
indicating unknown position</li>
<li>
<strong>snp.position</strong>: an <em>integer</em> (saved in
<em>Int32</em> format) representing the base position of each SNV on the
chromosome, and 0 for unknown position; it does not allow NA.</li>
<li>
<strong>snp.allele</strong>: a <em>character</em> string (saved as
<em>Str8</em> format) representing the reference allele and alternative
allele for each of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>snp.AF</strong>: a <em>numeric</em> value between 0 and 1
(saved as <em>PackedReal24</em> format) representing the allelic
frequency of the alternative allele in the general population for each
of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>snp.EAS_AF</strong>: a <em>numeric</em> value between 0 and
1 (saved as <em>PackedReal24</em> format) representing the allelic
frequency of the alternative allele in the East Asian population for
each of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>snp.EUR_AF</strong>: a <em>numeric</em> value between 0 and
1 (saved as <em>PackedReal24</em> format) representing the allelic
frequency of the alternative allele in the European population for each
of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>snp.AFR_AF</strong>: a <em>numeric</em> value between 0 and
1 (saved as <em>PackedReal24</em> format) representing the allelic
frequency of the alternative allele in the African population for each
of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>snp.AMR_AF</strong>: a <em>numeric</em> value between 0 and
1 (saved as <em>PackedReal24</em> format) representing the allelic
frequency of the alternative allele in the American population for each
of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>snp.SAS_AF</strong>: a <em>numeric</em> value between 0 and
1 (saved as <em>PackedReal24</em> format) representing the allelic
frequency of the alternative allele in the South Asian population for
each of the SNVs present in the <em>snp.id</em> field</li>
<li>
<strong>genotype</strong>: a SNV genotypic <em>matrix</em> of
<em>integer</em> values (saved in <em>Bit2</em> format) (i.e., the
number of A alleles) with SNVs as rows and samples as columns (number of
SNVs × number of Samples)</li>
<li>
<strong>sample.ref</strong>: an <em>integer</em> (saved in
<em>Bit1</em> format) indicating if the sample is retained to be used as
reference (=1) or removed (=0) as related samples have to be
discarded</li>
</ul>
<p><br></p>
<p>This following example shows how to create a <em>Population GDS
Reference file</em>. This example is for demonstration purpose only and
use hard coded values. A working <em>Population GDS Reference file</em>
would have to contain multiple samples from each continental population
and would also have to contain the SNVs from the entire genome.</p>
<p>To generate a real <em>Population GDS Reference file</em>, the
pipeline to process the information would depend of the selected source.
If the source files are in VCF format, you can use Bioconductor <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="external-link">VariationAnnotation</a>
package to extract the genotypic information (beware it may use a lot of
memory). Often, you will need to parse metadata files to get information
such as the sex and population of the profiles. In addition, the
Bioconductor <a href="https://bioconductor.org/packages/release/bioc/html/GENESIS.html" class="external-link">GENESIS</a>
package can be used to compute kinship coefficients to identify the
unrelated profiles.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create a temporary GDS Reference file in the temporary directory</span></span>
<span><span class="va">fileNewReferenceGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"reference_DEMO.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdsRefNew</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/createfn.gds.html" class="external-link">createfn.gds</a></span><span class="op">(</span><span class="va">fileNewReferenceGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The entry 'sample.id' contain the unique identifiers of 10 samples </span></span>
<span><span class="co">## that constitute the reference dataset</span></span>
<span><span class="va">sample.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HG00243"</span>, <span class="st">"HG00150"</span>, <span class="st">"HG00149"</span>, <span class="st">"HG00246"</span>, <span class="st">"HG00138"</span>, </span>
<span>                    <span class="st">"HG01334"</span>, <span class="st">"HG00268"</span>, <span class="st">"HG00275"</span>, <span class="st">"HG00290"</span>, <span class="st">"HG00364"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"sample.id"</span>, val<span class="op">=</span><span class="va">sample.id</span>, </span>
<span>            storage<span class="op">=</span><span class="st">"string"</span>, check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## A data frame containing the information about the 10 samples </span></span>
<span><span class="co">## (in the same order than in the 'sample.id') is created and added to </span></span>
<span><span class="co">## the 'sample.annot' entry</span></span>
<span><span class="co">## The data frame must contain those columns: </span></span>
<span><span class="co">##     'sex': '1'=male, '2'=female</span></span>
<span><span class="co">##     'pop.group': acronym for the population (ex: GBR, CDX, MSL, ASW, etc..)</span></span>
<span><span class="co">##     'superPop': acronym for the super-population (ex: AFR, EUR, etc...)</span></span>
<span><span class="co">##     'batch': number identifying the batch of provenance </span></span>
<span><span class="va">sampleInformation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sex<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"1"</span>, <span class="st">"1"</span>, <span class="st">"1"</span>, </span>
<span>        <span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"2"</span>, <span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span>, pop.group<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"GBR"</span>, <span class="fl">6</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FIN"</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>, </span>
<span>        superPop<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"EUR"</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>, batch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"sample.annot"</span>, val<span class="op">=</span><span class="va">sampleInformation</span>, </span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The identifier of each SNV is added in the 'snp.id' entry</span></span>
<span><span class="va">snvID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s29603"</span>, <span class="st">"s29605"</span>, <span class="st">"s29633"</span>, <span class="st">"s29634"</span>, <span class="st">"s29635"</span>, <span class="st">"s29637"</span>, </span>
<span>            <span class="st">"s29638"</span>, <span class="st">"s29663"</span>, <span class="st">"s29664"</span>, <span class="st">"s29666"</span>, <span class="st">"s29667"</span>, <span class="st">"s29686"</span>, </span>
<span>            <span class="st">"s29687"</span>, <span class="st">"s29711"</span>, <span class="st">"s29741"</span>, <span class="st">"s29742"</span>, <span class="st">"s29746"</span>, <span class="st">"s29750"</span>, </span>
<span>            <span class="st">"s29751"</span>, <span class="st">"s29753"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.id"</span>, val<span class="op">=</span><span class="va">snvID</span>, check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The chromosome of each SNV is added to the 'snp.chromosome' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvChrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.chromosome"</span>, val<span class="op">=</span><span class="va">snvChrom</span>, storage<span class="op">=</span><span class="st">"uint16"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The position on the chromosome of each SNV is added to </span></span>
<span><span class="co">## the 'snp.position' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvPos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3467333</span>, <span class="fl">3467428</span>, <span class="fl">3469375</span>, <span class="fl">3469387</span>, <span class="fl">3469502</span>, <span class="fl">3469527</span>, </span>
<span>                    <span class="fl">3469737</span>, <span class="fl">3471497</span>, <span class="fl">3471565</span>, <span class="fl">3471618</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.position"</span>, val<span class="op">=</span><span class="va">snvPos</span>, storage<span class="op">=</span><span class="st">"int32"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele information of each SNV is added to the 'snp.allele' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.allele' entry</span></span>
<span><span class="va">snvAllele</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A/G"</span>, <span class="st">"C/G"</span>, <span class="st">"C/T"</span>, <span class="st">"C/T"</span>, <span class="st">"T/G"</span>, <span class="st">"C/T"</span>, </span>
<span>                    <span class="st">"G/A"</span>, <span class="st">"A/G"</span>, <span class="st">"G/A"</span>, <span class="st">"G/A"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.allele"</span>, val<span class="op">=</span><span class="va">snvAllele</span>, storage<span class="op">=</span><span class="st">"string"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele frequency in the general population (between 0 and 1) of each </span></span>
<span><span class="co">## SNV is added to the 'snp.AF' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.86</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.AF"</span>, val<span class="op">=</span><span class="va">snvAF</span>, storage<span class="op">=</span><span class="st">"packedreal24"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele frequency in the East Asian population (between 0 and 1) of each </span></span>
<span><span class="co">## SNV is added to the 'snp.EAS_AF' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.80</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.02</span>, <span class="fl">0.00</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.EAS_AF"</span>, val<span class="op">=</span><span class="va">snvAF</span>, storage<span class="op">=</span><span class="st">"packedreal24"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele frequency in the European population (between 0 and 1) of each </span></span>
<span><span class="co">## SNV is added to the 'snp.EUR_AF' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.91</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.03</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.EUR_AF"</span>, val<span class="op">=</span><span class="va">snvAF</span>, storage<span class="op">=</span><span class="st">"packedreal24"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele frequency in the African population (between 0 and 1) of each </span></span>
<span><span class="co">## SNV is added to the 'snp.AFR_AF' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.85</span>, <span class="fl">0.04</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.AFR_AF"</span>, val<span class="op">=</span><span class="va">snvAF</span>, storage<span class="op">=</span><span class="st">"packedreal24"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele frequency in the American population (between 0 and 1) of each </span></span>
<span><span class="co">## SNV is added to the 'snp.AMR_AF' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.83</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.02</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.AMR_AF"</span>, val<span class="op">=</span><span class="va">snvAF</span>, storage<span class="op">=</span><span class="st">"packedreal24"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The allele frequency in the South Asian population (between 0 and 1) of each </span></span>
<span><span class="co">## SNV is added to the 'snp.SAS_AF' entry</span></span>
<span><span class="co">## The order of the SNVs is the same than in the 'snp.id' entry</span></span>
<span><span class="va">snvAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.89</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.05</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span>, <span class="fl">0.01</span>, <span class="fl">0.00</span>, <span class="fl">0.00</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"snp.SAS_AF"</span>, val<span class="op">=</span><span class="va">snvAF</span>, storage<span class="op">=</span><span class="st">"packedreal24"</span>,</span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The genotype of each SNV for each sample is added to the 'genotype' entry</span></span>
<span><span class="co">## The genotype correspond to the number of A alleles</span></span>
<span><span class="co">## The rows represent the SNVs is the same order than in 'snp.id' entry</span></span>
<span><span class="co">## The columns represent the samples is the same order than in 'sample.id' entry</span></span>
<span><span class="va">genotypeInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, </span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, </span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">10</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"genotype"</span>, val<span class="op">=</span><span class="va">genotypeInfo</span>, </span>
<span>            storage<span class="op">=</span><span class="st">"bit2"</span>, check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The entry 'sample.ref' is filled with 1 indicating that all 10 </span></span>
<span><span class="co">## samples are retained to be used as reference</span></span>
<span><span class="co">## The order of the samples is the same than in the 'sample.id' entry</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefNew</span>, name<span class="op">=</span><span class="st">"sample.ref"</span>, val<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">10</span><span class="op">)</span>, </span>
<span>            storage<span class="op">=</span><span class="st">"bit1"</span>, check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show the file format</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gdsRefNew</span><span class="op">)</span></span>
<span><span class="co">## File: /tmp/Rtmp4S5RLw/reference_DEMO.gds <span style="color: #555555;">(1.6K)</span></span></span>
<span><span class="co">## +    [  ]</span></span>
<span><span class="co">## |--+ sample.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 80B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ sample.annot   [ <span style="color: #555555;">data.frame</span> ] *</span></span>
<span><span class="co">## |  |--+ sex   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 20B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ pop.group   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ superPop   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  \--+ batch   <span style="color: #555555;">{</span> <span style="color: #555555;">Float64</span> <span style="color: #555555;">10, 80B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">20, 140B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.chromosome   <span style="color: #555555;">{</span> <span style="color: #555555;">UInt16</span> <span style="color: #555555;">20, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.position   <span style="color: #555555;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">10, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.allele   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">10, 40B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">10, 30B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">10, 30B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EUR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">10, 30B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AFR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">10, 30B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AMR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">10, 30B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.SAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">10, 30B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ genotype   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">10x10, 25B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## \--+ sample.ref   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit1</span> <span style="color: #555555;">10, 2B</span> <span style="color: #555555;">}</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRefNew</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">fileNewReferenceGDS</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="population-reference-annotation-gds-file">Population Reference Annotation GDS file<a class="anchor" aria-label="anchor" href="#population-reference-annotation-gds-file"></a>
</h2>
<p>The <em>Population Reference Annotation GDS file</em> contains phase
information and block group information for all the SNVs present in
<em>Population Reference GDS file</em>.<br>
If the source files are in VCF format, you can use Bioconductor <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="external-link">VariationAnnotation</a>
package to extract the phase information (beware it may use a lot of
memory). A block can be a linkage disequelibrium block relative to a
population or a gene. A bioconductor package like <a href="https://bioconductor.org/packages/release/bioc/html/GENESIS.html" class="external-link">GENESIS</a>
can be used to get the block information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tests"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileReferenceAnnotGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdsRefAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileReferenceAnnotGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show the file format</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gdsRefAnnot</span><span class="op">)</span></span>
<span><span class="co">## File: /__w/_temp/Library/RAIDS/extdata/tests/ex1_good_small_1KG.gds <span style="color: #555555;">(830.0K)</span></span></span>
<span><span class="co">## +    [  ] *</span></span>
<span><span class="co">## |--+ sample.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 1.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ sample.annot   [ <span style="color: #555555;">data.frame</span> ] *</span></span>
<span><span class="co">## |  |--+ sex   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 312B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ pop.group   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 624B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ superPop   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 624B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  \--+ batch   <span style="color: #555555;">{</span> <span style="color: #555555;">Float64</span> <span style="color: #555555;">156, 1.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">11000, 103.5K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.chromosome   <span style="color: #555555;">{</span> <span style="color: #555555;">UInt16</span> <span style="color: #555555;">11000, 21.5K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.position   <span style="color: #555555;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">11000, 43.0K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.allele   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">11000, 43.0K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EUR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AFR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AMR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.SAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ genotype   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">11000x156, 418.9K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## \--+ sample.ref   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit1</span> <span style="color: #555555;">156, 20B</span> <span style="color: #555555;">}</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRefAnnot</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>This output lists all variables stored in the <em>Population
Reference Annotation GDS file</em>. At the first level, it stores
variables <em>phase</em>, <em>block.annot</em>, etc. The additional
information displayed in the braces indicate the data type, size,
compressed or not + compression ratio.</p>
<p>The mandatory fields are:</p>
<ul>
<li>
<strong>phase</strong>: a <em>integer</em> (saved in <em>Bit2</em>
format) representing the phase of the SNVs in the <em>Population
Annotation GDS file</em>; 0 means the first allele is a reference; 1
means the first allele is the alternative and 3 means unknown. The first
allele combine with the genotype of the variant determine the phase for
a biallelic variant. The SNVs (rows) and samples (columns) in phase are
in the same order as in the <em>Population Annotation GDS
file</em>.</li>
<li>
<strong>block.annot</strong>: a <em>data.frame</em> containing those
columns:
<ul>
<li>
<strong>block.id</strong>: a <em>character</em> string (saved in
<em>Str8</em> format) representing an identifier of block group. A block
can be linkage disequilibrium block relative to a population or a
gene.</li>
<li>
<strong>block.desc</strong>: a <em>character</em> string (saved in
<em>Str8</em> format) describing the block group.</li>
</ul>
</li>
<li>
<strong>bloc</strong>: a <em>matrix</em> of <em>integer</em> values
(saved in <em>Int32</em> format) where each row representing a SNV in
the <em>Population Annotation GDS file</em> in the same order. The
columns are the block groups described in <em>block.annot</em>. Each
<em>integer</em> in the <em>matrix</em> representing a specific
block.</li>
</ul>
<p><br><br></p>
<p>This following example shows how to create a <em>Population Reference
Annotation GDS file</em>. This example is for demonstration purpose
only. A working <em>Population Reference Annotation GDS file</em> would
have to contain multiple samples from each continental population and
would also have to contain the SNVs from the entire genome.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create a temporary GDS Reference file in the temporary directory</span></span>
<span><span class="va">fileNewReferenceAnnotGDS</span> <span class="op">&lt;-</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"reference_SNV_Annotation_DEMO.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdsRefAnnotNew</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/createfn.gds.html" class="external-link">createfn.gds</a></span><span class="op">(</span><span class="va">fileNewReferenceAnnotGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The entry 'phase' contain the phase of the SNVs in the</span></span>
<span><span class="co">## Population Annotation GDS file</span></span>
<span><span class="co">## 0 means the first allele is a reference; 1 means the first allele is</span></span>
<span><span class="co">## the alternative and 3 means unknown</span></span>
<span><span class="co">## The SNVs (rows) and samples (columns) in phase are in the same order as</span></span>
<span><span class="co">## in the Population Annotation GDS file.</span></span>
<span><span class="va">phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>,</span>
<span>                        <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>,</span>
<span>                        <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">10</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefAnnotNew</span>, name<span class="op">=</span><span class="st">"phase"</span>, val<span class="op">=</span><span class="va">phase</span>, storage<span class="op">=</span><span class="st">"bit2"</span>, </span>
<span>                check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The entry 'blockAnnot' contain the information for each group of blocks</span></span>
<span><span class="co">## that are present in the 'block' entry.</span></span>
<span><span class="va">blockAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>block.id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EAS.0.05.500k"</span>, <span class="st">"EUR.0.05.500k"</span>,</span>
<span>                    <span class="st">"AFR.0.05.500k"</span>, <span class="st">"AMR.0.05.500k"</span>, <span class="st">"SAS.0.05.500k"</span><span class="op">)</span>,</span>
<span>                block.desc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                    <span class="st">"EAS populationblock base on SNP 0.05 and windows 500k"</span>,</span>
<span>                    <span class="st">"EUR populationblock base on SNP 0.05 and windows 500k"</span>,</span>
<span>                    <span class="st">"AFR populationblock base on SNP 0.05 and windows 500k"</span>,</span>
<span>                    <span class="st">"AMR populationblock base on SNP 0.05 and windows 500k"</span>,</span>
<span>                    <span class="st">"SAS populationblock base on SNP 0.05 and windows 500k"</span><span class="op">)</span>,</span>
<span>                stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefAnnotNew</span>, name<span class="op">=</span><span class="st">"block.annot"</span>, val<span class="op">=</span><span class="va">blockAnnot</span>, check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The entry 'block' contain the block information for the SNVs in the</span></span>
<span><span class="co">## Population Annotation GDS file.</span></span>
<span><span class="co">## The SNVs (rows) are in the same order as in </span></span>
<span><span class="co">## the Population Annotation GDS file.</span></span>
<span><span class="co">## The block groups (columns) are in the same order as in </span></span>
<span><span class="co">## the 'block.annot' entry.</span></span>
<span><span class="va">block</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>,</span>
<span>                        <span class="op">-</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">2</span>,  <span class="fl">1</span>, <span class="op">-</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">2</span>,</span>
<span>                        <span class="op">-</span><span class="fl">2</span>,  <span class="fl">1</span>,  <span class="fl">1</span>,  <span class="fl">1</span>, <span class="op">-</span><span class="fl">2</span>,</span>
<span>                        <span class="op">-</span><span class="fl">2</span>,  <span class="fl">1</span>,  <span class="fl">1</span>,  <span class="fl">1</span>, <span class="op">-</span><span class="fl">2</span>,</span>
<span>                        <span class="op">-</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">3</span>, <span class="op">-</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">3</span>, <span class="op">-</span><span class="fl">2</span>,</span>
<span>                         <span class="fl">1</span>,  <span class="fl">2</span>,  <span class="fl">2</span>,  <span class="fl">2</span>,  <span class="fl">1</span>,</span>
<span>                         <span class="fl">1</span>,  <span class="fl">2</span>,  <span class="fl">2</span>,  <span class="fl">2</span>,  <span class="fl">1</span>,</span>
<span>                        <span class="op">-</span><span class="fl">3</span>, <span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">3</span>, <span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">3</span>,</span>
<span>                         <span class="fl">2</span>, <span class="op">-</span><span class="fl">4</span>,  <span class="fl">3</span>, <span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">3</span>,</span>
<span>                         <span class="fl">2</span>, <span class="op">-</span><span class="fl">4</span>,  <span class="fl">3</span>, <span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">3</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">5</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/add.gdsn.html" class="external-link">add.gdsn</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">gdsRefAnnotNew</span>, name<span class="op">=</span><span class="st">"block"</span>, val<span class="op">=</span><span class="va">block</span>, storage<span class="op">=</span><span class="st">"int32"</span>, </span>
<span>            check<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show the file format</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gdsRefAnnotNew</span><span class="op">)</span></span>
<span><span class="co">## File: /tmp/Rtmp4S5RLw/reference_SNV_Annotation_DEMO.gds <span style="color: #555555;">(427B)</span></span></span>
<span><span class="co">## +    [  ]</span></span>
<span><span class="co">## |--+ phase   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">10x10, 25B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ block.annot   [ <span style="color: #555555;">data.frame</span> ] *</span></span>
<span><span class="co">## |  |--+ block.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">5, 70B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  \--+ block.desc   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">5, 270B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## \--+ block   <span style="color: #555555;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">10x5, 200B</span> <span style="color: #555555;">}</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRefAnnotNew</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">fileNewReferenceAnnotGDS</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="pre-processed-files-from-1000-genomes-in-hg38-are-available">Pre-processed files, from 1000 Genomes in hg38, are available<a class="anchor" aria-label="anchor" href="#pre-processed-files-from-1000-genomes-in-hg38-are-available"></a>
</h2>
<p>Pre-processed files used in the RAIDS associated publication, are
available at this address:</p>
<p><a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a></p>
<p>Beware that some of those files are voluminous.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] RAIDS_1.3.2      GENESIS_2.34.0   SNPRelate_1.38.1 gdsfmt_1.40.2   </span></span>
<span><span class="co">## [5] knitr_1.48       BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.9              shape_1.4.6.1              </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              jomo_2.7-6                 </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.56.0      logistf_1.26.0             </span></span>
<span><span class="co">##   [7] nloptr_2.1.1                rmarkdown_2.28             </span></span>
<span><span class="co">##   [9] fs_1.6.4                    BiocIO_1.14.0              </span></span>
<span><span class="co">##  [11] zlibbioc_1.50.0             ragg_1.3.2                 </span></span>
<span><span class="co">##  [13] vctrs_0.6.5                 memoise_2.0.1              </span></span>
<span><span class="co">##  [15] minqa_1.2.8                 Rsamtools_2.20.0           </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.16             quantsmooth_1.70.0         </span></span>
<span><span class="co">##  [19] htmltools_0.5.8.1           S4Arrays_1.4.1             </span></span>
<span><span class="co">##  [21] curl_5.2.3                  broom_1.0.7                </span></span>
<span><span class="co">##  [23] pROC_1.18.5                 SparseArray_1.4.8          </span></span>
<span><span class="co">##  [25] mitml_0.4-5                 sass_0.4.9                 </span></span>
<span><span class="co">##  [27] bslib_0.8.0                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">##  [29] desc_1.4.3                  plyr_1.8.9                 </span></span>
<span><span class="co">##  [31] sandwich_3.1-1              zoo_1.8-12                 </span></span>
<span><span class="co">##  [33] cachem_1.1.0                GenomicAlignments_1.40.0   </span></span>
<span><span class="co">##  [35] lifecycle_1.0.4             iterators_1.0.14           </span></span>
<span><span class="co">##  [37] pkgconfig_2.0.3             Matrix_1.7-0               </span></span>
<span><span class="co">##  [39] R6_2.5.1                    fastmap_1.2.0              </span></span>
<span><span class="co">##  [41] GenomeInfoDbData_1.2.12     MatrixGenerics_1.16.0      </span></span>
<span><span class="co">##  [43] digest_0.6.37               colorspace_2.1-1           </span></span>
<span><span class="co">##  [45] GWASTools_1.50.0            AnnotationDbi_1.66.0       </span></span>
<span><span class="co">##  [47] S4Vectors_0.42.1            textshaping_0.4.0          </span></span>
<span><span class="co">##  [49] GenomicRanges_1.56.2        RSQLite_2.3.7              </span></span>
<span><span class="co">##  [51] fansi_1.0.6                 httr_1.4.7                 </span></span>
<span><span class="co">##  [53] abind_1.4-8                 mgcv_1.9-1                 </span></span>
<span><span class="co">##  [55] compiler_4.4.1              bit64_4.5.2                </span></span>
<span><span class="co">##  [57] backports_1.5.0             BiocParallel_1.38.0        </span></span>
<span><span class="co">##  [59] DBI_1.2.3                   pan_1.9                    </span></span>
<span><span class="co">##  [61] MASS_7.3-61                 quantreg_5.98              </span></span>
<span><span class="co">##  [63] DelayedArray_0.30.1         rjson_0.2.23               </span></span>
<span><span class="co">##  [65] DNAcopy_1.78.0              tools_4.4.1                </span></span>
<span><span class="co">##  [67] lmtest_0.9-40               nnet_7.3-19                </span></span>
<span><span class="co">##  [69] glue_1.8.0                  restfulr_0.0.15            </span></span>
<span><span class="co">##  [71] nlme_3.1-165                grid_4.4.1                 </span></span>
<span><span class="co">##  [73] generics_0.1.3              gtable_0.3.5               </span></span>
<span><span class="co">##  [75] operator.tools_1.6.3        BSgenome_1.72.0            </span></span>
<span><span class="co">##  [77] formula.tools_1.7.1         class_7.3-22               </span></span>
<span><span class="co">##  [79] tidyr_1.3.1                 ensembldb_2.28.1           </span></span>
<span><span class="co">##  [81] data.table_1.16.2           utf8_1.2.4                 </span></span>
<span><span class="co">##  [83] XVector_0.44.0              BiocGenerics_0.50.0        </span></span>
<span><span class="co">##  [85] GWASExactHW_1.2             stringr_1.5.1              </span></span>
<span><span class="co">##  [87] foreach_1.5.2               pillar_1.9.0               </span></span>
<span><span class="co">##  [89] splines_4.4.1               dplyr_1.1.4                </span></span>
<span><span class="co">##  [91] lattice_0.22-6              survival_3.7-0             </span></span>
<span><span class="co">##  [93] rtracklayer_1.64.0          bit_4.5.0                  </span></span>
<span><span class="co">##  [95] SparseM_1.84-2              tidyselect_1.2.1           </span></span>
<span><span class="co">##  [97] SeqVarTools_1.42.0          Biostrings_2.72.1          </span></span>
<span><span class="co">##  [99] bookdown_0.41               ProtGenerics_1.36.0        </span></span>
<span><span class="co">## [101] IRanges_2.38.1              SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [103] stats4_4.4.1                xfun_0.48                  </span></span>
<span><span class="co">## [105] Biobase_2.64.0              matrixStats_1.4.1          </span></span>
<span><span class="co">## [107] stringi_1.8.4               UCSC.utils_1.0.0           </span></span>
<span><span class="co">## [109] lazyeval_0.2.2              yaml_2.3.10                </span></span>
<span><span class="co">## [111] boot_1.3-30                 evaluate_1.0.1             </span></span>
<span><span class="co">## [113] codetools_0.2-20            tibble_3.2.1               </span></span>
<span><span class="co">## [115] BiocManager_1.30.25         cli_3.6.3                  </span></span>
<span><span class="co">## [117] rpart_4.1.23                systemfonts_1.1.0          </span></span>
<span><span class="co">## [119] munsell_0.5.1               jquerylib_0.1.4            </span></span>
<span><span class="co">## [121] Rcpp_1.0.13                 GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">## [123] png_0.1-8                   XML_3.99-0.17              </span></span>
<span><span class="co">## [125] parallel_4.4.1              MatrixModels_0.5-3         </span></span>
<span><span class="co">## [127] ggplot2_3.5.1               pkgdown_2.1.1              </span></span>
<span><span class="co">## [129] blob_1.2.4                  AnnotationFilter_1.28.0    </span></span>
<span><span class="co">## [131] bitops_1.0-9                lme4_1.1-35.5              </span></span>
<span><span class="co">## [133] glmnet_4.1-8                SeqArray_1.44.3            </span></span>
<span><span class="co">## [135] VariantAnnotation_1.50.0    scales_1.3.0               </span></span>
<span><span class="co">## [137] purrr_1.0.2                 crayon_1.5.3               </span></span>
<span><span class="co">## [139] rlang_1.1.4                 KEGGREST_1.44.1            </span></span>
<span><span class="co">## [141] mice_3.16.0</span></span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Zheng2012" class="csl-entry">
Zheng, Xiuwen, David Levine, Jess Shen, Stephanie M. Gogarten, Cathy
Laurie, and Bruce S. Weir. 2012. <span>“<span class="nocase">A
high-performance computing toolset for relatedness and principal
component analysis of SNP data</span>.”</span> <em>Bioinformatics</em>
28 (24): 3326–28. <a href="https://doi.org/10.1093/bioinformatics/bts606" class="external-link">https://doi.org/10.1093/bioinformatics/bts606</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
