<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Population reference dataset GDS files • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Population reference dataset GDS files">
<meta property="og:description" content="RAIDS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Ancestry_Inference_Step_by_Step.html">Genetic Ancestry Inference Step by Step (optional)</a>
    </li>
    <li>
      <a href="../articles/Create_1KG_GDS_File.html">Formatting the information from the population reference dataset (optional)</a>
    </li>
    <li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Population reference dataset GDS files</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/HEAD/vignettes/Create_Reference_GDS_File.Rmd" class="external-link"><code>vignettes/Create_Reference_GDS_File.Rmd</code></a></small>
      <div class="hidden name"><code>Create_Reference_GDS_File.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 0.99.6<br><strong>Compiled date</strong>: 2023-08-17<br><strong>License</strong>: Apache License (&gt;= 2)<br></p>
<p><br><br></p>
<p>This section explains in further details the format of the population
reference GDS files that are needed to run the ancestry inference
tool.</p>
<p>Three different files are generated from the reference dataset:</p>
<ul>
<li>The population reference GDS File</li>
<li>The population reference SNV Annotation GDS file</li>
<li>The population reference SNV Retained VCF file</li>
</ul>
<div class="section level2">
<h2 id="population-reference-gds-file">Population Reference GDS File<a class="anchor" aria-label="anchor" href="#population-reference-gds-file"></a>
</h2>
<p>All related samples must be removed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileReferenceGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"gdsRef"</span>, <span class="st">"ex1kg.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdsRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileReferenceGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show the file format</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gdsRef</span><span class="op">)</span></span>
<span><span class="co">## File: /__w/_temp/Library/RAIDS/extdata/example/gdsRef/ex1kg.gds <span style="color: #555555;">(830.0K)</span></span></span>
<span><span class="co">## +    [  ] *</span></span>
<span><span class="co">## |--+ sample.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 1.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ sample.annot   [ <span style="color: #555555;">data.frame</span> ] *</span></span>
<span><span class="co">## |  |--+ sex   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 312B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ pop.group   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 624B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  |--+ superPop   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">156, 624B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  \--+ batch   <span style="color: #555555;">{</span> <span style="color: #555555;">Float64</span> <span style="color: #555555;">156, 1.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">11000, 103.5K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.chromosome   <span style="color: #555555;">{</span> <span style="color: #555555;">UInt16</span> <span style="color: #555555;">11000, 21.5K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.position   <span style="color: #555555;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">11000, 43.0K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.allele   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">11000, 43.0K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.EUR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AFR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.AMR_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ snp.SAS_AF   <span style="color: #555555;">{</span> <span style="color: #555555;">PackedReal24</span> <span style="color: #555555;">11000, 32.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ genotype   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">11000x156, 418.9K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## \--+ sample.ref   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit1</span> <span style="color: #555555;">156, 20B</span> <span style="color: #555555;">}</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRef</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="population-reference-annotation-gds-file">Population Reference Annotation GDS file<a class="anchor" aria-label="anchor" href="#population-reference-annotation-gds-file"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SNPRelate" class="external-link">SNPRelate</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example/"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileReferenceAnnotGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"gdsRef"</span>, <span class="st">"exAnnot1kg.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gdsRefAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/openfn.gds.html" class="external-link">openfn.gds</a></span><span class="op">(</span><span class="va">fileReferenceAnnotGDS</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show the file format</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gdsRefAnnot</span><span class="op">)</span></span>
<span><span class="co">## File: /__w/_temp/Library/RAIDS/extdata/example/gdsRef/exAnnot1kg.gds <span style="color: #555555;">(159.8K)</span></span></span>
<span><span class="co">## +    [  ]</span></span>
<span><span class="co">## |--+ phase   <span style="color: #555555;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">11000x156</span> <span style="color: #555555;">LZ4_ra(21.5%), 90.2K</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |--+ block.annot   [ <span style="color: #555555;">data.frame</span> ] *</span></span>
<span><span class="co">## |  |--+ block.id   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">7, 123B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## |  \--+ block.desc   <span style="color: #555555;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">7, 388B</span> <span style="color: #555555;">}</span></span></span>
<span><span class="co">## \--+ block   <span style="color: #555555;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">11000x7</span> <span style="color: #555555;">LZ4_ra(22.6%), 68.1K</span> <span style="color: #555555;">}</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gdsRefAnnot</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pre-processed-files-from-1000-genomes-in-hg38-are-available">Pre-processed files, from 1000 Genomes in hg38, are available<a class="anchor" aria-label="anchor" href="#pre-processed-files-from-1000-genomes-in-hg38-are-available"></a>
</h2>
<p>Pre-processed files used in the RAIDS associated publication, are
available at this address:</p>
<p><a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a></p>
<p>Beware that some of those files are voluminous.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] RAIDS_0.99.6     GENESIS_2.30.0   SNPRelate_1.34.1 gdsfmt_1.36.1   </span></span>
<span><span class="co">## [5] knitr_1.43       BiocStyle_2.28.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.7              shape_1.4.6                </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              jomo_2.7-6                 </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.52.1      logistf_1.25.0             </span></span>
<span><span class="co">##   [7] nloptr_2.0.3                rmarkdown_2.24             </span></span>
<span><span class="co">##   [9] fs_1.6.3                    BiocIO_1.10.0              </span></span>
<span><span class="co">##  [11] zlibbioc_1.46.0             ragg_1.2.5                 </span></span>
<span><span class="co">##  [13] vctrs_0.6.3                 memoise_2.0.1              </span></span>
<span><span class="co">##  [15] minqa_1.2.5                 Rsamtools_2.16.0           </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.12             quantsmooth_1.66.0         </span></span>
<span><span class="co">##  [19] htmltools_0.5.6             S4Arrays_1.0.5             </span></span>
<span><span class="co">##  [21] progress_1.2.2              curl_5.0.2                 </span></span>
<span><span class="co">##  [23] broom_1.0.5                 pROC_1.18.4                </span></span>
<span><span class="co">##  [25] mitml_0.4-5                 sass_0.4.7                 </span></span>
<span><span class="co">##  [27] bslib_0.5.1                 desc_1.4.2                 </span></span>
<span><span class="co">##  [29] plyr_1.8.8                  sandwich_3.0-2             </span></span>
<span><span class="co">##  [31] zoo_1.8-12                  cachem_1.0.8               </span></span>
<span><span class="co">##  [33] GenomicAlignments_1.36.0    lifecycle_1.0.3            </span></span>
<span><span class="co">##  [35] iterators_1.0.14            pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [37] Matrix_1.6-0                R6_2.5.1                   </span></span>
<span><span class="co">##  [39] fastmap_1.1.1               GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">##  [41] MatrixGenerics_1.12.3       digest_0.6.33              </span></span>
<span><span class="co">##  [43] GWASTools_1.46.0            AnnotationDbi_1.62.2       </span></span>
<span><span class="co">##  [45] S4Vectors_0.38.1            rprojroot_2.0.3            </span></span>
<span><span class="co">##  [47] textshaping_0.3.6           GenomicRanges_1.52.0       </span></span>
<span><span class="co">##  [49] RSQLite_2.3.1               filelock_1.0.2             </span></span>
<span><span class="co">##  [51] fansi_1.0.4                 httr_1.4.7                 </span></span>
<span><span class="co">##  [53] abind_1.4-5                 mgcv_1.8-42                </span></span>
<span><span class="co">##  [55] compiler_4.3.0              bit64_4.0.5                </span></span>
<span><span class="co">##  [57] backports_1.4.1             BiocParallel_1.34.2        </span></span>
<span><span class="co">##  [59] DBI_1.1.3                   pan_1.8                    </span></span>
<span><span class="co">##  [61] biomaRt_2.56.1              MASS_7.3-59                </span></span>
<span><span class="co">##  [63] quantreg_5.96               rappdirs_0.3.3             </span></span>
<span><span class="co">##  [65] DelayedArray_0.26.7         rjson_0.2.21               </span></span>
<span><span class="co">##  [67] DNAcopy_1.74.1              tools_4.3.0                </span></span>
<span><span class="co">##  [69] lmtest_0.9-40               nnet_7.3-18                </span></span>
<span><span class="co">##  [71] glue_1.6.2                  restfulr_0.0.15            </span></span>
<span><span class="co">##  [73] nlme_3.1-162                grid_4.3.0                 </span></span>
<span><span class="co">##  [75] generics_0.1.3              operator.tools_1.6.3       </span></span>
<span><span class="co">##  [77] BSgenome_1.68.0             formula.tools_1.7.1        </span></span>
<span><span class="co">##  [79] class_7.3-21                tidyr_1.3.0                </span></span>
<span><span class="co">##  [81] ensembldb_2.24.0            hms_1.1.3                  </span></span>
<span><span class="co">##  [83] data.table_1.14.8           xml2_1.3.5                 </span></span>
<span><span class="co">##  [85] utf8_1.2.3                  XVector_0.40.0             </span></span>
<span><span class="co">##  [87] BiocGenerics_0.46.0         GWASExactHW_1.01           </span></span>
<span><span class="co">##  [89] foreach_1.5.2               pillar_1.9.0               </span></span>
<span><span class="co">##  [91] stringr_1.5.0               splines_4.3.0              </span></span>
<span><span class="co">##  [93] dplyr_1.1.2                 BiocFileCache_2.8.0        </span></span>
<span><span class="co">##  [95] lattice_0.21-8              survival_3.5-5             </span></span>
<span><span class="co">##  [97] rtracklayer_1.60.1          bit_4.0.5                  </span></span>
<span><span class="co">##  [99] SparseM_1.81                tidyselect_1.2.0           </span></span>
<span><span class="co">## [101] SeqVarTools_1.38.0          Biostrings_2.68.1          </span></span>
<span><span class="co">## [103] bookdown_0.35               ProtGenerics_1.32.0        </span></span>
<span><span class="co">## [105] IRanges_2.34.1              SummarizedExperiment_1.30.2</span></span>
<span><span class="co">## [107] stats4_4.3.0                xfun_0.40                  </span></span>
<span><span class="co">## [109] Biobase_2.60.0              matrixStats_1.0.0          </span></span>
<span><span class="co">## [111] stringi_1.7.12              lazyeval_0.2.2             </span></span>
<span><span class="co">## [113] yaml_2.3.7                  boot_1.3-28.1              </span></span>
<span><span class="co">## [115] evaluate_0.21               codetools_0.2-19           </span></span>
<span><span class="co">## [117] tibble_3.2.1                BiocManager_1.30.22        </span></span>
<span><span class="co">## [119] cli_3.6.1                   rpart_4.1.19               </span></span>
<span><span class="co">## [121] systemfonts_1.0.4           jquerylib_0.1.4            </span></span>
<span><span class="co">## [123] Rcpp_1.0.11                 GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [125] dbplyr_2.3.3                png_0.1-8                  </span></span>
<span><span class="co">## [127] XML_3.99-0.14               parallel_4.3.0             </span></span>
<span><span class="co">## [129] MatrixModels_0.5-2          pkgdown_2.0.7              </span></span>
<span><span class="co">## [131] blob_1.2.4                  prettyunits_1.1.1          </span></span>
<span><span class="co">## [133] AnnotationFilter_1.24.0     bitops_1.0-7               </span></span>
<span><span class="co">## [135] lme4_1.1-34                 glmnet_4.1-7               </span></span>
<span><span class="co">## [137] SeqArray_1.40.1             purrr_1.0.2                </span></span>
<span><span class="co">## [139] crayon_1.5.2                rlang_1.1.1                </span></span>
<span><span class="co">## [141] KEGGREST_1.40.0             mice_3.16.0</span></span></code></pre>
<p><br><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
