<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate Inference of Genetic Ancestry from Cancer-derived Sequences • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate Inference of Genetic Ancestry from Cancer-derived Sequences">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
    <li>
      <a href="../articles/Wrappers.html">Using wrappper functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate Inference of Genetic Ancestry from
Cancer-derived Sequences</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/vignettes/RAIDS.Rmd" class="external-link"><code>vignettes/RAIDS.Rmd</code></a></small>
      <div class="hidden name"><code>RAIDS.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), David A. Tuveson
[aut] (<a href="https://orcid.org/0000-0002-8017-2712" class="external-link uri">https://orcid.org/0000-0002-8017-2712</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 1.3.2<br><strong>Compiled date</strong>: 2024-10-16<br><strong>License</strong>: Apache License (&gt;= 2)<br></p>
<div class="section level2">
<h2 id="licensing">Licensing<a class="anchor" aria-label="anchor" href="#licensing"></a>
</h2>
<p>The <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em>
package and the underlying <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em> code are
distributed under<br>
the <a href="https://opensource.org/licenses/Apache-2.0" class="external-link uri">https://opensource.org/licenses/Apache-2.0</a> license. You
are free to use and redistribute this software.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>If you use the <strong>RAIDS</strong> package for a publication, we
would ask you to cite the following:</p>
<blockquote>
<p>Pascal Belleau, Astrid Deschênes, Nyasha Chambwe, David A. Tuveson,
Alexander Krasnitz; Genetic Ancestry Inference from Cancer-Derived
Molecular Data across Genomic and Transcriptomic Platforms. Cancer Res 1
January 2023; 83 (1): 49–58. <a href="https://doi.org/10.1158/0008-5472.CAN-22-0682" class="external-link uri">https://doi.org/10.1158/0008-5472.CAN-22-0682</a> <br><br></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple methods have been implemented to infer ancestry from
germline DNA sequence <span class="citation">(Price et al. 2006;
Pritchard, Stephens, and Donnelly 2000; Alexander, Novembre, and Lange
2009)</span>. However, genotyping of DNA from matched normal specimens
is not part of standard clinical practice and is not performed routinely
outside academic clinical centers. In sum, matched germline DNA sequence
is often missing for cancer-derived molecular data. In such cases,
having the possibility to infer ancestry from tumor-derived data would
be beneficial.</p>
<p>The <strong>RAIDS</strong> package implements an inference procedure
that has been specifically developed to accurately infer genetic
ancestry from cancer-derived sequences. The current version can handle
cancer-derived sequences of:</p>
<ul>
<li>tumor exomes</li>
<li>targeted gene panels</li>
<li>RNA</li>
</ul>
<p>The <strong>RAIDS</strong> package implements a data synthesis method
that, for any given cancer-derived sequence profile, enables on the one
hand, profile-specific inference parameter optimization and on the other
hand, a profile-specific inference accuracy estimate.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install this package from <a href="https://bioconductor.org/packages/RAIDS" class="external-link">Bioconductor</a>, start R
(version 4.3 or later) and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"RAIDS"</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
<p><a name="mains"></a></p>
</div>
<div class="section level2">
<h2 id="main-steps">Main Steps<a class="anchor" aria-label="anchor" href="#main-steps"></a>
</h2>
<p>This is an overview of the genetic ancestry inference from
cancer-derived molecular data:</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_v05.png" alt="An overview of the genetic ancestry inference process." width="130%"><p class="caption">
An overview of the genetic ancestry inference process.
</p>
</div>
<p>The main steps are:</p>
<p><strong>Step 1.</strong> Set-up and provide population reference
files</p>
<p><strong>Step 2</strong> Sample the reference data for donor
genotypes, to be used for synthesis and optimize ancestry inference
parameters</p>
<p><strong>Step 3</strong> Infer ancestry for the subjects of the
external study</p>
<p><strong>Step 4</strong> Present and interpret the results</p>
<p>These steps are described in detail in the following.</p>
<p><br><br></p>
<div class="section level3">
<h3 id="step-1--set-up-and-provide-population-reference-files">Step 1. Set-up and provide population reference files<a class="anchor" aria-label="anchor" href="#step-1--set-up-and-provide-population-reference-files"></a>
</h3>
<div class="section level4">
<h4 id="create-a-directory-structure">1.1 Create a directory structure<a class="anchor" aria-label="anchor" href="#create-a-directory-structure"></a>
</h4>
<p>First, a specific directory structure should be created. The
structure must<br>
correspond to this:</p>
<pre><code><span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Working directory structure</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">workingDirectory</span><span class="op">/</span>  </span>
<span>    <span class="va">data</span><span class="op">/</span>  </span>
<span>        <span class="va">refGDS</span>  </span>
<span>        <span class="va">profileGDS</span> </span></code></pre>
<p><br></p>
<p>This following running example creates a temporary working directory
structure when the example will be run.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Create a temporary working directory structure</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathWorkingDirectory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"workingDirectory"</span><span class="op">)</span></span>
<span><span class="va">pathWorkingDirectoryData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectory</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">pathWorkingDirectory</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathWorkingDirectory</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span>, <span class="st">"refGDS"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span>, <span class="st">"profileGDS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="download-the-population-reference-files">1.2 Download the population reference files<a class="anchor" aria-label="anchor" href="#download-the-population-reference-files"></a>
</h4>
<p>The population reference files should be downloaded in the
<em>data/refGDS</em> sub-directory. This following code downloads the
complete pre-processed files for 1000 Genomes (1KG), in hg38. The size
of the 1KG GDS file is 15GB.</p>
<pre><code>
#############################################################################
## How to download the pre-processed files for 1000 Genomes (1KG) (15 GB)
#############################################################################
cd workingDirectory
cd data/refGDS

wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matGeno1000g.gds
wget https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper/matAnnot1000g.gds
cd -
</code></pre>
<p><br></p>
<p>For demonstration purpose, a small <strong>population reference GDS
file</strong> (called <em>ex1_good_small_1KG.gds</em>) and a small
<strong>population reference SNV Annotation GDS file</strong> (called
<em>ex1_good_small_1KG_Annot.gds</em>) are included in this package.
Beware that those two files should not be used to run a real ancestry
inference. The results obtained with those files won’t be reliable.</p>
<p>In this running example, the demonstration files are copied in the
required <em>data/refGDS</em> directory.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load RAIDS package</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>   </span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The population reference GDS file and SNV Annotation GDS file</span></span>
<span><span class="co">## need to be located in the same sub-directory.</span></span>
<span><span class="co">## Note that the population reference GDS file used for this example is a</span></span>
<span><span class="co">## simplified version and CANNOT be used for any real analysis</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Path to the demo 1KG GDS file is located in this package</span></span>
<span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"tests"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span>
<span><span class="va">fileAnnotGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG_Annot.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="va">fileGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span>, <span class="st">"refGDS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="va">fileAnnotGDS</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span>, <span class="st">"refGDS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] TRUE</span></span></code></pre></div>
<p><br><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-2-ancestry-inference-with-raids">Step 2 Ancestry inference with RAIDS<a class="anchor" aria-label="anchor" href="#step-2-ancestry-inference-with-raids"></a>
</h3>
<div class="section level4">
<h4 id="set-up-required-directories">2.1 Set-up required directories<a class="anchor" aria-label="anchor" href="#set-up-required-directories"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The file path to the population reference GDS file </span></span>
<span><span class="co">##     is required (refGenotype will be used as input later)</span></span>
<span><span class="co">## The file path to the population reference SNV Annotation GDS file</span></span>
<span><span class="co">##     is also required (refAnnotation will be used as input later)</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span>, <span class="st">"refGDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">refGenotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span>
<span><span class="va">refAnnotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG_Annot.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The output profileGDS directory, inside workingDirectory/data, must be </span></span>
<span><span class="co">##    created (pathProfileGDS will be used as input later)</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathProfileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathWorkingDirectoryData</span>, <span class="st">"profileGDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">pathProfileGDS</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="sample-reference-donor-profiles-from-the-reference-data">2.2 Sample reference donor profiles from the reference data<a class="anchor" aria-label="anchor" href="#sample-reference-donor-profiles-from-the-reference-data"></a>
</h4>
<p>With the 1KG reference, we recommend sampling 30 donor profiles per
population. For reproducibility, be sure to use the same random-number
generator seed.</p>
<p>In the following code, only 2 profiles per population are sampled
from the demo population GDS file:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Fix seed to ensure reproducible results</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Select the profiles from the population reference GDS file for </span></span>
<span><span class="co">## the synthetic data.</span></span>
<span><span class="co">## Here we select 2 profiles from the simplified 1KG GDS for each </span></span>
<span><span class="co">## subcontinental-level.</span></span>
<span><span class="co">## Normally, we would use 30 profiles for each subcontinental-level.</span></span>
<span><span class="co">## The 1KG files in this example only have 6 profiles for each </span></span>
<span><span class="co">## subcontinental-level (for demo purpose only).</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPopForSynthetic.html">select1KGPopForSynthetic</a></span><span class="op">(</span>fileReferenceGDS<span class="op">=</span><span class="va">refGenotype</span>,</span>
<span>                                        nbProfiles<span class="op">=</span><span class="fl">2L</span><span class="op">)</span></span></code></pre></div>
<p>The output object is going to be used later at the ancestry inference
step.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="perform-the-ancestry-inference">2.3 Perform the ancestry inference<a class="anchor" aria-label="anchor" href="#perform-the-ancestry-inference"></a>
</h4>
<p>Within a single function call, data synthesis is performed, the
synthetic data are used to optimize the inference parameters and, with
these, the ancestry of the input profile donor is inferred.</p>
<p>According to the type of input data (RNA or DNA), a specific function
is available.</p>
<p>The <em>inferAncestry()</em> function is used for DNA profiles while
the <em>inferAncestryGeneAware()</em> function is RNA specific.</p>
<p>In this example, the profile is from DNA source and requires the use
of the <em>inferAncestry()</em> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="co">## GenomeInfoDb and BSgenome are required libraries to run this example</span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GenomeInfoDb"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="co">#######################################################################</span></span>
<span>    <span class="co">## Chromosome length information is required</span></span>
<span>    <span class="co">## chr23 is chrX, chr24 is chrY and chrM is 25</span></span>
<span>    <span class="co">#######################################################################</span></span>
<span>    <span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span></span>
<span>    <span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co">#######################################################################</span></span>
<span>    <span class="co">## The SNP VCF file of the DNA profile donor</span></span>
<span>    <span class="co">#######################################################################</span></span>
<span>    <span class="va">fileDonorVCF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"snpPileup"</span>, <span class="st">"ex1.vcf.gz"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">#######################################################################</span></span>
<span>    <span class="co">## The ancestry inference call</span></span>
<span>    <span class="co">#######################################################################</span></span>
<span>    <span class="va">resOut</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/inferAncestry.html">inferAncestry</a></span><span class="op">(</span>profileFile<span class="op">=</span><span class="va">fileDonorVCF</span>, </span>
<span>        pathProfileGDS<span class="op">=</span><span class="va">pathProfileGDS</span>,</span>
<span>        fileReferenceGDS<span class="op">=</span><span class="va">refGenotype</span>,</span>
<span>        fileReferenceAnnotGDS<span class="op">=</span><span class="va">refAnnotation</span>,</span>
<span>        chrInfo<span class="op">=</span><span class="va">chrInfo</span>,</span>
<span>        syntheticRefDF<span class="op">=</span><span class="va">dataRef</span>,</span>
<span>        genoSource<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VCF"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>A profile GDS file is created in the <em>pathProfileGDS</em>
directory while all the ancestry and optimal parameters information are
integrated in the output object.</p>
<p>At last, all temporary files created in this example should be
deleted.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#######################################################################</span></span>
<span><span class="co">## Remove temporary files created for this demo</span></span>
<span><span class="co">#######################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathWorkingDirectory</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>      </span></code></pre></div>
<p><br><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-3--examine-the-value-of-the-inference-call">Step 3. Examine the value of the inference call<a class="anchor" aria-label="anchor" href="#step-3--examine-the-value-of-the-inference-call"></a>
</h3>
<p>The inferred ancestry and the optimal parameters are present in the
<em>list</em> object generated by the <em>inferAncestry()</em> and
<em>inferAncestryGeneAware()</em> functions.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="co">## The output is a list object with multiple entries</span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">resOut</span><span class="op">)</span></span>
<span><span class="co">## [1] "list"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">resOut</span><span class="op">)</span></span>
<span><span class="co">## [1] "pcaSample"    "paraSample"   "KNNSample"    "KNNSynthetic" "Ancestry"</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="inspect-the-inference-and-the-optimal-parameters">3.1 Inspect the inference and the optimal parameters<a class="anchor" aria-label="anchor" href="#inspect-the-inference-and-the-optimal-parameters"></a>
</h4>
<p>For the global ancestry inference using PCA followed by nearest
neighbor classification these parameters are <em>D</em> (the number of
the top principal directions retained) and <em>k</em> (the number of
nearest neighbors).</p>
<p>The information is stored in the <em>Ancestry</em> entry as a
<em>data.frame</em> object. It is a contains those columns:</p>
<ul>
<li>
<em>sample.id</em>: The unique identifier of the sample</li>
<li>
<em>D</em>: The optimal PCA dimension value used to infer the
ancestry</li>
<li>
<em>k</em>: The optimal number of neighbors value used to infer the
ancestry</li>
<li>
<em>SuperPop</em>: The inferred ancestry</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="co">## The ancestry information is stored in the 'Ancestry' entry </span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">resOut</span><span class="op">$</span><span class="va">Ancestry</span><span class="op">)</span></span>
<span><span class="co">##     sample.id  D K SuperPop</span></span>
<span><span class="co">## 173       ex1 14 6      AFR</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="visualize-the-raids-performance-for-the-synthetic-data">3.2 Visualize the RAIDS performance for the synthetic data<a class="anchor" aria-label="anchor" href="#visualize-the-raids-performance-for-the-synthetic-data"></a>
</h4>
<p>The <em>createAUROCGraph()</em> function enable the visualization of
RAIDS performance for the synthetic data, as a function of <em>D</em>
and <em>k</em>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###########################################################################</span></span>
<span><span class="co">## Create a graph showing the perfomance for the synthetic data</span></span>
<span><span class="co">## The output is a ggplot object</span></span>
<span><span class="co">###########################################################################</span></span>
<span><span class="fu"><a href="../reference/createAUROCGraph.html">createAUROCGraph</a></span><span class="op">(</span>dfAUROC<span class="op">=</span><span class="va">resOut</span><span class="op">$</span><span class="va">paraSample</span><span class="op">$</span><span class="va">dfAUROC</span>, title<span class="op">=</span><span class="st">"Example ex1"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="RAIDS_files/figure-html/visualize-1.png" alt="RAIDS performance for the synthtic data." width="700"><p class="caption">
RAIDS performance for the synthtic data.
</p>
</div>
<p>In this specific example, the performances are lower than expected
with a real profile and a complete reference population file.</p>
<p><br><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="format-population-reference-dataset-optional">Format population reference dataset (optional)<a class="anchor" aria-label="anchor" href="#format-population-reference-dataset-optional"></a>
</h2>
<div class="figure" style="text-align: center">
<img src="Step1_population_file_v01.png" alt="Step 1 - Provide population reference data" width="120%"><p class="caption">
Step 1 - Provide population reference data
</p>
</div>
<p>A population reference dataset with known ancestry is required to
infer ancestry.</p>
<p>Three important reference files, containing formatted information
about the reference dataset, are required:</p>
<ul>
<li>The population reference GDS File</li>
<li>The population reference SNV Annotation GDS file</li>
<li>The population reference SNV Retained VCF file (optional)</li>
</ul>
<p>The format of those files are described the <a href="Create_Reference_GDS_File.html">Population reference dataset GDS
files</a> vignette.</p>
<p><span style="color:darkred;font-weight:bold">The reference files
associated to the Cancer Research associated paper are available. Note
that these pre-processed files are for 1000 Genomes (1KG), in hg38. The
files are available here:</span></p>
<p><span style="color:red"> <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a>
</span></p>
<p><span style="color:darkred;font-weight:bold">The size of the 1KG GDS
file is 15GB.</span></p>
<p><span style="color:darkred;font-weight:bold">The 1KG GDS file is
mapped on hg38 <span class="citation">(Lowy-Gallego et al.
2019)</span>.</span></p>
<p><span style="color:darkred;font-weight:bold">This section can be
skipped if you choose to use the pre-processed files.</span></p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> in the environment
in which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] Matrix_1.7-0     RAIDS_1.3.2      GENESIS_2.34.0   SNPRelate_1.38.1</span></span>
<span><span class="co">## [5] gdsfmt_1.40.2    knitr_1.48       BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.9                    shape_1.4.6.1                    </span></span>
<span><span class="co">##   [3] magrittr_2.0.3                    jomo_2.7-6                       </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.56.0            farver_2.1.2                     </span></span>
<span><span class="co">##   [7] logistf_1.26.0                    nloptr_2.1.1                     </span></span>
<span><span class="co">##   [9] rmarkdown_2.28                    fs_1.6.4                         </span></span>
<span><span class="co">##  [11] BiocIO_1.14.0                     zlibbioc_1.50.0                  </span></span>
<span><span class="co">##  [13] ragg_1.3.2                        vctrs_0.6.5                      </span></span>
<span><span class="co">##  [15] memoise_2.0.1                     minqa_1.2.8                      </span></span>
<span><span class="co">##  [17] Rsamtools_2.20.0                  RCurl_1.98-1.16                  </span></span>
<span><span class="co">##  [19] quantsmooth_1.70.0                htmltools_0.5.8.1                </span></span>
<span><span class="co">##  [21] S4Arrays_1.4.1                    curl_5.2.3                       </span></span>
<span><span class="co">##  [23] broom_1.0.7                       pROC_1.18.5                      </span></span>
<span><span class="co">##  [25] SparseArray_1.4.8                 mitml_0.4-5                      </span></span>
<span><span class="co">##  [27] sass_0.4.9                        bslib_0.8.0                      </span></span>
<span><span class="co">##  [29] htmlwidgets_1.6.4                 desc_1.4.3                       </span></span>
<span><span class="co">##  [31] plyr_1.8.9                        sandwich_3.1-1                   </span></span>
<span><span class="co">##  [33] zoo_1.8-12                        cachem_1.1.0                     </span></span>
<span><span class="co">##  [35] GenomicAlignments_1.40.0          lifecycle_1.0.4                  </span></span>
<span><span class="co">##  [37] iterators_1.0.14                  pkgconfig_2.0.3                  </span></span>
<span><span class="co">##  [39] R6_2.5.1                          fastmap_1.2.0                    </span></span>
<span><span class="co">##  [41] GenomeInfoDbData_1.2.12           MatrixGenerics_1.16.0            </span></span>
<span><span class="co">##  [43] digest_0.6.37                     colorspace_2.1-1                 </span></span>
<span><span class="co">##  [45] GWASTools_1.50.0                  AnnotationDbi_1.66.0             </span></span>
<span><span class="co">##  [47] S4Vectors_0.42.1                  textshaping_0.4.0                </span></span>
<span><span class="co">##  [49] GenomicRanges_1.56.2              RSQLite_2.3.7                    </span></span>
<span><span class="co">##  [51] labeling_0.4.3                    fansi_1.0.6                      </span></span>
<span><span class="co">##  [53] httr_1.4.7                        abind_1.4-8                      </span></span>
<span><span class="co">##  [55] mgcv_1.9-1                        compiler_4.4.1                   </span></span>
<span><span class="co">##  [57] withr_3.0.1                       bit64_4.5.2                      </span></span>
<span><span class="co">##  [59] backports_1.5.0                   BiocParallel_1.38.0              </span></span>
<span><span class="co">##  [61] DBI_1.2.3                         highr_0.11                       </span></span>
<span><span class="co">##  [63] pan_1.9                           MASS_7.3-61                      </span></span>
<span><span class="co">##  [65] quantreg_5.98                     DelayedArray_0.30.1              </span></span>
<span><span class="co">##  [67] rjson_0.2.23                      DNAcopy_1.78.0                   </span></span>
<span><span class="co">##  [69] tools_4.4.1                       lmtest_0.9-40                    </span></span>
<span><span class="co">##  [71] nnet_7.3-19                       glue_1.8.0                       </span></span>
<span><span class="co">##  [73] restfulr_0.0.15                   nlme_3.1-165                     </span></span>
<span><span class="co">##  [75] grid_4.4.1                        generics_0.1.3                   </span></span>
<span><span class="co">##  [77] gtable_0.3.5                      operator.tools_1.6.3             </span></span>
<span><span class="co">##  [79] BSgenome_1.72.0                   formula.tools_1.7.1              </span></span>
<span><span class="co">##  [81] class_7.3-22                      tidyr_1.3.1                      </span></span>
<span><span class="co">##  [83] ensembldb_2.28.1                  data.table_1.16.2                </span></span>
<span><span class="co">##  [85] utf8_1.2.4                        XVector_0.44.0                   </span></span>
<span><span class="co">##  [87] BiocGenerics_0.50.0               GWASExactHW_1.2                  </span></span>
<span><span class="co">##  [89] stringr_1.5.1                     foreach_1.5.2                    </span></span>
<span><span class="co">##  [91] pillar_1.9.0                      splines_4.4.1                    </span></span>
<span><span class="co">##  [93] dplyr_1.1.4                       lattice_0.22-6                   </span></span>
<span><span class="co">##  [95] survival_3.7-0                    rtracklayer_1.64.0               </span></span>
<span><span class="co">##  [97] bit_4.5.0                         SparseM_1.84-2                   </span></span>
<span><span class="co">##  [99] BSgenome.Hsapiens.UCSC.hg38_1.4.5 tidyselect_1.2.1                 </span></span>
<span><span class="co">## [101] SeqVarTools_1.42.0                Biostrings_2.72.1                </span></span>
<span><span class="co">## [103] bookdown_0.41                     ProtGenerics_1.36.0              </span></span>
<span><span class="co">## [105] IRanges_2.38.1                    SummarizedExperiment_1.34.0      </span></span>
<span><span class="co">## [107] stats4_4.4.1                      xfun_0.48                        </span></span>
<span><span class="co">## [109] Biobase_2.64.0                    matrixStats_1.4.1                </span></span>
<span><span class="co">## [111] stringi_1.8.4                     UCSC.utils_1.0.0                 </span></span>
<span><span class="co">## [113] lazyeval_0.2.2                    yaml_2.3.10                      </span></span>
<span><span class="co">## [115] boot_1.3-30                       evaluate_1.0.1                   </span></span>
<span><span class="co">## [117] codetools_0.2-20                  tibble_3.2.1                     </span></span>
<span><span class="co">## [119] BiocManager_1.30.25               cli_3.6.3                        </span></span>
<span><span class="co">## [121] rpart_4.1.23                      systemfonts_1.1.0                </span></span>
<span><span class="co">## [123] munsell_0.5.1                     jquerylib_0.1.4                  </span></span>
<span><span class="co">## [125] Rcpp_1.0.13                       GenomeInfoDb_1.40.1              </span></span>
<span><span class="co">## [127] png_0.1-8                         XML_3.99-0.17                    </span></span>
<span><span class="co">## [129] parallel_4.4.1                    MatrixModels_0.5-3               </span></span>
<span><span class="co">## [131] ggplot2_3.5.1                     pkgdown_2.1.1                    </span></span>
<span><span class="co">## [133] blob_1.2.4                        AnnotationFilter_1.28.0          </span></span>
<span><span class="co">## [135] bitops_1.0-9                      lme4_1.1-35.5                    </span></span>
<span><span class="co">## [137] glmnet_4.1-8                      SeqArray_1.44.3                  </span></span>
<span><span class="co">## [139] VariantAnnotation_1.50.0          scales_1.3.0                     </span></span>
<span><span class="co">## [141] purrr_1.0.2                       crayon_1.5.3                     </span></span>
<span><span class="co">## [143] rlang_1.1.4                       KEGGREST_1.44.1                  </span></span>
<span><span class="co">## [145] mice_3.16.0</span></span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Alexander2009" class="csl-entry">
Alexander, D. H., J. Novembre, and K. Lange. 2009. <span>“Fast
Model-Based Estimation of Ancestry in Unrelated Individuals.”</span>
<em>Genome Res</em> 19 (9): 1655–64. <a href="https://doi.org/10.1101/gr.094052.109" class="external-link">https://doi.org/10.1101/gr.094052.109</a>.
</div>
<div id="ref-Lowy-Gallego2019a" class="csl-entry">
Lowy-Gallego, Ernesto, Susan Fairley, Xiangqun Zheng-Bradley, Magali
Ruffier, Laura Clarke, and Paul Flicek. 2019. <span>“<span class="nocase">Variant calling on the grch38 assembly with the data from
phase three of the 1000 genomes project [version 2; peer review: 1
approved, 1 not approved]</span>.”</span> <em>Wellcome Open
Research</em> 4: 1–42. <a href="https://doi.org/10.12688/wellcomeopenres.15126.1" class="external-link">https://doi.org/10.12688/wellcomeopenres.15126.1</a>.
</div>
<div id="ref-Price2006" class="csl-entry">
Price, A. L., N. J. Patterson, R. M. Plenge, M. E. Weinblatt, N. A.
Shadick, and D. Reich. 2006. <span>“Principal Components Analysis
Corrects for Stratification in Genome-Wide Association Studies.”</span>
<em>Nat Genet</em> 38 (8): 904–9. <a href="https://doi.org/10.1038/ng1847" class="external-link">https://doi.org/10.1038/ng1847</a>.
</div>
<div id="ref-Pritchard2000" class="csl-entry">
Pritchard, J. K., M. Stephens, and P. Donnelly. 2000. <span>“Inference
of Population Structure Using Multilocus Genotype Data.”</span>
<em>Genetics</em> 155 (2): 945–59. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10835412" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/10835412</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
