<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate Inference of Genetic Ancestry from Cancer-derived Sequences • RAIDS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate Inference of Genetic Ancestry from Cancer-derived Sequences">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RAIDS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RAIDS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Create_Reference_GDS_File.html">Population reference dataset GDS files</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KrasnitzLab/RAIDS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate Inference of Genetic Ancestry from
Cancer-derived Sequences</h1>
                        <h4 data-toc-skip class="author">Pascal Belleau,
Astrid Deschênes and Alexander Krasnitz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KrasnitzLab/RAIDS/blob/main/vignettes/RAIDS.Rmd" class="external-link"><code>vignettes/RAIDS.Rmd</code></a></small>
      <div class="hidden name"><code>RAIDS.Rmd</code></div>

    </div>

    
    
<p><br><strong>Package</strong>: <em>RAIDS</em><br><strong>Authors</strong>: Pascal Belleau [cre, aut] (<a href="https://orcid.org/0000-0002-0802-1071" class="external-link uri">https://orcid.org/0000-0002-0802-1071</a>), Astrid Deschênes
[aut] (<a href="https://orcid.org/0000-0001-7846-6749" class="external-link uri">https://orcid.org/0000-0001-7846-6749</a>), David A. Tuveson
[aut] (<a href="https://orcid.org/0000-0002-8017-2712" class="external-link uri">https://orcid.org/0000-0002-8017-2712</a>), Alexander
Krasnitz [aut]<br><strong>Version</strong>: 1.3.1<br><strong>Compiled date</strong>: 2024-09-05<br><strong>License</strong>: Apache License (&gt;= 2)<br></p>
<div class="section level2">
<h2 id="licensing">Licensing<a class="anchor" aria-label="anchor" href="#licensing"></a>
</h2>
<p>The <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em>
package and the underlying <em><a href="https://github.com/KrasnitzLab/RAIDS" class="external-link">RAIDS</a></em> code are
distributed under<br>
the <a href="https://opensource.org/licenses/Apache-2.0" class="external-link uri">https://opensource.org/licenses/Apache-2.0</a> license. You
are free to use and redistribute this software.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>If you use the <strong>RAIDS</strong> package for a publication, we
would ask you to cite the following:</p>
<blockquote>
<p>Pascal Belleau, Astrid Deschênes, Nyasha Chambwe, David A. Tuveson,
Alexander Krasnitz; Genetic Ancestry Inference from Cancer-Derived
Molecular Data across Genomic and Transcriptomic Platforms. Cancer Res 1
January 2023; 83 (1): 49–58. <a href="https://doi.org/10.1158/0008-5472.CAN-22-0682" class="external-link uri">https://doi.org/10.1158/0008-5472.CAN-22-0682</a> <br><br></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Multiple methods have been implemented to infer ancestry from
germline DNA sequence <span class="citation">(Price et al. 2006;
Pritchard, Stephens, and Donnelly 2000; Alexander, Novembre, and Lange
2009)</span>. However, genotyping of DNA from matched normal specimens
is not part of standard clinical practice and is not performed routinely
outside academic clinical centers. In sum, matched germline DNA sequence
is often missing for cancer-derived molecular data. In such cases,
having the possibility to infer ancestry from tumor-derived data would
be beneficial.</p>
<p>The <strong>RAIDS</strong> package implements an inference procedure
that has been specifically developed to accurately infer genetic
ancestry from cancer-derived sequences. The current version can handle
cancer-derived sequences of:</p>
<ul>
<li>tumor exomes</li>
<li>targeted gene panels</li>
<li>RNA</li>
</ul>
<p>The <strong>RAIDS</strong> package implements a data synthesis method
that, for any given cancer-derived sequence profile, enables on the one
hand, profile-specific inference parameter optimization and on the other
hand, a profile-specific inference accuracy estimate.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install this package from <a href="https://bioconductor.org/packages/RAIDS" class="external-link">Bioconductor</a>, start R
(version 4.3 or later) and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"RAIDS"</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
<p><a name="mains"></a></p>
</div>
<div class="section level2">
<h2 id="main-steps">Main Steps<a class="anchor" aria-label="anchor" href="#main-steps"></a>
</h2>
<p>This is an overview of genetic ancestry inference from cancer-derived
molecular data:</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_v04.png" alt="An overview of the genetic ancestry inference process." width="130%"><p class="caption">
An overview of the genetic ancestry inference process.
</p>
</div>
<p>The main steps are:</p>
<p><strong>Step 1.</strong> Format reference data from the population
reference dataset (optional)</p>
<p><strong>Step 2.1</strong> Optimize ancestry inference parameters</p>
<p><strong>Step 2.2</strong> Infer ancestry for the subjects of the
external study</p>
<p>These steps are described in detail in the following. Steps 2.1 and
2.2 can be run together using one wrapper function.</p>
<p><br><br></p>
<div class="section level3">
<h3 id="main-step---ancestry-inference">Main Step - Ancestry Inference<a class="anchor" aria-label="anchor" href="#main-step---ancestry-inference"></a>
</h3>
<p>A wrapper function encapsulates multiple steps of the workflow.</p>
<div class="figure" style="text-align: center">
<img src="MainSteps_Wrapper_v04.png" alt="Final step - The wrapper function encapsulates multiple steps of the workflow." width="120%"><p class="caption">
Final step - The wrapper function encapsulates multiple steps of the
workflow.
</p>
</div>
<p>In summary, the wrapper function generates the synthetic dataset and
uses it to selected the optimal parameters before calling the genetic
ancestry on the current profiles.</p>
<p>According to the type of input data (RNA or DNA), a specific wrapper
function is available.</p>
<p><br></p>
<div class="section level4">
<h4 id="dna-data---wrapper-function-to-run-ancestry-inference-on-dna-data">DNA Data - Wrapper function to run ancestry inference on DNA
data<a class="anchor" aria-label="anchor" href="#dna-data---wrapper-function-to-run-ancestry-inference-on-dna-data"></a>
</h4>
<p>The wrapper function, called <em>runExomeAncestry()</em>, requires 4
files as input:</p>
<ul>
<li>The <strong>population reference GDS file</strong>
</li>
<li>The <strong>population reference SNV Annotation GDS
file</strong>
</li>
<li>The <strong>Profile SNP file</strong> (one per sample present in the
study)</li>
<li>The <strong>Profile PED RDS file</strong> (one file with information
for all profiles in the study)</li>
</ul>
<p>In addition, a <em>data.frame</em> containing the general information
about the study is also required. The <em>data.frame</em> must contain
those 3 columns:</p>
<ul>
<li>
<em>study.id</em>: The study identifier (example: TCGA-BRCA).</li>
<li>
<em>study.desc</em>: The description of the study.</li>
<li>
<em>study.platform</em>: The type of sequencing (example:
RNA-seq).</li>
</ul>
<p><br></p>
<div class="section level5">
<h5 id="population-reference-files">
<strong>Population reference files</strong><a class="anchor" aria-label="anchor" href="#population-reference-files"></a>
</h5>
<p>For demonstration purpose, a small <strong>population reference GDS
file</strong> (called <em>ex1_good_small_1KG.gds</em>) and a small
<strong>population reference SNV Annotation GDS file</strong> (called
<em>ex1_good_small_1KG_Annot.gds</em>) are included in this package.
Beware that those two files should not be used to run a real ancestry
inference.The results obtained with those files won’t be reliable.</p>
<p>The required <strong>population reference GDS file</strong> and
<strong>population reference SNV Annotation GDS file</strong> should be
stored in the same directory. In the example below, this directory is
referred to as <strong>pathReference</strong>.</p>
<p><br></p>
</div>
<div class="section level5">
<h5 id="profile-snp-file">
<strong>Profile SNP file</strong><a class="anchor" aria-label="anchor" href="#profile-snp-file"></a>
</h5>
<p>The <strong>Profile SNP file</strong> can be either in a VCF format
or in a generic format.</p>
<p>The <strong>Profile SNP VCF file</strong> follows the VCF standard
with at least those genotype fields: <em>GT</em>, <em>AD</em> and
<em>DP</em>. The identifier of the genotype in the VCF file must
correspond to the profile identifier <em>Name.ID</em>. The SNVs must be
germline variants and should include the genotype of the wild-type
homozygous at the selected positions in the reference. One file per
profile is need and the VCF file must be gzipped.</p>
<p>Note that the name assigned to the <strong>Profile SNP VCF
file</strong> has to correspond to the profile identifier
<em>Name.ID</em> in the following analysis. For example, a SNP file
called “Sample.01.vcf.gz” would be associated to the “Sample.01”
profile.</p>
<p>A generic SNP file can replace the VCF file. The <strong>Profile SNP
Generic file</strong> format is coma separated and the mandatory columns
are:</p>
<ul>
<li>
<em>Chromosome</em>: The name of the chromosome</li>
<li>
<em>Position</em>: The position on the chromosome</li>
<li>
<em>Ref</em>: The reference nucleotide</li>
<li>
<em>Alt</em>: The aternative nucleotide</li>
<li>
<em>Count</em>: The total count</li>
<li>
<em>File1R</em>: The count for the reference nucleotide</li>
<li>
<em>File1A</em>: The count for the alternative nucleotide</li>
</ul>
<p>Beware that the starting position in the <strong>population reference
GDS File</strong> is zero (like BED files). The <strong>Profile SNP
Generic file</strong> should also start at position zero.</p>
<p>Note that the name assigned to the <strong>Profile SNP Generic
file</strong> has to correspond to the profile identifier
<em>Name.ID</em> in the following analysis. For example, a SNP file
called “Sample.01.generic.txt.gz” would be associated to the “Sample.01”
profile.</p>
<p><br></p>
</div>
<div class="section level5">
<h5 id="profile-ped-rds-file">
<strong>Profile PED RDS file</strong><a class="anchor" aria-label="anchor" href="#profile-ped-rds-file"></a>
</h5>
<p>The <strong>Profile PED RDS file</strong> must contain a
<em>data.frame</em> describing all the profiles to be analyzed. These 5
mandatory columns:</p>
<ul>
<li>
<em>Name.ID</em>: The unique sample identifier. The associated
<strong>profile SNP file</strong> should be called
“Name.ID.txt.gz”.</li>
<li>
<em>Case.ID</em>: The patient identifier associated to the
sample.</li>
<li>
<em>Sample.Type</em>: The information about the profile tissue
source (primary tumor, metastatic tumor, normal, etc..).</li>
<li>
<em>Diagnosis</em>: The donor’s diagnosis.</li>
<li>
<em>Source</em>: The source of the profile sequence data (example:
dbGAP_XYZ).</li>
</ul>
<p>Important: The row names of the <em>data.frame</em> must be the
profiles <em>Name.ID</em>.</p>
<p>This file is referred to as the <strong>Profile PED RDS file</strong>
(PED for pedigree). Alternatively, the PED information can be saved in
another type of file (CVS, etc..) as long as the <em>data.frame</em>
information can be regenerated in R (with <em>read.csv()</em> or
else).</p>
<p><br></p>
</div>
<div class="section level5">
<h5 id="example">
<strong>Example</strong><a class="anchor" aria-label="anchor" href="#example"></a>
</h5>
<p>This example run an ancestry inference on an exome sample. Both
population reference files are demonstration files and should not be
used for a real ancestry inference. Beware that running an ancestry
inference on real data will take longer to run.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Path to the demo 1KG GDS file is located in this package</span></span>
<span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load the information about the profile</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">demoPedigreeEx1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demoPedigreeEx1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Name.ID Case.ID   Sample.Type Diagnosis     Source</span></span>
<span><span class="co">## ex1     ex1     ex1 Primary Tumor    Cancer Databank B</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## The population reference GDS file and SNV Annotation GDS file</span></span>
<span><span class="co">## need to be located in the same directory.</span></span>
<span><span class="co">## Note that the population reference GDS file used for this example is a</span></span>
<span><span class="co">## simplified version and CANNOT be used for any real analysis</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"tests"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span>
<span><span class="va">fileAnnotGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG_Annot.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## A data frame containing general information about the study</span></span>
<span><span class="co">## is also required. The data frame must have</span></span>
<span><span class="co">## those 3 columns: "study.id", "study.desc", "study.platform"</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id<span class="op">=</span><span class="st">"MYDATA"</span>,</span>
<span>                   study.desc<span class="op">=</span><span class="st">"Description"</span>,</span>
<span>                   study.platform<span class="op">=</span><span class="st">"PLATFORM"</span>,</span>
<span>                   stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The Sample SNP VCF files (one per sample) need</span></span>
<span><span class="co">## to be all located in the same directory.</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathGeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"snpPileup"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Fix RNG seed to ensure reproducible results</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Select the profiles from the population reference GDS file for </span></span>
<span><span class="co">## the synthetic data.</span></span>
<span><span class="co">## Here we select 2 profiles from the simplified 1KG GDS for each </span></span>
<span><span class="co">## subcontinental-level.</span></span>
<span><span class="co">## Normally, we use 30 profile for each </span></span>
<span><span class="co">## subcontinental-level but it is too big for the example.</span></span>
<span><span class="co">## The 1KG files in this example only have 6 profiles for each </span></span>
<span><span class="co">## subcontinental-level (for demo purpose only).</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds1KG</span>, nbProfiles<span class="op">=</span><span class="fl">2L</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## GenomeInfoDb and BSgenome are required libraries to run this example</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GenomeInfoDb"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="co">## Chromosome length information</span></span>
<span>    <span class="co">## chr23 is chrX, chr24 is chrY and chrM is 25</span></span>
<span>    <span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co">###########################################################################</span></span>
<span>    <span class="co">## The path where the Sample GDS files (one per sample)</span></span>
<span>    <span class="co">## will be created needs to be specified.</span></span>
<span>    <span class="co">###########################################################################</span></span>
<span>    <span class="va">pathProfileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleDNA"</span>, <span class="st">"out.tmp"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">###########################################################################</span></span>
<span>    <span class="co">## The path where the result files will be created needs to </span></span>
<span>    <span class="co">## be specified</span></span>
<span>    <span class="co">###########################################################################</span></span>
<span>    <span class="va">pathOut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleDNA"</span>, <span class="st">"res.out"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Example can only be run if the current directory is in writing mode</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleDNA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleDNA"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span></span>
<span>    </span>
<span>        <span class="co">#########################################################################</span></span>
<span>        <span class="co">## The wrapper function generates the synthetic dataset and uses it </span></span>
<span>        <span class="co">## to selected the optimal parameters before calling the genetic </span></span>
<span>        <span class="co">## ancestry on the current profiles.</span></span>
<span>        <span class="co">## All important information, for each step, are saved in </span></span>
<span>        <span class="co">## multiple output files.</span></span>
<span>        <span class="co">## The 'genoSource' parameter has 2 options depending on how the </span></span>
<span>        <span class="co">##   SNP files have been generated: </span></span>
<span>        <span class="co">##   SNP VCF files have been generated: </span></span>
<span>        <span class="co">##  "VCF" or "generic" (other software)</span></span>
<span>        <span class="co">##</span></span>
<span>        <span class="co">#########################################################################</span></span>
<span>        <span class="fu"><a href="../reference/runExomeAncestry.html">runExomeAncestry</a></span><span class="op">(</span>pedStudy<span class="op">=</span><span class="va">demoPedigreeEx1</span>, studyDF<span class="op">=</span><span class="va">studyDF</span>,</span>
<span>                 pathProfileGDS<span class="op">=</span><span class="va">pathProfileGDS</span>,</span>
<span>                 pathGeno<span class="op">=</span><span class="va">pathGeno</span>,</span>
<span>                 pathOut<span class="op">=</span><span class="va">pathOut</span>,</span>
<span>                 fileReferenceGDS<span class="op">=</span><span class="va">fileGDS</span>,</span>
<span>                 fileReferenceAnnotGDS<span class="op">=</span><span class="va">fileAnnotGDS</span>,</span>
<span>                 chrInfo<span class="op">=</span><span class="va">chrInfo</span>,</span>
<span>                 syntheticRefDF<span class="op">=</span><span class="va">dataRef</span>,</span>
<span>                 genoSource<span class="op">=</span><span class="st">"VCF"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, <span class="va">demoPedigreeEx1</span><span class="op">$</span><span class="va">Name.ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co">#######################################################################</span></span>
<span>        <span class="co">## The file containing the ancestry inference (SuperPop column) and </span></span>
<span>        <span class="co">## optimal number of PCA component (D column)</span></span>
<span>        <span class="co">## optimal number of neighbours (K column)</span></span>
<span>        <span class="co">#######################################################################</span></span>
<span>        <span class="va">resAncestry</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">demoPedigreeEx1</span><span class="op">$</span><span class="va">Name.ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">".Ancestry.csv"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">resAncestry</span><span class="op">)</span></span>
<span></span>
<span>        <span class="co">## Remove temporary files created for this demo</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathOut</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleDNA"</span><span class="op">)</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample.id  D K SuperPop</span></span>
<span><span class="co">## 1       ex1 14 6      AFR</span></span></code></pre>
<p><br><br></p>
<p>The <em>runExomeAncestry()</em> function generates 3 types of files
in the <em>pathOut</em> directory.</p>
<ul>
<li>The ancestry inference CSV file (<strong>“.Ancestry.csv”</strong>
file)</li>
<li>The inference information RDS file (<strong>“.infoCall.rds”</strong>
file)</li>
<li>The parameter information RDS files from the synthetic inference
(<strong>“KNN.synt.</strong>*<strong>.rds”</strong> files in a
sub-directory)</li>
</ul>
<p>In addition, a sub-directory (named using the <em>profile ID</em>) is
also created.</p>
<p>The inferred ancestry is stored in the ancestry inference CSV file
(<strong>“.Ancestry.csv”</strong> file) which also contains those
columns:</p>
<ul>
<li>
<em>sample.id</em>: The unique identifier of the sample</li>
<li>
<em>D</em>: The optimal PCA dimension value used to infer the
ancestry</li>
<li>
<em>k</em>: The optimal number of neighbors value used to infer the
ancestry</li>
<li>
<em>SuperPop</em>: The inferred ancestry</li>
</ul>
<p><br><br></p>
</div>
</div>
<div class="section level4">
<h4 id="rna-data---wrapper-function-to-run-ancestry-inference-on-rna-data">RNA data - Wrapper function to run ancestry inference on RNA
data<a class="anchor" aria-label="anchor" href="#rna-data---wrapper-function-to-run-ancestry-inference-on-rna-data"></a>
</h4>
<p>The process is the same as for the DNA but use the wrapper function
called <em>runRNAAncestry()</em>. Internally the data is process
differently. It requires 4 files as input:</p>
<ul>
<li>The <strong>population reference GDS file</strong>
</li>
<li>The <strong>population reference SNV Annotation GDS
file</strong>
</li>
<li>The <strong>Profile SNP file</strong> (one per sample present in the
study)</li>
<li>The <strong>Profile PED RDS file</strong> (one file with information
for all profiles in the study)</li>
</ul>
<p>A <em>data.frame</em> containing the general information about the
study is also required. The <em>data.frame</em> must contain those 3
columns:</p>
<ul>
<li>
<em>study.id</em>: The study identifier (example: TCGA-BRCA).</li>
<li>
<em>study.desc</em>: The description of the study.</li>
<li>
<em>study.platform</em>: The type of sequencing (example:
RNA-seq).</li>
</ul>
<p><br></p>
<div class="section level5">
<h5 id="population-reference-files-1">
<strong>Population reference files</strong><a class="anchor" aria-label="anchor" href="#population-reference-files-1"></a>
</h5>
<p>For demonstration purpose, a small <strong>population reference GDS
file</strong> (called <em>ex1_good_small_1KG.gds</em>) and a small
<strong>population reference SNV Annotation GDS file</strong> (called
<em>ex1_good_small_1KG_Annot.gds</em>) are included in this package.
Beware that those two files should not be used to run a real ancestry
inference.The results obtained with those files won’t be reliable.</p>
<p>The required <strong>population reference GDS file</strong> and
<strong>population reference SNV Annotation GDS file</strong> should be
stored in the same directory. In the example below, this directory is
referred to as <strong>pathReference</strong>.</p>
<p><br></p>
</div>
<div class="section level5">
<h5 id="profile-snp-file-1">
<strong>Profile SNP file</strong><a class="anchor" aria-label="anchor" href="#profile-snp-file-1"></a>
</h5>
<p>The <strong>Profile SNP file</strong> can be either in a VCF format
or in a generic format.</p>
<p>The <strong>Profile SNP VCF file</strong> follows the VCF standard
with at least those genotype fields: <em>GT</em>, <em>AD</em> and
<em>DP</em>. The identifier of the genotype in the VCF file must
correspond to the profile identifier <em>Name.ID</em>. The SNVs must be
germline variants and should include the genotype of the wild-type
homozygous at the selected positions in the reference. One file per
profile is need and the VCF file must be gzipped.</p>
<p>Note that the name assigned to the <strong>Profile SNP VCF
file</strong> has to correspond to the profile identifier
<em>Name.ID</em> in the following analysis. For example, a SNP file
called “Sample.01.vcf.gz” would be associated to the “Sample.01”
profile.</p>
<p>A generic SNP file can replace the VCF file. The <strong>Profile SNP
Generic file</strong> format is coma separated and the mandatory columns
are:</p>
<ul>
<li>
<em>Chromosome</em>: The name of the chromosome</li>
<li>
<em>Position</em>: The position on the chromosome</li>
<li>
<em>Ref</em>: The reference nucleotide</li>
<li>
<em>Alt</em>: The aternative nucleotide</li>
<li>
<em>Count</em>: The total count</li>
<li>
<em>File1R</em>: The count for the reference nucleotide</li>
<li>
<em>File1A</em>: The count for the alternative nucleotide</li>
</ul>
<p>Beware that the starting position in the <strong>population reference
GDS File</strong> is zero (like BED files). The <strong>Profile SNP
Generic file</strong> should also start at position zero.</p>
<p>Note that the name assigned to the <strong>Profile SNP Generic
file</strong> has to correspond to the profile identifier
<em>Name.ID</em> in the following analysis. For example, a SNP file
called “Sample.01.generic.txt.gz” would be associated to the “Sample.01”
profile.</p>
<p><br></p>
</div>
<div class="section level5">
<h5 id="profile-ped-rds-file-1">
<strong>Profile PED RDS file</strong><a class="anchor" aria-label="anchor" href="#profile-ped-rds-file-1"></a>
</h5>
<p>The <strong>Profile PED RDS file</strong> must contain a
<em>data.frame</em> describing all the profiles to be analyzed. These 5
mandatory columns:</p>
<ul>
<li>
<em>Name.ID</em>: The unique sample identifier. The associated
<strong>profile SNP file</strong><br>
should be called “Name.ID.txt.gz”.</li>
<li>
<em>Case.ID</em>: The patient identifier associated to the
sample.</li>
<li>
<em>Sample.Type</em>: The information about the profile tissue
source (primary tumor, metastatic tumor, normal, etc..).</li>
<li>
<em>Diagnosis</em>: The donor’s diagnosis.</li>
<li>
<em>Source</em>: The source of the profile sequence data (example:
dbGAP_XYZ).</li>
</ul>
<p>Important: The row names of the <em>data.frame</em> must be the
profiles <em>Name.ID</em>.</p>
<p>This file is referred to as the <strong>Profile PED RDS file</strong>
(PED for pedigree). Alternatively, the PED information can be saved in
another type of file (CVS, etc..) as long as the <em>data.frame</em>
information can be regenerated in R (with <em>read.csv()</em> or
else).</p>
<p><br></p>
</div>
<div class="section level5">
<h5 id="example-1">
<strong>Example</strong><a class="anchor" aria-label="anchor" href="#example-1"></a>
</h5>
<p>This example run an ancestry inference on an RNA sample. Both
population reference files are demonstration files and should not be
used for a real ancestry inference. Beware that running an ancestry
inference on real data will take longer to run.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load required packages</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://krasnitzlab.github.io/RAIDS/" class="external-link">RAIDS</a></span><span class="op">)</span>    </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/gdsfmt" class="external-link">gdsfmt</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Path to the demo 1KG GDS file is located in this package</span></span>
<span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"RAIDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Load the information about the profile</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">demoPedigreeEx1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demoPedigreeEx1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Name.ID Case.ID   Sample.Type Diagnosis     Source</span></span>
<span><span class="co">## ex1     ex1     ex1 Primary Tumor    Cancer Databank B</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#############################################################################</span></span>
<span><span class="co">## The population reference GDS file and SNV Annotation GDS file</span></span>
<span><span class="co">## need to be located in the same directory.</span></span>
<span><span class="co">## Note that the population reference GDS file used for this example is a</span></span>
<span><span class="co">## simplified version and CANNOT be used for any real analysis</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathReference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"tests"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG.gds"</span><span class="op">)</span></span>
<span><span class="va">fileAnnotGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathReference</span>, <span class="st">"ex1_good_small_1KG_Annot.gds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## A data frame containing general information about the study</span></span>
<span><span class="co">## is also required. The data frame must have</span></span>
<span><span class="co">## those 3 columns: "study.id", "study.desc", "study.platform"</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">studyDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>study.id<span class="op">=</span><span class="st">"MYDATA"</span>,</span>
<span>                   study.desc<span class="op">=</span><span class="st">"Description"</span>,</span>
<span>                   study.platform<span class="op">=</span><span class="st">"PLATFORM"</span>,</span>
<span>                   stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## The Sample SNP VCF files (one per sample) need</span></span>
<span><span class="co">## to be all located in the same directory.</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">pathGeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dataDir</span>, <span class="st">"example"</span>, <span class="st">"snpPileupRNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Fix RNG seed to ensure reproducible results</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3043</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="co">## Select the profiles from the population reference GDS file for </span></span>
<span><span class="co">## the synthetic data.</span></span>
<span><span class="co">## Here we select 2 profiles from the simplified 1KG GDS for each </span></span>
<span><span class="co">## subcontinental-level.</span></span>
<span><span class="co">## Normally, we use 30 profile for each </span></span>
<span><span class="co">## subcontinental-level but it is too big for the example.</span></span>
<span><span class="co">## The 1KG files in this example only have 6 profiles for each </span></span>
<span><span class="co">## subcontinental-level (for demo purpose only).</span></span>
<span><span class="co">#############################################################################</span></span>
<span><span class="va">gds1KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsOpen.html" class="external-link">snpgdsOpen</a></span><span class="op">(</span><span class="va">fileGDS</span><span class="op">)</span></span>
<span><span class="va">dataRef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select1KGPop.html">select1KGPop</a></span><span class="op">(</span><span class="va">gds1KG</span>, nbProfiles<span class="op">=</span><span class="fl">2L</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gdsfmt/man/closefn.gds.html" class="external-link">closefn.gds</a></span><span class="op">(</span><span class="va">gds1KG</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## GenomeInfoDb and BSgenome are required libraries to run this example</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GenomeInfoDb"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="co">## Chromosome length information</span></span>
<span>  <span class="co">## chr23 is chrX, chr24 is chrY and chrM is 25</span></span>
<span>  <span class="va">chrInfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="fu">BSgenome.Hsapiens.UCSC.hg38</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html" class="external-link">Hsapiens</a></span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span>
<span></span>
<span>  <span class="co">#############################################################################</span></span>
<span>  <span class="co">## The path where the Sample GDS files (one per sample)</span></span>
<span>  <span class="co">## will be created needs to be specified.</span></span>
<span>  <span class="co">#############################################################################</span></span>
<span>  <span class="va">pathProfileGDS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleRNA"</span>, <span class="st">"outRNA.tmp"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">#############################################################################</span></span>
<span>  <span class="co">## The path where the result files will be created needs to </span></span>
<span>  <span class="co">## be specified</span></span>
<span>  <span class="co">#############################################################################</span></span>
<span>  <span class="va">pathOut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleRNA"</span>, <span class="st">"resRNA.out"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co">## Example can only be run if the current directory is in writing mode</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleRNA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"exampleRNA"</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathProfileGDS</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span></span>
<span>    </span>
<span>      <span class="co">#########################################################################</span></span>
<span>      <span class="co">## The wrapper function generates the synthetic dataset and uses it </span></span>
<span>      <span class="co">## to selected the optimal parameters before calling the genetic </span></span>
<span>      <span class="co">## ancestry on the current profiles.</span></span>
<span>      <span class="co">## All important information, for each step, are saved in </span></span>
<span>      <span class="co">## multiple output files.</span></span>
<span>      <span class="co">## The 'genoSource' parameter has 2 options depending on how the </span></span>
<span>      <span class="co">##   SNP files have been generated: </span></span>
<span>      <span class="co">##   SNP VCF files have been generated: </span></span>
<span>      <span class="co">##  "VCF" or "generic" (other software)</span></span>
<span>      <span class="co">#########################################################################</span></span>
<span>      <span class="fu"><a href="../reference/runRNAAncestry.html">runRNAAncestry</a></span><span class="op">(</span>pedStudy<span class="op">=</span><span class="va">demoPedigreeEx1</span>, studyDF<span class="op">=</span><span class="va">studyDF</span>,</span>
<span>                    pathProfileGDS<span class="op">=</span><span class="va">pathProfileGDS</span>,</span>
<span>                    pathGeno<span class="op">=</span><span class="va">pathGeno</span>,</span>
<span>                    pathOut<span class="op">=</span><span class="va">pathOut</span>,</span>
<span>                    fileReferenceGDS<span class="op">=</span><span class="va">fileGDS</span>,</span>
<span>                    fileReferenceAnnotGDS<span class="op">=</span><span class="va">fileAnnotGDS</span>,</span>
<span>                    chrInfo<span class="op">=</span><span class="va">chrInfo</span>,</span>
<span>                    syntheticRefDF<span class="op">=</span><span class="va">dataRef</span>,</span>
<span>                    blockTypeID<span class="op">=</span><span class="st">"GeneS.Ensembl.Hsapiens.v86"</span>,</span>
<span>                    genoSource<span class="op">=</span><span class="st">"VCF"</span><span class="op">)</span></span>
<span>    </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">pathOut</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, <span class="va">demoPedigreeEx1</span><span class="op">$</span><span class="va">Name.ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>      <span class="co">#########################################################################</span></span>
<span>      <span class="co">## The file containing the ancestry inference (SuperPop column) and </span></span>
<span>      <span class="co">## optimal number of PCA component (D column)</span></span>
<span>      <span class="co">## optimal number of neighbours (K column)</span></span>
<span>      <span class="co">#########################################################################</span></span>
<span>      <span class="va">resAncestry</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pathOut</span>, </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">demoPedigreeEx1</span><span class="op">$</span><span class="va">Name.ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">".Ancestry.csv"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">resAncestry</span><span class="op">)</span></span>
<span></span>
<span>      <span class="co">## Remove temporary files created for this demo</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathProfileGDS</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">pathOut</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"example"</span><span class="op">)</span>, recursive<span class="op">=</span><span class="cn">TRUE</span>, force<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample.id  D K SuperPop</span></span>
<span><span class="co">## 1       ex1 14 6      AFR</span></span></code></pre>
<p><br><br></p>
<p>The <em>runRNAAncestry()</em> function generates 3 types of files in
the <em>pathOut</em> directory.</p>
<ul>
<li>The ancestry inference CSV file (<strong>“.Ancestry.csv”</strong>
file)</li>
<li>The inference information RDS file (<strong>“.infoCall.rds”</strong>
file)</li>
<li>The parameter information RDS files from the synthetic inference
(<strong>“KNN.synt.</strong>*<strong>.rds”</strong> files in a
sub-directory)</li>
</ul>
<p>In addition, a sub-directory (named using the <em>profile ID</em>) is
also created.</p>
<p>The inferred ancestry is stored in the ancestry inference CSV file
(<strong>“.Ancestry.csv”</strong> file) which also contains those
columns:</p>
<ul>
<li>
<em>sample.id</em>: The unique identifier of the sample</li>
<li>
<em>D</em>: The optimal PCA dimension value used to infer the
ancestry</li>
<li>
<em>k</em>: The optimal number of neighbors value used to infer the
ancestry</li>
<li>
<em>SuperPop</em>: The inferred ancestry</li>
</ul>
<p><br><br></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="format-population-reference-dataset-optional">Format population reference dataset (optional)<a class="anchor" aria-label="anchor" href="#format-population-reference-dataset-optional"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="MainSteps_Step1_v04.png" alt="Step 1 - Formatting the information from the population reference dataset (optional)" width="120%"><p class="caption">
Step 1 - Formatting the information from the population reference
dataset (optional)
</p>
</div>
<p>A population reference dataset with known ancestry is required to
infer ancestry. The population must be large enough to ensure ???</p>
<p>Three important reference files, containing formatted information
about the reference dataset, are required:</p>
<ul>
<li>The population reference GDS File</li>
<li>The population reference SNV Annotation GDS file</li>
<li>The population reference SNV Retained VCF file</li>
</ul>
<p>The format of those files are described the <a href="Create_Reference_GDS_File.html">Population reference dataset GDS
files</a> vignette.</p>
<p><span style="color:darkred;font-weight:bold">The reference files
associated to the Cancer Research associated paper are available. Note
that these pre-processed files are for 1000 Genomes (1KG), in hg38. The
files are available here:</span></p>
<p><span style="color:red"> <a href="https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper" class="external-link">https://labshare.cshl.edu/shares/krasnitzlab/aicsPaper</a>
</span></p>
<p><span style="color:darkred;font-weight:bold">The size of the 1KG GDS
file is 15GB.</span></p>
<p><span style="color:darkred;font-weight:bold">The 1KG GDS file is
mapped on hg38 <span class="citation">(Lowy-Gallego et al.
2019)</span>.</span></p>
<p><span style="color:darkred;font-weight:bold">This section can be
skipped if you choose to use the pre-processed files.</span></p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> in the environment
in which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] Matrix_1.7-0     RAIDS_1.3.1      GENESIS_2.34.0   SNPRelate_1.38.0</span></span>
<span><span class="co">## [5] gdsfmt_1.40.1    knitr_1.48       BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.8                    shape_1.4.6.1                    </span></span>
<span><span class="co">##   [3] magrittr_2.0.3                    jomo_2.7-6                       </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.56.0            logistf_1.26.0                   </span></span>
<span><span class="co">##   [7] nloptr_2.1.1                      rmarkdown_2.28                   </span></span>
<span><span class="co">##   [9] fs_1.6.4                          BiocIO_1.14.0                    </span></span>
<span><span class="co">##  [11] zlibbioc_1.50.0                   ragg_1.3.2                       </span></span>
<span><span class="co">##  [13] vctrs_0.6.5                       memoise_2.0.1                    </span></span>
<span><span class="co">##  [15] minqa_1.2.8                       Rsamtools_2.20.0                 </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.16                   quantsmooth_1.70.0               </span></span>
<span><span class="co">##  [19] htmltools_0.5.8.1                 S4Arrays_1.4.1                   </span></span>
<span><span class="co">##  [21] curl_5.2.2                        broom_1.0.6                      </span></span>
<span><span class="co">##  [23] pROC_1.18.5                       SparseArray_1.4.8                </span></span>
<span><span class="co">##  [25] mitml_0.4-5                       sass_0.4.9                       </span></span>
<span><span class="co">##  [27] bslib_0.8.0                       htmlwidgets_1.6.4                </span></span>
<span><span class="co">##  [29] desc_1.4.3                        plyr_1.8.9                       </span></span>
<span><span class="co">##  [31] sandwich_3.1-0                    zoo_1.8-12                       </span></span>
<span><span class="co">##  [33] cachem_1.1.0                      GenomicAlignments_1.40.0         </span></span>
<span><span class="co">##  [35] lifecycle_1.0.4                   iterators_1.0.14                 </span></span>
<span><span class="co">##  [37] pkgconfig_2.0.3                   R6_2.5.1                         </span></span>
<span><span class="co">##  [39] fastmap_1.2.0                     GenomeInfoDbData_1.2.12          </span></span>
<span><span class="co">##  [41] MatrixGenerics_1.16.0             digest_0.6.37                    </span></span>
<span><span class="co">##  [43] colorspace_2.1-1                  GWASTools_1.50.0                 </span></span>
<span><span class="co">##  [45] AnnotationDbi_1.66.0              S4Vectors_0.42.1                 </span></span>
<span><span class="co">##  [47] textshaping_0.4.0                 GenomicRanges_1.56.1             </span></span>
<span><span class="co">##  [49] RSQLite_2.3.7                     fansi_1.0.6                      </span></span>
<span><span class="co">##  [51] httr_1.4.7                        abind_1.4-5                      </span></span>
<span><span class="co">##  [53] mgcv_1.9-1                        compiler_4.4.1                   </span></span>
<span><span class="co">##  [55] bit64_4.0.5                       backports_1.5.0                  </span></span>
<span><span class="co">##  [57] BiocParallel_1.38.0               DBI_1.2.3                        </span></span>
<span><span class="co">##  [59] highr_0.11                        pan_1.9                          </span></span>
<span><span class="co">##  [61] MASS_7.3-61                       quantreg_5.98                    </span></span>
<span><span class="co">##  [63] DelayedArray_0.30.1               rjson_0.2.22                     </span></span>
<span><span class="co">##  [65] DNAcopy_1.78.0                    tools_4.4.1                      </span></span>
<span><span class="co">##  [67] lmtest_0.9-40                     nnet_7.3-19                      </span></span>
<span><span class="co">##  [69] glue_1.7.0                        restfulr_0.0.15                  </span></span>
<span><span class="co">##  [71] nlme_3.1-165                      grid_4.4.1                       </span></span>
<span><span class="co">##  [73] generics_0.1.3                    gtable_0.3.5                     </span></span>
<span><span class="co">##  [75] operator.tools_1.6.3              BSgenome_1.72.0                  </span></span>
<span><span class="co">##  [77] formula.tools_1.7.1               class_7.3-22                     </span></span>
<span><span class="co">##  [79] tidyr_1.3.1                       ensembldb_2.28.1                 </span></span>
<span><span class="co">##  [81] data.table_1.16.0                 utf8_1.2.4                       </span></span>
<span><span class="co">##  [83] XVector_0.44.0                    BiocGenerics_0.50.0              </span></span>
<span><span class="co">##  [85] GWASExactHW_1.2                   stringr_1.5.1                    </span></span>
<span><span class="co">##  [87] foreach_1.5.2                     pillar_1.9.0                     </span></span>
<span><span class="co">##  [89] splines_4.4.1                     dplyr_1.1.4                      </span></span>
<span><span class="co">##  [91] lattice_0.22-6                    survival_3.7-0                   </span></span>
<span><span class="co">##  [93] rtracklayer_1.64.0                bit_4.0.5                        </span></span>
<span><span class="co">##  [95] SparseM_1.84-2                    BSgenome.Hsapiens.UCSC.hg38_1.4.5</span></span>
<span><span class="co">##  [97] tidyselect_1.2.1                  SeqVarTools_1.42.0               </span></span>
<span><span class="co">##  [99] Biostrings_2.72.1                 bookdown_0.40                    </span></span>
<span><span class="co">## [101] ProtGenerics_1.36.0               IRanges_2.38.1                   </span></span>
<span><span class="co">## [103] SummarizedExperiment_1.34.0       stats4_4.4.1                     </span></span>
<span><span class="co">## [105] xfun_0.47                         Biobase_2.64.0                   </span></span>
<span><span class="co">## [107] matrixStats_1.4.0                 stringi_1.8.4                    </span></span>
<span><span class="co">## [109] UCSC.utils_1.0.0                  lazyeval_0.2.2                   </span></span>
<span><span class="co">## [111] yaml_2.3.10                       boot_1.3-30                      </span></span>
<span><span class="co">## [113] evaluate_0.24.0                   codetools_0.2-20                 </span></span>
<span><span class="co">## [115] tibble_3.2.1                      BiocManager_1.30.25              </span></span>
<span><span class="co">## [117] cli_3.6.3                         rpart_4.1.23                     </span></span>
<span><span class="co">## [119] systemfonts_1.1.0                 munsell_0.5.1                    </span></span>
<span><span class="co">## [121] jquerylib_0.1.4                   Rcpp_1.0.13                      </span></span>
<span><span class="co">## [123] GenomeInfoDb_1.40.1               png_0.1-8                        </span></span>
<span><span class="co">## [125] XML_3.99-0.17                     parallel_4.4.1                   </span></span>
<span><span class="co">## [127] MatrixModels_0.5-3                ggplot2_3.5.1                    </span></span>
<span><span class="co">## [129] pkgdown_2.1.0                     blob_1.2.4                       </span></span>
<span><span class="co">## [131] AnnotationFilter_1.28.0           bitops_1.0-8                     </span></span>
<span><span class="co">## [133] lme4_1.1-35.5                     glmnet_4.1-8                     </span></span>
<span><span class="co">## [135] SeqArray_1.44.1                   VariantAnnotation_1.50.0         </span></span>
<span><span class="co">## [137] scales_1.3.0                      purrr_1.0.2                      </span></span>
<span><span class="co">## [139] crayon_1.5.3                      rlang_1.1.4                      </span></span>
<span><span class="co">## [141] KEGGREST_1.44.1                   mice_3.16.0</span></span></code></pre>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Alexander2009" class="csl-entry">
Alexander, D. H., J. Novembre, and K. Lange. 2009. <span>“Fast
Model-Based Estimation of Ancestry in Unrelated Individuals.”</span>
<em>Genome Res</em> 19 (9): 1655–64. <a href="https://doi.org/10.1101/gr.094052.109" class="external-link">https://doi.org/10.1101/gr.094052.109</a>.
</div>
<div id="ref-Lowy-Gallego2019a" class="csl-entry">
Lowy-Gallego, Ernesto, Susan Fairley, Xiangqun Zheng-Bradley, Magali
Ruffier, Laura Clarke, and Paul Flicek. 2019. <span>“<span class="nocase">Variant calling on the grch38 assembly with the data from
phase three of the 1000 genomes project [version 2; peer review: 1
approved, 1 not approved]</span>.”</span> <em>Wellcome Open
Research</em> 4: 1–42. <a href="https://doi.org/10.12688/wellcomeopenres.15126.1" class="external-link">https://doi.org/10.12688/wellcomeopenres.15126.1</a>.
</div>
<div id="ref-Price2006" class="csl-entry">
Price, A. L., N. J. Patterson, R. M. Plenge, M. E. Weinblatt, N. A.
Shadick, and D. Reich. 2006. <span>“Principal Components Analysis
Corrects for Stratification in Genome-Wide Association Studies.”</span>
<em>Nat Genet</em> 38 (8): 904–9. <a href="https://doi.org/10.1038/ng1847" class="external-link">https://doi.org/10.1038/ng1847</a>.
</div>
<div id="ref-Pritchard2000" class="csl-entry">
Pritchard, J. K., M. Stephens, and P. Donnelly. 2000. <span>“Inference
of Population Structure Using Multilocus Genotype Data.”</span>
<em>Genetics</em> 155 (2): 945–59. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10835412" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/10835412</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Pascal Belleau, Astrid Deschênes, David A. Tuveson, Alexander Krasnitz.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
